--- a/nss_dp_main.c
+++ b/nss_dp_main.c
@@ -295,8 +295,6 @@ static int nss_dp_close(struct net_devic
 		return -EAGAIN;
 	}
 
-	if (dp_priv->phydev)
-		phy_stop(dp_priv->phydev);
 	dp_priv->link_state = __NSS_DP_LINK_DOWN;
 
 #if defined(NSS_DP_PPE_SUPPORT)
@@ -434,7 +432,12 @@ static int nss_dp_open(struct net_device
 		netif_carrier_on(netdev);
 	} else {
 		dp_priv->link_state = __NSS_DP_LINK_DOWN;
-		phy_start(dp_priv->phydev);
+		if (phy_is_started(dp_priv->phydev)) {
+			dp_priv->phydev->state = PHY_UP;
+			phy_trigger_machine(dp_priv->phydev);
+		} else {
+			phy_start(dp_priv->phydev);
+		}
 		phy_start_aneg(dp_priv->phydev);
 	}
 
@@ -1024,6 +1027,9 @@ static int nss_dp_remove(struct platform
 		dp_ops = dp_priv->data_plane_ops;
 		hal_ops = dp_priv->gmac_hal_ops;
 
+		if (phy_is_started(dp_priv->phydev))
+			phy_stop(dp_priv->phydev);
+
 		if (dp_priv->phydev)
 			phy_disconnect(dp_priv->phydev);
 
