--- a/src/drivers/driver_nl80211_event.c	2025-07-11 09:37:29.772326959 +0800
+++ b/src/drivers/driver_nl80211_event.c	2025-07-14 15:27:22.079961399 +0800
@@ -4904,6 +4904,10 @@
 	dl_list_for_each_safe(drv, tmp, &global->interfaces,
 			      struct wpa_driver_nl80211_data, list) {
 		for (bss = drv->first_bss; bss; bss = bss->next) {
+			if (gnlh->cmd == NL80211_CMD_AWGN_DETECT && 
+					(nla_get_u32(tb[NL80211_ATTR_WIPHY_FREQ])) != bss->flink->freq)
+				continue;
+
 			if (wiphy_idx_set)
 				wiphy_idx = nl80211_get_wiphy_index(bss);
 			if ((ifidx == -1 && !wiphy_idx_set && !wdev_id_set) ||
