--- a/src/drivers/driver_nl80211.c
+++ b/src/drivers/driver_nl80211.c
@@ -10577,11 +10577,6 @@ static int wpa_driver_nl80211_get_survey
 					 &arg, NULL, NULL);
 	} while (err > 0);
 
-	if (!bss->scan_link) {
-		wpa_printf(MSG_ERROR, "nl80211: scan_link empty");
-		return -1;
-	}
-
 	if (err)
 		wpa_printf(MSG_ERROR, "nl80211: Failed to process survey data");
 	else if (bss->scan_link && bss->scan_link->ctx)
