--- a/src/ap/vlan_full.c
+++ b/src/ap/vlan_full.c
@@ -772,6 +772,7 @@ full_dynamic_vlan_init(struct hostapd_da
 {
 	struct sockaddr_nl local;
 	struct full_dynamic_vlan *priv;
+	int rxbuf = 262144;
 
 	priv = os_zalloc(sizeof(*priv));
 	if (priv == NULL)
@@ -790,6 +791,7 @@ full_dynamic_vlan_init(struct hostapd_da
 		os_free(priv);
 		return NULL;
 	}
+	setsockopt(priv->s, SOL_SOCKET, SO_RCVBUF, &rxbuf, sizeof(rxbuf));
 
 	os_memset(&local, 0, sizeof(local));
 	local.nl_family = AF_NETLINK;
--- a/src/drivers/netlink.c
+++ b/src/drivers/netlink.c
@@ -92,6 +92,7 @@ struct netlink_data * netlink_init(struc
 {
 	struct netlink_data *netlink;
 	struct sockaddr_nl local;
+	int rxbuf = 262144;
 
 	netlink = os_zalloc(sizeof(*netlink));
 	if (netlink == NULL)
@@ -105,6 +106,7 @@ struct netlink_data * netlink_init(struc
 		return NULL;
 	}
 
+	setsockopt(netlink->sock, SOL_SOCKET, SO_RCVBUF, &rxbuf, sizeof(rxbuf));
 	os_memset(&local, 0, sizeof(local));
 	local.nl_family = AF_NETLINK;
 	local.nl_groups = RTMGRP_LINK;
