--- a/src/ap/drv_callbacks.c
+++ b/src/ap/drv_callbacks.c
@@ -1975,6 +1975,14 @@
 {
	int i;

+	if (hapd->iface->conf->he_mu_edca.he_mu_ac_be_param[2] != 255 ||
+		hapd->iface->conf->he_mu_edca.he_mu_ac_bk_param[2] != 255 ||
+		hapd->iface->conf->he_mu_edca.he_mu_ac_vi_param[2] != 255 ||
+		hapd->iface->conf->he_mu_edca.he_mu_ac_vo_param[2] != 255) {
+		wpa_printf(MSG_DEBUG, "User-configured MU-EDCA parameters detected, ignoring firmware update");
+		return;
+	}
+
	/* Update current MU-EDCA parameters */
	for (i = 0; i < 3; i++) {
		hapd->iface->conf->he_mu_edca.he_mu_ac_be_param[i] =
