--- a/src/ap/beacon.c
+++ b/src/ap/beacon.c
@@ -2517,6 +2517,10 @@ int ieee802_11_build_ap_params(struct ho
 		if (hapd->iconf->num_bss == 1) {
 			params->mbssid.mbssid_tx_iface = hapd->conf->iface;
 			params->mbssid.mbssid_index = hostapd_mbssid_get_bss_index(hapd);
+			if (tx_bss->conf->mld_ap)
+				params->mbssid.mbssid_tx_iface_linkid = tx_bss->mld_link_id;
+			else
+				params->mbssid.mbssid_tx_iface_linkid = -1;
 			complete = true;
 		} else {
 			if (ieee802_11_build_ap_params_mbssid(hapd, params)) {
