Index: backports-20220404-5.4.164-f40abb4788/drivers/net/wireless/ath/ath11k/qmi.h
===================================================================
--- backports-20220404-5.4.164-f40abb4788.orig/drivers/net/wireless/ath/ath11k/qmi.h
+++ backports-20220404-5.4.164-f40abb4788/drivers/net/wireless/ath/ath11k/qmi.h
@@ -8,6 +8,7 @@
 
 #include <linux/mutex.h>
 #include <linux/notifier.h>
+#include <linux/device.h>
 #include <linux/soc/qcom/qmi.h>
 #include <linux/remoteproc/qcom_rproc.h>
 
