Index: backports-20220822-5.4.213-ef7197996efe/drivers/net/wireless/ath/ath12k/core.c
===================================================================
--- backports-20220822-5.4.213-ef7197996efe.orig/drivers/net/wireless/ath/ath12k/core.c
+++ backports-20220822-5.4.213-ef7197996efe/drivers/net/wireless/ath/ath12k/core.c
@@ -48,7 +48,7 @@ unsigned int ath12k_ftm_mode;
 module_param_named(ftm_mode, ath12k_ftm_mode, uint, 0444);
 MODULE_PARM_DESC(ftm_mode, "Boots up in factory test mode");
 
-unsigned int ath12k_mlo_capable = true;
+unsigned int ath12k_mlo_capable = false;
 module_param_named(mlo_capable, ath12k_mlo_capable, uint, 0644);
 MODULE_PARM_DESC(mlo_capable, "mlo capable: 0-disable, 1-enable");
 
@@ -474,7 +474,7 @@ const struct firmware *ath12k_core_firmw
 
 	ath12k_core_create_firmware_path(ab, file, path, sizeof(path));
 
-	ret = request_firmware_direct(&fw, path, ab->dev);
+	ret = request_firmware(&fw, path, ab->dev);
 	if (ret)
 		return ERR_PTR(ret);
 
@@ -734,6 +734,8 @@ int ath12k_core_fetch_bdf(struct ath12k_
 
 	ab->bd_api = 1;
 	ret = ath12k_core_fetch_board_data_api_1(ab, bd, ATH12K_DEFAULT_BOARD_FILE);
+	if (!ret)
+		goto success;
 	if (ret) {
 		ath12k_err(ab, "failed to fetch board-2.bin from %s\n",
 			   ab->hw_params->fw.dir);
