diff --git a/nat46/modules/nat46-core.c b/nat46/modules/nat46-core.c
index 8279dae..3411850 100644
--- a/nat46/modules/nat46-core.c
+++ b/nat46/modules/nat46-core.c
@@ -922,10 +922,15 @@ __sum16 csum_ipv6_udp_remagic(struct ipv6hdr *ip6hdr, u32 csum) {
 __sum16 csum_ipv4_unmagic(__be32 saddr, __be32 daddr,
                         u32 csum) {
   u32 s;
+  u16 addr_csum_fold;
   uint32_t addr_csum;
   csum = ntohs(~csum);
   addr_csum =  (saddr & 0xffff)  + (saddr >> 16) + (daddr & 0xffff) + (daddr >> 16);
-  s= csum + ntohs(~addr_csum);
+  addr_csum_fold = ((addr_csum >> 16) & 0xffff) + (addr_csum & 0xffff);
+  /* By subtracting, undo the saddr and daddr which was previously included with
+   * pseudo header calculation
+   */
+  s= csum + ntohs(~addr_csum_fold);
   s = ((s >> 16) & 0xffff) + (s & 0xffff);
   s += ((s >> 16) & 0xffff);
   return htons((u16)(~s));
