From 2353829cb2674327ef6f23661acf5d161c82523f Mon Sep 17 00:00:00 2001
From: Feng Pei <quic_fpei@quicinc.com>
Date: Sat, 13 Apr 2024 01:00:42 +0800
Subject: [PATCH] nat46: fix kernel warning impacting performance

Signed-off-by: Feng Pei <quic_fpei@quicinc.com>
---
 nat46/modules/nat46-core.c | 4 ----
 1 file changed, 4 deletions(-)

diff --git a/nat46/modules/nat46-core.c b/nat46/modules/nat46-core.c
index 352385f..0bf227b 100644
--- a/nat46/modules/nat46-core.c
+++ b/nat46/modules/nat46-core.c
@@ -2045,10 +2045,6 @@ int nat46_ipv6_input(struct sk_buff *old_skb) {
   fill_v4hdr_from_v6hdr(iph, ver_class_flow, hop_limit, v4saddr, v4daddr, frag_id, frag_off, proto, l3_infrag_payload_len);
   old_skb->protocol = htons(ETH_P_IP);
 
-  if (ntohs(iph->tot_len) >= 2000) {
-    nat46debug(0, "Too big IP len: %d", ntohs(iph->tot_len));
-  }
-
   nat46debug(5, "about to send v4 packet, flags: %02x",  IPCB(old_skb)->flags);
   nat46_netdev_count_xmit(old_skb, old_skb->dev);
 
-- 
2.42.0

