Index: bridger-2023-05-12-3159bbe0/bpf.c
===================================================================
--- bridger-2023-05-12-3159bbe0.orig/bpf.c
+++ bridger-2023-05-12-3159bbe0/bpf.c
@@ -42,6 +42,10 @@ void bridger_bpf_flow_upload(struct brid
 	if (bpf_map_lookup_elem(map_offload, &flow->key, &val) == 0)
 		flow->offload.packets = val.packets;
 
+	if ((flow->key.ifindex == flow->offload.target_port) &&
+	    (flow->key.vlan == flow->offload.vlan)) {
+		return;
+	}
 	bpf_map_update_elem(map_offload, &flow->key, &flow->offload, BPF_ANY);
 }
 
