diff -Nur bridger-2024.04.22~40b1c5b6-orig/flow.c bridger-2024.04.22~40b1c5b6/flow.c
--- bridger-2024.04.22~40b1c5b6-orig/flow.c	2026-09-03 14:54:50.729749756 +0800
+++ bridger-2024.04.22~40b1c5b6/flow.c	2026-09-03 14:57:55.319087099 +0800
@@ -143,7 +143,7 @@
 	flow->offload.redirect_flags = fdb_out->dev->br ? BPF_F_INGRESS : 0;
 
 	bridger_bpf_flow_upload(flow);
-	if (!fdb_in->dev->br && !fdb_out->dev->br)
+	if (!fdb_in->dev->br && !fdb_out->dev->br && !bridger_ubus_dev_blacklisted(fdb_out->dev))
 		bridger_nl_flow_offload_add(flow);
 
 	avl_insert(&sorted_flows, &flow->sort_node);
