--- a/drivers/net/wireless/ath/ath11k/ahb.c
+++ b/drivers/net/wireless/ath/ath11k/ahb.c
@@ -530,6 +530,7 @@ static int ath11k_ahb_ext_irq_config(str
 	int irq;
 	int ret;
 	bool nss_offload;
+	static int devidx = 0;
 
 	/* TCL Completion, REO Dest, ERR, Exception and h2rxdma rings are offloaded
 	 * to nss when its enabled, hence don't enable these interrupts
@@ -543,6 +544,9 @@ static int ath11k_ahb_ext_irq_config(str
 		irq_grp->ab = ab;
 		irq_grp->grp_id = i;
 		init_dummy_netdev(&irq_grp->napi_ndev);
+		snprintf(irq_grp->napi_ndev.name, sizeof(irq_grp->napi_ndev.name), "%s%d:%d",
+			 "ath11k_ahb", devidx, i);
+		irq_grp->napi_ndev.threaded = 1;
 		netif_napi_add(&irq_grp->napi_ndev, &irq_grp->napi,
 			       ath11k_ahb_ext_grp_napi_poll, NAPI_POLL_WEIGHT);
 
@@ -608,6 +612,8 @@ static int ath11k_ahb_ext_irq_config(str
 		}
 	}
 
+	devidx++;
+
 	return 0;
 }
 
--- a/drivers/net/wireless/ath/ath11k/pci.c
+++ b/drivers/net/wireless/ath/ath11k/pci.c
@@ -703,10 +703,9 @@ static int ath11k_pci_ext_grp_napi_poll(
 	int work_done;
 
 	work_done = ath11k_dp_service_srng(ab, irq_grp, budget);
-	if (work_done < budget) {
-		napi_complete_done(napi, work_done);
+	if (work_done < budget &&
+	    napi_complete_done(napi, work_done))
 		ath11k_pci_ext_grp_enable(irq_grp);
-	}
 
 	if (work_done > budget)
 		work_done = budget;
@@ -753,6 +752,7 @@ static int ath11k_pci_ext_irq_config(str
 	int i, j, ret, num_vectors = 0;
 	u32 user_base_data = 0, base_vector = 0, base_idx;
 	struct ath11k_pci *ab_pci = ath11k_pci_priv(ab);
+	static int devidx = 0;
 	u8 domain_id;
 
 	domain_id = ath11k_pci_get_domain_id(ab);
@@ -770,6 +770,9 @@ static int ath11k_pci_ext_irq_config(str
 		irq_grp->ab = ab;
 		irq_grp->grp_id = i;
 		init_dummy_netdev(&irq_grp->napi_ndev);
+		snprintf(irq_grp->napi_ndev.name, sizeof(irq_grp->napi_ndev.name), "%s%d:%d",
+			 "ath11k_pci", devidx, i);
+		irq_grp->napi_ndev.threaded = 1;
 		netif_napi_add(&irq_grp->napi_ndev, &irq_grp->napi,
 			       ath11k_pci_ext_grp_napi_poll, NAPI_POLL_WEIGHT);
 
@@ -816,6 +819,8 @@ static int ath11k_pci_ext_irq_config(str
 		}
 	}
 
+	devidx++;
+
 	return 0;
 }
 
