Index: backports-20210222_001-4.4.60-b157d2276/drivers/net/wireless/ath/ath11k/mac.c
===================================================================
--- backports-20210222_001-4.4.60-b157d2276.orig/drivers/net/wireless/ath/ath11k/mac.c
+++ backports-20210222_001-4.4.60-b157d2276/drivers/net/wireless/ath/ath11k/mac.c
@@ -9160,7 +9160,7 @@ static int ath11k_mac_setup_iface_combin
 	limits[0].max = 1;
 	limits[0].types |= BIT(NL80211_IFTYPE_STATION);
 
-	limits[1].max = 16;
+	limits[1].max = 8;
 	limits[1].types |= BIT(NL80211_IFTYPE_AP);
 
 	if (IS_ENABLED(CPTCFG_MAC80211_MESH) &&
@@ -9169,7 +9169,7 @@ static int ath11k_mac_setup_iface_combin
 
 	combinations[0].limits = limits;
 	combinations[0].n_limits = n_limits;
-	combinations[0].max_interfaces = 16;
+	combinations[0].max_interfaces = 8;
 	combinations[0].num_different_channels = 1;
 	combinations[0].beacon_int_infra_match = true;
 	combinations[0].beacon_int_min_gcd = 100;
