--- a/net/mac80211/tx.c
+++ b/net/mac80211/tx.c
@@ -3978,7 +3978,8 @@ void __ieee80211_subif_start_xmit(struct
 			ap_sdata = sdata;
 
 		if (ap_sdata->vif.offload_flags & IEEE80211_OFFLOAD_ENCAP_ENABLED &&
-		    !is_multicast_ether_addr(skb->data)) {
+		    !is_multicast_ether_addr(skb->data) &&
+		    (sta || ieee80211_hw_check(&local->hw, SUPPORTS_NSS_OFFLOAD))) {
 		    	if (sta)
 				key = rcu_dereference(sta->ptk[sta->ptk_idx]);
 			ieee80211_8023_xmit(sdata, dev, sta, key, skb);
