--- a/drivers/net/wireless/ath/ath11k/mac.c
+++ b/drivers/net/wireless/ath/ath11k/mac.c
@@ -6009,7 +6009,7 @@ static int __ath11k_set_antenna(struct a
 
 	if (ar->state != ATH11K_STATE_ON &&
 	    ar->state != ATH11K_STATE_RESTARTED)
-		return 0;
+		goto out;
 
 	ret = ath11k_wmi_pdev_set_param(ar, WMI_PDEV_PARAM_TX_CHAIN_MASK,
 					tx_ant, ar->pdev->pdev_id);
@@ -6019,8 +6019,6 @@ static int __ath11k_set_antenna(struct a
 		return ret;
 	}
 
-	ar->num_tx_chains = get_num_chains(tx_ant);
-
 	ret = ath11k_wmi_pdev_set_param(ar, WMI_PDEV_PARAM_RX_CHAIN_MASK,
 					rx_ant, ar->pdev->pdev_id);
 	if (ret) {
@@ -6029,6 +6027,8 @@ static int __ath11k_set_antenna(struct a
 		return ret;
 	}
 
+out:
+	ar->num_tx_chains = get_num_chains(tx_ant);
 	ar->num_rx_chains = get_num_chains(rx_ant);
 
 	/* Reload HT/VHT/HE capability */
