From 23839df49e7b57d22a329bcd6b7cbf4b0636ec8b Mon Sep 17 00:00:00 2001
From: "himanshu.goyal" <himanshu.goyal@mediatek.com>
Date: Fri, 17 Mar 2023 17:36:01 +0800
Subject: [PATCH 024/115] mtk: nl80211: Mark DFS channel as available for CSA.

---
 net/wireless/nl80211.c | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/net/wireless/nl80211.c b/net/wireless/nl80211.c
index bf2e036c..b198ebda 100644
--- a/net/wireless/nl80211.c
+++ b/net/wireless/nl80211.c
@@ -11092,6 +11092,11 @@ skip_beacons:
 	if (err)
 		goto free;
 
+	/* Use RADAR_BACKGROUND attribute here for skipping CAC */
+	if (info->attrs[NL80211_ATTR_RADAR_BACKGROUND]) {
+		cfg80211_set_dfs_state(&rdev->wiphy, &params.chandef, NL80211_DFS_AVAILABLE);
+	}
+
 	if (!cfg80211_reg_can_beacon_relax(&rdev->wiphy, &params.chandef,
 					   wdev->iftype)) {
 		err = -EINVAL;
-- 
2.45.2

