From 17d6bf54e8a499d5c652f8dcea03ffa2f5c0a6e4 Mon Sep 17 00:00:00 2001
From: ye he <ye.he@mediatek.com>
Date: Wed, 22 Feb 2023 16:09:32 +0800
Subject: [PATCH 028/115] mtk: mac80211: add send bar action when recieve addba
 rsp

Signed-off-by: ye he <ye.he@mediatek.com>
---
 net/mac80211/agg-tx.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/net/mac80211/agg-tx.c b/net/mac80211/agg-tx.c
index dfc6e619..060d3d81 100644
--- a/net/mac80211/agg-tx.c
+++ b/net/mac80211/agg-tx.c
@@ -1047,7 +1047,8 @@ void ieee80211_process_addba_resp(struct ieee80211_local *local,
 
 		tid_tx->buf_size = buf_size;
 		tid_tx->amsdu = amsdu;
-
+		ieee80211_send_bar(&sta->sdata->vif, sta->sta.addr,
+					   tid, 0);
 		if (test_bit(HT_AGG_STATE_DRV_READY, &tid_tx->state))
 			ieee80211_agg_tx_operational(local, sta, tid);
 
-- 
2.45.2

