From f28f363b8432c2a2d667f27051cb0839083c42fb Mon Sep 17 00:00:00 2001
From: Shayne Chen <shayne.chen@mediatek.com>
Date: Wed, 9 Oct 2024 00:20:20 +0800
Subject: [PATCH 082/115] tmp disable ndo_fill_receive_path ops for 6.6

---
 net/mac80211/iface.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/net/mac80211/iface.c b/net/mac80211/iface.c
index 7cd947ec..400baf31 100644
--- a/net/mac80211/iface.c
+++ b/net/mac80211/iface.c
@@ -1003,6 +1003,7 @@ out:
 	return ret;
 }
 
+#if LINUX_VERSION_IS_LESS(6,6,0)
 static int ieee80211_netdev_fill_receive_path(struct net_device_path_ctx *ctx,
 					      struct net_device_path *path)
 {
@@ -1024,6 +1025,7 @@ static int ieee80211_netdev_fill_receive_path(struct net_device_path_ctx *ctx,
 
 	return ret;
 }
+#endif
 
 static const struct net_device_ops ieee80211_dataif_8023_ops = {
 	.ndo_open		= ieee80211_open,
@@ -1033,7 +1035,9 @@ static const struct net_device_ops ieee80211_dataif_8023_ops = {
 	.ndo_set_rx_mode	= ieee80211_set_multicast_list,
 	.ndo_set_mac_address	= ieee80211_change_mac,
 	.ndo_fill_forward_path	= ieee80211_netdev_fill_forward_path,
+#if LINUX_VERSION_IS_LESS(6,6,0)
 	.ndo_fill_receive_path	= ieee80211_netdev_fill_receive_path,
+#endif
 	.ndo_setup_tc		= ieee80211_netdev_setup_tc,
 #if LINUX_VERSION_IS_LESS(6,9,0)
 	.ndo_get_stats64	= ieee80211_get_stats64,
-- 
2.45.2

