--- a/net/mac80211/iface.c	2025-12-09 08:04:20.370546940 +0000
+++ b/net/mac80211/iface.c	2025-12-09 08:02:40.203854591 +0000
@@ -592,7 +592,8 @@ static void ieee80211_do_stop(struct iee
 		list_del(&sdata->u.vlan.list);
 		RCU_INIT_POINTER(sdata->vif.bss_conf.chanctx_conf, NULL);
 		/* see comment in the default case below */
-		ieee80211_free_keys(sdata, true);
+		/* skip free keys for crash issue of ECCLOUD-7528 */
+		//ieee80211_free_keys(sdata, true);
 		/* no need to tell driver */
 		break;
 	case NL80211_IFTYPE_MONITOR:
