Index: linux-5.4.164-qsdk-26349818b464f8c7b52d59ce73579d9f3dd6bd5d/arch/arm64/boot/dts/qcom/ipq5018.dtsi
===================================================================
--- linux-5.4.164-qsdk-26349818b464f8c7b52d59ce73579d9f3dd6bd5d.orig/arch/arm64/boot/dts/qcom/ipq5018.dtsi
+++ linux-5.4.164-qsdk-26349818b464f8c7b52d59ce73579d9f3dd6bd5d/arch/arm64/boot/dts/qcom/ipq5018.dtsi
@@ -1265,7 +1265,7 @@
 
 	qti,scm_restart_reason {
 		compatible = "qti,scm_restart_reason";
-		dload_status = <0>;
+		dload_status = <1>;
 		dload_warm_reset = <0>;
 	};
 
