Index: a/arch/arm64/boot/dts/qcom/ipq5018-memory.dtsi
===================================================================
--- a/arch/arm64/boot/dts/qcom/ipq5018-memory.dtsi
+++ b/arch/arm64/boot/dts/qcom/ipq5018-memory.dtsi
@@ -199,6 +199,14 @@
 			no-map;
 			reg = <0x0 0x07000000 0x0 0x58000>;
 		};
+
+		ramoops@4fd00000 {
+			compatible = "ramoops";
+			reg = <0x0 0x4fd00000 0x0 0x100000>;
+			record-size = <0x1000>;
+			console-size = <0x1000>;
+			pmsg-size = <0x1000>;
+		};
 	};
 	#endif
 };
