diff --git a/src/wm2/tests/unit.mk b/src/wm2/tests/unit.mk
index e98529d..6b67334 100644
--- a/src/wm2/tests/unit.mk
+++ b/src/wm2/tests/unit.mk
@@ -33,3 +33,4 @@ UNIT_DEPS := src/lib/common
 UNIT_DEPS += src/lib/osa
 UNIT_DEPS += src/lib/schema
 UNIT_DEPS += src/lib/ovsdb
+UNIT_LDFLAGS += -luci
diff --git a/src/wm2/unit.mk b/src/wm2/unit.mk
index 9b27ad9..302b64f 100644
--- a/src/wm2/unit.mk
+++ b/src/wm2/unit.mk
@@ -44,6 +44,7 @@ UNIT_LDFLAGS += -ljansson
 UNIT_LDFLAGS += -ldl
 UNIT_LDFLAGS += -lev
 UNIT_LDFLAGS += -lrt
+UNIT_LDFLAGS += -luci
 
 UNIT_EXPORT_CFLAGS := $(UNIT_CFLAGS)
 UNIT_EXPORT_LDFLAGS := $(UNIT_LDFLAGS)
