Index: opensync-2.0.5.0/src/wm2/tests/unit.mk
===================================================================
--- opensync-2.0.5.0.orig/src/wm2/tests/unit.mk
+++ opensync-2.0.5.0/src/wm2/tests/unit.mk
@@ -33,3 +33,5 @@ UNIT_DEPS := src/lib/common
 UNIT_DEPS += src/lib/osa
 UNIT_DEPS += src/lib/schema
 UNIT_DEPS += src/lib/ovsdb
+UNIT_LDFLAGS += -luci
+UNIT_LDFLAGS += -lnl-tiny -lubox -lubus -lblobmsg_json
Index: opensync-2.0.5.0/src/wm2/unit.mk
===================================================================
--- opensync-2.0.5.0.orig/src/wm2/unit.mk
+++ opensync-2.0.5.0/src/wm2/unit.mk
@@ -46,6 +46,8 @@ UNIT_LDFLAGS += -ljansson
 UNIT_LDFLAGS += -ldl
 UNIT_LDFLAGS += -lev
 UNIT_LDFLAGS += -lrt
+UNIT_LDFLAGS += -luci
+UNIT_LDFLAGS += -lnl-tiny -lubox -lubus -lblobmsg_json
 
 UNIT_EXPORT_CFLAGS := $(UNIT_CFLAGS)
 UNIT_EXPORT_LDFLAGS := $(UNIT_LDFLAGS)
Index: opensync-2.0.5.0/src/sm/unit.mk
===================================================================
--- opensync-2.0.5.0.orig/src/sm/unit.mk
+++ opensync-2.0.5.0/src/sm/unit.mk
@@ -55,6 +55,7 @@ UNIT_LDFLAGS += -ldl
 UNIT_LDFLAGS += -lev
 UNIT_LDFLAGS += -lrt
 UNIT_LDFLAGS += -lz
+UNIT_LDFLAGS += -lnl-tiny -lubox
 
 UNIT_DEPS    := src/lib/ovsdb
 UNIT_DEPS    += src/lib/pjs
Index: opensync-2.0.5.0/src/um/unit.mk
===================================================================
--- opensync-2.0.5.0.orig/src/um/unit.mk
+++ opensync-2.0.5.0/src/um/unit.mk
@@ -40,6 +40,7 @@ UNIT_SRC += src/um_ovsdb.c
 UNIT_CFLAGS  := -I$(UNIT_PATH)/inc
 UNIT_LDFLAGS := -lpthread
 UNIT_LDFLAGS += -lev
+UNIT_LDFLAGS += -lcurl
 
 UNIT_EXPORT_CFLAGS := $(UNIT_CFLAGS)
 UNIT_EXPORT_LDFLAGS := $(UNIT_LDFLAGS)
