Index: opensync-2.0.5.0/src/nm2/unit.mk
===================================================================
--- opensync-2.0.5.0.orig/src/nm2/unit.mk
+++ opensync-2.0.5.0/src/nm2/unit.mk
@@ -27,7 +27,8 @@
 # Network manager
 #
 ###############################################################################
-UNIT_DISABLE := $(if $(CONFIG_MANAGER_NM),n,y)
+UNIT_DISABLE := y
+#$(if $(CONFIG_MANAGER_NM),n,y)
 
 UNIT_NAME := nm
 
