Index: opensync-1.4.0.1/build/flags.mk
===================================================================
--- opensync-1.4.0.1.orig/build/flags.mk
+++ opensync-1.4.0.1/build/flags.mk
@@ -23,12 +23,13 @@
 # SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 
 # C FLAGS
-CFLAGS   := -Wall -Wextra -Werror
+CFLAGS   ?=
+CFLAGS   += -Wall -Wextra -Werror
 CFLAGS   += -Wno-unused-parameter
 CFLAGS   += -Wno-unused-label
 CFLAGS   += -Wno-strict-aliasing
 CFLAGS   += -fasynchronous-unwind-tables
-
+CFLAGS   += $(OWRT_CFLAGS)
 DEBUGFLAGS := -g
 OPTIMIZE :=
 
