diff --git a/src/wm2/tests/unit.mk b/src/wm2/tests/unit.mk
index 6b67334..66e9679 100644
--- a/src/wm2/tests/unit.mk
+++ b/src/wm2/tests/unit.mk
@@ -34,3 +34,4 @@ UNIT_DEPS += src/lib/osa
 UNIT_DEPS += src/lib/schema
 UNIT_DEPS += src/lib/ovsdb
 UNIT_LDFLAGS += -luci
+UNIT_LDFLAGS += -liwinfo
diff --git a/src/wm2/unit.mk b/src/wm2/unit.mk
index 302b64f..319c663 100644
--- a/src/wm2/unit.mk
+++ b/src/wm2/unit.mk
@@ -45,6 +45,7 @@ UNIT_LDFLAGS += -ldl
 UNIT_LDFLAGS += -lev
 UNIT_LDFLAGS += -lrt
 UNIT_LDFLAGS += -luci
+UNIT_LDFLAGS += -liwinfo

 UNIT_EXPORT_CFLAGS := $(UNIT_CFLAGS)
 UNIT_EXPORT_LDFLAGS := $(UNIT_LDFLAGS)

