Index: opensync-1.4.0.1/src/wm2/src/wm2_clients.c
===================================================================
--- opensync-1.4.0.1.orig/src/wm2/src/wm2_clients.c
+++ opensync-1.4.0.1/src/wm2/src/wm2_clients.c
@@ -69,7 +69,7 @@ wm2_clients_oftag_from_key_id(const char
 {
     struct schema_Wifi_VIF_Config vconf;
     ovsdb_table_t table_Wifi_VIF_Config;
-    char oftagkey[32];
+    char oftagkey[38];
     char *ptr;
     bool ok;
 
