Index: opensync-2.0.4.0/src/lib/osn/unit.mk
===================================================================
--- opensync-2.0.4.0.orig/src/lib/osn/unit.mk
+++ opensync-2.0.4.0/src/lib/osn/unit.mk
@@ -55,7 +55,7 @@ UNIT_SRC += $(if $(CONFIG_OSN_BACKEND_DH
 UNIT_SRC += $(if $(CONFIG_OSN_BACKEND_DHCPV4_SERVER_DNSMASQ),src/osn_dhcp_server_dnsmasq.c)
 
 UNIT_SRC += $(if $(CONFIG_OSN_BACKEND_UPNP_NULL),src/osn_upnp_null.c)
-UNIT_SRC += $(if $(CONFIG_OSN_BACKEND_UPNP_MINIUPNPD),src/osn_upnp_mupnp.c)
+#UNIT_SRC += $(if $(CONFIG_OSN_BACKEND_UPNP_MINIUPNPD),src/osn_upnp_mupnp.c)
 
 UNIT_SRC += $(if $(CONFIG_OSN_BACKEND_IPV6_NULL),src/osn_ip6_null.c)
 UNIT_SRC += $(if $(CONFIG_OSN_BACKEND_IPV6_LINUX),src/osn_ip6_linux.c)
@@ -79,7 +79,7 @@ UNIT_SRC += $(if $(CONFIG_OSN_LINUX_IPV6
 UNIT_SRC += $(if $(CONFIG_OSN_LINUX_NETIF),src/linux/lnx_netif.c)
 UNIT_SRC += $(if $(CONFIG_OSN_LINUX_NETLINK),src/linux/lnx_netlink.c)
 UNIT_SRC += $(if $(CONFIG_OSN_LINUX_ROUTE),src/linux/lnx_route.c)
-UNIT_SRC += $(if $(CONFIG_OSN_MINIUPNPD),src/linux/mupnp_server.c)
+#UNIT_SRC += $(if $(CONFIG_OSN_MINIUPNPD),src/linux/mupnp_server.c)
 UNIT_SRC += $(if $(CONFIG_OSN_ODHCP6),src/linux/odhcp6_client.c)
 UNIT_SRC += $(if $(CONFIG_OSN_UDHCPC),src/linux/udhcp_client.c)
 
