Index: opensync-2.0.5.0/interfaces/opensync.ovsschema
===================================================================
--- opensync-2.0.5.0.orig/interfaces/opensync.ovsschema
+++ opensync-2.0.5.0/interfaces/opensync.ovsschema
@@ -8525,6 +8525,92 @@
             }
         },
         "isRoot": true
-    }
+    },
+    "Command_Config": {
+        "columns": {
+            "command": {
+               "type": {
+                  "key": {
+                       "type": "string"
+		    }
+                }
+            },
+            "timestamp": {
+               "type": {
+                  "key": {
+                     "type": "integer"
+                  }
+               }
+            },
+            "delay": {
+               "type": {
+                  "key": {
+                     "type": "integer"
+                  }
+               }
+            },
+            "duration": {
+               "type": {
+                  "key": {
+                     "type": "integer"
+                  }
+               }
+            },
+            "payload": {
+                "type": {
+                    "key": "string",
+                    "value": "string",
+                    "min": 0,
+                    "max": "unlimited"
+                }
+            }
+        },
+        "isRoot": true
+    },
+    "Command_State": {
+        "columns": {
+             "cmd_uuid": {
+                "type": {
+                   "key": {
+                     "type": "uuid",
+                     "refTable": "Command_Config"
+                   }
+                }
+             },
+            "state": {
+               "type": {
+                  "key": {
+                    "type": "string",
+                    "enum": [
+                      "set",
+                      [
+                       "waiting",
+                       "pending",
+                       "running",
+                       "complete",
+                       "failed"
+                      ]
+                    ]
+                  }
+                }
+             },
+            "timestamp": {
+               "type": {
+                  "key": {
+                     "type": "integer"
+                  }
+               }
+            },
+            "result": {
+                "type": {
+                    "key": "string",
+                    "value": "string",
+                    "min": 0,
+                    "max": "unlimited"
+                }
+            }
+      },
+        "isRoot": true
+      }
   }
 }
