--- a/src/lib/version/version-gen
+++ b/src/lib/version/version-gen
@@ -101,6 +101,7 @@ function matrix()
     if [ -n "${IMAGE_DEPLOYMENT_PROFILE}" -a "${IMAGE_DEPLOYMENT_PROFILE}" != "none" ]; then
         echo "FW_PROFILE:${IMAGE_DEPLOYMENT_PROFILE}"
     fi
+    echo "FW_IMAGE_ACTIVE:$FW_PKG_NAME"
     echo "DATE:${VER_DATE}"
     echo "HOST:${USERNAME}@${HOSTNAME}"
     for LAYER in .. $LAYER_LIST $SDK_DIR $SDK_BASE; do
--- a/src/lib/version/version-lib.sh
+++ b/src/lib/version/version-lib.sh
@@ -104,6 +104,9 @@ fi
 
 OSYNC_VERSION=`cat .version`
 
+PKG_FILE="$VENDOR_DIR/.pkgname"
+FW_PKG_NAME=`cat $PKG_FILE`
+
 DIRTY_STRING=""
 if [ ${DIRTY} -ne 0 ]; then
     DIRTY_STRING="-mods"
--- a/interfaces/opensync.ovsschema
+++ b/interfaces/opensync.ovsschema
@@ -178,7 +178,10 @@
         "version_matrix": {
           "type": {
             "key": "string",
-            "value": "string",
+            "value": {
+               "type": "string",
+               "maxLength": 128
+            },
             "min": 0,
             "max": "unlimited"
           }
