--- a/src/cm2/unit.mk
+++ b/src/cm2/unit.mk
@@ -52,6 +52,7 @@ UNIT_LDFLAGS += -ljansson
 UNIT_LDFLAGS += -ldl
 UNIT_LDFLAGS += -lev
 UNIT_LDFLAGS += -lrt
+UNIT_LDFLAGS += -lnl-tiny
 
 
 ifneq ($(CONFIG_MANAGER_CM),y)
@@ -73,6 +74,7 @@ UNIT_EXPORT_LDFLAGS := $(UNIT_LDFLAGS)
 UNIT_DEPS := src/lib/ovsdb
 UNIT_DEPS += src/lib/pjs
 UNIT_DEPS += src/lib/schema
+UNIT_DEPS += src/lib/target
 ifeq ($(BUILD_HAVE_LIBCARES),y)
 UNIT_DEPS += src/lib/evx
 endif
