Index: opensync-2.0.5.0/interfaces/opensync.ovsschema
===================================================================
--- opensync-2.0.5.0.orig/interfaces/opensync.ovsschema
+++ opensync-2.0.5.0/interfaces/opensync.ovsschema
@@ -8634,6 +8634,13 @@
                    }
                 }
              },
+            "command": {
+               "type": {
+                  "key": {
+                       "type": "string"
+		    }
+                }
+            },
             "state": {
                "type": {
                   "key": {
