Index: opensync-2.0.5.0/src/lib/schema/inc/schema_consts.h
===================================================================
--- opensync-2.0.5.0.orig/src/lib/schema/inc/schema_consts.h
+++ opensync-2.0.5.0/src/lib/schema/inc/schema_consts.h
@@ -166,4 +166,16 @@ typedef enum {
 #define SCHEMA_CONSTS_LOGIN_SUCCESS_TEXT           "login_success_text"
 #define SCHEMA_CONSTS_PAGE_TITLE                   "page_title"
 #define SCHEMA_CONSTS_USERPASS_FILE                "username_password_file"
+
+/* AX options */
+#define SCHEMA_CONSTS_HE_SU_BEAMFORMER		"he_su_beamformer"
+#define SCHEMA_CONSTS_HE_MU_BEAMFORMER		"he_mu_beamformer"
+#define SCHEMA_CONSTS_HE_RTS_THRESHOLD		"he_rts_threshold"
+#define SCHEMA_CONSTS_HE_BSS_COLOR		"he_bss_color"
+#define SCHEMA_CONSTS_HE_TWT_REQUIRED		"he_twt_required"
+#define SCHEMA_CONSTS_HE_MULTIPLE_BSSID		"multiple_bssid"
+#define SCHEMA_CONSTS_HE_CO_LOCATE		"co_locate"
+#define SCHEMA_CONSTS_HE_RNR_BEACON		"rnr_beacon"
+#define SCHEMA_CONSTS_HE_EMA_BEACON		"ema_beacon"
+
 #endif /* SCHEMA_CONSTS_H_INCLUDED */
