Index: opensync-2.0.5.0/interfaces/opensync.ovsschema
===================================================================
--- opensync-2.0.5.0.orig/interfaces/opensync.ovsschema
+++ opensync-2.0.5.0/interfaces/opensync.ovsschema
@@ -1732,7 +1732,10 @@
         "channels": {
           "type": {
             "key": "string",
-            "value": "string",
+            "value": {
+               "type": "string",
+               "maxLength": 256
+            },
             "min": 0,
             "max": "unlimited"
           }
