Index: opensync-2.0.5.0/interfaces/opensync_stats.proto
===================================================================
--- opensync-2.0.5.0.orig/interfaces/opensync_stats.proto
+++ opensync-2.0.5.0/interfaces/opensync_stats.proto
@@ -805,8 +805,80 @@ message EventReport {
         optional ClientConnectEvent     client_connect_event                = 10;
     }
 
+    /* DHCP Common Data */
+    message DhcpCommonData {
+        required uint32                 x_id                                = 1;
+        required uint32                 vlan_id                             = 2;
+        required bytes                  dhcp_server_ip                      = 3;
+        required bytes                  client_ip                           = 4;
+        required bytes                  relay_ip                            = 5;
+        required string                 device_mac_address                  = 6;
+        required uint64                 timestamp_ms                        = 7;
+    }
+
+    /* DHCP Ack Event */
+    message DhcpAckEvent {
+        required DhcpCommonData         dhcp_common_data                    = 1;
+        optional bytes                  subnet_mask                         = 2;
+        optional bytes                  primary_dns                         = 3;
+        optional bytes                  secondary_dns                       = 4;
+        optional uint32                 lease_time                          = 5;
+        optional uint32                 renewal_time                        = 6;
+        optional uint32                 rebinding_time                      = 7;
+        optional uint32                 time_offset                         = 8;
+        optional bytes                  gateway_ip                          = 9;
+    }
+
+    /* DHCP Nak Event */
+    message DhcpNakEvent {
+        required DhcpCommonData         dhcp_common_data                    = 1;
+        optional bool                   from_internal                       = 2;
+    }
+
+    /* DHCP Offer Event */
+    message DhcpOfferEvent {
+        required DhcpCommonData         dhcp_common_data                    = 1;
+        optional bool                   from_internal                       = 2;
+    }
+
+    /* DHCP Inform Event */
+    message DhcpInformEvent {
+        required DhcpCommonData         dhcp_common_data                    = 1;
+    }
+
+    /* DHCP Decline Event */
+    message DhcpDeclineEvent {
+        required DhcpCommonData         dhcp_common_data                    = 1;
+    }
+
+    /* DHCP Request Event */
+    message DhcpRequestEvent {
+        required DhcpCommonData         dhcp_common_data                    = 1;
+        optional string                 hostname                            = 2;
+    }
+
+    /* DHCP Discover Event */
+    message DhcpDiscoverEvent {
+        required DhcpCommonData         dhcp_common_data                    = 1;
+        optional string                 hostname                            = 2;
+    }
+
+    /* DHCP Transaction */
+    message DhcpTransaction {
+        required uint32                 x_id                                = 1;
+        optional DhcpAckEvent           dhcp_ack_event                      = 2;
+        optional DhcpNakEvent           dhcp_nak_event                      = 3;
+        optional DhcpOfferEvent         dhcp_offer_event                    = 4;
+        optional DhcpInformEvent        dhcp_inform_event                   = 5;
+        optional DhcpDeclineEvent       dhcp_decline_event                  = 6;
+        optional DhcpRequestEvent       dhcp_request_event                  = 7;
+        optional DhcpDiscoverEvent      dhcp_discover_event                 = 8;
+    }
+
     /* Multiple Client Sessions */
     repeated ClientSession              client_session                      = 1;
+    /* Multiple DHCP Transactions */
+    repeated DhcpTransaction            dhcp_transaction                    = 3;
 }
 
 ////////////////////////////////////////////////////////////////////////////////
