Index: opensync-2.0.5.0/src/lib/datapipeline/src/dppline.c
===================================================================
--- opensync-2.0.5.0.orig/src/lib/datapipeline/src/dppline.c
+++ opensync-2.0.5.0/src/lib/datapipeline/src/dppline.c
@@ -276,10 +276,22 @@ static void dppline_free_stat(dppline_st
                 break;
             case DPP_T_UCC:
                 free(&s->u.ucc.record);
-                break;	case DPP_T_EVENTS:
-			free(s->u.events.list);
-			break;
                 break;
+			case DPP_T_EVENTS:
+				for (i = 0; i < s->u.events.qty; i++) {
+					free(s->u.events.list[i].session_list);
+					free(s->u.events.list[i].assoc_list);
+					free(s->u.events.list[i].auth_list);
+					free(s->u.events.list[i].disconnect_list);
+					free(s->u.events.list[i].connect_list);
+					free(s->u.events.list[i].failure_list);
+					free(s->u.events.list[i].first_data_list);
+					free(s->u.events.list[i].id_list);
+					free(s->u.events.list[i].ip_list);
+					free(s->u.events.list[i].timeout_list);
+				}
+				free(s->u.events.list);
+				break;
             default:;
         }

@@ -2974,6 +2986,10 @@ static void dppline_add_stat_rssi(Sts__R
 	}
 }

+static void ip_string_to_ip_bytes(const char *ip_string, uint8_t *ip_bytes, size_t *size) {
+	*size = sscanf(ip_string, "%hhd.%hhd.%hhd.%hhd", &ip_bytes[0], &ip_bytes[1], &ip_bytes[2], &ip_bytes[3]);
+}
+
 static void dppline_add_stat_events(Sts__Report *r, dppline_stats_t *s)
 {
 	Sts__EventReport *sr = NULL;
@@ -3355,16 +3371,15 @@ static void dppline_add_stat_events(Sts_
 				}

 				if (srx->ip_addr) {
-					memcpy(drx->ip_addr.data, srx->ip_addr,
-					       16);
-					drx->ip_addr.len = 16;
+					uint8_t ip_tmp[4] = {0};
+					size_t ip_tmp_size = 0;
+					ip_string_to_ip_bytes(srx->ip_addr, &ip_tmp[0], &ip_tmp_size);
+					drx->ip_addr.data = ip_tmp;
+					drx->ip_addr.len = ip_tmp_size;
 					drx->has_ip_addr = true;
 				}

-				if (srx->clt_id) {
-					drx->clt_id = strcpy(drx->clt_id,
-							     srx->clt_id);
-				}
+				drx->clt_id = strdup(srx->clt_id);

 				if (srx->ev_time_bootup_in_us_auth) {
 					drx->ev_time_bootup_in_us_auth =
@@ -3665,9 +3680,11 @@ static void dppline_add_stat_events(Sts_
 				}

 				if (srx->ip_addr) {
-					memcpy(drx->ip_addr.data, srx->ip_addr,
-					       16);
-					drx->ip_addr.len = 16;
+					uint8_t ip_tmp[4] = {0};
+					size_t ip_tmp_size = 0;
+					ip_string_to_ip_bytes(srx->ip_addr, &ip_tmp[0], &ip_tmp_size);
+					drx->ip_addr.data = ip_tmp;
+					drx->ip_addr.len = ip_tmp_size;
 					drx->has_ip_addr = true;
 				}

