Index: opensync-2.0.5.0/src/lib/schema/inc/schema_consts.h
===================================================================
--- opensync-2.0.5.0.orig/src/lib/schema/inc/schema_consts.h
+++ opensync-2.0.5.0/src/lib/schema/inc/schema_consts.h
@@ -153,7 +153,9 @@ typedef enum {
 #define SCHEMA_CONSTS_DTIM_PERIOD	"dtim_period"
 #define SCHEMA_CONSTS_DISABLE_B_RATES	"disable_b_rates"
 #define SCHEMA_CONSTS_IEEE80211k	"ieee80211k"
-
+#define SCHEMA_CONSTS_DYNAMIC_VLAN	"dynamic_vlan"
+#define SCHEMA_CONSTS_VLAN_TAGGED_IFACE	"vlan_tagged_interface"
+#define SCHEMA_CONSTS_VLAN_FILE		"vlan_file"
 /* Captive Portal */
 #define SCHEMA_CONSTS_SESSION_TIMEOUT              "session_timeout"
 #define SCHEMA_CONSTS_BROWSER_TITLE                "browser_title"
