--- a/src/lib/schema/inc/schema_consts.h
+++ b/src/lib/schema/inc/schema_consts.h
@@ -176,4 +176,8 @@ typedef enum {
 #define SCHEMA_CONSTS_MESH_HOP_PENALTY     "hop_penalty"
 #define SCHEMA_CONSTS_MESH_MCAST_RATE      "mcast_rate"
 #define SCHEMA_CONSTS_MESH_MTU             "mtu"
+
+/* Max clients per radio */
+#define SCHEMA_CONSTS_MAX_CLIENTS	"max_clients"
+
 #endif /* SCHEMA_CONSTS_H_INCLUDED */
