Index: opensync-2.0.5.0/interfaces/opensync.ovsschema
===================================================================
--- opensync-2.0.5.0.orig/interfaces/opensync.ovsschema
+++ opensync-2.0.5.0/interfaces/opensync.ovsschema
@@ -9368,6 +9368,69 @@
             }
        },
       "isRoot": true
+    },
+    "APC_Config": {
+      "columns": {
+        "enabled": {
+          "type": {
+            "key": {
+              "type": "boolean"
+            },
+            "min": 0,
+            "max": 1
+          }
+        }
+      },
+      "isRoot": true,
+      "maxRows": 1
+    },
+    "APC_State": {
+      "columns": {
+        "dr_addr": {
+          "type": {
+            "key": {
+              "type": "string"
+            }
+          }
+        },
+        "bdr_addr": {
+          "type": {
+            "key": {
+              "type": "string"
+            }
+          }
+        },
+        "enabled": {
+          "type": {
+            "key": {
+              "type": "boolean"
+            },
+            "min": 0,
+            "max": 1
+          }
+        },
+        "mode": {
+          "type": {
+            "key": {
+              "type": "string",
+              "enum": [
+                "set",
+                [
+                  "DR",
+                  "BDR",
+                  "OR",
+                  "WT",
+                  "NC"
+                ]
+              ]
+            },
+            "min": 0,
+            "max": 1
+          }
+        }
+      },
+      "isRoot": true,
+      "maxRows": 1
     }
   }
 }
Index: opensync-2.0.5.0/src/lib/schema/inc/schema_consts.h
===================================================================
--- opensync-2.0.5.0.orig/src/lib/schema/inc/schema_consts.h
+++ opensync-2.0.5.0/src/lib/schema/inc/schema_consts.h
@@ -154,6 +154,7 @@ typedef enum {
 #define SCHEMA_CONSTS_DISABLE_B_RATES	"disable_b_rates"
 #define SCHEMA_CONSTS_IEEE80211k	"ieee80211k"
 #define SCHEMA_CONSTS_DYNAMIC_VLAN	"dynamic_vlan"
+#define SCHEMA_CONSTS_RADPROXY	"radproxy"
 
 /* radio Custom options */
 #define SCHEMA_CONSTS_LOCAL_PWR_CONSTRAINT "local_pwr_constraint"
