Index: opensync-2.0.5.0/src/lib/schema/inc/schema_consts.h
===================================================================
--- opensync-2.0.5.0.orig/src/lib/schema/inc/schema_consts.h
+++ opensync-2.0.5.0/src/lib/schema/inc/schema_consts.h
@@ -155,6 +155,8 @@ typedef enum {
 #define SCHEMA_CONSTS_IEEE80211k	"ieee80211k"
 #define SCHEMA_CONSTS_DYNAMIC_VLAN	"dynamic_vlan"
 #define SCHEMA_CONSTS_RADPROXY	"radproxy"
+#define SCHEMA_CONSTS_PROXY_ARP		"proxy_arp"
+#define SCHEMA_CONSTS_MCAST_TO_UCAST	"mcast_to_ucast"
 
 /* radio Custom options */
 #define SCHEMA_CONSTS_LOCAL_PWR_CONSTRAINT "local_pwr_constraint"
