--- a/interfaces/opensync.ovsschema
+++ b/interfaces/opensync.ovsschema
@@ -1880,7 +1880,15 @@
             "min": 0,
             "max": "unlimited"
           }
-        }
+        },
+        "channel_max_power": {
+          "type": {
+            "key": "integer",
+            "value": "integer",
+            "min": 0,
+            "max": "unlimited"
+          }
+        }
       },
       "isRoot": true,
       "maxRows": 256
