Index: opensync-2.0.5.0/src/lib/datapipeline/inc/dpp_types.h
===================================================================
--- opensync-2.0.5.0.orig/src/lib/datapipeline/inc/dpp_types.h
+++ opensync-2.0.5.0/src/lib/datapipeline/inc/dpp_types.h
@@ -148,10 +148,9 @@ typedef enum
     RADIO_SCAN_TYPE_FULL,
     RADIO_SCAN_TYPE_ONCHAN,
     RADIO_SCAN_TYPE_OFFCHAN,
+    RADIO_SCAN_MAX_TYPE_QTY
 } radio_scan_type_t;
 
-#define RADIO_SCAN_MAX_TYPE_QTY       3
-
 typedef enum
 {
     RADIO_QUEUE_TYPE_VI = 0,
