--- a/src/um/src/um_ovsdb.c
+++ b/src/um/src/um_ovsdb.c
@@ -356,7 +356,8 @@ static void callback_AWLAN_Node(
                 //TODO Is there something that needs to be done here?
             }
 
-            if(awlan_node->upgrade_timer_changed){
+            if(awlan_node->upgrade_timer_changed
+                || ((awlan_node->firmware_url_changed) && (strlen(awlan_node->firmware_url) > 0))) {
                 if (awlan_node->upgrade_timer > 0)
                 {
                     /* if there is active timer, stop it to set new value   */
