--- a/src/lib/ovsdb/src/ovsdb_table.c
+++ b/src/lib/ovsdb/src/ovsdb_table.c
@@ -560,7 +560,7 @@ void ovsdb_table_update_cb(ovsdb_update_
         return;
     }
 
-    LOG(INFO, "MON upd: %s table: %s row: %s", typestr, table->table_name, mon_uuid );
+    LOG(TRACE, "MON upd: %s table: %s row: %s", typestr, table->table_name, mon_uuid );
 
     if (LOG_SEVERITY_TRACE <= log_module_severity_get(MODULE_ID))
     {
