Index: opensync-2.0.5.0/interfaces/opensync.ovsschema
===================================================================
--- opensync-2.0.5.0.orig/interfaces/opensync.ovsschema
+++ opensync-2.0.5.0/interfaces/opensync.ovsschema
@@ -900,6 +900,17 @@
             "min": 0,
             "max": 64
           }
+        },
+        "eth_ports": {
+          "type": {
+            "key": {
+                "type": "string",
+                "minLength": 1,
+                "maxLength": 32
+            },
+            "min": 0,
+            "max": 1
+          }
         }
       },
       "isRoot": true,
@@ -1170,6 +1181,22 @@
             "min": 0,
             "max": 64
           }
+        },
+        "eth_ports": {
+          "type": {
+            "key": {
+                "type": "string",
+                "minLength": 1,
+                "maxLength": 32
+            },
+            "value": {
+                "type": "string",
+                "minLength": 1,
+                "maxLength": 64
+            },
+            "min": 0,
+            "max": 32
+          }
         }
       },
       "isRoot": true,
