Index: opensync-2.0.5.0/src/lib/datapipeline/inc/dpp_survey.h
===================================================================
--- opensync-2.0.5.0.orig/src/lib/datapipeline/inc/dpp_survey.h
+++ opensync-2.0.5.0/src/lib/datapipeline/inc/dpp_survey.h
@@ -70,6 +70,7 @@ typedef struct
     uint32_t                        chan_tx;
     uint32_t                        chan_noise;
     uint32_t                        duration_ms;
+    uint32_t                        chan_width;
 
     /* Linked list survey data */
     ds_dlist_node_t                 node;
Index: opensync-2.0.5.0/src/sm/src/sm_survey_report.c
===================================================================
--- opensync-2.0.5.0.orig/src/sm/src/sm_survey_report.c
+++ opensync-2.0.5.0/src/sm/src/sm_survey_report.c
@@ -526,7 +526,7 @@ bool sm_survey_report_calculate_raw (
         memcpy(report_entry, record_entry, sizeof(*report_entry));
 
         LOGD("Sending %s %s %u survey report "
-             "{busy=%u tx=%u self=%u rx=%u ext=%u noise=%u duration=%u}",
+             "{busy=%u tx=%u self=%u rx=%u ext=%u noise=%u duration=%u chan_width=%u}",
              radio_get_name_from_cfg(radio_cfg_ctx),
              radio_get_scan_name_from_type(scan_type),
              report_entry->info.chan,
@@ -536,7 +536,8 @@ bool sm_survey_report_calculate_raw (
              report_entry->chan_rx,
              report_entry->chan_busy_ext,
              report_entry->chan_noise,
-             report_entry->duration_ms);
+             report_entry->duration_ms,
+	     report_entry->chan_width);
 
         ds_dlist_insert_tail(report_list, report_entry);
     }
@@ -732,7 +733,7 @@ bool sm_survey_update_list_cb (
                 result_entry);
 
         LOGD("Processed %s %s %u survey percent "
-             "{busy=%u tx=%u self=%u rx=%u ext=%u noise=%u duration=%u}",
+             "{busy=%u tx=%u self=%u rx=%u ext=%u noise=%u duration=%u channel_bandwidth:%u}",
              radio_get_name_from_cfg(radio_cfg_ctx),
              radio_get_scan_name_from_type(scan_type),
              result_entry->info.chan,
@@ -742,7 +743,8 @@ bool sm_survey_update_list_cb (
              result_entry->chan_rx,
              result_entry->chan_busy_ext,
              result_entry->chan_noise,
-             result_entry->duration_ms);
+             result_entry->duration_ms,
+	     result_entry->chan_width);
 
         result_entry->info.timestamp_ms =
             request_ctx->reporting_timestamp - survey_ctx->report_ts +
Index: opensync-2.0.5.0/interfaces/opensync_stats.proto
===================================================================
--- opensync-2.0.5.0.orig/interfaces/opensync_stats.proto
+++ opensync-2.0.5.0/interfaces/opensync_stats.proto
@@ -308,6 +308,7 @@ message Survey {
         optional uint32     offset_ms       = 9;
         optional uint32     busy_ext        = 10;   /* 40MHz extention channel busy */
         optional uint32     noise           = 11;
+        optional uint32     chan_width      = 12;
     }
     message SurveyAvg {
         required uint32     channel         = 1;
