From ebd46629ad718be6811c66bed748f61d9ea67f1b Mon Sep 17 00:00:00 2001
From: Miles Hu <milehu@codeaurora.org>
Date: Tue, 5 Nov 2019 14:49:16 -0800
Subject: [PATCH] iw: add HE rate/gi/ltf support

Signed-off-by: Miles Hu <milehu@codeaurora.org>
---
 bitrate.c | 172 ++++++++++++++++++++++++++++++++++++++++++++++++++++++++------
 nl80211.h |  24 +++++++++
 2 files changed, 181 insertions(+), 15 deletions(-)

--- a/bitrate.c
+++ b/bitrate.c
@@ -4,13 +4,22 @@
 #include "iw.h"
 
 
-static int parse_vht_chunk(const char *arg, __u8 *nss, __u16 *mcs)
+static int parse_rate_chunk(const char *arg, __u8 *nss, __u16 *mcs, unsigned int mode)
 {
 	unsigned int count, i;
-	unsigned int inss, mcs_start, mcs_end, tab[10];
+	unsigned int inss, mcs_start, mcs_end, tab[12];
+	unsigned int max_mcs = 0, max_nss = 0;
 
 	*nss = 0; *mcs = 0;
 
+	if (mode == NL80211_TXRATE_HE) {
+		max_mcs = 11;
+		max_nss = NL80211_HE_NSS_MAX;
+	} else {
+		max_mcs = 9;
+		max_nss = NL80211_VHT_NSS_MAX;
+	}
+
 	if (strchr(arg, '-')) {
 		/* Format: NSS:MCS_START-MCS_END */
 		count = sscanf(arg, "%u:%u-%u", &inss, &mcs_start, &mcs_end);
@@ -18,13 +27,13 @@ static int parse_vht_chunk(const char *a
 		if (count != 3)
 			return 0;
 
-		if (inss < 1 || inss > NL80211_VHT_NSS_MAX)
+		if (inss < 1 || inss > max_nss)
 			return 0;
 
 		if (mcs_start > mcs_end)
 			return 0;
 
-		if (mcs_start > 9 || mcs_end > 9)
+		if (mcs_start > max_mcs || mcs_end > max_mcs)
 			return 0;
 
 		*nss = inss;
@@ -33,19 +42,26 @@ static int parse_vht_chunk(const char *a
 
 	} else {
 		/* Format: NSS:MCSx,MCSy,... */
-		count = sscanf(arg, "%u:%u,%u,%u,%u,%u,%u,%u,%u,%u,%u", &inss,
-			   &tab[0], &tab[1], &tab[2], &tab[3], &tab[4], &tab[5],
-			   &tab[6], &tab[7], &tab[8], &tab[9]);
+		if (mode == NL80211_TXRATE_HE) {
+			count = sscanf(arg, "%u:%u,%u,%u,%u,%u,%u,%u,%u,%u,%u,%u,%u",
+				       &inss, &tab[0], &tab[1], &tab[2], &tab[3],
+				       &tab[4], &tab[5], &tab[6], &tab[7], &tab[8],
+				       &tab[9], &tab[10], &tab[11]);
+		} else {
+			count = sscanf(arg, "%u:%u,%u,%u,%u,%u,%u,%u,%u,%u,%u", &inss,
+				       &tab[0], &tab[1], &tab[2], &tab[3], &tab[4],
+				       &tab[5], &tab[6], &tab[7], &tab[8], &tab[9]);
+		}
 
 		if (count < 2)
 			return 0;
 
-		if (inss < 1 || inss > NL80211_VHT_NSS_MAX)
+		if (inss < 1 || inss > max_nss)
 			return 0;
 
 		*nss = inss;
 		for (i = 0; i < count - 1; i++) {
-			if (tab[i] > 9)
+			if (tab[i] > max_mcs)
 				return 0;
 			*mcs |= 1 << tab[i];
 		}
@@ -54,6 +70,16 @@ static int parse_vht_chunk(const char *a
 	return 1;
 }
 
+static int parse_vht_chunk(const char *arg, __u8 *nss, __u16 *mcs)
+{
+	return parse_rate_chunk(arg, nss, mcs, NL80211_TXRATE_VHT);
+}
+
+static int parse_he_chunk(const char *arg, __u8 *nss, __u16 *mcs)
+{
+	return parse_rate_chunk(arg, nss, mcs, NL80211_TXRATE_HE);
+}
+
 static int setup_vht(struct nl80211_txrate_vht *txrate_vht,
 		     int argc, char **argv)
 {
@@ -74,6 +100,45 @@ static int setup_vht(struct nl80211_txra
 	return 1;
 }
 
+static int setup_he(struct nl80211_txrate_he *txrate_he,
+		    int argc, char **argv)
+{
+	__u8 nss;
+	__u16 mcs;
+	int i;
+
+	memset(txrate_he, 0, sizeof(*txrate_he));
+
+	for (i = 0; i < argc; i++) {
+		if (!parse_he_chunk(argv[i], &nss, &mcs))
+			return 0;
+
+		nss--;
+		txrate_he->mcs[nss] |= mcs;
+	}
+
+	return 1;
+}
+
+#define HE_GI_STR_MAX	16
+#define HE_GI_08_STR "0.8"
+#define HE_GI_16_STR "1.6"
+#define HE_GI_32_STR "3.2"
+int parse_he_gi(char *he_gi)
+{
+	if (he_gi == NULL)
+		return 0;
+
+	if (!strncmp(he_gi, HE_GI_08_STR, sizeof(HE_GI_08_STR)))
+		return NL80211_RATE_INFO_HE_GI_0_8;
+	if (!strncmp(he_gi, HE_GI_16_STR, sizeof(HE_GI_16_STR)))
+		return NL80211_RATE_INFO_HE_GI_1_6;
+	if (!strncmp(he_gi, HE_GI_32_STR, sizeof(HE_GI_32_STR)))
+		return NL80211_RATE_INFO_HE_GI_3_2;
+
+	return -1;
+}
+
 #define VHT_ARGC_MAX	100
 
 static int handle_bitrates(struct nl80211_state *state,
@@ -90,24 +155,35 @@ static int handle_bitrates(struct nl8021
 	int *n_legacy = NULL;
 	bool have_ht_mcs_24 = false, have_ht_mcs_5 = false;
 	bool have_vht_mcs_24 = false, have_vht_mcs_5 = false;
+	bool have_he_mcs_24 = false, have_he_mcs_5 = false;
 	uint8_t ht_mcs_24[77], ht_mcs_5[77];
 	int n_ht_mcs_24 = 0, n_ht_mcs_5 = 0;
 	struct nl80211_txrate_vht txrate_vht_24 = {};
 	struct nl80211_txrate_vht txrate_vht_5 = {};
+	struct nl80211_txrate_he txrate_he_24 = {};
+	struct nl80211_txrate_he txrate_he_5 = {};
 	uint8_t *mcs = NULL;
 	int *n_mcs = NULL;
 	char *vht_argv_5[VHT_ARGC_MAX] = {}; char *vht_argv_24[VHT_ARGC_MAX] = {};
-	char **vht_argv = NULL;
+	char *he_argv_5[VHT_ARGC_MAX] = {}; char *he_argv_24[VHT_ARGC_MAX] = {};
+	char **vht_argv = NULL, **he_argv = NULL;
 	int vht_argc_5 = 0; int vht_argc_24 = 0;
-	int *vht_argc = NULL;
+	int he_argc_5 = 0; int he_argc_24 = 0;
+	int *vht_argc = NULL, *he_argc = NULL;
 	int sgi_24 = 0, sgi_5 = 0, lgi_24 = 0, lgi_5 = 0;
+	int has_he_gi_24 = 0, has_he_gi_5 = 0, has_he_ltf_24 = 0, has_he_ltf_5 = 0;
+	int he_gi = 0, he_ltf = 0;
+	char *he_gi_argv = NULL;
 
 	enum {
 		S_NONE,
 		S_LEGACY,
 		S_HT,
 		S_VHT,
+		S_HE,
 		S_GI,
+		S_HE_GI,
+		S_HE_LTF,
 	} parser_state = S_NONE;
 
 	for (i = 0; i < argc; i++) {
@@ -158,6 +234,20 @@ static int handle_bitrates(struct nl8021
 			vht_argv = vht_argv_5;
 			vht_argc = &vht_argc_5;
 			have_vht_mcs_5 = true;
+		} else if (strcmp(argv[i], "he-mcs-2.4") == 0) {
+			if (have_he_mcs_24)
+				return 1;
+			parser_state = S_HE;
+			he_argv = he_argv_24;
+			he_argc = &he_argc_24;
+			have_he_mcs_24 = true;
+		} else if (strcmp(argv[i], "he-mcs-5") == 0) {
+			if (have_he_mcs_5)
+				return 1;
+			parser_state = S_HE;
+			he_argv = he_argv_5;
+			he_argc = &he_argc_5;
+			have_he_mcs_5 = true;
 		} else if (strcmp(argv[i], "sgi-2.4") == 0) {
 			sgi_24 = 1;
 			parser_state = S_GI;
@@ -170,6 +260,18 @@ static int handle_bitrates(struct nl8021
 		} else if (strcmp(argv[i], "lgi-5") == 0) {
 			lgi_5 = 1;
 			parser_state = S_GI;
+		} else if (strcmp(argv[i], "he-gi-2.4") == 0) {
+			has_he_gi_24 = 1;
+			parser_state = S_HE_GI;
+		} else if (strcmp(argv[i], "he-gi-5") == 0) {
+			has_he_gi_5 = 1;
+			parser_state = S_HE_GI;
+		} else if (strcmp(argv[i], "he-ltf-2.4") == 0) {
+			has_he_ltf_24 = 1;
+			parser_state = S_HE_LTF;
+		} else if (strcmp(argv[i], "he-ltf-5") == 0) {
+			has_he_ltf_5 = 1;
+			parser_state = S_HE_LTF;
 		} else switch (parser_state) {
 		case S_LEGACY:
 			tmpd = strtod(argv[i], &end);
@@ -192,8 +294,24 @@ static int handle_bitrates(struct nl8021
 				return 1;
 			vht_argv[(*vht_argc)++] = argv[i];
 			break;
+		case S_HE:
+			if (*he_argc >= VHT_ARGC_MAX)
+				return 1;
+			he_argv[(*he_argc)++] = argv[i];
+			break;
 		case S_GI:
 			break;
+		case S_HE_GI:
+			he_gi_argv = argv[i];
+			break;
+		case S_HE_LTF:
+			he_ltf = strtol(argv[i], &end, 0);
+			if (*end != '\0')
+				return 1;
+			if (he_ltf < 0 || he_ltf > 4)
+				return 1;
+			he_ltf = he_ltf >> 1;
+			break;
 		default:
 			return 1;
 		}
@@ -207,17 +325,32 @@ static int handle_bitrates(struct nl8021
 		if(!setup_vht(&txrate_vht_5, vht_argc_5, vht_argv_5))
 			return -EINVAL;
 
+	if (have_he_mcs_24)
+		if (!setup_he(&txrate_he_24, he_argc_24, he_argv_24))
+			return -EINVAL;
+
+	if (have_he_mcs_5)
+		if (!setup_he(&txrate_he_5, he_argc_5, he_argv_5))
+			return -EINVAL;
+
 	if (sgi_5 && lgi_5)
 		return 1;
 
 	if (sgi_24 && lgi_24)
 		return 1;
 
+	if (he_gi_argv) {
+		he_gi = parse_he_gi(he_gi_argv);
+		if (he_gi < 0)
+			return 1;
+	}
+
 	nl_rates = nla_nest_start(msg, NL80211_ATTR_TX_RATES);
 	if (!nl_rates)
 		goto nla_put_failure;
 
-	if (have_legacy_24 || have_ht_mcs_24 || have_vht_mcs_24 || sgi_24 || lgi_24) {
+	if (have_legacy_24 || have_ht_mcs_24 || have_vht_mcs_24 || have_he_mcs_24 ||
+	    sgi_24 || lgi_24 || has_he_gi_24 || has_he_ltf_24) {
 		nl_band = nla_nest_start(msg, NL80211_BAND_2GHZ);
 		if (!nl_band)
 			goto nla_put_failure;
@@ -227,14 +360,22 @@ static int handle_bitrates(struct nl8021
 			nla_put(msg, NL80211_TXRATE_HT, n_ht_mcs_24, ht_mcs_24);
 		if (have_vht_mcs_24)
 			nla_put(msg, NL80211_TXRATE_VHT, sizeof(txrate_vht_24), &txrate_vht_24);
+		if (have_he_mcs_24)
+			nla_put(msg, NL80211_TXRATE_HE, sizeof(txrate_he_24),
+				&txrate_he_24);
 		if (sgi_24)
 			nla_put_u8(msg, NL80211_TXRATE_GI, NL80211_TXRATE_FORCE_SGI);
 		if (lgi_24)
 			nla_put_u8(msg, NL80211_TXRATE_GI, NL80211_TXRATE_FORCE_LGI);
+		if (has_he_gi_24)
+			nla_put_u8(msg, NL80211_TXRATE_HE_GI, he_gi);
+		if (has_he_ltf_24)
+			nla_put_u8(msg, NL80211_TXRATE_HE_LTF, he_ltf);
 		nla_nest_end(msg, nl_band);
 	}
 
-	if (have_legacy_5 || have_ht_mcs_5 || have_vht_mcs_5 || sgi_5 || lgi_5) {
+	if (have_legacy_5 || have_ht_mcs_5 || have_vht_mcs_5 || have_he_mcs_5 ||
+	    sgi_5 || lgi_5 || has_he_gi_5 || has_he_ltf_5) {
 		nl_band = nla_nest_start(msg, NL80211_BAND_5GHZ);
 		if (!nl_band)
 			goto nla_put_failure;
@@ -244,10 +385,17 @@ static int handle_bitrates(struct nl8021
 			nla_put(msg, NL80211_TXRATE_HT, n_ht_mcs_5, ht_mcs_5);
 		if (have_vht_mcs_5)
 			nla_put(msg, NL80211_TXRATE_VHT, sizeof(txrate_vht_5), &txrate_vht_5);
+		if (have_he_mcs_5)
+			nla_put(msg, NL80211_TXRATE_HE, sizeof(txrate_he_5),
+				&txrate_he_5);
 		if (sgi_5)
 			nla_put_u8(msg, NL80211_TXRATE_GI, NL80211_TXRATE_FORCE_SGI);
 		if (lgi_5)
 			nla_put_u8(msg, NL80211_TXRATE_GI, NL80211_TXRATE_FORCE_LGI);
+		if (has_he_gi_5)
+			nla_put_u8(msg, NL80211_TXRATE_HE_GI, he_gi);
+		if (has_he_ltf_5)
+			nla_put_u8(msg, NL80211_TXRATE_HE_LTF, he_ltf);
 		nla_nest_end(msg, nl_band);
 	}
 
@@ -259,9 +407,9 @@ static int handle_bitrates(struct nl8021
 }
 
 #define DESCR_LEGACY "[legacy-<2.4|5> <legacy rate in Mbps>*]"
-#define DESCR DESCR_LEGACY " [ht-mcs-<2.4|5> <MCS index>*] [vht-mcs-<2.4|5> <NSS:MCSx,MCSy... | NSS:MCSx-MCSy>*] [sgi-2.4|lgi-2.4] [sgi-5|lgi-5]"
+#define DESCR DESCR_LEGACY " [ht-mcs-<2.4|5> <MCS index>*] [vht-mcs-<2.4|5>  [he-mcs-<2.4|5> <NSS:MCSx,MCSy... | NSS:MCSx-MCSy>*] [sgi-2.4|lgi-2.4] [sgi-5|lgi-5]"
 
-COMMAND(set, bitrates, "[legacy-<2.4|5> <legacy rate in Mbps>*] [ht-mcs-<2.4|5> <MCS index>*] [vht-mcs-<2.4|5> <NSS:MCSx,MCSy... | NSS:MCSx-MCSy>*] [sgi-2.4|lgi-2.4] [sgi-5|lgi-5]",
+COMMAND(set, bitrates, "[legacy-<2.4|5> <legacy rate in Mbps>*] [ht-mcs-<2.4|5> <MCS index>*] [vht-mcs-<2.4|5> [he-mcs-<2.4|5> <NSS:MCSx,MCSy... | NSS:MCSx-MCSy>*] [sgi-2.4|lgi-2.4] [sgi-5|lgi-5] [he-gi-2.4:0.8/1.6/3.2] [he-gi-5:0.8/1.6/3.2] [he-ltf-2.4:1/2/4] [he-ltf-5:1/2/4]",
 	NL80211_CMD_SET_TX_BITRATE_MASK, 0, CIB_NETDEV, handle_bitrates,
 	"Sets up the specified rate masks.\n"
 	"Not passing any arguments would clear the existing mask (if any).");
--- a/nl80211.h
+++ b/nl80211.h
@@ -3059,6 +3059,18 @@ enum nl80211_he_gi {
 };
 
 /**
+ * enum nl80211_he_ltf - HE long training field
+ * @NL80211_RATE_INFO_HE_1xLTF: 3.2 usec
+ * @NL80211_RATE_INFO_HE_2xLTF: 6.4 usec
+ * @NL80211_RATE_INFO_HE_4xLTF: 12.8 usec
+ */
+enum nl80211_he_ltf {
+	NL80211_RATE_INFO_HE_1XLTF,
+	NL80211_RATE_INFO_HE_2XLTF,
+	NL80211_RATE_INFO_HE_4XLTF,
+};
+
+/**
  * enum nl80211_he_ru_alloc - HE RU allocation values
  * @NL80211_RATE_INFO_HE_RU_ALLOC_26: 26-tone RU allocation
  * @NL80211_RATE_INFO_HE_RU_ALLOC_52: 52-tone RU allocation
@@ -4570,6 +4582,10 @@ enum nl80211_key_attributes {
  *	in an array of MCS numbers.
  * @NL80211_TXRATE_VHT: VHT rates allowed for TX rate selection,
  *	see &struct nl80211_txrate_vht
+ * @NL80211_TXRATE_HE: HE rates allowed for TX rate selection,
+ *	see &struct nl80211_txrate_he
+ * @NL80211_TXRATE_HE_GI: configure HE GI, 0.8us, 1.6us and 3.2us.
+ * @NL80211_TXRATE_HE_LTF: configure HE LTF, 1XLTF, 2XLTF and 4XLTF.
  * @NL80211_TXRATE_GI: configure GI, see &enum nl80211_txrate_gi
  * @__NL80211_TXRATE_AFTER_LAST: internal
  * @NL80211_TXRATE_MAX: highest TX rate attribute
@@ -4580,6 +4596,9 @@ enum nl80211_tx_rate_attributes {
 	NL80211_TXRATE_HT,
 	NL80211_TXRATE_VHT,
 	NL80211_TXRATE_GI,
+	NL80211_TXRATE_HE,
+	NL80211_TXRATE_HE_GI,
+	NL80211_TXRATE_HE_LTF,
 
 	/* keep last */
 	__NL80211_TXRATE_AFTER_LAST,
@@ -4597,6 +4616,15 @@ struct nl80211_txrate_vht {
 	__u16 mcs[NL80211_VHT_NSS_MAX];
 };
 
+#define NL80211_HE_NSS_MAX		8
+/**
+ * struct nl80211_txrate_he - HE MCS/NSS txrate bitmap
+ * @mcs: MCS bitmap table for each NSS (array index 0 for 1 stream, etc.)
+ */
+struct nl80211_txrate_he {
+	__u16 mcs[NL80211_HE_NSS_MAX];
+};
+
 enum nl80211_txrate_gi {
 	NL80211_TXRATE_DEFAULT_GI,
 	NL80211_TXRATE_FORCE_SGI,
