--- a/lib/tls/openssl/openssl-x509.c	2026-03-20 14:44:19.544290115 +0800
+++ b/lib/tls/openssl/openssl-x509.c	2026-03-20 14:45:37.501112849 +0800
@@ -77,7 +77,8 @@ lws_tls_openssl_cert_info(X509 *x509, en
 {
 	X509_NAME *xn;
 #if !defined(LWS_PLAT_OPTEE)
-	char *p;
+	char *p, *p1;
+	size_t rl;
 #endif
 
 	if (!x509)
@@ -112,8 +113,16 @@ lws_tls_openssl_cert_info(X509 *x509, en
 			return -1;
 		X509_NAME_oneline(xn, buf->ns.name, (int)len - 2);
 		p = strstr(buf->ns.name, "/CN=");
-		if (p)
-			memmove(buf->ns.name, p + 4, strlen(p + 4) + 1);
+		if (p) {
+			p += 4;
+			p1 = strchr(p, '/');
+			if (p1)
+				rl = (size_t)(p1 - p);
+			else
+				rl = strlen(p);
+			memmove(buf->ns.name, p, rl);
+			buf->ns.name[rl] = '\0';
+		}
 		buf->ns.len = (int)strlen(buf->ns.name);
 		return 0;
 #endif
