From 132689b25103a48ac7c8fb0af64a6a99080ca192 Mon Sep 17 00:00:00 2001
From: John Crispin <john@phrozen.org>
Date: Sun, 22 May 2022 16:56:10 +0200
Subject: [PATCH 04/68] include/image-commands.mk: build mmc qdsk images

Signed-off-by: John Crispin <john@phrozen.org>
---
 include/image-commands.mk | 7 +++++++
 1 file changed, 7 insertions(+)

diff --git a/include/image-commands.mk b/include/image-commands.mk
index 08adbb6454..bd20f9608c 100644
--- a/include/image-commands.mk
+++ b/include/image-commands.mk
@@ -499,6 +499,13 @@ define Build/qsdk-ipq-factory-nor
 	@mv $@.new $@
 endef
 
+define Build/qsdk-ipq-factory-mmc
+	$(TOPDIR)/scripts/mkits-qsdk-ipq-image.sh \
+		$@.its hlos $(IMAGE_KERNEL) rootfs $(IMAGE_ROOTFS)
+	PATH=$(LINUX_DIR)/scripts/dtc:$(PATH) mkimage -f $@.its $@.new
+	@mv $@.new $@
+endef
+
 define Build/seama
 	$(STAGING_DIR_HOST)/bin/seama -i $@ \
 		-m "dev=/dev/mtdblock/$(SEAMA_MTDBLOCK)" -m "type=firmware"
-- 
2.34.1

