From c7923aafa72d82f9b2439314c80a5345ffb4e847 Mon Sep 17 00:00:00 2001
From: John Crispin <john@phrozen.org>
Date: Mon, 22 Jan 2024 12:21:02 +0100
Subject: [PATCH 05/68] rstrip.sh: do not strip FW binaries

Signed-off-by: John Crispin <john@phrozen.org>
---
 scripts/rstrip.sh | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/scripts/rstrip.sh b/scripts/rstrip.sh
index e47960a941..2aa7e96fda 100755
--- a/scripts/rstrip.sh
+++ b/scripts/rstrip.sh
@@ -20,7 +20,7 @@ TARGETS=$*
   exit 1
 }
 
-find $TARGETS -type f -a -exec file {} \; | \
+find $TARGETS -not -path \*/lib/firmware/\* -a -type f -a -exec file {} \; | \
   sed -n -e 's/^\(.*\):.*ELF.*\(executable\|relocatable\|shared object\).*,.*/\1:\2/p' | \
 (
   IFS=":"
-- 
2.34.1

