From 5ae30d4b5bf67788493b96a53d343c3518b9a56b Mon Sep 17 00:00:00 2001
From: John Crispin <john@phrozen.org>
Date: Mon, 2 Oct 2023 11:14:11 +0200
Subject: [PATCH 07/68] treewide: replace libustream-mbedtls with -openssl

Signed-off-by: John Crispin <john@phrozen.org>
---
 include/target.mk | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/include/target.mk b/include/target.mk
index bef904ffbc..4677ef5c49 100644
--- a/include/target.mk
+++ b/include/target.mk
@@ -17,7 +17,7 @@ DEFAULT_PACKAGES:=\
 	fstools \
 	libc \
 	libgcc \
-	libustream-mbedtls \
+	libustream-openssl \
 	logd \
 	mtd \
 	netifd \
-- 
2.34.1

