From efdbdcd3503315022a3270886826656be116345b Mon Sep 17 00:00:00 2001
From: John Crispin <john@phrozen.org>
Date: Sat, 19 Jun 2021 16:00:43 +0200
Subject: [PATCH 38/38] ugps: update to latest HEAD

Signed-off-by: John Crispin <john@phrozen.org>
---
 package/utils/ugps/Makefile | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/package/utils/ugps/Makefile b/package/utils/ugps/Makefile
index 4ce64d9631..e9a7f4fa8f 100644
--- a/package/utils/ugps/Makefile
+++ b/package/utils/ugps/Makefile
@@ -12,9 +12,9 @@ PKG_RELEASE:=1
 
 PKG_SOURCE_URL=$(PROJECT_GIT)/project/ugps.git
 PKG_SOURCE_PROTO:=git
-PKG_SOURCE_DATE:=2020-08-29
-PKG_SOURCE_VERSION:=511a5b3c84fa715ef0305cf26c98619c12a4867a
-PKG_MIRROR_HASH:=551d57b4c18da4c4695fa2f7a959c25a03be35d14cf872785e7db740f12366e2
+PKG_SOURCE_DATE:=2021-07-19
+PKG_SOURCE_VERSION:=350b960b6f8b08409dcef719ef1e1e0c8911040f
+PKG_MIRROR_HASH:=94d489072f8fccc26f624c3d3c65e7cb40e4b6267df5fb39d4e0e78469006bde
 
 PKG_MAINTAINER:=John Crispin <john@phrozen.org>
 PKG_LICENSE:=GPL-2.0+
-- 
2.25.1

