From 5b61c5de9334f24302741c2c964580023ee3c611 Mon Sep 17 00:00:00 2001
From: John Crispin <john@phrozen.org>
Date: Sun, 16 Jul 2023 17:15:05 +0200
Subject: [PATCH 24/43] elfutils: GCC5 compile fix

Signed-off-by: John Crispin <john@phrozen.org>
---
 package/libs/elfutils/Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/package/libs/elfutils/Makefile b/package/libs/elfutils/Makefile
index f7364c36be..1e7e86771f 100644
--- a/package/libs/elfutils/Makefile
+++ b/package/libs/elfutils/Makefile
@@ -87,7 +87,7 @@ TARGET_CFLAGS += \
 	-Wno-unused-result \
 	-Wno-format-nonliteral
 
-ifneq ($(CONFIG_GCC_USE_VERSION_11),y)
+ifneq ($(CONFIG_GCC_USE_VERSION_11)$(CONFIG_GCC_VERSION_5),y)
 TARGET_CFLAGS += \
 	-Wno-error=use-after-free
 endif
-- 
2.34.1

