From fa611ce803cbb53138cc184cf3779cd4edc6fd5d Mon Sep 17 00:00:00 2001
From: John Crispin <john@phrozen.org>
Date: Fri, 3 Nov 2023 09:36:58 +0100
Subject: [PATCH 30/68] netifd: update to latest HEAD

Signed-off-by: John Crispin <john@phrozen.org>
---
 package/network/config/netifd/Makefile | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/package/network/config/netifd/Makefile b/package/network/config/netifd/Makefile
index e60e592098..ffbc2613f5 100644
--- a/package/network/config/netifd/Makefile
+++ b/package/network/config/netifd/Makefile
@@ -5,9 +5,9 @@ PKG_RELEASE:=1
 
 PKG_SOURCE_PROTO:=git
 PKG_SOURCE_URL=$(PROJECT_GIT)/project/netifd.git
+PKG_MIRROR_HASH:=b13e0c9d34fece8d349cc512a91103e997329ae263afb6f365a3d7cb34137a62
 PKG_SOURCE_DATE:=2023-09-19
-PKG_SOURCE_VERSION:=7a58b995fdbecd9beed57e4d66d42cb3cf66aee2
-PKG_MIRROR_HASH:=a460a3b912047f8802eb24bb737084a08dad65b2dd520e5f5e7459379d1fcf8c
+PKG_SOURCE_VERSION:=8587c074f1eb2064c42adb0a6aa5073f695ab89d
 PKG_MAINTAINER:=Felix Fietkau <nbd@nbd.name>
 
 PKG_LICENSE:=GPL-2.0
-- 
2.34.1

