From aebe1b77a68ee3e438651185916b6e6fe093bed9 Mon Sep 17 00:00:00 2001
From: John Crispin <john@phrozen.org>
Date: Sun, 16 Jul 2023 17:15:23 +0200
Subject: [PATCH 40/43] ipq807x: package/linux hack

Signed-off-by: John Crispin <john@phrozen.org>
---
 package/Makefile | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/package/Makefile b/package/Makefile
index 4b8df7f484..128a0814a0 100644
--- a/package/Makefile
+++ b/package/Makefile
@@ -11,7 +11,11 @@ include $(INCLUDE_DIR)/feeds.mk
 include $(INCLUDE_DIR)/rootfs.mk
 
 -include $(TMP_DIR)/.packagedeps
+ifeq ($(CONFIG_TARGET_ipq807x_ipq807x)$(CONFIG_TARGET_ipq807x_ipq60xx)$(CONFIG_TARGET_ipq807x_ipq50xx),y)
+package-y +=feeds/ipq807x/linux
+else
 package-y += kernel/linux
+endif
 $(curdir)/autoremove:=1
 $(curdir)/builddirs:=$(sort $(package-) $(package-y) $(package-m))
 $(curdir)/builddirs-default:=. $(sort $(package-y) $(package-m))
-- 
2.34.1

