From 2a53eaa80156c56a8ea7881ba6b4295613faf399 Mon Sep 17 00:00:00 2001
From: John Crispin <john@phrozen.org>
Date: Sun, 16 Jul 2023 17:30:49 +0200
Subject: [PATCH 42/43] ipq807x: drop nand.sh from base-files

Signed-off-by: John Crispin <john@phrozen.org>
---
 package/base-files/Makefile | 7 +++++++
 1 file changed, 7 insertions(+)

diff --git a/package/base-files/Makefile b/package/base-files/Makefile
index dc89de2035..dd1d0760f8 100644
--- a/package/base-files/Makefile
+++ b/package/base-files/Makefile
@@ -138,6 +138,12 @@ ifeq ($(CONFIG_NAND_SUPPORT),)
   endef
 endif
 
+ifeq ($(CONFIG_TARGET_ipq807x),y)
+  define Package/base-files/ipq807x-support
+	rm -f $(1)/lib/upgrade/nand.sh
+  endef
+endif
+
 ifeq ($(CONFIG_EMMC_SUPPORT),)
   define Package/base-files/emmc-support
 	rm -f $(1)/lib/upgrade/emmc.sh
@@ -155,6 +161,7 @@ define Package/base-files/install
 	$(CP) ./files/* $(1)/
 	$(Package/base-files/install-key)
 	$(Package/base-files/nand-support)
+	$(Package/base-files/ipq807x-support)
 	$(Package/base-files/legacy-sdcard-support)
 	$(Package/base-files/emmc-support)
 	if [ -d $(GENERIC_PLATFORM_DIR)/base-files/. ]; then \
-- 
2.34.1

