From e886d2090484b46dc7759afe9a229ba906d80d57 Mon Sep 17 00:00:00 2001
From: John Crispin <john@phrozen.org>
Date: Sat, 15 Jul 2023 11:58:50 +0200
Subject: [PATCH 52/68] generic: enable pstore support

Signed-off-by: John Crispin <john@phrozen.org>
---
 target/linux/generic/config-5.15 | 16 ++++++++--------
 1 file changed, 8 insertions(+), 8 deletions(-)

diff --git a/target/linux/generic/config-5.15 b/target/linux/generic/config-5.15
index 214a1e6ecd..177562ad17 100644
--- a/target/linux/generic/config-5.15
+++ b/target/linux/generic/config-5.15
@@ -4844,21 +4844,21 @@ CONFIG_PROC_SYSCTL=y
 # CONFIG_PSAMPLE is not set
 # CONFIG_PSB6970_PHY is not set
 # CONFIG_PSI is not set
-# CONFIG_PSTORE is not set
+CONFIG_PSTORE=y
 # CONFIG_PSTORE_842_COMPRESS is not set
 # CONFIG_PSTORE_BLK is not set
-# CONFIG_PSTORE_COMPRESS is not set
+CONFIG_PSTORE_COMPRESS=y
+CONFIG_PSTORE_COMPRESS_DEFAULT="deflate"
 # CONFIG_PSTORE_CONSOLE is not set
-CONFIG_PSTORE_DEFAULT_KMSG_BYTES=10240
-# CONFIG_PSTORE_DEFLATE_COMPRESS is not set
-# CONFIG_PSTORE_DEFLATE_COMPRESS_DEFAULT is not set
-# CONFIG_PSTORE_FTRACE is not set
+CONFIG_PSTORE_DEFLATE_COMPRESS=y
+CONFIG_PSTORE_DEFLATE_COMPRESS_DEFAULT=y
+# CONFIG_PSTORE_ZSTD_COMPRESS is not set
 # CONFIG_PSTORE_LZ4HC_COMPRESS is not set
 # CONFIG_PSTORE_LZ4_COMPRESS is not set
 # CONFIG_PSTORE_LZO_COMPRESS is not set
 # CONFIG_PSTORE_PMSG is not set
-# CONFIG_PSTORE_RAM is not set
-# CONFIG_PSTORE_ZSTD_COMPRESS is not set
+CONFIG_PSTORE_RAM=y
+CONFIG_PSTORE_DEFAULT_KMSG_BYTES=10240
 # CONFIG_PTDUMP_DEBUGFS is not set
 # CONFIG_PTP_1588_CLOCK is not set
 # CONFIG_PTP_1588_CLOCK_IDT82P33 is not set
-- 
2.34.1

