From 978f5ba35a6a27c3abc500f13704275456d524c0 Mon Sep 17 00:00:00 2001
From: Venkat Chimata <venkatch@gmail.com>
Date: Tue, 30 Nov 2021 10:26:13 +0530
Subject: [PATCH] Minim Unum dependencies

---
 config/Config-devel.in | 12 ++++++++++++
 1 file changed, 12 insertions(+)

diff --git a/config/Config-devel.in b/config/Config-devel.in
index 126462bfc3..1999cac934 100644
--- a/config/Config-devel.in
+++ b/config/Config-devel.in
@@ -136,3 +136,15 @@ menuconfig DEVEL
 		default "-fno-caller-saves"
 		help
 		  Extra target-independent optimizations to use when building for the target.
+
+	config VERSION_PRODUCT
+		string
+		prompt "Minim unum agent product name"
+		help
+		  This specifies the product name to pass to Minim's unum agent
+
+	config VERSION_HWREV
+		string
+		prompt "Minim unum agent hardware revision"
+		help
+		  This specifies the hardware version to pass to Minim's unum agent
-- 
2.25.1

