From 3425faa1a72f1aef67d6dc2d229940e6be78d3a0 Mon Sep 17 00:00:00 2001
From: John Crispin <john@phrozen.org>
Date: Thu, 16 Nov 2023 16:05:32 +0100
Subject: [PATCH] libubox: update to latest HEAD

Signed-off-by: John Crispin <john@phrozen.org>
---
 package/libs/libubox/Makefile | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/package/libs/libubox/Makefile b/package/libs/libubox/Makefile
index c62278adc5..28d0d5f9ce 100644
--- a/package/libs/libubox/Makefile
+++ b/package/libs/libubox/Makefile
@@ -5,9 +5,9 @@ PKG_RELEASE=1
 
 PKG_SOURCE_PROTO:=git
 PKG_SOURCE_URL=$(PROJECT_GIT)/project/libubox.git
-PKG_MIRROR_HASH:=f22de22a784a0135cc2869fe81ff30e52136dca36863ee713503b4be5be01869
+PKG_MIRROR_HASH:=a0db9ed4ed26f2346c28902a78cd138750f3cad18a1a8b1bb960a2a6c397c773
 PKG_SOURCE_DATE:=2023-05-23
-PKG_SOURCE_VERSION:=75a3b870cace1171faf57bd55e5a9a2f1564f757
+PKG_SOURCE_VERSION:=82fa6480de7a85d0ced0701ab7c8825e31b90770
 PKG_ABI_VERSION:=$(call abi_version_str,$(PKG_SOURCE_DATE))
 CMAKE_INSTALL:=1
 
-- 
2.34.1

