From 9afc63f356f10942f924aaa6361f314f06727495 Mon Sep 17 00:00:00 2001
From: John Crispin <john@phrozen.org>
Date: Mon, 21 Jul 2025 13:27:07 +0200
Subject: [PATCH] include/download.mk: switch to using git:// instead of
 https://

Signed-off-by: John Crispin <john@phrozen.org>
---
 include/download.mk | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/include/download.mk b/include/download.mk
index 9ab0b6c08f..430bd89307 100644
--- a/include/download.mk
+++ b/include/download.mk
@@ -3,7 +3,7 @@
 # Copyright (C) 2006-2012 OpenWrt.org
 # Copyright (C) 2016 LEDE project
 
-PROJECT_GIT = https://git.openwrt.org
+PROJECT_GIT = git://git.openwrt.org
 
 OPENWRT_GIT = $(PROJECT_GIT)
 LEDE_GIT = $(PROJECT_GIT)
-- 
2.34.1

