From 0f590999ea33fb91160aec123a9a74c31623584e Mon Sep 17 00:00:00 2001
From: Paul White <paul@shasta.cloud>
Date: Mon, 21 Jul 2025 22:41:52 +0000
Subject: [PATCH] feeds.default.conf: switch to using git:// instead of
 https://

Signed-off-by: Paul White <paul@shasta.cloud>
---
 feeds.conf.default | 8 ++++----
 1 file changed, 4 insertions(+), 4 deletions(-)

diff --git a/feeds.conf.default b/feeds.conf.default
index d467db5627..bafcf8aa98 100644
--- a/feeds.conf.default
+++ b/feeds.conf.default
@@ -1,4 +1,4 @@
-src-git packages https://git.openwrt.org/feed/packages.git;openwrt-23.05
-src-git luci https://git.openwrt.org/project/luci.git;openwrt-23.05
-src-git routing https://git.openwrt.org/feed/routing.git;openwrt-23.05
-src-git telephony https://git.openwrt.org/feed/telephony.git;openwrt-23.05
+src-git packages git://git.openwrt.org/feed/packages.git;openwrt-23.05
+src-git luci git://git.openwrt.org/project/luci.git;openwrt-23.05
+src-git routing git://git.openwrt.org/feed/routing.git;openwrt-23.05
+src-git telephony git://git.openwrt.org/feed/telephony.git;openwrt-23.05
-- 
2.43.0

