From 197f527c738bf083017c1010dbca8e448e9c0f3f Mon Sep 17 00:00:00 2001
From: "wenij.chang" <wenij.chang@cybertan.com.tw>
Date: Thu, 16 Mar 2023 15:57:24 +0800
Subject: [PATCH] add profile can run_cmd before start gen_config

---
 scripts/gen_config.py | 5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

diff --git a/scripts/gen_config.py b/scripts/gen_config.py
index 01c3f4050d..1fa1e1c3a1 100755
--- a/scripts/gen_config.py
+++ b/scripts/gen_config.py
@@ -70,7 +70,7 @@ def load_yaml(fname: str, profile: dict, include=True):
 
     new = yaml.safe_load(profile_file.read_text())
     for n in new:
-        if n in {"profile", "target", "subtarget", "external_target", "image"}:
+        if n in {"profile", "target", "subtarget", "external_target", "image", "run_cmd"}:
             if profile.get(n):
                 die(f"Duplicate tag found {n}")
             profile.update({n: new.get(n)})
@@ -222,6 +222,9 @@ if __name__ == "__main__":
     for d in profile.get("description"):
         print(f" - {d}")
 
+    if profile.get("run_cmd", False):
+        run(["sh",profile["run_cmd"]])
+
     clean_tree()
     setup_feeds(profile)
     generate_config(profile)
-- 
2.17.1

