From 40894005e2c114664a44abb0ffadb1cb945a88e2 Mon Sep 17 00:00:00 2001
From: Antonio Wu <antonio.wu@cybertan.com.tw>
Date: Mon, 4 Aug 2025 05:10:41 +0000
Subject: [PATCH] add KERNEL_IPQ_MEM_PROFILE for IPQ53XX 

---
 config/Config-kernel.in | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/config/Config-kernel.in b/config/Config-kernel.in
index 3529696559..b7f13eccbe 100644
--- a/config/Config-kernel.in
+++ b/config/Config-kernel.in
@@ -6,7 +6,7 @@ config KERNEL_IPQ_MEM_PROFILE
 	int "Different memory profile "
 	range 0 1024
 	default 512
-	depends on TARGET_ipq807x || TARGET_ipq60xx || TARGET_ipq50xx || TARGET_ipq95xx
+	depends on TARGET_ipq807x || TARGET_ipq60xx || TARGET_ipq50xx || TARGET_ipq95xx || TARGET_ipq53xx
 	help
 	  This option select memory profile to be used,which defines
 	  the reserved memory configuration used in device tree.
-- 
2.17.1

