From 363e5f1c2ea75772b238a0e37d6a4b31f596d1a2 Mon Sep 17 00:00:00 2001
From: Owen <owen.chen@cybertan.com.tw>
Date: Fri, 21 Jul 2023 13:55:36 +0800
Subject: [PATCH] Add run_cmd_2 to apply patches after all feeds ready.

---
 scripts/gen_config.py | 6 +++++-
 1 file changed, 5 insertions(+), 1 deletion(-)

diff --git a/scripts/gen_config.py b/scripts/gen_config.py
index 1fa1e1c..348bd4f 100755
--- a/scripts/gen_config.py
+++ b/scripts/gen_config.py
@@ -70,7 +70,7 @@ def load_yaml(fname: str, profile: dict, include=True):
 
     new = yaml.safe_load(profile_file.read_text())
     for n in new:
-        if n in {"profile", "target", "subtarget", "external_target", "image", "run_cmd"}:
+        if n in {"profile", "target", "subtarget", "external_target", "image", "run_cmd", "run_cmd_2"}:
             if profile.get(n):
                 die(f"Duplicate tag found {n}")
             profile.update({n: new.get(n)})
@@ -227,6 +227,10 @@ if __name__ == "__main__":
 
     clean_tree()
     setup_feeds(profile)
+    
+    if profile.get("run_cmd_2", False):
+        run(["sh",profile["run_cmd_2"]])
+    
     generate_config(profile)
     run(["rm", "-rf", "tmp/"])
     print("Running make defconfig")
-- 
2.7.4

