From ee2f29586b15891ccb99fc8f5fc89d0d0418ca0d Mon Sep 17 00:00:00 2001
From: "Crychic.Tan" <Crychic.Tan@cybertan.com.tw>
Date: Mon, 22 Dec 2025 09:17:38 +0000
Subject: [PATCH] sonicfi_rap630w-211g: import v6.6 kernel + drivers from
 OpenWrt 24.10

---
 include/image-commands.mk | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/include/image-commands.mk b/include/image-commands.mk
index b7a0d98d3d..cb3a0d38c4 100644
--- a/include/image-commands.mk
+++ b/include/image-commands.mk
@@ -293,7 +293,7 @@ define Build/initrd_compression
 	$(if $(CONFIG_TARGET_INITRAMFS_COMPRESSION_ZSTD),.zstd)
 endef
 
-ifeq ($(CONFIG_TARGET_mediatek_mt7981_DEVICE_sonicfi_rap630w_211g),y)
+ifeq ($(CONFIG_TARGET_mediatek_filogic_DEVICE_sonicfi_rap630w_211g),y)
 define Build/fit
 	$(TOPDIR)/scripts/mkits_sonicfi_rap630w_211g.sh \
 		-D $(DEVICE_NAME) -o $@.its -k $@ \
-- 
2.17.1

