--- a/drivers/bus/mhi/test/mhitest_pci.c
+++ b/drivers/bus/mhi/test/mhitest_pci.c
@@ -547,11 +547,6 @@ int mhitest_pci_register_mhi(struct mhit
 	mplat->mhi_ctrl = mhi_ctrl;
 	dev_set_drvdata(&pci_dev->dev, mplat);
 	mhi_ctrl->cntrl_dev = &pci_dev->dev;
-
-	if (!mplat->fw_name) {
-		MHITEST_ERR("fw_name is NULLL\n");
-		return -EINVAL;
-	}
 	mhi_ctrl->fw_image = mplat->fw_name;
 
 	mhi_ctrl->regs = mplat->bar;
