Index: linux-5.4.164-qsdk-26349818b464f8c7b52d59ce73579d9f3dd6bd5d/drivers/mtd/nand/raw/nand_ids.c
===================================================================
--- linux-5.4.164-qsdk-26349818b464f8c7b52d59ce73579d9f3dd6bd5d.orig/drivers/mtd/nand/raw/nand_ids.c
+++ linux-5.4.164-qsdk-26349818b464f8c7b52d59ce73579d9f3dd6bd5d/drivers/mtd/nand/raw/nand_ids.c
@@ -86,6 +86,12 @@ struct nand_flash_dev nand_flash_ids[] =
 	{"GD5F1GQ5REYIH SPI NAND 1G 1.8V",
 		{ .id = {0xc8, 0x21} },
 		SZ_2K, SZ_128, SZ_128K, 0, 2, 64, NAND_ECC_INFO(4, SZ_512), 0},
+	{"GD5F1GM7REYIG SPI NAND 1G-BIT",
+		{.id = {0xc8, 0x81}},
+		SZ_2K, SZ_128, SZ_128K, 0, 2, 128, NAND_ECC_INFO(8,SZ_512), 0},
+	{"GD5F2GM7REYIG SPI NAND 2G 8-bit",
+		{.id = {0xc8, 0x82}},
+		SZ_2K, SZ_256, SZ_128K, 0, 2, 128, NAND_ECC_INFO(8,SZ_512), 0},
 	{"W25N01JW SPI NAND 1.8V 1G-BIT",
 		{ .id = {0xef, 0xbc} },
 		SZ_2K, SZ_128, SZ_128K, 0, 2, 64, NAND_ECC_INFO(4, SZ_512), 0},
