--- a/watchdog.c
+++ b/watchdog.c
@@ -54,12 +54,7 @@ static enum hrtimer_restart morse_watchd
 static void watchdog_timer_start(struct morse *mors)
 {
 	ktime_t interval = ktime_set(mors->watchdog.interval_secs, 0);
-
-#if defined(MAC80211_BACKPORT_VERSION_CODE) && (KERNEL_VERSION(4, 10, 0) <= MAC80211_VERSION_CODE)
-	hrtimer_start(&mors->watchdog.timer, interval.tv64, HRTIMER_MODE_REL);
-#else
 	hrtimer_start(&mors->watchdog.timer, interval, HRTIMER_MODE_REL);
-#endif
 }

 int morse_watchdog_start(struct morse *mors)
