Index: morse_driver-rel_1_12_3_2024_May_23/sdio.c
===================================================================
--- morse_driver-rel_1_12_3_2024_May_23.orig/sdio.c
+++ morse_driver-rel_1_12_3_2024_May_23/sdio.c
@@ -572,7 +572,7 @@ static int morse_sdio_reset(int reset_pi
 
 	sdio_claim_host(func);
 	sdio_disable_func(func);
-#if KERNEL_VERSION(5, 18, 0) > LINUX_VERSION_CODE
+#if KERNEL_VERSION(5, 18, 0) > MAC80211_VERSION_CODE
 	mmc_hw_reset(func->card->host);
 #else
 	mmc_hw_reset(func->card);
