--- a/debug.h
+++ b/debug.h
@@ -162,7 +162,7 @@ void morse_log_set_default(u8 lvl);
  *
  * @returns	True if output would be generated and false otherwise.
  */
-bool morse_log_is_enabled(u32 id, u8 level);
+bool morse_log_is_enabled(enum morse_feature_id id, u8 level);
 
 /**
  * Set the default logging level for all features.
--- a/firmware.c
+++ b/firmware.c
@@ -850,7 +850,7 @@ static int morse_firmware_get_firmware(struct morse *mors, const char *fw_name)
 	return ret;
 }
 
-int morse_firmware_init(struct morse *mors, enum morse_config_test_mode test_mode)
+int morse_firmware_init(struct morse *mors, uint test_mode)
 {
 	int n;
 	int ret = 0;
