From aca295cc5d38ede242c39781ebce56924e621a66 Mon Sep 17 00:00:00 2001
From: Lei Wei <quic_leiwei@quicinc.com>
Date: Thu, 21 Dec 2023 13:39:56 +0800
Subject: [PATCH] nat46: fix sleep calltrace warning in proc file show

Signed-off-by: Lei Wei <quic_leiwei@quicinc.com>
---
 nat46/modules/nat46-netdev.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/nat46/modules/nat46-netdev.c b/nat46/modules/nat46-netdev.c
index 78ad44c..561f9e2 100644
--- a/nat46/modules/nat46-netdev.c
+++ b/nat46/modules/nat46-netdev.c
@@ -394,7 +394,7 @@ void nat64_show_all_configs(struct seq_file *m) {
 		if(is_nat46(dev)) {
 			nat46_instance_t *nat46 = netdev_nat46_instance(dev);
 			int ipair = -1;
-			char *buf = kmalloc(NAT46_CFG_BUFLEN + 1, GFP_KERNEL);
+			char *buf = kmalloc(NAT46_CFG_BUFLEN + 1, GFP_ATOMIC);
 			seq_printf(m, "add %s\n", dev->name);
 			if(buf) {
 				for(ipair = 0; ipair < nat46->npairs; ipair++) {
-- 
2.42.0

