From: Sven Eckelmann <sven@narfation.org>
Date: Mon, 14 Mar 2022 15:23:40 +0100
Subject: om-fwupgradecfg-gen: Add support for Plasma Cloud PAX1800 v1/v2

Signed-off-by: Sven Eckelmann <sven@narfation.org>

diff --git a/scripts/om-fwupgradecfg-gen.sh b/scripts/om-fwupgradecfg-gen.sh
index bf7c047c5ba8e78ad4e43fa7edff2bb4bb3f4388..61617a0b772e1661a46f559380ffaf0b021c524d 100755
--- a/scripts/om-fwupgradecfg-gen.sh
+++ b/scripts/om-fwupgradecfg-gen.sh
@@ -45,6 +45,13 @@ case $CE_TYPE in
 		SIZE_FACTOR=1024
 		SIZE_FORMAT="0x%08x"
 		;;
+	PAX1800|\
+	PAX1800v2)
+		MAX_PART_SIZE=28672
+		KERNEL_FLASH_ADDR=0x3a0000
+		SIZE_FACTOR=1024
+		SIZE_FORMAT="0x%08x"
+		;;
 	*)
 		echo "Error - unsupported ce type: $CE_TYPE"
 		exit 1
