From 398ceb5dee1a6e1e7e30e02d4f77316a83fdc5b1 Mon Sep 17 00:00:00 2001
From: John Crispin <john@phrozen.org>
Date: Sun, 22 May 2022 16:56:10 +0200
Subject: [PATCH] include/image-commands.mk: build mmc qdsk images

Signed-off-by: John Crispin <john@phrozen.org>
---
 include/image-commands.mk | 7 +++++++
 1 file changed, 7 insertions(+)

diff --git a/include/image-commands.mk b/include/image-commands.mk
index d4bce80e68..541ec4f575 100644
--- a/include/image-commands.mk
+++ b/include/image-commands.mk
@@ -354,6 +354,13 @@ define Build/qsdk-ipq-factory-nor
 	@mv $@.new $@
 endef
 
+define Build/qsdk-ipq-factory-mmc
+	$(TOPDIR)/scripts/mkits-qsdk-ipq-image.sh \
+		$@.its hlos $(IMAGE_KERNEL) rootfs $(IMAGE_ROOTFS)
+	PATH=$(LINUX_DIR)/scripts/dtc:$(PATH) mkimage -f $@.its $@.new
+	@mv $@.new $@
+endef
+
 define Build/seama
 	$(STAGING_DIR_HOST)/bin/seama -i $@ \
 		-m "dev=/dev/mtdblock/$(SEAMA_MTDBLOCK)" -m "type=firmware"
-- 
2.25.1

