From 3cf8656c8bd6ca7d9a0a8ef3dbba81557ce0eba3 Mon Sep 17 00:00:00 2001
From: Brett Mastbergen <brettm@minim.com>
Date: Thu, 4 Aug 2022 14:11:58 -0400
Subject: [PATCH] linux/modules: Fix more netfilter 4.4 deps

---
 include/netfilter.mk | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/include/netfilter.mk b/include/netfilter.mk
index cee40ee008..9341a3ba6f 100644
--- a/include/netfilter.mk
+++ b/include/netfilter.mk
@@ -231,6 +231,8 @@ $(eval $(call nf_add,NF_NATHELPER_EXTRA,CONFIG_NF_CONNTRACK_TFTP, $(P_XT)nf_conn
 $(eval $(call nf_add,NF_NATHELPER_EXTRA,CONFIG_NF_NAT_TFTP, $(P_XT)nf_nat_tftp))
 $(eval $(call nf_add,NF_NATHELPER_EXTRA,CONFIG_NF_CONNTRACK_IRC, $(P_XT)nf_conntrack_irc))
 $(eval $(call nf_add,NF_NATHELPER_EXTRA,CONFIG_NF_NAT_IRC, $(P_XT)nf_nat_irc))
+$(eval $(call nf_add,NF_NATHELPER_EXTRA,CONFIG_NF_CT_PROTO_GRE, $(P_XT)nf_conntrack_proto_gre))
+$(eval $(call nf_add,NF_NATHELPER_EXTRA,CONFIG_NF_NAT_PROTO_GRE, $(P_V4)nf_nat_proto_gre))
 
 
 # ulog
-- 
2.30.2

