Index: wlan-ap/openwrt/package/kernel/linux/modules/netdevices.mk
===================================================================
--- wlan-ap.orig/openwrt/package/kernel/linux/modules/netdevices.mk
+++ wlan-ap/openwrt/package/kernel/linux/modules/netdevices.mk
@@ -323,7 +323,7 @@ define KernelPackage/switch-rtl8366-smi
   SUBMENU:=$(NETWORK_DEVICES_MENU)
   TITLE:=Realtek RTL8366 SMI switch interface support
   DEPENDS:=@GPIO_SUPPORT +kmod-swconfig +(TARGET_armvirt||TARGET_bcm27xx_bcm2708||TARGET_tegra):kmod-of-mdio
-  KCONFIG:=CONFIG_RTL8366_SMI
+  KCONFIG:=CONFIG_RTL8366_SMI=y
   FILES:=$(LINUX_DIR)/drivers/net/phy/rtl8366_smi.ko
   AUTOLOAD:=$(call AutoLoad,42,rtl8366_smi,1)
 endef
@@ -371,7 +371,7 @@ define KernelPackage/switch-rtl8367b
   SUBMENU:=$(NETWORK_DEVICES_MENU)
   TITLE:=Realtek RTL8367R/B switch support
   DEPENDS:=+kmod-switch-rtl8366-smi
-  KCONFIG:=CONFIG_RTL8367B_PHY
+  KCONFIG:=CONFIG_RTL8367B_PHY=y
   FILES:=$(LINUX_DIR)/drivers/net/phy/rtl8367b.ko
   AUTOLOAD:=$(call AutoLoad,43,rtl8367b,1)
 endef
