Index: backports-6.5-rc3/net/wireless/nl80211.c
===================================================================
--- backports-6.5-rc3.orig/net/wireless/nl80211.c
+++ backports-6.5-rc3/net/wireless/nl80211.c
@@ -10270,14 +10296,15 @@ static int nl80211_trigger_scan(struct s
 		}
 	}
 
-	if (!i) {
-		err = -EINVAL;
-		goto out_free;
+	if (!chandef_found) {
+		if (!i) {
+			err = -EINVAL;
+			goto out_free;
+		}
+		request->n_channels = i;
+		request->scan_with_freq_info = freq_info_provided;
 	}
 
-	request->n_channels = i;
-	request->scan_with_freq_info = freq_info_provided;
-
 	wdev_lock(wdev);
 	for (i = 0; i < request->n_channels; i++) {
 		struct ieee80211_channel *chan = request->channels[i];
