Index: libtcmd-11.5/libtcmd.h
===================================================================
--- libtcmd-11.5.orig/libtcmd.h
+++ libtcmd-11.5/libtcmd.h
@@ -71,7 +71,9 @@ struct tcmd_cfg {
 	struct sigevent sev;
 	timer_t timer;
 	bool timeout;
-} tcmd_cfg;
+};
+
+extern struct tcmd_cfg tcmd_cfg;
 
 /* WLAN API */
 #ifdef WLAN_API_NL80211
Index: libtcmd-11.5/nl80211.c
===================================================================
--- libtcmd-11.5.orig/nl80211.c
+++ libtcmd-11.5/nl80211.c
@@ -23,6 +23,7 @@
 #endif
 
 int cb_ret;
+struct tcmd_cfg tcmd_cfg;
 
 #ifdef LIBNL_2
 static inline struct nl_sock *nl_handle_alloc(void)
