From a5e0fc0938a7b457b3bfdfa246ba823106e01289 Mon Sep 17 00:00:00 2001
From: Balamurugan Ramar <quic_bramar@quicinc.com>
Date: Thu, 10 Apr 2025 23:25:18 +0530
Subject: [PATCH] mac80211: disable signed regulatory

Signed-off-by: Balamurugan Ramar <quic_bramar@quicinc.com>
---
 net/wireless/Kconfig | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/net/wireless/Kconfig b/net/wireless/Kconfig
index 9e2ce3b..9737c46 100644
--- a/net/wireless/Kconfig
+++ b/net/wireless/Kconfig
@@ -90,7 +90,7 @@ config CFG80211_CERTIFICATION_ONUS
 
 config CFG80211_REQUIRE_SIGNED_REGDB
 	bool "require regdb signature" if CFG80211_CERTIFICATION_ONUS
-	default y
+	default n
 	depends on SYSTEM_DATA_VERIFICATION
 	help
 	  Require that in addition to the "regulatory.db" file a
-- 
2.34.1

