From f191ff5b8b620be4b31b34186746c92ca246d4af Mon Sep 17 00:00:00 2001
From: Balamurugan Ramar <quic_bramar@quicinc.com>
Date: Thu, 10 Apr 2025 23:25:41 +0530
Subject: [PATCH] mac80211: fix compilation issues with Pkg Upgrd

Signed-off-by: Aditya Kumar Singh <quic_adisi@quicinc.com>
Signed-off-by: Balamurugan Ramar <quic_bramar@quicinc.com>
---
 net/mac80211/cfg.c | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/net/mac80211/cfg.c b/net/mac80211/cfg.c
index 2120e18..774296b 100644
--- a/net/mac80211/cfg.c
+++ b/net/mac80211/cfg.c
@@ -616,10 +616,12 @@ static struct ieee80211_key *
 ieee80211_lookup_key(struct ieee80211_sub_if_data *sdata, int link_id,
 		     u8 key_idx, bool pairwise, const u8 *mac_addr)
 {
-	struct ieee80211_local *local __maybe_unused = sdata->local;
+	struct ieee80211_local *local __maybe_unused;
 	struct ieee80211_link_data *link = &sdata->deflink;
 	struct ieee80211_key *key;
 
+	local = sdata->local;
+
 	if (link_id >= 0) {
 		link = sdata_dereference(sdata->link[link_id], sdata);
 		if (!link)
-- 
2.34.1

