From e514833e2ec808c6998407a553db0d1e3fb8a167 Mon Sep 17 00:00:00 2001
From: Balamurugan Ramar <quic_bramar@quicinc.com>
Date: Thu, 10 Apr 2025 23:27:07 +0530
Subject: [PATCH] mac80211: QSDK wifi add Kconfig for ATHMEMDEBUG infra

Signed-off-by: Hari Chandrakanthan <quic_haric@quicinc.com>
Signed-off-by: Raj Kumar Bhagat <quic_rajkbhag@quicinc.com>
Signed-off-by: Balamurugan Ramar <quic_bramar@quicinc.com>
---
 local-symbols        | 1 +
 net/mac80211/Kconfig | 9 +++++++++
 2 files changed, 10 insertions(+)

diff --git a/local-symbols b/local-symbols
index a78db6b..5b141a2 100644
--- a/local-symbols
+++ b/local-symbols
@@ -61,6 +61,7 @@ MAC80211_NSS_SUPPORT=
 MAC80211_PPE_SUPPORT=
 MAC80211_DS_SUPPORT=
 MAC80211_SFE_SUPPORT=
+MAC80211_ATHMEMDEBUG=
 WLAN=
 ATH_COMMON=
 WLAN_VENDOR_ATH=
diff --git a/net/mac80211/Kconfig b/net/mac80211/Kconfig
index b58cbf2..c9b6971 100644
--- a/net/mac80211/Kconfig
+++ b/net/mac80211/Kconfig
@@ -243,6 +243,15 @@ config MAC80211_PS_DEBUG
 
 	  Do not select this option.
 
+config MAC80211_ATHMEMDEBUG
+	bool "ATH memory debugging"
+	depends on ATHMEMDEBUG
+	default n
+	help
+	  Select this option to enable a debug build. This debug
+	  build helps to track memory allocation, deallocation
+	  for ATH related sub-system (cfg/mac80211, ath).
+
 config MAC80211_MPL_DEBUG
 	bool "Verbose mesh peer link debugging"
 	depends on MAC80211_DEBUG_MENU
-- 
2.34.1

