From: Karthikeyan Periyasamy <periyasa@codeaurora.org>
Subject: [PATCH] QSDK: ath11k: Add debug logs for the CR2849877 and CR2854245

Patch-dependency: None

Signed-off-by: Karthikeyan Periyasamy <periyasa@codeaurora.org>
Signed-off-by: Gautham Kumar Senthilkumaran <quic_gauthamk@quicinc.com>
---
--- a/net/mac80211/rx.c
+++ b/net/mac80211/rx.c
@@ -5342,12 +5342,16 @@ void ieee80211_rx_list(struct ieee80211_
 		}
 	}
 
-	if (WARN_ON(status->band >= NUM_NL80211_BANDS))
+	if (WARN_ON(status->band >= NUM_NL80211_BANDS)) {
+		printk("invalid status band %d freq %d\n", status->band, status->freq);
 		goto drop;
+	}
 
 	sband = local->hw.wiphy->bands[status->band];
-	if (WARN_ON(!sband))
+	if (WARN_ON(!sband)) {
+		printk("NULL sband status band %d freq %d\n", status->band, status->freq);
 		goto drop;
+	}
 
 	/*
 	 * If we're suspending, it is possible although not too likely
--- a/net/mac80211/scan.c
+++ b/net/mac80211/scan.c
@@ -647,9 +647,16 @@ static void ieee80211_scan_state_send_pr
 	int i;
 	struct ieee80211_sub_if_data *sdata;
 	struct cfg80211_scan_request *scan_req;
-	enum nl80211_band band = local->hw.conf.chandef.chan->band;
+	enum nl80211_band band;
 	u32 flags = 0, tx_flags;
 
+	if (!local->hw.conf.chandef.chan) {
+		printk("channel NULL\n");
+		WARN_ON(1);
+		return;
+	}
+
+	band = local->hw.conf.chandef.chan->band;
 	scan_req = rcu_dereference_protected(local->scan_req,
 					     lockdep_is_held(&local->mtx));
 
--- a/net/mac80211/cfg.c
+++ b/net/mac80211/cfg.c
@@ -4062,6 +4062,7 @@ int ieee80211_attach_ack_skb(struct ieee
 
 	if (id < 0) {
 		kfree_skb(ack_skb);
+		printk(KERN_ERR "attach ack fail %d\n", id);
 		return -ENOMEM;
 	}
 
--- a/net/mac80211/tx.c
+++ b/net/mac80211/tx.c
@@ -2562,6 +2562,7 @@ static u16 ieee80211_store_ack_skb(struc
 			}
 		} else {
 			kfree_skb(ack_skb);
+			printk(KERN_ERR "store ack fail %d\n", id);
 		}
 	}
 
