From 84373952feff6cc92669847a1360e79ad2ff97b1 Mon Sep 17 00:00:00 2001
From: Balamurugan Mahalingam <quic_bmahalin@quicinc.com>
Date: Thu, 17 Mar 2022 18:21:22 -0700
Subject: [PATCH 1/2] Add support to display EHT rx rate info in iw station

Change-Id: I0e3ca7ce694e7054b72d33eb834027a7986d2219
Signed-off-by: Balamurugan Mahalingam <quic_bmahalin@quicinc.com>

--- a/net/mac80211/sta_info.c
+++ b/net/mac80211/sta_info.c
@@ -2473,6 +2473,7 @@ static void sta_stats_decode_rate(struct
 		rinfo->nss = STA_STATS_GET(EHT_NSS, rate);
 		rinfo->eht_gi = STA_STATS_GET(EHT_GI, rate);
 		rinfo->eht_ru_alloc = STA_STATS_GET(EHT_RU, rate);
+		rinfo->he_dcm = STA_STATS_GET(HE_DCM, rate);
 		break;
 	}
 }
--- a/include/net/mac80211.h
+++ b/include/net/mac80211.h
@@ -1545,6 +1545,7 @@ enum mac80211_rx_encoding {
 	RX_ENC_HT,
 	RX_ENC_VHT,
 	RX_ENC_HE,
+	RX_ENC_S1G,
 	RX_ENC_EHT,
 };
 
