From c0685a8836b5075a51877342386d288f98a9eac9 Mon Sep 17 00:00:00 2001
From: Hari Chandrakanthan <quic_haric@quicinc.com>
Date: Wed, 14 Aug 2024 15:50:15 +0530
Subject: [PATCH] QSDK: wifi: mac80211: add Kconfig for ATHMEMDEBUG infra

Signed-off-by: Hari Chandrakanthan <quic_haric@quicinc.com>
Signed-off-by: Raj Kumar Bhagat <quic_rajkbhag@quicinc.com>
---
 local-symbols        | 1 +
 net/mac80211/Kconfig | 9 +++++++++
 2 files changed, 10 insertions(+)

diff --git a/local-symbols b/local-symbols
index 594a989..d8e7731 100644
--- a/local-symbols
+++ b/local-symbols
@@ -69,6 +69,7 @@ MAC80211_NSS_SUPPORT=
 MAC80211_PPE_SUPPORT=
 MAC80211_DS_SUPPORT=
 MAC80211_SFE_SUPPORT=
+MAC80211_ATHMEMDEBUG=
 QCOM_AOSS_QMP=
 QCOM_COMMAND_DB=
 QCOM_CPR=
diff --git a/net/mac80211/Kconfig b/net/mac80211/Kconfig
index f277c5d..d31109d 100644
--- a/net/mac80211/Kconfig
+++ b/net/mac80211/Kconfig
@@ -232,6 +232,15 @@ config MAC80211_PS_DEBUG
 
 	  Do not select this option.
 
+config MAC80211_ATHMEMDEBUG
+	bool "ATH memory debugging"
+	depends on ATHMEMDEBUG
+	default n
+	help
+	  Select this option to enable a debug build. This debug
+	  build helps to track memory allocation, deallocation
+	  for ATH related sub-system (cfg/mac80211, ath).
+
 config MAC80211_MPL_DEBUG
 	bool "Verbose mesh peer link debugging"
 	depends on MAC80211_DEBUG_MENU
-- 
2.34.1

