--- a/arch/x86/Kconfig
+++ b/arch/x86/Kconfig
@@ -1508,7 +1508,6 @@
 config X86_DIRECT_GBPAGES
 	def_bool y
 	depends on X86_64
-	depends on COMPILE_TEST || !CPU_MITIGATIONS # wait for LASS
 	help
 	  Certain kernel features effectively disable kernel
 	  linear 1 GB mappings (even if the CPU otherwise
--- a/drivers/clk/qcom/clk-alpha-pll.c
+++ b/drivers/clk/qcom/clk-alpha-pll.c
@@ -217,16 +217,6 @@
 		[PLL_OFF_TEST_CTL_U] = 0x34,
 		[PLL_OFF_STATUS] = 0x28,
 	},
-	[CLK_ALPHA_PLL_TYPE_NSS_HUAYRA] =  {
-		[PLL_OFF_L_VAL] = 0x04,
-		[PLL_OFF_ALPHA_VAL] = 0x08,
-		[PLL_OFF_TEST_CTL] = 0x0c,
-		[PLL_OFF_TEST_CTL_U] = 0x10,
-		[PLL_OFF_USER_CTL] = 0x14,
-		[PLL_OFF_CONFIG_CTL] = 0x18,
-		[PLL_OFF_CONFIG_CTL_U] = 0x1c,
-		[PLL_OFF_STATUS] = 0x20,
-	},
 	[CLK_ALPHA_PLL_TYPE_STROMER_PLUS] =  {
 		[PLL_OFF_L_VAL] = 0x04,
 		[PLL_OFF_USER_CTL] = 0x08,
--- a/drivers/clk/qcom/clk-alpha-pll.h
+++ b/drivers/clk/qcom/clk-alpha-pll.h
@@ -28,7 +28,6 @@
 	CLK_ALPHA_PLL_TYPE_DEFAULT_EVO,
 	CLK_ALPHA_PLL_TYPE_BRAMMO_EVO,
 	CLK_ALPHA_PLL_TYPE_STROMER,
-	CLK_ALPHA_PLL_TYPE_NSS_HUAYRA,
 	CLK_ALPHA_PLL_TYPE_STROMER_PLUS,
 	CLK_ALPHA_PLL_TYPE_NSS_HUAYRA,
 	CLK_ALPHA_PLL_TYPE_MAX,
--- a/drivers/clk/qcom/clk-rcg2.c
+++ b/drivers/clk/qcom/clk-rcg2.c
@@ -1083,16 +1075,10 @@
 	struct clk_rcg2 *rcg = to_clk_rcg2(hw);
 	const struct freq_tbl *f;
 
-	switch (policy) {
-	case FLOOR:
-		f = qcom_find_freq_floor(rcg->freq_tbl, rate);
-		break;
-	case CEIL:
-		f = qcom_find_freq(rcg->freq_tbl, rate);
-		break;
-	default:
+	f = clk_rcg2_find_best_freq(hw, rcg->freq_tbl, rate, CEIL);
+
+	if (!f)
 		return -EINVAL;
-	}
 
 	/*
 	 * In case clock is disabled, update the M, N and D registers, cache
--- a/drivers/clk/qcom/gcc-ipq5332.c
+++ b/drivers/clk/qcom/gcc-ipq5332.c
@@ -2710,7 +2710,6 @@
 	[GCC_Q6_AXIM_CLK_SRC] = &gcc_q6_axim_clk_src.clkr,
 #ifdef CONFIG_CORESIGHT
 	[GCC_QDSS_AT_CLK_SRC] = &gcc_qdss_at_clk_src.clkr,
-	[GCC_QDSS_TSCTR_CLK_SRC] = &gcc_qdss_tsctr_clk_src.clkr,
 #endif
 	[GCC_QPIC_AHB_CLK] = &gcc_qpic_ahb_clk.clkr,
 	[GCC_QPIC_CLK] = &gcc_qpic_clk.clkr,
@@ -2724,15 +2723,15 @@
 	[GCC_SNOC_LPASS_CFG_CLK] = &gcc_snoc_lpass_cfg_clk.clkr,
 	[GCC_SNOC_NSSNOC_1_CLK] = &gcc_snoc_nssnoc_1_clk.clkr,
 	[GCC_SNOC_NSSNOC_CLK] = &gcc_snoc_nssnoc_clk.clkr,
+#ifdef CONFIG_PCI
 	[GCC_SNOC_PCIE3_1LANE_1_M_CLK] = &gcc_snoc_pcie3_1lane_1_m_clk.clkr,
 	[GCC_SNOC_PCIE3_1LANE_1_S_CLK] = &gcc_snoc_pcie3_1lane_1_s_clk.clkr,
 	[GCC_SNOC_PCIE3_1LANE_M_CLK] = &gcc_snoc_pcie3_1lane_m_clk.clkr,
 	[GCC_SNOC_PCIE3_1LANE_S_CLK] = &gcc_snoc_pcie3_1lane_s_clk.clkr,
 	[GCC_SNOC_PCIE3_2LANE_M_CLK] = &gcc_snoc_pcie3_2lane_m_clk.clkr,
 	[GCC_SNOC_PCIE3_2LANE_S_CLK] = &gcc_snoc_pcie3_2lane_s_clk.clkr,
+#endif
 	[GCC_SNOC_USB_CLK] = &gcc_snoc_usb_clk.clkr,
-	[GCC_SYS_NOC_AT_CLK] = &gcc_sys_noc_at_clk.clkr,
-	[GCC_SYS_NOC_WCSS_AHB_CLK] = &gcc_sys_noc_wcss_ahb_clk.clkr,
 	[GCC_SYSTEM_NOC_BFDCD_CLK_SRC] = &gcc_system_noc_bfdcd_clk_src.clkr,
 	[GCC_UNIPHY0_AHB_CLK] = &gcc_uniphy0_ahb_clk.clkr,
 	[GCC_UNIPHY0_SYS_CLK] = &gcc_uniphy0_sys_clk.clkr,
@@ -2753,26 +2752,16 @@
 	[GCC_USB0_PIPE_CLK] = &gcc_usb0_pipe_clk.clkr,
 	[GCC_USB0_SLEEP_CLK] = &gcc_usb0_sleep_clk.clkr,
 	[GCC_WCSS_AHB_CLK_SRC] = &gcc_wcss_ahb_clk_src.clkr,
-	[GCC_WCSS_AXIM_CLK] = &gcc_wcss_axim_clk.clkr,
-	[GCC_WCSS_AXIS_CLK] = &gcc_wcss_axis_clk.clkr,
-	[GCC_WCSS_DBG_IFC_APB_BDG_CLK] = &gcc_wcss_dbg_ifc_apb_bdg_clk.clkr,
-	[GCC_WCSS_DBG_IFC_APB_CLK] = &gcc_wcss_dbg_ifc_apb_clk.clkr,
-	[GCC_WCSS_DBG_IFC_ATB_BDG_CLK] = &gcc_wcss_dbg_ifc_atb_bdg_clk.clkr,
-	[GCC_WCSS_DBG_IFC_ATB_CLK] = &gcc_wcss_dbg_ifc_atb_clk.clkr,
-	[GCC_WCSS_DBG_IFC_NTS_BDG_CLK] = &gcc_wcss_dbg_ifc_nts_bdg_clk.clkr,
-	[GCC_WCSS_DBG_IFC_NTS_CLK] = &gcc_wcss_dbg_ifc_nts_clk.clkr,
-	[GCC_WCSS_ECAHB_CLK] = &gcc_wcss_ecahb_clk.clkr,
-	[GCC_WCSS_MST_ASYNC_BDG_CLK] = &gcc_wcss_mst_async_bdg_clk.clkr,
-	[GCC_WCSS_SLV_ASYNC_BDG_CLK] = &gcc_wcss_slv_async_bdg_clk.clkr,
-	[GCC_XO_CLK] = &gcc_xo_clk.clkr,
 	[GCC_XO_CLK_SRC] = &gcc_xo_clk_src.clkr,
 	[GCC_XO_DIV4_CLK] = &gcc_xo_div4_clk.clkr,
 	[GCC_IM_SLEEP_CLK] = &gcc_im_sleep_clk.clkr,
 	[GCC_NSSNOC_PCNOC_1_CLK] = &gcc_nssnoc_pcnoc_1_clk.clkr,
 	[GCC_SNOC_QOSGEN_EXTREF_DIV_CLK_SRC] = &gcc_snoc_qosgen_extref_div_clk_src.clkr,
+#ifdef CONFIG_PCI
 	[GCC_PCIE3X2_PIPE_CLK_SRC] = &gcc_pcie3x2_pipe_clk_src.clkr,
 	[GCC_PCIE3X1_0_PIPE_CLK_SRC] = &gcc_pcie3x1_0_pipe_clk_src.clkr,
 	[GCC_PCIE3X1_1_PIPE_CLK_SRC] = &gcc_pcie3x1_1_pipe_clk_src.clkr,
+#endif
 	[GCC_USB0_PIPE_CLK_SRC] = &gcc_usb0_pipe_clk_src.clkr,
 };
 
@@ -2866,6 +2855,7 @@
 	[GCC_Q6SS_ATBM_CLK_ARES] = { 0x2501c, 2 },
 	[GCC_Q6SS_PCLKDBG_CLK_ARES] = { 0x25024, 2 },
 	[GCC_Q6SS_TRIG_CLK_ARES] = { 0x250a0, 2 },
+#ifdef CONFIG_CORESIGHT
 	[GCC_QDSS_APB2JTAG_CLK_ARES] = { 0x2d060, 2 },
 	[GCC_QDSS_AT_CLK_ARES] = { 0x2d038, 2 },
 	[GCC_QDSS_BCR] = { 0x2d000 },
@@ -2882,6 +2872,7 @@
 	[GCC_QDSS_TSCTR_DIV3_CLK_ARES] = { 0x2d04c, 2 },
 	[GCC_QDSS_TSCTR_DIV4_CLK_ARES] = { 0x2d050, 2 },
 	[GCC_QDSS_TSCTR_DIV8_CLK_ARES] = { 0x2d054, 2 },
+#endif
 	[GCC_QPIC_AHB_CLK_ARES] = { 0x32010, 2 },
 	[GCC_QPIC_CLK_ARES] = { 0x32014, 2 },
 	[GCC_QPIC_BCR] = { 0x32000 },
@@ -2895,7 +2886,9 @@
 	[GCC_SNOC_LPASS_CFG_CLK_ARES] = { 0x2e028, 2 },
 	[GCC_SNOC_NSSNOC_1_CLK_ARES] = { 0x17090, 2 },
 	[GCC_SNOC_NSSNOC_CLK_ARES] = { 0x17084, 2 },
+#ifdef CONFIG_CORESIGHT
 	[GCC_SYS_NOC_QDSS_STM_AXI_CLK_ARES] = { 0x2e034, 2 },
+#endif
 	[GCC_SYS_NOC_WCSS_AHB_CLK_ARES] = { 0x2e030, 2 },
 	[GCC_UNIPHY0_AHB_CLK_ARES] = { 0x16010, 2 },
 	[GCC_UNIPHY0_BCR] = { 0x16000 },
@@ -2964,11 +2957,6 @@
 	&gpll0_div2.hw,
 	&gcc_xo_div4_clk_src.hw,
 	&gcc_system_noc_bfdcd_div2_clk_src.hw,
-	&gcc_qdss_tsctr_div2_clk_src.hw,
-	&gcc_qdss_tsctr_div3_clk_src.hw,
-	&gcc_qdss_tsctr_div4_clk_src.hw,
-	&gcc_qdss_tsctr_div8_clk_src.hw,
-	&gcc_qdss_tsctr_div16_clk_src.hw,
 	&gcc_eud_at_div_clk_src.hw,
 };
 
--- a/drivers/iio/pressure/bmp280-core.c
+++ b/drivers/iio/pressure/bmp280-core.c
@@ -795,7 +795,7 @@
 
 static const int bmp280_oversampling_avail[] = { 1, 2, 4, 8, 16 };
 static const int bmp390_oversampling_avail[] = { 1, 2, 4, 8, 16, 32};
-static const u8 bmp280_chip_ids[] = { BMP280_CHIP_ID };
+static const u16 bmp280_chip_ids[] = { BMP280_CHIP_ID };
 
 const struct bmp280_chip_info bmp280_chip_info = {
 	.id_reg = BMP280_REG_ID,
@@ -848,7 +848,7 @@
 	return bmp280_chip_config(data);
 }
 
-static const u8 bme280_chip_ids[] = { BME280_CHIP_ID };
+static const u16 bme280_chip_ids[] = { BME280_CHIP_ID };
 
 const struct bmp280_chip_info bme280_chip_info = {
 	.id_reg = BMP280_REG_ID,
@@ -1229,7 +1229,7 @@
 
 static const int bmp380_oversampling_avail[] = { 1, 2, 4, 8, 16, 32 };
 static const int bmp380_iir_filter_coeffs_avail[] = { 1, 2, 4, 8, 16, 32, 64, 128};
-static const u8 bmp380_chip_ids[] = { BMP380_CHIP_ID };
+static const u16 bmp380_chip_ids[] = { BMP380_CHIP_ID };
 
 const struct bmp280_chip_info bmp380_chip_info = {
 	.id_reg = BMP380_REG_ID,
@@ -1738,7 +1738,7 @@
 }
 
 static const int bmp580_oversampling_avail[] = { 1, 2, 4, 8, 16, 32, 64, 128 };
-static const u8 bmp580_chip_ids[] = { BMP580_CHIP_ID, BMP580_CHIP_ID_ALT };
+static const u16 bmp580_chip_ids[] = { BMP580_CHIP_ID, BMP580_CHIP_ID_ALT };
 
 const struct bmp280_chip_info bmp580_chip_info = {
 	.id_reg = BMP580_REG_CHIP_ID,
@@ -2003,7 +2003,7 @@
 
 static const int bmp180_oversampling_temp_avail[] = { 1 };
 static const int bmp180_oversampling_press_avail[] = { 1, 2, 4, 8 };
-static const u8 bmp180_chip_ids[] = { BMP180_CHIP_ID };
+static const u16 bmp180_chip_ids[] = { BMP180_CHIP_ID };
 
 const struct bmp280_chip_info bmp180_chip_info = {
 	.id_reg = BMP280_REG_ID,
@@ -2144,7 +2144,7 @@
 	partial_data2 = (u64)(calib->T2 * partial_data1);
 	partial_data3 = (u64)(partial_data1 * partial_data1);
 	partial_data4 = (s64)(((s64)partial_data3) * ((s64)calib->T3));
-	partial_data5 = ((s64)(((s64)partial_data2) * 262144)  (s64)partial_data4);
+	partial_data5 = ((s64)(((s64)partial_data2) * 262144) + (s64)partial_data4);
 	partial_data6 = (s64)(((s64)partial_data5) / 4294967296U);
 
 	/*Update the compensated temperature in calib structure since this is needed
@@ -2175,20 +2175,20 @@
 	partial_data4 = (calib->P8 * partial_data3) / 32;
 	partial_data5 = (calib->P7 * partial_data1) * 16;
 	partial_data6 = (calib->P6 * data->t_fine) * 4194304;
-	offset = (s64)((s64)(calib->P5) * (s64)140737488355328U)  partial_data4  partial_data5  partial_data6;
+	offset = (s64)((s64)(calib->P5) * (s64)140737488355328U) + partial_data4 + partial_data5 + partial_data6;
 	partial_data2 = (((s64)calib->P4) * partial_data3) / 32;
 	partial_data4 = (calib->P3 * partial_data1) * 4;
 	partial_data5 = ((s64)(calib->P2) - 16384) * ((s64)data->t_fine) * 2097152;
-	sensitivity = (((s64)(calib->P1) - 16384) * (s64)70368744177664U)  partial_data2  partial_data4  partial_data5;
+	sensitivity = (((s64)(calib->P1) - 16384) * (s64)70368744177664U) + partial_data2 + partial_data4 + partial_data5;
 	partial_data1 = (sensitivity / 16777216) * uncomp_press;
 	partial_data2 = (s64)(calib->P10) * (s64)data->t_fine;
-	partial_data3 = partial_data2  (65536 * (s64)(calib->P9));
+	partial_data3 = partial_data2 + (65536 * (s64)(calib->P9));
 	partial_data4 = (partial_data3 * uncomp_press) / 8192;
 	partial_data5 = (uncomp_press * partial_data4) / 512;
 	partial_data6 = (s64)((u64)uncomp_press * (u64)uncomp_press);
 	partial_data2 = ((s64)(calib->P11) * (s64)(partial_data6)) / 65536;
 	partial_data3 = (partial_data2 * uncomp_press) / 128;
-	partial_data4 = (offset / 4)  partial_data1  partial_data5  partial_data3;
+	partial_data4 = (offset / 4) + partial_data1 + partial_data5 + partial_data3;
 	comp_press = (((u64)partial_data4 * 25) / (u64)1099511627776U);
 
 	return comp_press;
@@ -2361,7 +2361,7 @@
 			return ret;
 		}
 		/* Wait for 2ms */
-		usleep_range(data->start_up_time, data->start_up_time  100);
+		usleep_range(data->start_up_time, data->start_up_time + 100);
 
 		/* Read for Command error status */
 		ret = regmap_read(data->regmap, BMP390_REG_ERR, &cmd_err_status);
@@ -2379,11 +2379,14 @@
 	return 0;
 }
 
+static const u16 bmp390_chip_ids[] = { BMP390_CHIP_ID };
+
 const struct bmp280_chip_info bmp390_chip_info = {
 	.channels = bmp280_channels,
 	.num_channels = 2,
 	.id_reg = BMP390_REG_ID,
-	.chip_id = BMP390_CHIP_ID,
+	.chip_id = bmp390_chip_ids,
+	.num_chip_id = ARRAY_SIZE(bmp390_chip_ids),
 	.regmap_config = &bmp390_regmap_config,
 	.start_up_time = 2000,
 	.oversampling_temp_avail = bmp390_oversampling_avail,
@@ -2490,26 +2493,9 @@
 
 	data->regmap = regmap;
 
-	/* BMP3xx requires soft-reset as part of initialization */
-	if (data->chip_info->chip_id == BMP380_CHIP_ID) {
-		ret = bmp380_cmd(data, BMP380_CMD_SOFT_RESET);
-		if (ret < 0)
-			return ret;
-	}
-	else if (data->chip_info->chip_id == BMP390_CHIP_ID) {
-		ret = regmap_read(regmap, BMP390_REG_ID, &chip_id);
-		if (ret < 0)
-			return ret;
-		ret = regmap_read(regmap, BMP390_REG_REV, &rev_id);
-		if (ret < 0)
-			return ret;
-		pr_info("CHIP_ID: %x, REV_ID: %x\n", chip_id, rev_id);
-		/* Concatenating the chip identification code and mask revision of the ASIC
-		 * as BME280 and BMP390 share the same chip id (0x60)*/
-		chip_id = ((chip_id << 8) | rev_id);
-		/* Trigger soft reset */
-		ret = bmp390_soft_reset(data);
-	}
+	ret = regmap_read(regmap, data->chip_info->id_reg, &chip_id);
+	if (ret < 0)
+		return ret;
 
 	for (i = 0; i < data->chip_info->num_chip_id; i++) {
 		if (chip_id == data->chip_info->chip_id[i]) {
@@ -2521,6 +2507,29 @@
 	if (i == data->chip_info->num_chip_id)
 		dev_warn(dev, "bad chip id: 0x%x is not a known chip id\n", chip_id);
 
+	for (i = 0; i < data->chip_info->num_chip_id; i++) {
+		/* BMP3xx requires soft-reset as part of initialization */
+		if (data->chip_info->chip_id[i] == BMP380_CHIP_ID) {
+			ret = bmp380_cmd(data, BMP380_CMD_SOFT_RESET);
+			if (ret < 0)
+				return ret;
+		}
+		else if (data->chip_info->chip_id[i] == BMP390_CHIP_ID) {
+			ret = regmap_read(regmap, BMP390_REG_ID, &chip_id);
+			if (ret < 0)
+				return ret;
+			ret = regmap_read(regmap, BMP390_REG_REV, &rev_id);
+			if (ret < 0)
+				return ret;
+			pr_info("CHIP_ID: %x, REV_ID: %x\n", chip_id, rev_id);
+			/* Concatenating the chip identification code and mask revision of the ASIC
+			 * as BME280 and BMP390 share the same chip id (0x60)*/
+			chip_id = ((chip_id << 8) | rev_id);
+			/* Trigger soft reset */
+			ret = bmp390_soft_reset(data);
+		}
+	}
+
 	if (data->chip_info->preinit) {
 		ret = data->chip_info->preinit(data);
 		if (ret)
--- a/drivers/iio/pressure/bmp280.h
+++ b/drivers/iio/pressure/bmp280.h
@@ -353,8 +353,8 @@
  * CHIP_ID: 0x6001
  */
 #define BMP390_CHIP_ID_VALUE		0x60
-#define BMP390_REV_ID				0x01
-#define BMP390_CHIP_ID				((BMP390_CHIP_ID_VALUE << 8) | (BMP390_REV_ID))
+#define BMP390_REV_ID			0x01
+#define BMP390_CHIP_ID			((BMP390_CHIP_ID_VALUE << 8) | (BMP390_REV_ID))
 #define BMP280_SOFT_RESET_VAL		0xB6
 #define BMP390_SOFT_RESET_VAL		0xB6
 
@@ -500,7 +500,7 @@
 
 struct bmp280_chip_info {
 	unsigned int id_reg;
-	const u8 *chip_id;
+	const u16 *chip_id;
 	int num_chip_id;
 
 	const struct regmap_config *regmap_config;
--- a/drivers/mmc/host/sdhci-msm.c
+++ b/drivers/mmc/host/sdhci-msm.c
@@ -1924,17 +1924,15 @@
 	struct sdhci_msm_host *msm_host = sdhci_pltfm_priv(pltfm_host);
 	union cqhci_crypto_cap_entry cap;
 
-	if (!(cfg->config_enable & CQHCI_CRYPTO_CONFIGURATION_ENABLE))
-		return qcom_ice_evict_key(msm_host->ice, slot);
-
 	cap = cq_host->crypto_cap_array[cfg->crypto_cap_idx];
 
-	return qcom_ice_program_key(msm_host->ice,
-					cap.algorithm_id,
-					cap.key_size,
-				    cfg->crypto_key,
-				    cfg->data_unit_size, slot,
-					cq_host->use_hwkey);
+	if (cfg->config_enable & CQHCI_CRYPTO_CONFIGURATION_ENABLE)
+		return qcom_ice_program_key(msm_host->ice,
+					    cap.algorithm_id,
+					    cap.key_size,
+					    cfg->crypto_key,
+					    cfg->data_unit_size, slot,
+					    cq_host->use_hwkey);
 	else if (!cq_host->use_hwkey)
 		return qcom_ice_evict_key(msm_host->ice, slot);
 	else
--- a/drivers/net/mdio/mdio-ipq4019.c
+++ b/drivers/net/mdio/mdio-ipq4019.c
@@ -37,8 +37,8 @@
 
 /* MDIO clock source frequency is fixed to 100M */
 #define IPQ_MDIO_CLK_RATE	100000000
-#define IPQ_UNIPHY_AHB_CLK_RATE		100000000
-#define IPQ_UNIPHY_SYS_CLK_RATE		24000000
+#define IPQ_UNIPHY_AHB_CLK_RATE	100000000
+#define IPQ_UNIPHY_SYS_CLK_RATE	24000000
 
 #define IPQ_PHY_SET_DELAY_US	100000
 
@@ -70,7 +70,7 @@
 #define PHY_ICC_EFUSE_REG			0x280
 #define PHY_10BT_SG_THRESH_REG			0x3380
 #define PHY_MMD1_CTRL2ANA_OPTION2_REG		0x40018102
-#define ETH_LDO_RDY_CNT					3
+#define ETH_LDO_RDY_CNT				3
 
 #define CMN_PLL_REFCLK_INDEX			GENMASK(3, 0)
 #define CMN_PLL_REFCLK_EXTERNAL			BIT(9)
@@ -123,6 +123,7 @@
 	"gcc_mdio_ahb_clk", "uniphy0_ahb_clk", "uniphy0_sys_clk",
 	"uniphy1_ahb_clk", "uniphy1_sys_clk",
 	"uniphy2_ahb_clk", "uniphy2_sys_clk"
+};
 
 static int ipq4019_mdio_wait_busy(struct mii_bus *bus)
 {
@@ -874,8 +875,8 @@
 	ipq_cmn_clk_reset(bus);
 
 	/* For the platform ipq5332, the uniphy clock should be configured for resetting
-     * the connected device such as qca8386 switch or qca8081 PHY.
-     */
+	 * the connected device such as qca8386 switch or qca8081 PHY.
+	 */
 	if (of_machine_is_compatible("qcom,ipq5332")) {
 		unsigned long rate = 0;
 
@@ -901,38 +902,36 @@
 			ret = clk_prepare_enable(priv->clk[i]);
 		}
 	}
-		fsleep(IPQ_PHY_SET_DELAY_US);
-	}
 
 	/* To indicate CMN_PLL that ethernet_ldo has been ready if the additional
-     * platform resources are specified in the device tree.
-     */
-    for (i = 0; i < ETH_LDO_RDY_CNT; i++) {
-        if (priv->eth_ldo_rdy[i]) {
-            val = readl(priv->eth_ldo_rdy[i]);
-            val |= BIT(0);
-            writel(val, priv->eth_ldo_rdy[i]);
-            fsleep(IPQ_PHY_SET_DELAY_US);
-        }
-    }
-
-    /* Do the optional reset on the devices connected with MDIO bus */
-    if (priv->reset_gpios) {
-        unsigned long *values = bitmap_zalloc(priv->reset_gpios->ndescs, GFP_KERNEL);
-        if (!values)
-            return -ENOMEM;
-
-        bitmap_fill(values, priv->reset_gpios->ndescs);
-        gpiod_set_array_value_cansleep(priv->reset_gpios->ndescs, priv->reset_gpios->desc,
-                priv->reset_gpios->info, values);
+	 * platform resources are specified in the device tree.
+	 */
+	for (i = 0; i < ETH_LDO_RDY_CNT; i++) {
+		if (priv->eth_ldo_rdy[i]) {
+			val = readl(priv->eth_ldo_rdy[i]);
+			val |= BIT(0);
+			writel(val, priv->eth_ldo_rdy[i]);
+			fsleep(IPQ_PHY_SET_DELAY_US);
+		}
+	}
+
+	/* Do the optional reset on the devices connected with MDIO bus */
+	if (priv->reset_gpios) {
+		unsigned long *values = bitmap_zalloc(priv->reset_gpios->ndescs, GFP_KERNEL);
+		if (!values)
+			return -ENOMEM;
+
+		bitmap_fill(values, priv->reset_gpios->ndescs);
+		gpiod_set_array_value_cansleep(priv->reset_gpios->ndescs, priv->reset_gpios->desc,
+				priv->reset_gpios->info, values);
 
 		fsleep(IPQ_PHY_SET_DELAY_US);
 
 		bitmap_zero(values, priv->reset_gpios->ndescs);
-        gpiod_set_array_value_cansleep(priv->reset_gpios->ndescs, priv->reset_gpios->desc,
-                priv->reset_gpios->info, values);
-        bitmap_free(values);
-    }
+		gpiod_set_array_value_cansleep(priv->reset_gpios->ndescs, priv->reset_gpios->desc,
+				priv->reset_gpios->info, values);
+		bitmap_free(values);
+	}
 
 	/* Configure MDIO clock source frequency if clock is specified in the device tree */
 	ret = clk_set_rate(priv->clk[MDIO_CLK_MDIO_AHB], IPQ_MDIO_CLK_RATE);
@@ -997,8 +996,8 @@
 	priv->reset_gpios= devm_gpiod_get_array_optional(&pdev->dev, "phy-reset", GPIOD_OUT_LOW);
 	if (IS_ERR(priv->reset_gpios)) {
 		ret = dev_err_probe(&pdev->dev, PTR_ERR(priv->reset_gpios),
-					"mii_bus %s couldn't get reset GPIO\n",
-					bus->id);
+				    "mii_bus %s couldn't get reset GPIO\n",
+				    bus->id);
 		return ret;
 	}
 
--- a/drivers/pci/controller/dwc/pcie-qcom.c
+++ b/drivers/pci/controller/dwc/pcie-qcom.c
@@ -2020,7 +2020,6 @@
 	{ .compatible = "qcom,pcie-ipq8074", .data = &cfg_2_3_3 },
 	{ .compatible = "qcom,pcie-ipq8074-gen3", .data = &cfg_2_9_0 },
 	{ .compatible = "qti,pcie-ipq9574", .data = &cfg_1_27_0 },
-	{ .compatible = "qcom,pcie-ipq9574", .data = &cfg_2_9_0 },
 	{ .compatible = "qcom,pcie-msm8996", .data = &cfg_2_3_2 },
 	{ .compatible = "qcom,pcie-qcs404", .data = &cfg_2_4_0 },
 	{ .compatible = "qcom,pcie-sa8540p", .data = &cfg_1_9_0 },
--- a/drivers/scsi/sd.c
+++ b/drivers/scsi/sd.c
@@ -3185,7 +3185,7 @@
 	rcu_read_lock();
 	vpd = rcu_dereference(sdkp->device->vpd_pgb2);
 
-	if (!vpd || vpd->len <= 8) {
+	if (!vpd || vpd->len < 8) {
 		rcu_read_unlock();
 		return;
 	}
--- a/drivers/soc/qcom/ice.c
+++ b/drivers/soc/qcom/ice.c
@@ -213,8 +213,8 @@
 }
 
 static int qcom_ice_get_algo_mode(struct qcom_ice *ice, u8 algorithm_id,
-				 u8 key_size, enum qcom_scm_ice_cipher *cipher,
-				 u32 *key_len, bool use_hwkey)
+				  u8 key_size, enum qcom_scm_ice_cipher *cipher,
+				  u32 *key_len, bool use_hwkey)
 {
 	struct device *dev = ice->dev;
 
@@ -229,24 +229,24 @@
 	}
 
 	switch (algorithm_id) {
-    case QCOM_ICE_CRYPTO_ALG_AES_XTS:
-        if (key_size == QCOM_ICE_CRYPTO_KEY_SIZE_256) {
-            *cipher = QCOM_SCM_ICE_CIPHER_AES_256_XTS;
-            *key_len = AES_256_XTS_KEY_SIZE;
-        } else {
-            *cipher = QCOM_SCM_ICE_CIPHER_AES_128_XTS;
-            *key_len = AES_128_XTS_KEY_SIZE;
-        }
-        break;
+	case QCOM_ICE_CRYPTO_ALG_AES_XTS:
+		if (key_size == QCOM_ICE_CRYPTO_KEY_SIZE_256) {
+			*cipher = QCOM_SCM_ICE_CIPHER_AES_256_XTS;
+			*key_len = AES_256_XTS_KEY_SIZE;
+		} else {
+			*cipher = QCOM_SCM_ICE_CIPHER_AES_128_XTS;
+			*key_len = AES_128_XTS_KEY_SIZE;
+		}
+		break;
 	case QCOM_ICE_CRYPTO_ALG_BITLOCKER_AES_CBC:
-        if (key_size == QCOM_ICE_CRYPTO_KEY_SIZE_256) {
-            *cipher = QCOM_SCM_ICE_CIPHER_AES_256_CBC;
-            *key_len = AES_256_CBC_KEY_SIZE;
-        } else {
-            *cipher = QCOM_SCM_ICE_CIPHER_AES_128_CBC;
-            *key_len = AES_128_CBC_KEY_SIZE;
-        }
-        break;
+		if (key_size == QCOM_ICE_CRYPTO_KEY_SIZE_256) {
+			*cipher = QCOM_SCM_ICE_CIPHER_AES_256_CBC;
+			*key_len = AES_256_CBC_KEY_SIZE;
+		} else {
+			*cipher = QCOM_SCM_ICE_CIPHER_AES_128_CBC;
+			*key_len = AES_128_CBC_KEY_SIZE;
+		}
+		break;
 	case QCOM_ICE_CRYPTO_ALG_AES_ECB:
 		/* ECB mode only supports for HW key slot */
 		if (!use_hwkey) {
@@ -265,13 +265,13 @@
 		break;
 	default:
 		dev_err_ratelimited(dev, "Unhandled crypto capability; algorithm_id=%d, key_size=%d\n",
-					algorithm_id, key_size);
+				    algorithm_id, key_size);
 		return -EINVAL;
 	}
 
 	dev_info(dev, "cipher: %d key_size: %d", *cipher, *key_len);
 	return 0;
-}	
+}
 
 int qcom_ice_program_key(struct qcom_ice *ice,
 			 u8 algorithm_id, u8 key_size,
--- a/drivers/tty/serial/qcom_geni_serial.c
+++ b/drivers/tty/serial/qcom_geni_serial.c
@@ -145,9 +145,8 @@
 static struct uart_driver qcom_geni_console_driver;
 static struct uart_driver qcom_geni_uart_driver;
 
-static int qcom_geni_serial_port_setup(struct uart_port *uport);
-
 static void qcom_geni_serial_cancel_tx_cmd(struct uart_port *uport);
+static int qcom_geni_serial_port_setup(struct uart_port *uport);
 
 static inline struct qcom_geni_serial_port *to_dev_port(struct uart_port *uport)
 {
--- a/include/linux/ath9k_platform.h
+++ b/include/linux/ath9k_platform.h
@@ -46,6 +46,13 @@
 	int (*external_reset)(void);
 
 	bool use_eeprom;
+
+	int num_leds;
+	const struct gpio_led *leds;
+
+	unsigned num_btns;
+	const struct gpio_keys_button *btns;
+	unsigned btn_poll_interval;
 };
 
 #endif /* _LINUX_ATH9K_PLATFORM_H */
--- a/include/linux/compiler.h
+++ b/include/linux/compiler.h
@@ -202,6 +202,8 @@
 	__v;								\
 })
 
+#include <asm/rwonce.h>
+
 #endif /* __KERNEL__ */
 
 /**
@@ -258,6 +260,4 @@
  */
 #define prevent_tail_call_optimization()	mb()
 
-#include <asm/rwonce.h>
-
 #endif /* __LINUX_COMPILER_H */
--- a/net/core/sock.c
+++ b/net/core/sock.c
@@ -1615,10 +1615,20 @@
 			break;
 		}
 
+		if (val == sk->offload_appid) {
+			ret = -EPERM;
+			break;
+		}
+
 		if (val)
 			sk->offload_appid = val;
 
 		sk->offload = val ? 1 : 0;
+
+		mutex_lock(&sock_notifier.mutex);
+		ret = raw_notifier_call_chain(&sock_notifier.notify,
+			SOCK_OFFLOAD_NOTIFY, (void *)sk);
+		mutex_unlock(&sock_notifier.mutex);
 		break;
 
 	default:
--- a/net/ipv4/ipmr.c
+++ b/net/ipv4/ipmr.c
@@ -449,8 +449,8 @@
 
 	mrt->vif_table[vif].pkt_in += pkts_in;
 	mrt->vif_table[vif].bytes_in += bytes_in;
-	cache->_c.mfc_un.res.pkt  += pkts_out;
-	cache->_c.mfc_un.res.bytes += bytes_out;
+	atomic_long_add(pkts_out, &cache->_c.mfc_un.res.pkt);
+	atomic_long_add(bytes_out, &cache->_c.mfc_un.res.bytes);
 
 	for (vifi = cache->_c.mfc_un.res.minvif;
 			vifi < cache->_c.mfc_un.res.maxvif; vifi++) {
--- a/net/ipv6/ip6mr.c
+++ b/net/ipv6/ip6mr.c
@@ -915,8 +915,8 @@
 
 	mrt->vif_table[vif].pkt_in += pkts_in;
 	mrt->vif_table[vif].bytes_in += bytes_in;
-	cache->_c.mfc_un.res.pkt += pkts_out;
-	cache->_c.mfc_un.res.bytes += bytes_out;
+	atomic_long_add(pkts_out, &cache->_c.mfc_un.res.pkt);
+	atomic_long_add(bytes_out, &cache->_c.mfc_un.res.bytes);
 
 	for (vifi = cache->_c.mfc_un.res.minvif;
 			vifi < cache->_c.mfc_un.res.maxvif; vifi++) {
--- a/net/ipv6/netfilter/nf_reject_ipv6.c
+++ b/net/ipv6/netfilter/nf_reject_ipv6.c
@@ -269,6 +269,7 @@
 		    int hook)
 {
 	const struct ipv6hdr *oip6h = ipv6_hdr(oldskb);
+	struct ipv6hdr *ip6h;
 	struct dst_entry *dst = NULL;
 	const struct tcphdr *otcph;
 	struct sk_buff *nskb;
@@ -300,7 +301,15 @@
 		skb_dst_set(oldskb, dst);
 	}
 
-	fl6.flowi6_oif = l3mdev_master_ifindex(skb_dst(oldskb)->dev);
+	/* For forwarding packet, the skb->skb_iif is the incoming device's
+	 * ifindex, but it is 0 for local out skb, use dst->dev's ifindex
+	 * instead.
+	 */
+	if (oldskb->skb_iif != 0)
+		fl6.flowi6_oif = oldskb->skb_iif;
+	else
+		fl6.flowi6_oif = l3mdev_master_ifindex(skb_dst(oldskb)->dev);
+
 	fl6.flowi6_mark = IP6_REPLY_MARK(net, oldskb->mark);
 	security_skb_classify_flow(oldskb, flowi6_to_flowi_common(&fl6));
 	dst = ip6_route_output(net, NULL, &fl6);
@@ -327,7 +336,8 @@
 	nskb->mark = fl6.flowi6_mark;
 
 	skb_reserve(nskb, LL_MAX_HEADER);
-	nf_reject_ip6hdr_put(nskb, oldskb, IPPROTO_TCP, ip6_dst_hoplimit(dst));
+	ip6h = nf_reject_ip6hdr_put(nskb, oldskb, IPPROTO_TCP,
+				    ip6_dst_hoplimit(dst));
 	nf_reject_ip6_tcphdr_put(nskb, oldskb, otcph, otcplen);
 
 	nf_ct_attach(nskb, oldskb);
--- a/net/netfilter/xt_connmark.c
+++ b/net/netfilter/xt_connmark.c
@@ -50,7 +50,7 @@
 				dscp = ipv4_get_dsfield(ip_hdr(skb)) >> 2;
 			else if (skb->protocol == htons(ETH_P_IPV6))
 				dscp = ipv6_get_dsfield(ipv6_hdr(skb)) >> 2;
-			else
+			else	/* protocol doesn't have diffserv */
 				break;
 
 			newmark = (newmark & ~info->ctmark) |
@@ -99,7 +99,7 @@
 		.ctmask	= info->ctmask,
 		.nfmask	= info->nfmask,
 		.mode	= info->mode,
-		.func   = XT_CONNMARK_VALUE
+		.func	= XT_CONNMARK_VALUE
 	};
 
 	return connmark_tg_shift(skb, &info3);
@@ -110,11 +110,11 @@
 {
 	const struct xt_connmark_tginfo2 *info = par->targinfo;
 	const struct xt_connmark_tginfo3 info3 = {
-		.ctmark = info->ctmark,
-		.ctmask = info->ctmask,
-		.nfmask = info->nfmask,
-		.mode   = info->mode,
-		.func   = XT_CONNMARK_VALUE
+		.ctmark	= info->ctmark,
+		.ctmask	= info->ctmask,
+		.nfmask	= info->nfmask,
+		.mode	= info->mode,
+		.func	= XT_CONNMARK_VALUE
 	};
 
 	return connmark_tg_shift(skb, &info3);
@@ -198,7 +198,7 @@
 	{
 		.name           = "CONNMARK",
 		.revision       = 3,
-		.family         = NFPROTO_UNSPEC,
+		.family         = NFPROTO_IPV4,
 		.checkentry     = connmark_tg_check,
 		.target         = connmark_tg_v3,
 		.targetsize     = sizeof(struct xt_connmark_tginfo3),
@@ -226,6 +226,16 @@
 		.destroy        = connmark_tg_destroy,
 		.me             = THIS_MODULE,
 	},
+	{
+		.name           = "CONNMARK",
+		.revision       = 3,
+		.family         = NFPROTO_IPV6,
+		.checkentry     = connmark_tg_check,
+		.target         = connmark_tg_v3,
+		.targetsize     = sizeof(struct xt_connmark_tginfo3),
+		.destroy        = connmark_tg_destroy,
+		.me             = THIS_MODULE,
+	},
 #endif
 };
 
--- a/net/sched/sch_fq_codel.c
+++ b/net/sched/sch_fq_codel.c
@@ -304,20 +304,20 @@
 			    &flow->cvars, &q->cstats, qdisc_pkt_len,
 			    codel_get_enqueue_time, drop_func, dequeue_func);
 
-   /* If our qlen is 0 qdisc_tree_reduce_backlog() will deactivate
-    * parent class, dequeue in parent qdisc will do the same if we
-    * return skb. Temporary increment qlen if we have skb.
-    */
-   if (q->stats.drop_count) {
-       if (skb)
-           sch->q.qlen++;
-       qdisc_tree_reduce_backlog(sch, q->stats.drop_count,
-                     q->stats.drop_len);
-       if (skb)
-           sch->q.qlen--;
-       q->stats.drop_count = 0;
-       q->stats.drop_len = 0;
-   }
+	/* If our qlen is 0 qdisc_tree_reduce_backlog() will deactivate
+	 * parent class, dequeue in parent qdisc will do the same if we
+	 * return skb. Temporary increment qlen if we have skb.
+	 */
+	if (q->cstats.drop_count) {
+		if (skb)
+			sch->q.qlen++;
+		qdisc_tree_reduce_backlog(sch, q->cstats.drop_count,
+					  q->cstats.drop_len);
+		if (skb)
+			sch->q.qlen--;
+		q->cstats.drop_count = 0;
+		q->cstats.drop_len = 0;
+	}
 
 	if (!skb) {
 		/* force a pass through old_flows to prevent starvation */
--- a/include/uapi/linux/netfilter/xt_connmark.h
+++ b/include/uapi/linux/netfilter/xt_connmark.h
@@ -15,8 +15,8 @@
 };
 
 enum {
-	XT_CONNMARK_VALUE = (1 << 0),
-	XT_CONNMARK_DSCP =  (1 << 1)
+	XT_CONNMARK_VALUE =	(1 << 0),
+	XT_CONNMARK_DSCP = 	(1 << 1)
 };
 
 enum {
--- a/net/core/sock.c    2025-07-08 12:30:06.220379796 +0530
+++ b/net/core/sock.c 2025-07-08 12:31:42.656304940 +0530
@@ -1629,6 +1629,12 @@
		ret = raw_notifier_call_chain(&sock_notifier.notify,
			SOCK_OFFLOAD_NOTIFY, (void *)sk);
		mutex_unlock(&sock_notifier.mutex);
+
+		ret = (ret == NOTIFY_OK) ? 0 : -EOPNOTSUPP;
+		if (ret) {
+			sk->offload_appid = 0;
+			sk->offload = 0;
+		}
		break;

	default:

