From 55a52fc6e5020c5fe450c2e7ea68fee47a1a1b32 Mon Sep 17 00:00:00 2001
From: huangyunxiang <huangyunxiang@cigtech.com>
Date: Mon, 14 Apr 2025 14:59:15 +0800
Subject: [PATCH] add for hw watchdog

---
 drivers/misc/Makefile     | 1 +
 1 file changed, 1 insertion(+)

diff --git a/drivers/misc/Makefile b/drivers/misc/Makefile
index 35f9f79be..9862806c6 100644
--- a/drivers/misc/Makefile
+++ b/drivers/misc/Makefile
@@ -68,3 +68,4 @@ obj-$(CONFIG_TMR_INJECT)	+= xilinx_tmr_inject.o
 obj-$(CONFIG_TPS6594_ESM)	+= tps6594-esm.o
 obj-$(CONFIG_TPS6594_PFSM)	+= tps6594-pfsm.o
 obj-$(CONFIG_QSEECOM)		+= qseecom.o
+obj-y				+= hw_watchdog.o
-- 
2.34.1

