diff --git a/drivers/net/wireless/ath/ath12k/dp_tx.c b/drivers/net/wireless/ath/ath12k/dp_tx.c
index d6fac62..708951d 100644
--- a/drivers/net/wireless/ath/ath12k/dp_tx.c
+++ b/drivers/net/wireless/ath/ath12k/dp_tx.c
@@ -1400,6 +1400,9 @@ int ath12k_dp_tx_completion_handler(struct ath12k_base *ab, int ring_id,
 			continue;
 		}
 
+		prefetch(msdu);
+		prefetch( (u8 *) msdu + 64);
+
 		/* Find the HAL_WBM_RELEASE_INFO0_REL_SRC_MODULE value */
 		buf_rel_source = tx_status->info0 & HAL_WBM_RELEASE_INFO0_REL_SRC_MASK;
 		ab->soc_stats.tx_wbm_rel_source[buf_rel_source]++;
