From: Lorenzo Bianconi <lorenzo@kernel.org>
Date: Sat, 16 Mar 2024 12:35:24 +0100
Subject: [PATCH] hostapd: export hostapd_is_usable_chans utility routine

This is a preliminary patch to introduce AFC support.

Tested-by: Allen Ye <allen.ye@mediatek.com>
Signed-off-by: Lorenzo Bianconi <lorenzo@kernel.org>
---

--- a/src/ap/hw_features.c
+++ b/src/ap/hw_features.c
@@ -953,7 +953,7 @@ static int hostapd_is_usable_edmg(struct
 }
 
 
-static int hostapd_is_usable_chans(struct hostapd_iface *iface)
+int hostapd_is_usable_chans(struct hostapd_iface *iface)
 {
 	int secondary_freq;
 	struct hostapd_channel_data *pri_chan;
--- a/src/ap/hw_features.h
+++ b/src/ap/hw_features.h
@@ -28,6 +28,8 @@ int hostapd_prepare_rates(struct hostapd
 void hostapd_stop_setup_timers(struct hostapd_iface *iface);
 int hostapd_hw_skip_mode(struct hostapd_iface *iface,
 			 struct hostapd_hw_modes *mode);
+int hostapd_is_usable_chans(struct hostapd_iface *iface);
+
 #else /* NEED_AP_MLME */
 static inline void
 hostapd_free_hw_features(struct hostapd_hw_modes *hw_features,
@@ -91,6 +93,11 @@ static inline int hostapd_check_he_6ghz_
 	return 0;
 }
 
+static inline int hostapd_is_usable_chans(struct hostapd_iface *iface)
+{
+	return 1;
+}
+
 #endif /* NEED_AP_MLME */
 
 #endif /* HW_FEATURES_H */
