From f8e1949f3cfb29e5742c5156b6870e259eafefcd Mon Sep 17 00:00:00 2001
From: John Crispin <john@phrozen.org>
Date: Sat, 7 May 2022 19:48:29 +0200
Subject: [PATCH 01/11] rtkmipsel: add kernel-version

Signed-off-by: John Crispin <john@phrozen.org>
---
 include/kernel-version.mk | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/include/kernel-version.mk b/include/kernel-version.mk
index 40e6048..907fdab 100644
--- a/include/kernel-version.mk
+++ b/include/kernel-version.mk
@@ -6,6 +6,7 @@ ifdef CONFIG_TESTING_KERNEL
   KERNEL_PATCHVER:=$(KERNEL_TESTING_PATCHVER)
 endif
 
+LINUX_VERSION-3.18 = .29
 LINUX_VERSION-4.14 = .193
 ifeq ($(CONFIG_TARGET_ipq807x_ipq807x)$(CONFIG_TARGET_ipq807x_ipq60xx)$(CONFIG_TARGET_ipq807x_ipq50xx),y)
 LINUX_VERSION-5.4 = .164
@@ -14,6 +15,7 @@ LINUX_VERSION-5.4 = .158
 endif
 LINUX_VERSION-5.10 = .27
 
+LINUX_KERNEL_HASH-3.18.29 = d099072fcbd6aefea8c787c0325eaa56c65b35b15c52579c8a613349bcba425f
 LINUX_KERNEL_HASH-4.14.193 = 0b0fb41d4430e1a42738b341cbfd2f41951aa5cd02acabbd53f076119c8b9f03
 LINUX_KERNEL_HASH-5.4.158 = 6e018fecdc8fc24553756e582d83b82d65b10a6b03ef36262a24911f839b8d59
 LINUX_KERNEL_HASH-5.4.164 = 925f59a86b9549afd5e4b88b468789ffac3a5dc7b4a8a0c9ca718405675bd3e8
-- 
2.7.4

