From a53155dd79b1114533e30272ccbc216feaa8b755 Mon Sep 17 00:00:00 2001
From: John Crispin <john@phrozen.org>
Date: Wed, 3 Nov 2021 14:39:21 +0100
Subject: [PATCH 51/51] toolchain: do not enable mips16 on realtek AP kernel

Signed-off-by: John Crispin <john@phrozen.org>
---
 toolchain/Config.in | 1 +
 1 file changed, 1 insertion(+)

diff --git a/toolchain/Config.in b/toolchain/Config.in
index 6dda9af92d..b9638a1b69 100644
--- a/toolchain/Config.in
+++ b/toolchain/Config.in
@@ -28,6 +28,7 @@ menuconfig TARGET_OPTIONS
 	config USE_MIPS16
 		bool "Build packages with MIPS16 instructions" if TARGET_OPTIONS
 		depends on HAS_MIPS16
+		depends on !TARGET_rtkmipsel
 		default y
 		help
 		  If your target CPU does support the MIPS16 instruction set
-- 
2.25.1

