From aa6938c36045da952053965eefdb7a29049641e9 Mon Sep 17 00:00:00 2001
From: John Crispin <john@phrozen.org>
Date: Sun, 8 May 2022 07:03:08 +0200
Subject: [PATCH] rtkmipsel: select gcc-5 as the compiler

Signed-off-by: John Crispin <john@phrozen.org>
---
 toolchain/gcc/Config.in | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/toolchain/gcc/Config.in b/toolchain/gcc/Config.in
index 6625e02114..3ba07aaa63 100644
--- a/toolchain/gcc/Config.in
+++ b/toolchain/gcc/Config.in
@@ -4,6 +4,7 @@ choice
 	prompt "GCC compiler Version" if TOOLCHAINOPTS
 	default GCC_USE_VERSION_8
 	default GCC_USE_VERSION_5 if TARGET_ipq807x
+	default GCC_USE_VERSION_5 if TARGET_rtkmipsel
 	help
 	  Select the version of gcc you wish to use.
 
@@ -15,18 +16,22 @@ choice
 		bool "gcc 7.x"
 		depends on !arc
 		depends on !TARGET_ipq807x
+		depends on !TARGET_rtkmipsel
 
 	config GCC_USE_VERSION_8
 		bool "gcc 8.x"
 		depends on !TARGET_ipq807x
+		depends on !TARGET_rtkmipsel
 
 	config GCC_USE_VERSION_9
 		bool "gcc 9.x"
 		depends on !TARGET_ipq807x
+		depends on !TARGET_rtkmipsel
 
 	config GCC_USE_VERSION_10
 		bool "gcc 10.x"
 		depends on !TARGET_ipq807x
+		depends on !TARGET_rtkmipsel
 endchoice
 
 config GCC_USE_GRAPHITE
-- 
2.25.1

