Index: ucentralgw-2021-03-22-220dd3aa/CMakeLists.txt
===================================================================
--- ucentralgw-2021-03-22-220dd3aa.orig/CMakeLists.txt
+++ ucentralgw-2021-03-22-220dd3aa/CMakeLists.txt
@@ -54,6 +54,10 @@ add_executable( ucentral
                 src/RESTAPI_default_configuration.h src/RESTAPI_default_configurations.cpp src/RESTAPI_default_configurations.h src/RESTAPI_commands.cpp
                 src/RESTAPI_commands.h src/uCommandManager.cpp src/uCommandManager.h src/RESTAPI_command.cpp src/RESTAPI_command.h)
 
+INSTALL(TARGETS ucentral
+	RUNTIME DESTINATION /usr/bin
+)
+
 if(SMALL_BUILD)
 target_link_libraries(ucentral PRIVATE yaml-cpp
                         ${Poco_LIBRARIES} ${Boost_LIBRARIES} ${ZLIB_LIBRARIES})
@@ -61,4 +65,4 @@ else()
     target_link_libraries(ucentral PRIVATE yaml-cpp
             ${Poco_LIBRARIES} ${Boost_LIBRARIES} ${PostgreSQL_LIBRARIES}
             ${MySQL_LIBRARIES} ${ODBC_LIBRARIES} ${ZLIB_LIBRARIES})
-endif()
\ No newline at end of file
+endif()
