Index: ucentralgw-2021-03-28-555b5fef/cert_scripts/clean.sh
===================================================================
--- ucentralgw-2021-03-28-555b5fef.orig/cert_scripts/clean.sh
+++ ucentralgw-2021-03-28-555b5fef/cert_scripts/clean.sh
@@ -1,3 +1,3 @@
-#!/bin/bash
+#!/bin/sh
 
 rm *.pem *.csr
Index: ucentralgw-2021-03-28-555b5fef/cert_scripts/create_certificates.sh
===================================================================
--- ucentralgw-2021-03-28-555b5fef.orig/cert_scripts/create_certificates.sh
+++ ucentralgw-2021-03-28-555b5fef/cert_scripts/create_certificates.sh
@@ -1,15 +1,27 @@
-#!/bin/bash
+#!/bin/sh
 
-hn=$(hostname)
+mkdir -p /etc/ucentral/certs/cas
+
+hn=$(cat /proc/sys/kernel/hostname)
 howmany=10
-cert_life=365
-subject="/C=CA/ST=British Columbia/L=Vancouver/O=Arilia Wireless/OU=Engineering/CN=$hn/emailAddress=support@example.com"
+cert_life=1825
+subject="/C=uC/ST=uCentral/L=uCentral/O=uCentral/OU=uCentral/CN=$hn/emailAddress=support@example.com"
+
+openssl genrsa -out /etc/ucentral/certs/server-key.pem 2048
+openssl req -new -key /etc/ucentral/certs/server-key.pem -subj "$subject" -out /etc/ucentral/certs/server.csr
+openssl x509 -req -days $cert_life -in /etc/ucentral/certs/server.csr -signkey /etc/ucentral/certs/server-key.pem -out /etc/ucentral/certs/server-cert.pem
+
+for a in clientcas.pem issuer.pem restapi-ca.pem restapi-cert.pem root.pem websocket-cert.pem; do
+	cp /etc/ucentral/certs/server-cert.pem /etc/ucentral/certs/$a
+done
+
+for a in restapi-key.pem websocket-key.pem; do
+	cp /etc/ucentral/certs/server-key.pem /etc/ucentral/certs/$a
+done
 
-openssl genrsa -out server-key.pem 2048
-openssl req -new -key server-key.pem -subj "$subject" -out server.csr
-openssl x509 -req -days $cert_life -in server.csr -signkey server-key.pem -out server-cert.pem
+cp  /etc/ucentral/certs/server-cert.pem /etc/ucentral/certs/cas/
 
-for i in `eval echo {1..$howmany}`
+for i in `seq 1 $howmany`
 do
-  openssl x509 -signkey server-key.pem -in server.csr -req -days $cert_life -out dev-$i-cert.pem
+  openssl x509 -signkey /etc/ucentral/certs/server-key.pem -in /etc/ucentral/certs/server.csr -req -days $cert_life -out /etc/ucentral/certs/dev-$i-cert.pem
 done
Index: ucentralgw-2021-03-28-555b5fef/cert_scripts/more_devices.sh
===================================================================
--- ucentralgw-2021-03-28-555b5fef.orig/cert_scripts/more_devices.sh
+++ ucentralgw-2021-03-28-555b5fef/cert_scripts/more_devices.sh
@@ -1,4 +1,4 @@
-#!/usr/bin/env bash
+#!/bin/sh
 
 start=11
 finish=50
@@ -7,4 +7,4 @@ cert_life=365
 for i in `eval echo {$start..$finish}`
 do
   openssl x509 -signkey server-key.pem -in server.csr -req -days $cert_life -out dev-$i-cert.pem
-done
\ No newline at end of file
+done
