Index: ucentralgw-2021-03-28-555b5fef/CMakeLists.txt
===================================================================
--- ucentralgw-2021-03-28-555b5fef.orig/CMakeLists.txt
+++ ucentralgw-2021-03-28-555b5fef/CMakeLists.txt
@@ -40,7 +40,7 @@ find_package(ZLIB REQUIRED)
 find_package(Lua REQUIRED)
 
 if(SMALL_BUILD)
-    find_package(Poco REQUIRED COMPONENTS Crypto Net Util NetSSL Data DataSQLite)
+    find_package(Poco REQUIRED COMPONENTS Crypto JWT Net Util NetSSL Data DataSQLite)
 else()
     find_package(CppKafka REQUIRED)
     find_package(PostgreSQL REQUIRED)
@@ -81,4 +81,4 @@ if(NOT SMALL_BUILD)
             ${MySQL_LIBRARIES} ${ODBC_LIBRARIES} ${ZLIB_LIBRARIES} ${LUA_LIBRARIES}
             CppKafka::cppkafka
              )
-endif()
\ No newline at end of file
+endif()
