--- a/src/hsl/phy/rtl826xb_phy.c
+++ b/src/hsl/phy/rtl826xb_phy.c
@@ -48,69 +48,46 @@ void rtl826xb_phy_lock_init(void)
 
 static a_uint16_t rtl826x_phy_mmd_read(a_uint32_t dev_id, a_uint32_t phy_id, a_uint16_t reg_mmd, a_uint16_t reg_id)
 {
-	a_uint16_t phy_data;
-	sw_error_t rv;
 	a_uint32_t reg_id_c45 = RTL826XB_REG_ADDRESS(reg_mmd, reg_id);
-	__HSL_REG_ENTRY_GEN_GET(rv, dev_id, phy_id, reg_id_c45, &phy_data, sizeof(a_uint32_t));
 
-	return phy_data;
+	return __hsl_phy_mii_reg_read(dev_id, phy_id, reg_id_c45);
 }
 
 
 static sw_error_t rtl826x_phy_mmd_write(a_uint32_t dev_id, a_uint32_t phy_id, a_uint16_t reg_mmd, a_uint16_t reg_id, a_uint16_t reg_val)
 {
-	sw_error_t rv;
 	a_uint32_t reg_id_c45 = RTL826XB_REG_ADDRESS(reg_mmd, reg_id);
-	__HSL_REG_ENTRY_GEN_SET(rv, dev_id,  phy_id, reg_id_c45, reg_val, sizeof(a_uint32_t));
 
-	return rv;
+	return __hsl_phy_mii_reg_write(dev_id, phy_id, reg_id_c45, reg_val);
 }
 
 
 static a_uint16_t rtl826x_phy_reg_read(a_uint32_t dev_id, a_uint32_t phy_id, a_uint32_t reg)
 {
-	a_uint16_t phy_data;
-	sw_error_t rv;
-	__HSL_REG_ENTRY_GEN_GET(rv, dev_id, phy_id, reg, &phy_data, sizeof(a_uint32_t));
-
-	return phy_data;
+	return __hsl_phy_mii_reg_read(dev_id, phy_id, reg);
 }
 
 
 static sw_error_t rtl826x_phy_reg_write(a_uint32_t dev_id, a_uint32_t phy_id, a_uint32_t reg, a_uint16_t reg_val)
 {
-	sw_error_t rv;
-	
-	__HSL_REG_ENTRY_GEN_SET(rv, dev_id,  phy_id, reg, reg_val, sizeof(a_uint32_t));
-
-	return rv;
+	return __hsl_phy_mii_reg_write(dev_id, phy_id, reg, reg_val);
 }
 
 
 static a_int16_t hal_miim_mmd_read(a_uint32_t dev_id, a_uint32_t phy_id, a_uint16_t mmdAddr, a_uint16_t mmdReg)
 {
-	a_uint16_t phy_data;
-	sw_error_t rv;
-	
 	a_uint32_t reg_id_c45 = RTL826XB_REG_ADDRESS(mmdAddr, mmdReg);
-	
-	__HSL_REG_ENTRY_GEN_GET(rv, dev_id, phy_id, reg_id_c45, &phy_data, sizeof(a_uint32_t));
-	
-	return phy_data;
+
+	return __hsl_phy_mii_reg_read(dev_id, phy_id, reg_id_c45);
 }
 
 
 
 static a_int32_t hal_miim_mmd_write(a_uint32_t dev_id, a_uint32_t phy_id, a_uint16_t mmdAddr, a_uint16_t mmdReg, a_uint16_t phy_data)
 {
-	sw_error_t rv;
-	
 	a_uint32_t reg_id_c45 = RTL826XB_REG_ADDRESS(mmdAddr, mmdReg);
-	
-	__HSL_REG_ENTRY_GEN_SET(rv, dev_id, phy_id, reg_id_c45, phy_data, sizeof(a_uint32_t));
-	
 
-	return rv;
+	return __hsl_phy_mii_reg_write(dev_id, phy_id, reg_id_c45, phy_data);
 }
 
 
