Index: linux-5.4.105/arch/arm/boot/dts/qcom-ipq4019.dtsi
===================================================================
--- linux-5.4.105.orig/arch/arm/boot/dts/qcom-ipq4019.dtsi
+++ linux-5.4.105/arch/arm/boot/dts/qcom-ipq4019.dtsi
@@ -32,6 +32,12 @@
 			reg = <0x87e80000 0x180000>;
 			no-map;
 		};
+
+		ramoops@87f00000 {
+			compatible = "ramoops";
+			reg = <0x87df0000 0x10000>;
+			record-size = <0x1000>;
+		};
 	};
 
 	aliases {
