diff -Nurb a/arch/arm/boot/dts/Makefile b/arch/arm/boot/dts/Makefile
--- a/arch/arm/boot/dts/Makefile	2022-03-31 03:03:23.379141391 +0000
+++ b/arch/arm/boot/dts/Makefile	2022-03-31 03:07:24.671320669 +0000
@@ -908,6 +908,7 @@
 	qcom-msm8974-sony-xperia-honami.dtb \
 	qcom-mdm9615-wp8548-mangoh-green.dtb \
 	qcom-ipq4019-um-325ac.dtb \
+	qcom-ipq4019-um-510ac-v3.dtb \
 	qcom-ipq4019-um-550ac.dtb
 dtb-$(CONFIG_ARCH_RDA) += \
 	rda8810pl-orangepi-2g-iot.dtb \
