Index: linux-5.4.111/arch/arm/boot/dts/Makefile
===================================================================
--- linux-5.4.111.orig/arch/arm/boot/dts/Makefile
+++ linux-5.4.111/arch/arm/boot/dts/Makefile
@@ -883,6 +883,7 @@ dtb-$(CONFIG_ARCH_QCOM) += \
 	qcom-ipq4019-u4019-32m.dtb \
 	qcom-ipq4019-wpj419.dtb \
 	qcom-ipq4019-wtr-m2133hp.dtb \
+	qcom-ipq4019-tp-link-ec420-g1.dtb \
 	qcom-ipq4028-wpj428.dtb \
 	qcom-ipq4029-ap-303.dtb \
 	qcom-ipq4029-ap-303h.dtb \
