Index: linux-5.4.158/arch/arm/boot/dts/Makefile
===================================================================
--- linux-5.4.158.orig/arch/arm/boot/dts/Makefile
+++ linux-5.4.158/arch/arm/boot/dts/Makefile
@@ -888,6 +888,7 @@ dtb-$(CONFIG_ARCH_QCOM) += \
 	qcom-ipq4019-wpj419.dtb \
 	qcom-ipq4019-wtr-m2133hp.dtb \
 	qcom-ipq4019-tp-link-ec420-g1.dtb \
+	qcom-ipq4019-hfcl-ion4.dtb \
 	qcom-ipq4028-wpj428.dtb \
 	qcom-ipq4029-ap-303.dtb \
 	qcom-ipq4029-ap-303h.dtb \
@@ -908,6 +909,7 @@ dtb-$(CONFIG_ARCH_QCOM) += \
 	qcom-mdm9615-wp8548-mangoh-green.dtb \
 	qcom-ipq4019-um-325ac.dtb \
 	qcom-ipq4019-um-510ac-v3.dtb \
+	qcom-ipq4018-udaya-a5-id2.dtb \
 	qcom-ipq4019-um-550ac.dtb
 dtb-$(CONFIG_ARCH_RDA) += \
 	rda8810pl-orangepi-2g-iot.dtb \
