Index: qca-nss-dp/nss_dp_main.c
===================================================================
--- qca-nss-dp.orig/nss_dp_main.c
+++ qca-nss-dp/nss_dp_main.c
@@ -32,6 +32,7 @@
 #include <net/switchdev.h>
 
 #include "nss_dp_hal.h"
+#include "fal/fal_port_ctrl.h"
 
 /*
  * Number of TX/RX queue supported is based on the number of host CPU
@@ -293,6 +294,7 @@ static int nss_dp_open(struct net_device
 		phy_start(dp_priv->phydev);
 		phy_start_aneg(dp_priv->phydev);
 	}
+	fal_port_power_on(0, dp_priv->macid);
 
 	return 0;
 }
