From 9c5dab549d93e90d4f15793128b648cf3f9dfc93 Mon Sep 17 00:00:00 2001
From: John Crispin <john@phrozen.org>
Date: Mon, 9 Nov 2020 09:49:52 +0100
Subject: [PATCH] netifd: update to latest rev

Signed-off-by: John Crispin <john@phrozen.org>
---
 package/network/config/netifd/Makefile | 6 ++++--
 1 file changed, 4 insertions(+), 2 deletions(-)

diff --git a/package/network/config/netifd/Makefile b/package/network/config/netifd/Makefile
index 2c26517f44..6a3f6de2f6 100644
--- a/package/network/config/netifd/Makefile
+++ b/package/network/config/netifd/Makefile
@@ -6,8 +6,10 @@ PKG_RELEASE:=1
 PKG_SOURCE_PROTO:=git
 PKG_SOURCE_URL=$(PROJECT_GIT)/project/netifd.git
 PKG_SOURCE_DATE:=2020-10-22
-PKG_SOURCE_VERSION:=24ce1eab4910869576406bafd0489daf0d3e6e28
-PKG_MIRROR_HASH:=aa73b4e470e81165baba0262144f410bb05a3ff141cf72ad451914f033868ea1
+PKG_SOURCE_VERSION:=ed11f0c0ffe4fdacfe3f8223049ef8a61d9c53e9
+#24ce1eab4910869576406bafd0489daf0d3e6e28
+PKG_MIRROR_HASH:=
+#aa73b4e470e81165baba0262144f410bb05a3ff141cf72ad451914f033868ea1
 PKG_MAINTAINER:=Felix Fietkau <nbd@nbd.name>
 
 PKG_LICENSE:=GPL-2.0
-- 
2.25.1

