Index: hostapd-2021-02-20-59e9794c/src/ap/hostapd.c
===================================================================
--- hostapd-2021-02-20-59e9794c.orig/src/ap/hostapd.c
+++ hostapd-2021-02-20-59e9794c/src/ap/hostapd.c
@@ -1034,6 +1034,23 @@ hostapd_das_disconnect(void *ctx, struct
 	struct hostapd_data *hapd = ctx;
 	struct sta_info *sta;
 	int multi;
+	int ubus_resp;
+	struct hostapd_ubus_request req = {
+		.type = HOSTAPD_UBUS_COA,
+		.mgmt_frame = 0,
+		.ssi_signal = 0,
+		.addr = attr->sta_addr,
+	};
+
+	if (hostapd_ubus_handle_event(hapd, &req)) {
+		wpa_printf(MSG_INFO, "DAS: disconnect due approved via ubus");
+		sta = ap_get_sta(hapd, attr->sta_addr);
+		if (sta) {
+			hostapd_drv_sta_deauth(hapd, attr->sta_addr, 2);
+			ap_sta_deauthenticate(hapd, sta, 2);
+		}
+		return RADIUS_DAS_SUCCESS;
+	}
 
 	if (hostapd_das_nas_mismatch(hapd, attr))
 		return RADIUS_DAS_NAS_MISMATCH;
