--- a/src/ap/ieee802_11_auth.c
+++ b/src/ap/ieee802_11_auth.c
@@ -149,6 +149,14 @@
 	if (add_common_radius_attr(hapd, hapd->conf->radius_auth_req_attr,
 				   NULL, msg) < 0)
 		goto fail;
+
+	if (hapd->conf->radius_request_cui &&
+	    !radius_msg_add_attr(msg,
+				 RADIUS_ATTR_CHARGEABLE_USER_IDENTITY,
+				 (const u8 *) "\0", 1)) {
+		wpa_printf(MSG_DEBUG, "Could not add CUI");
+		goto fail;
+	}

 	os_snprintf(buf, sizeof(buf), RADIUS_802_1X_ADDR_FORMAT,
 		    MAC2STR(addr));
