--- a/src/ap/beacon.c
+++ b/src/ap/beacon.c
@@ -934,8 +934,10 @@ void handle_probe_req(struct hostapd_dat
 	int ubus_response;
 
 	if (hapd->iconf->rssi_ignore_probe_request && ssi_signal &&
-	    ssi_signal < hapd->iconf->rssi_ignore_probe_request)
+	    ssi_signal < hapd->iconf->rssi_ignore_probe_request) {
+		hostapd_ubus_notify_rssi(hapd, "rssi-ignore-probe", mgmt->sa, ssi_signal);
 		return;
+	}
 
 	if (len < IEEE80211_HDRLEN)
 		return;
