--- a/scan.c
+++ b/scan.c
@@ -1100,14 +1100,12 @@ static void print_rsn_ie(const char *def
 	_print_rsn_ie(defcipher, defauth, len, data, 0);
 }

-#ifdef IW_FULL
 static void print_osen_ie(const char *defcipher, const char *defauth,
 			  uint8_t len, const uint8_t *data)
 {
 	printf("\n\t");
 	_print_rsn_ie(defcipher, defauth, len, data, 1);
 }
-#endif

 static void print_rsn(const uint8_t type, uint8_t len, const uint8_t *data,
 		      const struct print_ies_data *ie_buffer)
@@ -1125,7 +1123,6 @@ static void print_ht_capa(const uint8_t
 	print_ht_mcs(data + 3);
 }

-#ifdef IW_FULL
 static const char* ntype_11u(uint8_t t)
 {
 	switch (t) {
@@ -1140,7 +1137,6 @@ static const char* ntype_11u(uint8_t t)
 	default: return "Reserved";
 	}
 }
-#endif

 static const char* vgroup_11u(uint8_t t)
 {
