From 38a201d06c08af5f7b1755f740b557900074a861 Mon Sep 17 00:00:00 2001
From: Michael-CY Lee <michael-cy.lee@mediatek.com>
Date: Mon, 6 Jan 2025 14:23:30 +0800
Subject: [PATCH 058/123] mtk: mt76: mt7996: set neg-ttlm related action frames
 to ALTX queue

Signed-off-by: Michael-CY Lee <michael-cy.lee@mediatek.com>
---
 mt7996/mac.c | 29 +++++++++++++++++++++++++----
 1 file changed, 25 insertions(+), 4 deletions(-)

diff --git a/mt7996/mac.c b/mt7996/mac.c
index 2884cd49..e74152a8 100644
--- a/mt7996/mac.c
+++ b/mt7996/mac.c
@@ -768,6 +768,30 @@ mt7996_mac_write_txwi_8023(struct mt7996_dev *dev, __le32 *txwi,
 		txwi[3] |= cpu_to_le32(MT_TXD3_HW_AMSDU);
 }
 
+static bool mt7996_is_skb_altx(struct ieee80211_mgmt *mgmt)
+{
+	__le16 fc = mgmt->frame_control;
+
+	if (ieee80211_is_deauth(fc)) {
+		/* In WPA3 cert TC-4.8.1, the deauth must be transmitted without
+		 * considering PSM bit
+		 */
+		return true;
+	}
+
+	if (ieee80211_is_action(fc) &&
+	    mgmt->u.action.category == WLAN_CATEGORY_PROTECTED_EHT &&
+	    (mgmt->u.action.u.ttlm_req.action_code ==
+	    WLAN_PROTECTED_EHT_ACTION_TTLM_REQ ||
+	    mgmt->u.action.u.ttlm_req.action_code ==
+	    WLAN_PROTECTED_EHT_ACTION_TTLM_RES ||
+	    mgmt->u.action.u.ttlm_req.action_code ==
+	    WLAN_PROTECTED_EHT_ACTION_TTLM_TEARDOWN))
+		return true;
+
+	return false;
+}
+
 static void
 mt7996_mac_write_txwi_80211(struct mt7996_dev *dev, __le32 *txwi,
 			    struct sk_buff *skb,
@@ -784,10 +808,7 @@ mt7996_mac_write_txwi_80211(struct mt7996_dev *dev, __le32 *txwi,
 	u8 fc_type, fc_stype;
 	u32 val;
 
-	if (ieee80211_is_deauth(fc)) {
-		/* In WPA3 cert TC-4.8.1, the deauth must be transmitted without
-		 * considering PSM bit
-		 */
+	if (mt7996_is_skb_altx(mgmt)) {
 		txwi[0] &= ~cpu_to_le32(MT_TXD0_Q_IDX);
 		txwi[0] |= cpu_to_le32(FIELD_PREP(MT_TXD0_Q_IDX, MT_LMAC_ALTX0));
 	}
-- 
2.45.2

