From e785029baa7b628a087250a896ca3e96aeb9255f Mon Sep 17 00:00:00 2001
From: Michael-CY Lee <michael-cy.lee@mediatek.com>
Date: Thu, 5 Jun 2025 10:30:17 +0800
Subject: [PATCH 097/123] mtk: mt76: mt7996: allign FW BSS_INFO Adv-TTLM subtag
 format

Signed-off-by: Michael-CY Lee <michael-cy.lee@mediatek.com>
---
 mt7996/mcu.c | 2 +-
 mt7996/mcu.h | 7 ++++++-
 2 files changed, 7 insertions(+), 2 deletions(-)

diff --git a/mt7996/mcu.c b/mt7996/mcu.c
index b8366733..4937f73f 100644
--- a/mt7996/mcu.c
+++ b/mt7996/mcu.c
@@ -4404,7 +4404,7 @@ mt7996_mcu_beacon_ttlm(struct mt7996_dev *dev, struct ieee80211_bss_conf *conf,
 				     sizeof(*attlm_offset));
 	attlm_offset = (struct bss_bcn_attlm_offset_tlv *)tlv;
 	attlm_offset->valid_id_bitmap = BIT(0);
-	attlm_offset->offset = cpu_to_le16(offset);
+	attlm_offset->offset[0] = cpu_to_le16(offset);
 }
 
 static bool
diff --git a/mt7996/mcu.h b/mt7996/mcu.h
index 53110d20..144677a0 100644
--- a/mt7996/mcu.h
+++ b/mt7996/mcu.h
@@ -602,12 +602,17 @@ struct bss_bcn_ml_reconf_offset {
 	u8 rsv;
 } __packed;
 
+#define MAX_ATTLM_NUM 1
 struct bss_bcn_attlm_offset_tlv {
 	__le16 tag;
 	__le16 len;
 	u8 valid_id_bitmap;
 	u8 rsv;
-	__le16 offset;
+	__le16 offset[MAX_ATTLM_NUM];
+	__le16 non_tx_bitmap;
+	u8 rsv_2[2];
+	u8 non_tx_valid_id_bitmap[MAX_BEACON_NUM];
+	__le16 non_tx_offsets[MAX_BEACON_NUM][MAX_ATTLM_NUM];
 } __packed;
 
 struct bss_txcmd_tlv {
-- 
2.45.2

