From 92a9ab1d16d26b7a2f052be2dc8ffd5d2bbed117 Mon Sep 17 00:00:00 2001
From: Michael-CY Lee <michael-cy.lee@mediatek.com>
Date: Fri, 13 Jun 2025 15:54:19 +0800
Subject: [PATCH 107/123] mtk: mt76: mt7996: separate Adv-TTLM offset parsing
 into a function

Signed-off-by: Michael-CY Lee <michael-cy.lee@mediatek.com>
---
 mt7996/mcu.c | 34 ++++++++++++++++++++++------------
 1 file changed, 22 insertions(+), 12 deletions(-)

diff --git a/mt7996/mcu.c b/mt7996/mcu.c
index d1710e19..95adf2e9 100644
--- a/mt7996/mcu.c
+++ b/mt7996/mcu.c
@@ -4383,6 +4383,25 @@ mt7996_mcu_beacon_ml_reconf(struct mt7996_dev *dev,
 	}
 }
 
+static u16
+mt7996_get_attlm_offset(const struct element *elem, struct sk_buff *skb)
+{
+	if (!elem)
+		return 0;
+
+	if (ieee80211_tid_to_link_map_size_ok(elem->data + 1,
+					      elem->datalen - 1)) {
+		struct ieee80211_ttlm_elem *ttlm =
+			(struct ieee80211_ttlm_elem *)elem->data + 1;
+
+		if (!(ttlm->control & IEEE80211_TTLM_CONTROL_SWITCH_TIME_PRESENT) &&
+		    (ttlm->control & IEEE80211_TTLM_CONTROL_EXPECTED_DUR_PRESENT))
+			return (u16)((u8 *)elem - skb->data);
+	}
+
+	return 0;
+}
+
 static void
 mt7996_mcu_beacon_ttlm(struct mt7996_dev *dev, struct ieee80211_bss_conf *conf,
 		       struct sk_buff *rskb, struct sk_buff *skb,
@@ -4392,7 +4411,6 @@ mt7996_mcu_beacon_ttlm(struct mt7996_dev *dev, struct ieee80211_bss_conf *conf,
 	struct bss_bcn_attlm_offset_tlv *attlm_offset;
 	u8 *beacon_tail = skb->data + tail_offset;
 	const struct element *elem;
-	struct ieee80211_ttlm_elem *ttlm;
 	bool cntdown_ttlm = false;
 	struct tlv *tlv;
 
@@ -4401,17 +4419,9 @@ mt7996_mcu_beacon_ttlm(struct mt7996_dev *dev, struct ieee80211_bss_conf *conf,
 
 	for_each_element_extid(elem, WLAN_EID_EXT_TID_TO_LINK_MAPPING,
 			       beacon_tail, skb->len - tail_offset) {
-		if (ieee80211_tid_to_link_map_size_ok(elem->data + 1,
-						      elem->datalen - 1)) {
-			ttlm = (struct ieee80211_ttlm_elem *)elem->data + 1;
-			if (!(ttlm->control &
-			      IEEE80211_TTLM_CONTROL_SWITCH_TIME_PRESENT) &&
-			    (ttlm->control &
-			     IEEE80211_TTLM_CONTROL_EXPECTED_DUR_PRESENT)) {
-				offset = (u8 *)elem - skb->data;
-				cntdown_ttlm = true;
-				break;
-			}
+		if ((offset = mt7996_get_attlm_offset(elem, skb)) > 0) {
+			cntdown_ttlm = true;
+			break;
 		}
 	}
 
-- 
2.45.2

