From df3be75094c3192f8195a3d17350f169b0fdf6ed Mon Sep 17 00:00:00 2001
From: Peter Chiu <chui-hao.chiu@mediatek.com>
Date: Wed, 3 Sep 2025 10:07:28 +0800
Subject: [PATCH 120/123] mtk: mt76: mt7996: update agg session time when
 receiving txfree

Signed-off-by: Peter Chiu <chui-hao.chiu@mediatek.com>
---
 mt7996/mac.c | 17 +++++++++++++++++
 1 file changed, 17 insertions(+)

diff --git a/mt7996/mac.c b/mt7996/mac.c
index ea122f94..f059f4b4 100644
--- a/mt7996/mac.c
+++ b/mt7996/mac.c
@@ -1348,6 +1348,19 @@ out:
 	mt76_put_txwi(mdev, t);
 }
 
+static void mt7996_refresh_tx_agg_session_timer(struct ieee80211_sta *sta)
+{
+	struct mt7996_sta *msta = (struct mt7996_sta *)sta->drv_priv;
+	u8 tid;
+
+	rcu_read_lock();
+	for (tid = 0; tid < IEEE80211_NUM_TIDS; tid++) {
+		if (test_bit(tid, &msta->deflink.wcid.ampdu_state))
+			ieee80211_refresh_tx_agg_session_timer(sta, tid);
+	}
+	rcu_read_unlock();
+}
+
 static void
 mt7996_mac_tx_free(struct mt7996_dev *dev, void *data, int len)
 {
@@ -3168,6 +3181,7 @@ static int mt7996_mac_sta_poll(struct mt76_dev *dev)
 {
 	u16 sta_list[PER_STA_INFO_MAX_NUM];
 	struct mt7996_sta_link *msta_link;
+	struct ieee80211_sta *sta;
 	int i, ret;
 
 	spin_lock_bh(&dev->sta_poll_lock);
@@ -3180,6 +3194,9 @@ static int mt7996_mac_sta_poll(struct mt76_dev *dev)
 					 wcid.poll_list);
 		list_del_init(&msta_link->wcid.poll_list);
 		sta_list[i] = msta_link->wcid.idx;
+		sta = container_of((void *)msta_link->sta,
+				   struct ieee80211_sta, drv_priv);
+		mt7996_refresh_tx_agg_session_timer(sta);
 	}
 	spin_unlock_bh(&dev->sta_poll_lock);
 
-- 
2.45.2

