--- a/include/hsl/hsl.h
+++ b/include/hsl/hsl.h
@@ -75,6 +75,14 @@ extern "C" {
 #define __HSL_REG_FIELD_GEN_SET(rv, dev, regaddr, bitlength, bitoffset, value, val_len)\
 	rv = __qca_mii_field_set(dev,regaddr,bitoffset,bitlength, (a_uint8_t*)value,val_len);
 
+extern sw_error_t reduce_hsl_phy_get(a_uint32_t dev,a_uint32_t phy_addr,a_uint32_t reg,a_uint16_t* value);
+#define HSL_PHY_GET(rv, dev, phy_addr, reg, value) \
+                rv = reduce_hsl_phy_get(dev,phy_addr,reg,value);
+
+extern sw_error_t reduce_hsl_phy_set(a_uint32_t dev,a_uint32_t phy_addr,a_uint32_t reg,a_uint16_t value);
+#define HSL_PHY_SET(rv, dev, phy_addr, reg, value) \
+                rv = reduce_hsl_phy_set(dev,phy_addr,reg,value);
+
 /*APIs with mdio bus lock*/
 #define HSL_REG_ENTRY_GET(rv, dev, reg, index, value, val_len) \
 	rv = qca_mii_reg_get(dev,reg##_OFFSET + ((a_uint32_t)index) * reg##_E_OFFSET,value,val_len);
--- a/src/hsl/phy/rtl826xb_phy.c
+++ b/src/hsl/phy/rtl826xb_phy.c
@@ -51,7 +51,7 @@ static a_uint16_t rtl826x_phy_mmd_read(a
 	a_uint16_t phy_data;
 	sw_error_t rv;
 	a_uint32_t reg_id_c45 = RTL826XB_REG_ADDRESS(reg_mmd, reg_id);
-	HSL_PHY_GET(rv, dev_id, phy_id, reg_id_c45, &phy_data);
+	__HSL_REG_ENTRY_GEN_GET(rv, dev_id, phy_id, reg_id_c45, &phy_data, sizeof(a_uint32_t));
 
 	return phy_data;
 }
@@ -61,7 +61,7 @@ static sw_error_t rtl826x_phy_mmd_write(
 {
 	sw_error_t rv;
 	a_uint32_t reg_id_c45 = RTL826XB_REG_ADDRESS(reg_mmd, reg_id);
-	HSL_PHY_SET(rv, dev_id,  phy_id, reg_id_c45, reg_val);
+	__HSL_REG_ENTRY_GEN_SET(rv, dev_id,  phy_id, reg_id_c45, reg_val, sizeof(a_uint32_t));
 
 	return rv;
 }
@@ -71,7 +71,7 @@ static a_uint16_t rtl826x_phy_reg_read(a
 {
 	a_uint16_t phy_data;
 	sw_error_t rv;
-	HSL_PHY_GET(rv, dev_id, phy_id, reg, &phy_data);
+	__HSL_REG_ENTRY_GEN_GET(rv, dev_id, phy_id, reg, &phy_data, sizeof(a_uint32_t));
 
 	return phy_data;
 }
@@ -81,7 +81,7 @@ static sw_error_t rtl826x_phy_reg_write(
 {
 	sw_error_t rv;
 	
-	HSL_PHY_SET(rv, dev_id,  phy_id, reg, reg_val);
+	__HSL_REG_ENTRY_GEN_SET(rv, dev_id,  phy_id, reg, reg_val, sizeof(a_uint32_t));
 
 	return rv;
 }
@@ -94,7 +94,7 @@ static a_int16_t hal_miim_mmd_read(a_uin
 	
 	a_uint32_t reg_id_c45 = RTL826XB_REG_ADDRESS(mmdAddr, mmdReg);
 	
-	HSL_PHY_GET(rv, dev_id, phy_id, reg_id_c45, &phy_data);
+	__HSL_REG_ENTRY_GEN_GET(rv, dev_id, phy_id, reg_id_c45, &phy_data, sizeof(a_uint32_t));
 	
 	return phy_data;
 }
@@ -107,7 +107,7 @@ static a_int32_t hal_miim_mmd_write(a_ui
 	
 	a_uint32_t reg_id_c45 = RTL826XB_REG_ADDRESS(mmdAddr, mmdReg);
 	
-	HSL_PHY_SET(rv, dev_id, phy_id, reg_id_c45, phy_data);
+	__HSL_REG_ENTRY_GEN_SET(rv, dev_id, phy_id, reg_id_c45, phy_data, sizeof(a_uint32_t));
 	
 
 	return rv;
