From 41861d8524e61b1c350f00c67e0ca8534b73d033 Mon Sep 17 00:00:00 2001
From: huangyunxiang <huangyunxiang@cigtech.com>
Date: Wed, 9 Apr 2025 10:03:52 +0800
Subject: [PATCH] qca-ssdk:Fix QCA8385 Maximum support for 4096 VLAN settings

---
 include/init/ssdk_plat.h                                    | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/include/init/ssdk_plat.h b/init/ssdk_plat.h
index 414918a5bd..92596477af 100644
--- a/include/init/ssdk_plat.h
+++ b/include/init/ssdk_plat.h
@@ -174,7 +174,7 @@
 #define AR8327_NUM_PHYS                              5
 #define AR8327_PORT_CPU                              0
 #define AR8327_NUM_PORTS                             7
-#define AR8327_MAX_VLANS                             128
+#define AR8327_MAX_VLANS                             4096
 
 #define MII_PHYADDR_C45                              (1<<30)
 
-- 
2.34.1

