 src/hsl/phy/sfp_phy.c | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

--- a/src/hsl/phy/sfp_phy.c
+++ b/src/hsl/phy/sfp_phy.c
@@ -932,7 +932,7 @@ sfp_phy_rx_los_status_get(a_uint32_t dev
 	{
 		return SW_NOT_SUPPORTED;
 	}
-	*rx_los_status = gpio_get_value(rx_los_pin);
+	*rx_los_status = gpio_get_value_cansleep(rx_los_pin);
 	SSDK_DEBUG("port%d sfp rx_los_pin is %d, the rx_los_status is %d",
 		port_id, rx_los_pin, *rx_los_status);
 
@@ -985,7 +985,7 @@ sfp_phy_tx_dis_status_get(a_uint32_t dev
 	{
 		return SW_NOT_SUPPORTED;
 	}
-	*tx_dis_status = gpio_get_value(tx_dis_pin);
+	*tx_dis_status = gpio_get_value_cansleep(tx_dis_pin);
 	SSDK_DEBUG("port%d sfp tx_dis_pin is %d, the tx_dis_status is %d",
 		port_id, tx_dis_pin, *tx_dis_status);
 
@@ -1007,7 +1007,7 @@ sfp_phy_mod_present_status_get(a_uint32_
 	{
 		return SW_NOT_SUPPORTED;
 	}
-	mod_present_pin_status = gpio_get_value(mod_present_pin);
+	mod_present_pin_status = gpio_get_value_cansleep(mod_present_pin);
 	SSDK_DEBUG("port%d sfp mod_present_pin is %d, mod_present_pin_status is %d",
 		port_id, mod_present_pin, mod_present_pin_status);
 	if(mod_present_pin_status)
