Index: backports-6.12.6/net/mac80211/vht.c
===================================================================
--- backports-6.12.6.orig/net/mac80211/vht.c
+++ backports-6.12.6/net/mac80211/vht.c
@@ -365,7 +365,7 @@ _ieee80211_sta_cap_rx_bw(struct link_sta
 		enum nl80211_band band;
 		u8 info;
 
-		if (chandef) {
+		if (chandef && chandef->chan) {
 			band = chandef->chan->band;
 		} else {
 			struct ieee80211_bss_conf *link_conf;
