Index: linux-5.4.158/drivers/net/ethernet/Makefile
===================================================================
--- linux-5.4.158.orig/drivers/net/ethernet/Makefile
+++ linux-5.4.158/drivers/net/ethernet/Makefile
@@ -52,7 +52,7 @@ obj-$(CONFIG_KORINA) += korina.o
 obj-$(CONFIG_LANTIQ_ETOP) += lantiq_etop.o
 obj-$(CONFIG_LANTIQ_XRX200) += lantiq_xrx200.o
 obj-$(CONFIG_NET_VENDOR_MARVELL) += marvell/
-obj-$(CONFIG_NET_VENDOR_MEDIATEK) += mediatek/
+obj-$(CONFIG_NET_VENDOR_MEDIATEK) += mediatek-sdk/
 obj-$(CONFIG_NET_VENDOR_MELLANOX) += mellanox/
 obj-$(CONFIG_NET_VENDOR_MICREL) += micrel/
 obj-$(CONFIG_NET_VENDOR_MICROCHIP) += microchip/
Index: linux-5.4.158/drivers/net/ethernet/Kconfig
===================================================================
--- linux-5.4.158.orig/drivers/net/ethernet/Kconfig
+++ linux-5.4.158/drivers/net/ethernet/Kconfig
@@ -119,7 +119,7 @@ config LANTIQ_XRX200
 	  Lantiq / Intel VRX200 VDSL SoC
 
 source "drivers/net/ethernet/marvell/Kconfig"
-source "drivers/net/ethernet/mediatek/Kconfig"
+source "drivers/net/ethernet/mediatek-sdk/Kconfig"
 source "drivers/net/ethernet/mellanox/Kconfig"
 source "drivers/net/ethernet/micrel/Kconfig"
 source "drivers/net/ethernet/microchip/Kconfig"
