--- a/hostapd/ctrl_iface.c	2025-09-24 14:15:25.135668867 +0800
+++ b/hostapd/ctrl_iface.c	2025-09-24 15:32:46.082317382 +0800
@@ -2657,6 +2657,11 @@ static int hostapd_ctrl_iface_chan_switc
 		break;
 	}
 
+    /* Initialize HT/VHT/HE parameters */
+    settings.freq_params.ht_enabled = iface->conf->ieee80211n;
+    settings.freq_params.vht_enabled = iface->conf->ieee80211ac;
+    settings.freq_params.he_enabled = iface->conf->ieee80211ax;
+
 	if (settings.freq_params.center_freq1)
 		dfs_range += hostapd_is_dfs_overlap(
 			iface, bandwidth, settings.freq_params.center_freq1);
