From 5047d592b46a2d9516a220f6219f6f40168057ab Mon Sep 17 00:00:00 2001
From: John Crispin <john@phrozen.org>
Date: Sat, 19 Jun 2021 16:00:43 +0200
Subject: [PATCH 4/5] ugps: update to latest HEAD

Signed-off-by: John Crispin <john@phrozen.org>
---
 package/utils/ugps/Makefile | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/package/utils/ugps/Makefile b/package/utils/ugps/Makefile
index c70a7b7ea6..389138d1dd 100644
--- a/package/utils/ugps/Makefile
+++ b/package/utils/ugps/Makefile
@@ -12,9 +12,9 @@ PKG_RELEASE:=$(AUTORELEASE)
 
 PKG_SOURCE_URL=$(PROJECT_GIT)/project/ugps.git
 PKG_SOURCE_PROTO:=git
-PKG_SOURCE_DATE:=2021-06-08
-PKG_SOURCE_VERSION:=5e88403fc0d39ae8a270d2c6c6e9c8a4d5232cf3
-PKG_MIRROR_HASH:=a3dfc2cc6def3b634d8b7cda9139926be3d876cf2faa5251c4821987954442a5
+PKG_SOURCE_DATE:=2021-07-19
+PKG_SOURCE_VERSION:=350b960b6f8b08409dcef719ef1e1e0c8911040f
+PKG_MIRROR_HASH:=94d489072f8fccc26f624c3d3c65e7cb40e4b6267df5fb39d4e0e78469006bde
 
 PKG_MAINTAINER:=John Crispin <john@phrozen.org>
 PKG_LICENSE:=GPL-2.0+
-- 
2.25.1

