From 09386632b364cf1021db0225e0d717bce4c4e8c9 Mon Sep 17 00:00:00 2001
From: Shayne Chen <shayne.chen@mediatek.com>
Date: Thu, 19 Oct 2023 10:48:11 +0800
Subject: [PATCH 059/260] mtk: wpa_supplicant: correctly get assoc frequency

Signed-off-by: Shayne Chen <shayne.chen@mediatek.com>
---
 src/drivers/driver_nl80211_event.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/src/drivers/driver_nl80211_event.c b/src/drivers/driver_nl80211_event.c
index df4a666a6..5911e97ed 100644
--- a/src/drivers/driver_nl80211_event.c
+++ b/src/drivers/driver_nl80211_event.c
@@ -337,6 +337,7 @@ static void mlme_event_assoc(struct wpa_driver_nl80211_data *drv,
 			   wpa_ssid_txt(drv->ssid, drv->ssid_len));
 	}
 
+	drv->assoc_freq = nl80211_get_assoc_freq(drv);
 	event.assoc_info.freq = drv->assoc_freq;
 	drv->first_bss->flink->freq = drv->assoc_freq;
 
-- 
2.45.2

