From edc15467de8e19eee77f47aaa820ce6df877a7dc Mon Sep 17 00:00:00 2001
From: Michael-CY Lee <michael-cy.lee@mediatek.com>
Date: Mon, 9 Dec 2024 11:38:54 +0800
Subject: [PATCH 124/260] mtk: hostapd: WPS: use passphrase over psk file

Signed-off-by: Michael-CY Lee <michael-cy.lee@mediatek.com>
---
 src/ap/wps_hostapd.c | 7 ++++---
 1 file changed, 4 insertions(+), 3 deletions(-)

diff --git a/src/ap/wps_hostapd.c b/src/ap/wps_hostapd.c
index ce306a387..731e84f8c 100644
--- a/src/ap/wps_hostapd.c
+++ b/src/ap/wps_hostapd.c
@@ -1225,9 +1225,10 @@ int hostapd_init_wps(struct hostapd_data *hapd,
 		wps->auth_types |= WPS_AUTH_OPEN;
 	}
 
-	if (conf->ssid.wpa_psk_file) {
-		/* Use per-device PSKs */
-	} else if (conf->ssid.wpa_passphrase) {
+	if (conf->ssid.wpa_psk_file)
+		wpa_printf(MSG_DEBUG, "WPS: allow using per-device PSKs");
+
+	if (conf->ssid.wpa_passphrase) {
 		wps->network_key = (u8 *) os_strdup(conf->ssid.wpa_passphrase);
 		wps->network_key_len = os_strlen(conf->ssid.wpa_passphrase);
 	} else if (conf->ssid.wpa_psk) {
-- 
2.45.2

