From 55f0be6524be605dfc870c86456f80a5c31c935d Mon Sep 17 00:00:00 2001
From: Shayne Chen <shayne.chen@mediatek.com>
Date: Wed, 24 Sep 2025 15:49:44 +0800
Subject: [PATCH 214/260] Revert "Avoid set beacon operations when color change
 is in progress"

Signed-off-by: Shayne Chen <shayne.chen@mediatek.com>
---
 src/ap/beacon.c | 8 --------
 1 file changed, 8 deletions(-)

diff --git a/src/ap/beacon.c b/src/ap/beacon.c
index 31d42e0f4..94face924 100644
--- a/src/ap/beacon.c
+++ b/src/ap/beacon.c
@@ -2851,14 +2851,6 @@ static int __ieee802_11_set_beacon(struct hostapd_data *hapd)
 		return -1;
 	}
 
-#ifdef CONFIG_IEEE80211AX
-	if (hapd->cca_in_progress) {
-		wpa_printf(MSG_ERROR,
-			   "Cannot set beacons during CCA period");
-		return -1;
-	}
-#endif /* CONFIG_IEEE80211AX */
-
 	if (iface->cac_started) {
 		wpa_printf(MSG_DEBUG, "Ignore set beacons during CAC period");
 		return 0;
-- 
2.45.2

