From d331928b09130d22d15c6a5b1f5907c5da9f52a0 Mon Sep 17 00:00:00 2001
From: Shayne Chen <shayne.chen@mediatek.com>
Date: Fri, 3 Oct 2025 14:41:26 +0800
Subject: [PATCH 215/260] tmp add back hostapd_is_mld_ap

---
 src/ap/ieee802_11.c | 9 +++++++++
 src/ap/ieee802_11.h | 1 +
 2 files changed, 10 insertions(+)

diff --git a/src/ap/ieee802_11.c b/src/ap/ieee802_11.c
index 83b79983d..cc005ec07 100644
--- a/src/ap/ieee802_11.c
+++ b/src/ap/ieee802_11.c
@@ -5482,6 +5482,15 @@ bool hostapd_is_multiple_link_mld(struct hostapd_data *hapd)
 	return false;
 }
 
+
+bool hostapd_is_mld_ap(struct hostapd_data *hapd)
+{
+	if (!hapd->conf->mld_ap)
+		return false;
+
+	return true;
+}
+
 #endif /* CONFIG_IEEE80211BE */
 
 
diff --git a/src/ap/ieee802_11.h b/src/ap/ieee802_11.h
index fa127c8a8..f9d2c2987 100644
--- a/src/ap/ieee802_11.h
+++ b/src/ap/ieee802_11.h
@@ -301,6 +301,7 @@ bool hostapd_is_multiple_link_mld(struct hostapd_data *hapd);
 int sae_password_bind(struct hostapd_data *hapd, const u8 *addr,
 		      const char *password);
 size_t hostapd_eid_non_inheritance_len(struct hostapd_data *hapd);
+bool hostapd_is_mld_ap(struct hostapd_data *hapd);
 const char * sae_get_password(struct hostapd_data *hapd,
 			      struct sta_info *sta, const u8 *rx_id,
 			      size_t rx_id_len,
-- 
2.45.2

