From f3fa93ad3a50bd95b0c1cea2b94ef8f8813f06f9 Mon Sep 17 00:00:00 2001
From: Michael-CY Lee <michael-cy.lee@mediatek.com>
Date: Mon, 20 Oct 2025 15:06:06 +0800
Subject: [PATCH 216/260] mtk: hostapd: check for invalid mld_link_id in
 hostapd.conf

Signed-off-by: Michael-CY Lee <michael-cy.lee@mediatek.com>
---
 src/ap/ap_config.c | 8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/src/ap/ap_config.c b/src/ap/ap_config.c
index 81f111f2b..5f79168e7 100644
--- a/src/ap/ap_config.c
+++ b/src/ap/ap_config.c
@@ -1556,6 +1556,14 @@ static int hostapd_config_check_bss(struct hostapd_bss_config *bss,
 			   "Cannot enable mld_ap when IEEE 802.11be is disabled");
 		return -1;
 	}
+
+#ifdef CONFIG_TESTING_OPTIONS
+	if (full_config && (bss->mld_link_id < -1 ||
+	    bss->mld_link_id > MAX_NUM_MLD_LINKS)) {
+		wpa_printf(MSG_ERROR, "Invalid MLD link ID");
+		return -1;
+	}
+#endif /* CONFIG_TESTING_OPTIONS */
 #endif /* CONFIG_IEEE80211BE */
 
 	if (full_config && bss->ignore_broadcast_ssid && conf->mbssid) {
-- 
2.45.2

