--- a/src/radius/radius_das.h
+++ b/src/radius/radius_das.h
@@ -37,6 +37,8 @@ struct radius_das_attrs {
 	const u8 *cui;
 	size_t cui_len;
 
+	const u8 *proxy;
+	size_t proxy_len;
 	/* Authorization changes */
 	const u8 *hs20_t_c_filtering;
 };
--- a/src/radius/radius_das.c
+++ b/src/radius/radius_das.c
@@ -61,6 +61,9 @@ static struct radius_msg * radius_das_disconnect(struct radius_das_data *das,
 		RADIUS_ATTR_EVENT_TIMESTAMP,
 		RADIUS_ATTR_MESSAGE_AUTHENTICATOR,
 		RADIUS_ATTR_CHARGEABLE_USER_IDENTITY,
+		RADIUS_ATTR_VENDOR_SPECIFIC,
+		RADIUS_ATTR_CALLED_STATION_ID,
+		RADIUS_ATTR_PROXY_STATE,
 #ifdef CONFIG_IPV6
 		RADIUS_ATTR_NAS_IPV6_ADDRESS,
 #endif /* CONFIG_IPV6 */
@@ -157,6 +160,12 @@ static struct radius_msg * radius_das_disconnect(struct radius_das_data *das,
 		attrs.cui_len = len;
 	}
 
+	if (radius_msg_get_attr_ptr(msg, RADIUS_ATTR_PROXY_STATE,
+				    &buf, &len, NULL) == 0) {
+		attrs.proxy = buf;
+		attrs.proxy_len = len;
+	}
+
 	res = das->disconnect(das->ctx, &attrs);
 	switch (res) {
 	case RADIUS_DAS_NAS_MISMATCH:
@@ -165,10 +174,7 @@ static struct radius_msg * radius_das_disconnect(struct radius_das_data *das,
 		error = 403;
 		break;
 	case RADIUS_DAS_SESSION_NOT_FOUND:
-		wpa_printf(MSG_INFO, "DAS: Session not found for request from "
-			   "%s:%d", abuf, from_port);
-		error = 503;
-		break;
+		return NULL;
 	case RADIUS_DAS_MULTI_SESSION_MATCH:
 		wpa_printf(MSG_INFO,
 			   "DAS: Multiple sessions match for request from %s:%d",
@@ -195,6 +201,9 @@ fail:
 		return NULL;
 	}
 
+	if (attrs.proxy)
+		radius_msg_add_attr(reply, RADIUS_ATTR_PROXY_STATE, attrs.proxy, attrs.proxy_len);
+
 	if (error) {
 		if (!radius_msg_add_attr_int32(reply, RADIUS_ATTR_ERROR_CAUSE,
 					       error)) {
@@ -223,9 +232,9 @@ static struct radius_msg * radius_das_coa(struct radius_das_data *das,
 		RADIUS_ATTR_EVENT_TIMESTAMP,
 		RADIUS_ATTR_MESSAGE_AUTHENTICATOR,
 		RADIUS_ATTR_CHARGEABLE_USER_IDENTITY,
-#ifdef CONFIG_HS20
 		RADIUS_ATTR_VENDOR_SPECIFIC,
-#endif /* CONFIG_HS20 */
+		RADIUS_ATTR_CALLED_STATION_ID,
+		RADIUS_ATTR_PROXY_STATE,
 #ifdef CONFIG_IPV6
 		RADIUS_ATTR_NAS_IPV6_ADDRESS,
 #endif /* CONFIG_IPV6 */
@@ -351,6 +360,12 @@ static struct radius_msg * radius_das_coa(struct radius_das_data *das,
 	}
 #endif /* CONFIG_HS20 */
 
+	if (radius_msg_get_attr_ptr(msg, RADIUS_ATTR_PROXY_STATE,
+				    &buf, &len, NULL) == 0) {
+		attrs.proxy = buf;
+		attrs.proxy_len = len;
+	}
+
 	res = das->coa(das->ctx, &attrs);
 	switch (res) {
 	case RADIUS_DAS_NAS_MISMATCH:
@@ -391,6 +406,9 @@ fail:
 		return NULL;
 	}
 
+	if (attrs.proxy)
+		radius_msg_add_attr(reply, RADIUS_ATTR_PROXY_STATE, attrs.proxy, attrs.proxy_len);
+
 	if (error &&
 	    !radius_msg_add_attr_int32(reply, RADIUS_ATTR_ERROR_CAUSE, error)) {
 		radius_msg_free(reply);
