--- a/src/ap/beacon.c
+++ b/src/ap/beacon.c
@@ -1510,8 +1510,10 @@ void handle_probe_req(struct hostapd_dat
 	};
 
 	if (hapd->iconf->rssi_ignore_probe_request && ssi_signal &&
-	    ssi_signal < hapd->iconf->rssi_ignore_probe_request)
+	    ssi_signal < hapd->iconf->rssi_ignore_probe_request) {
+		hostapd_ubus_notify_rssi(hapd, "rssi-ignore-probe", mgmt->sa, ssi_signal);
 		return;
+	}
 
 	if (len < IEEE80211_HDRLEN)
 		return;
--- a/src/ap/ubus.c
+++ b/src/ap/ubus.c
@@ -1951,6 +1951,23 @@ void hostapd_ubus_notify(struct hostapd_
 	ubus_notify(ctx, &hapd->ubus.obj, type, b.head, -1);
 }
 
+void hostapd_ubus_notify_rssi(struct hostapd_data *hapd, const char *type,
+			      const u8 *addr, int rssi)
+{
+	if (!hapd->ubus.obj.has_subscribers)
+		return;
+
+	if (!addr)
+		return;
+
+	blob_buf_init(&b, 0);
+	blobmsg_add_macaddr(&b, "address", addr);
+	blobmsg_add_string(&b, "ifname", hapd->conf->iface);
+	blobmsg_add_u32(&b, "rssi", rssi);
+
+	ubus_notify(ctx, &hapd->ubus.obj, type, b.head, -1);
+}
+
 void hostapd_ubus_notify_authorized(struct hostapd_data *hapd, struct sta_info *sta,
 				    const char *auth_alg)
 {
--- a/src/ap/ubus.h
+++ b/src/ap/ubus.h
@@ -52,4 +52,5 @@ void hostapd_ubus_notify(struct hostapd_
 void hostapd_ubus_notify(struct hostapd_data *hapd, const char *type, const u8 *mac);
+void hostapd_ubus_notify_rssi(struct hostapd_data *hapd, const char *type, const u8 *addr, int rssi);
 void hostapd_ubus_notify_beacon_report(struct hostapd_data *hapd,
 				       const u8 *addr, u8 token, u8 rep_mode,
 				       struct rrm_measurement_beacon_report *rep,
@@ -116,6 +117,10 @@ static inline void hostapd_ubus_notify(s
 {
 }
 
+static inline void hostapd_ubus_notify_rssi(struct hostapd_data *hapd, const char *type, const u8 *addr, int rssi)
+{
+}
+
 static inline void hostapd_ubus_notify_beacon_report(struct hostapd_data *hapd,
 						     const u8 *addr, u8 token,
 						     u8 rep_mode,
