--- a/src/ap/ubus.h
+++ b/src/ap/ubus.h
@@ -12,6 +12,7 @@ enum hostapd_ubus_event_type {
 	HOSTAPD_UBUS_PROBE_REQ,
 	HOSTAPD_UBUS_AUTH_REQ,
 	HOSTAPD_UBUS_ASSOC_REQ,
+	HOSTAPD_UBUS_COA,
 	HOSTAPD_UBUS_TYPE_MAX
 };
 
--- a/src/ap/hostapd.c
+++ b/src/ap/hostapd.c
@@ -1344,6 +1344,22 @@ hostapd_das_disconnect(void *ctx, struct
 	struct hostapd_data *hapd = ctx;
 	struct sta_info *sta;
 	int multi;
+	struct hostapd_ubus_request req = {
+		.type = HOSTAPD_UBUS_COA,
+		.mgmt_frame = 0,
+		.ssi_signal = 0,
+		.addr = attr->sta_addr,
+	};
+
+	if (hostapd_ubus_handle_event(hapd, &req)) {
+		wpa_printf(MSG_INFO, "DAS: disconnect approved via ubus");
+		sta = ap_get_sta(hapd, attr->sta_addr);
+		if (sta) {
+			hostapd_drv_sta_deauth(hapd, attr->sta_addr, 2);
+			ap_sta_deauthenticate(hapd, sta, 2);
+		}
+		return RADIUS_DAS_SUCCESS;
+	}
 
 	if (hostapd_das_nas_mismatch(hapd, attr))
 		return RADIUS_DAS_NAS_MISMATCH;
