--- a/src/ap/ieee802_11.c
+++ b/src/ap/ieee802_11.c
@@ -7107,9 +7107,20 @@ static int robust_action_frame(u8 category)
 }
 
 
+static int
+ewma(int new, int old)
+{
+	#define ALPHA	10
+	if (!old)
+		return new;
+	if (new >= 0)
+		return old;
+	return ((ALPHA * new) + ((100 - ALPHA) * old)) / 100;
+}
+
 static int handle_action(struct hostapd_data *hapd,
 			 const struct ieee80211_mgmt *mgmt, size_t len,
-			 unsigned int freq)
+			 unsigned int freq, int ssi_signal)
 {
 	struct sta_info *sta;
 	u8 *action __maybe_unused;
@@ -7166,6 +7177,7 @@ static int handle_action(struct hostapd_data *hapd,
 
 		sta->last_seq_ctrl = seq_ctrl;
 		sta->last_subtype = WLAN_FC_STYPE_ACTION;
+		sta->signal_mgmt = ewma(ssi_signal, sta->signal_mgmt);
 	}
 
 	switch (mgmt->u.action.category) {
@@ -7352,6 +7364,8 @@ int ieee802_11_mgmt(struct hostapd_data *hapd, const u8 *buf, size_t len,
 	int ret = 0;
 	unsigned int freq;
 	int ssi_signal = fi ? fi->ssi_signal : 0;
+
+	hapd->signal_mgmt = ewma(ssi_signal, hapd->signal_mgmt);
 #ifdef CONFIG_NAN_USD
 	static const u8 nan_network_id[ETH_ALEN] =
 		{ 0x51, 0x6f, 0x9a, 0x01, 0x00, 0x00 };
@@ -7493,7 +7507,7 @@ int ieee802_11_mgmt(struct hostapd_data *hapd, const u8 *buf, size_t len,
 		break;
 	case WLAN_FC_STYPE_ACTION:
 		wpa_printf(MSG_DEBUG, "mgmt::action");
-		ret = handle_action(hapd, mgmt, len, freq);
+		ret = handle_action(hapd, mgmt, len, freq, ssi_signal);
 		break;
 	default:
 		hostapd_logger(hapd, mgmt->sa, HOSTAPD_MODULE_IEEE80211,
--- a/src/ap/sta_info.h
+++ b/src/ap/sta_info.h
@@ -327,6 +327,7 @@ struct sta_info {
 #ifdef CONFIG_PASN
 	struct pasn_data *pasn;
 #endif /* CONFIG_PASN */
+	int signal_mgmt;
 
 #ifdef CONFIG_IEEE80211BE
 	struct mld_info mld_info;
--- a/src/ap/hostapd.h
+++ b/src/ap/hostapd.h
@@ -549,6 +549,7 @@ struct hostapd_data {
 #ifdef CONFIG_CTRL_IFACE_UDP
        unsigned char ctrl_iface_cookie[CTRL_IFACE_COOKIE_LEN];
 #endif /* CONFIG_CTRL_IFACE_UDP */
+       int signal_mgmt;
 
 #ifdef CONFIG_IEEE80211BE
 	u8 eht_mld_bss_param_change;
