--- a/src/ap/ieee802_11.c
+++ b/src/ap/ieee802_11.c
@@ -519,6 +519,7 @@ static void handle_auth_ft_finish(void *ctx, const u8 *dst,
 
 	hostapd_logger(hapd, dst, HOSTAPD_MODULE_IEEE80211,
 		       HOSTAPD_LEVEL_DEBUG, "authentication OK (FT)");
+	hostapd_ubus_notify(hapd, "ft-finish", sta->addr);
 	sta->flags |= WLAN_STA_AUTH;
 	mlme_authenticate_indication(hapd, sta);
 }
@@ -6311,6 +6312,9 @@ static void handle_assoc(struct hostapd_data *hapd,
 	}
 
 	sta = ap_get_sta(hapd, mgmt->sa);
+	if (sta && reassoc)
+		os_memcpy(sta->origin_ap, mgmt->u.reassoc_req.current_ap,
+			  ETH_ALEN);
 
 #ifdef CONFIG_IEEE80211BE
 	/*
--- a/src/ap/sta_info.h
+++ b/src/ap/sta_info.h
@@ -93,6 +93,7 @@ struct sta_info {
 	struct sta_info *next; /* next entry in sta list */
 	struct sta_info *hnext; /* next entry in hash table list */
 	u8 addr[6];
+	u8 origin_ap[6];
 	u8 setup_link_addr[6];
 	be32 ipaddr;
 	struct dl_list ip6addr; /* list head for struct ip6addr */
--- a/src/ap/ubus.c
+++ b/src/ap/ubus.c
@@ -1947,6 +1947,7 @@ void hostapd_ubus_notify(struct hostapd_data *hapd, const char *type, const u8 *
 	blob_buf_init(&b, 0);
 	blobmsg_add_macaddr(&b, "address", addr);
 	blobmsg_add_string(&b, "ifname", hapd->conf->iface);
+	blobmsg_add_macaddr(&b, "target", hapd->conf->bssid);
 
 	ubus_notify(ctx, &hapd->ubus.obj, type, b.head, -1);
 }
@@ -1988,6 +1989,9 @@ void hostapd_ubus_notify_authorized(struct hostapd_data *hapd, struct sta_info *
 		blobmsg_add_u32(&b, "", sta->bandwidth[1]);
 		blobmsg_close_array(&b, r);
 	}
+	if (!is_zero_ether_addr(sta->origin_ap))
+		blobmsg_add_macaddr(&b, "origin", sta->origin_ap);
+	blobmsg_add_macaddr(&b, "target", hapd->conf->bssid);
 
 	ubus_notify(ctx, &hapd->ubus.obj, "sta-authorized", b.head, -1);
 }
