Index: backports-20200902_001-4.4.60-931c337125/drivers/net/wireless/ath/ath11k/mac.c
===================================================================
--- backports-20200902_001-4.4.60-931c337125.orig/drivers/net/wireless/ath/ath11k/mac.c
+++ backports-20200902_001-4.4.60-931c337125/drivers/net/wireless/ath/ath11k/mac.c
@@ -497,7 +497,7 @@ struct ath11k_vif *ath11k_mac_get_arvif(
 						   ath11k_mac_get_arvif_iter,
 						   &arvif_iter);
 	if (!arvif_iter.arvif) {
-		ath11k_warn(ar->ab, "No VIF found for vdev %d\n", vdev_id);
+		//ath11k_warn(ar->ab, "No VIF found for vdev %d\n", vdev_id);
 		return NULL;
 	}
 
Index: backports-20200902_001-4.4.60-931c337125/drivers/net/wireless/ath/ath11k/wmi.c
===================================================================
--- backports-20200902_001-4.4.60-931c337125.orig/drivers/net/wireless/ath/ath11k/wmi.c
+++ backports-20200902_001-4.4.60-931c337125/drivers/net/wireless/ath/ath11k/wmi.c
@@ -8490,7 +8490,7 @@ static void ath11k_wmi_tlv_op_rx(struct
 		break;
 	/* TODO: Add remaining events */
 	default:
-		ath11k_warn(ab, "Unknown eventid: 0x%x\n", id);
+		//ath11k_warn(ab, "Unknown eventid: 0x%x\n", id);
 		break;
 	}
 
