Index: backports-20200902_001-4.4.60-931c337125/compat/backport-genetlink.c
===================================================================
--- backports-20200902_001-4.4.60-931c337125.orig/compat/backport-genetlink.c
+++ backports-20200902_001-4.4.60-931c337125/compat/backport-genetlink.c
@@ -381,6 +381,8 @@ int genlmsg_multicast_netns(const struct
 			    u32 portid, unsigned int group,
 			    gfp_t flags)
 {
+	if (in_atomic())
+		return -EINVAL;
 	group = __backport_genl_group(family, group);
 	if (group == INVALID_GROUP)
 		return -EINVAL;
