Index: backports-20200902_001-4.4.60-931c337125/drivers/net/wireless/ath/ath11k/qmi.c
===================================================================
--- backports-20200902_001-4.4.60-931c337125.orig/drivers/net/wireless/ath/ath11k/qmi.c
+++ backports-20200902_001-4.4.60-931c337125/drivers/net/wireless/ath/ath11k/qmi.c
@@ -16,7 +16,7 @@
 #define HOST_CSTATE_BIT			0x04
 
 /* set the default max assoc sta to max supported by driver */
-bool enable_cold_boot_cal = 1;
+bool enable_cold_boot_cal = 0;
 EXPORT_SYMBOL(enable_cold_boot_cal);
 module_param(enable_cold_boot_cal, bool, 0644);
 MODULE_PARM_DESC(enable_cold_boot_cal, "cold boot calibration enable:1 disable:0");
