Index: backports-20200902_001-4.4.60-931c337125/drivers/net/wireless/ath/ath11k/mac.c
===================================================================
--- backports-20200902_001-4.4.60-931c337125.orig/drivers/net/wireless/ath/ath11k/mac.c
+++ backports-20200902_001-4.4.60-931c337125/drivers/net/wireless/ath/ath11k/mac.c
@@ -5919,7 +5919,7 @@ static int ath11k_mac_op_add_interface(s
 		case NL80211_IFTYPE_STATION:
 		case NL80211_IFTYPE_AP_VLAN:
 		case NL80211_IFTYPE_AP:
-			hw_encap = 1;
+			hw_encap = 0;
 			break;
 		default:
 			break;
