Index: backports-20210222_001-4.4.60-b157d2276/drivers/net/wireless/ath/ath11k/mac.c
===================================================================
--- backports-20210222_001-4.4.60-b157d2276.orig/drivers/net/wireless/ath/ath11k/mac.c
+++ backports-20210222_001-4.4.60-b157d2276/drivers/net/wireless/ath/ath11k/mac.c
@@ -5872,6 +5915,7 @@ static int ath11k_mac_copy_he_cap(struct
 				  int band)
 {
 	int i, idx = 0;
+	struct ath11k_base *ab = ar->ab;
 
 	for (i = 0; i < NUM_NL80211_IFTYPES; i++) {
 		struct ieee80211_sta_he_cap *he_cap = &data[idx].he_cap;
@@ -5904,6 +5948,17 @@ static int ath11k_mac_copy_he_cap(struct
 			~IEEE80211_HE_PHY_CAP5_BEAMFORMEE_NUM_SND_DIM_UNDER_80MHZ_MASK;
 		he_cap_elem->phy_cap_info[5] |= ar->num_tx_chains - 1;
 
+		he_cap_elem->phy_cap_info[0] &=
+			~IEEE80211_HE_PHY_CAP0_CHANNEL_WIDTH_SET_80PLUS80_MHZ_IN_5G;
+		switch (ab->hw_rev) {
+		case ATH11K_HW_QCN6122:
+		case ATH11K_HW_QCN9074_HW10:
+			break;
+		default:
+			he_cap_elem->phy_cap_info[0] &=
+				~IEEE80211_HE_PHY_CAP0_CHANNEL_WIDTH_SET_160MHZ_IN_5G;
+			break;
+		}
 		switch (i) {
 		case NL80211_IFTYPE_AP:
 			he_cap_elem->phy_cap_info[3] &=

