diff -Naur a/drivers/net/wireless/ath/ath11k/wmi.c b/drivers/net/wireless/ath/ath11k/wmi.c
--- a/drivers/net/wireless/ath/ath11k/wmi.c	2021-12-15 19:14:36.363332750 -0500
+++ b/drivers/net/wireless/ath/ath11k/wmi.c	2021-12-22 13:09:19.899202295 -0500
@@ -8127,6 +8127,7 @@
 		cc_freq_hz = (ch_info_ev.mac_clk_mhz * 1000);
 
 	if (ch_info_ev.cmd_flags == WMI_CHAN_INFO_START_RESP) {
+	#if 0
 		survey = &ar->survey[idx];
 		memset(survey, 0, sizeof(*survey));
 		survey->noise = ch_info_ev.noise_floor;
@@ -8137,6 +8138,9 @@
 		if (survey->time)
 			ar->hw->medium_busy = div_u64((survey->time_busy * 100),
 						      survey->time);
+		printk("DBGL: CHAN_info: freq:%d, nf:%d, time:%llu, time_busy:%llu\n",
+			ch_info_ev.freq, survey->noise, survey->time, survey->time_busy);
+	#endif
 	}
 exit:
 	spin_unlock_bh(&ar->data_lock);
@@ -8211,6 +8215,11 @@
 			       SURVEY_INFO_TIME_RX |
 			       SURVEY_INFO_TIME_TX |
 			       SURVEY_INFO_TIME_BSS_RX);
+	printk("DBGL: BSS_CHAN: freq:%d, nf:%d, time:%llu, time_busy:%llu, time_rx:%llu,"
+		" time_bss_rx:%llu, time_tx:%llu\n",
+		bss_ch_info_ev.freq, survey->noise, survey->time, survey->time_busy,
+		survey->time_rx, survey->time_bss_rx, survey->time_tx);
+
 exit:
 	spin_unlock_bh(&ar->data_lock);
 	complete(&ar->bss_survey_done);
