From c22892ec958b3bcb2d2faf126a792c6e7be19357 Mon Sep 17 00:00:00 2001
From: John Crispin <john@phrozen.org>
Date: Fri, 23 Sep 2022 10:18:53 +0200
Subject: [PATCH 10/11] procd: disable ujail on ipq807x

Signed-off-by: John Crispin <john@phrozen.org>
---
 package/system/procd/Makefile | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/package/system/procd/Makefile b/package/system/procd/Makefile
index 2c345e7b8f..0c3a644679 100644
--- a/package/system/procd/Makefile
+++ b/package/system/procd/Makefile
@@ -67,7 +67,8 @@ define Package/procd-ujail
   SECTION:=base
   CATEGORY:=Base system
   DEPENDS:=@KERNEL_NAMESPACES +@KERNEL_UTS_NS +@KERNEL_IPC_NS +@KERNEL_PID_NS \
-	  +libubox +libubus +libuci +libblobmsg-json
+	  +libubox +libubus +libuci +libblobmsg-json \
+	  @!TARGET_ipq807x
   TITLE:=OpenWrt process jail helper
 endef
 
-- 
2.25.1

