From e82d4f44910162690029cabb1363361e9a81f4e7 Mon Sep 17 00:00:00 2001
From: John Crispin <john@phrozen.org>
Date: Fri, 23 Sep 2022 10:19:09 +0200
Subject: [PATCH 11/11] ucode: add ipq807x related nl80211.h hack

Signed-off-by: John Crispin <john@phrozen.org>
---
 package/utils/ucode/Makefile | 6 +++++-
 1 file changed, 5 insertions(+), 1 deletion(-)

diff --git a/package/utils/ucode/Makefile b/package/utils/ucode/Makefile
index 025f997387..4415f4854a 100644
--- a/package/utils/ucode/Makefile
+++ b/package/utils/ucode/Makefile
@@ -92,7 +92,7 @@ endef
 define Package/ucode-mod-nl80211
   $(Package/ucode/default)
   TITLE+= (nl80211 module)
-  DEPENDS:=ucode +libnl-tiny
+  DEPENDS:=ucode +libnl-tiny +kmod-mac80211
 endef
 
 define Package/ucode-mod-nl80211/description
@@ -174,6 +174,10 @@ define Build/InstallDev
 	$(CP) $(PKG_INSTALL_DIR)/usr/lib/libucode.so* $(1)/usr/lib/
 endef
 
+define Build/Prepare
+	$(Build/Prepare/Default)
+	$(CP) $(STAGING_DIR)/usr/include/mac80211/uapi/linux/nl80211.h $(PKG_BUILD_DIR)/nl80211_copy.h
+endef
 
 define Package/ucode/install
 	$(INSTALL_DIR) $(1)/usr/bin
-- 
2.25.1

