From 68400d212c87de3a835cdfe8811e5372891e0627 Mon Sep 17 00:00:00 2001
From: Samul Hwang <samul_hwang@accton.com>
Date: Mon, 24 Nov 2025 03:44:59 +0000
Subject: [PATCH] add logread to pmsg0

---
 package/system/ubox/files/log.init | 19 +++++++++++++++++++
 1 file changed, 19 insertions(+)

diff --git a/package/system/ubox/files/log.init b/package/system/ubox/files/log.init
index d03605adb3..765470ca10 100644
--- a/package/system/ubox/files/log.init
+++ b/package/system/ubox/files/log.init
@@ -64,6 +64,24 @@ start_service_file()
 	procd_close_instance
 }
 
+start_service_pmsg()
+{
+	PIDCOUNT="$(( ${PIDCOUNT} + 1))"
+	local pid_file="/var/run/logread.${PIDCOUNT}.pid"
+	local pmsg_file="/dev/pmsg0"
+
+	[ "$2" = 0 ] || {
+		echo "validation failed"
+		return 1
+	}
+	[ -e "${pmsg_file}" ] || return
+
+	procd_open_instance pmsgfile
+	procd_set_param command "$PROG" -f -F "$pmsg_file" -p "$pid_file"
+	procd_append_param command -S "128"
+	procd_close_instance
+}
+
 start_service_remote()
 {
 	PIDCOUNT="$(( ${PIDCOUNT} + 1))"
@@ -106,6 +124,7 @@ start_service()
 	config_load system
 	config_foreach validate_log_daemon system start_service_daemon
 	config_foreach validate_log_section system start_service_file
+	config_foreach validate_log_section system start_service_pmsg
 	config_foreach validate_log_section system start_service_remote
 }
 
-- 
2.17.1

