From 6c8c0c0cd25f869fe6b587e9ffa4acba80d16bad Mon Sep 17 00:00:00 2001
From: "Justin.Guo" <guoxijun@actiontec.com>
Date: Mon, 13 Apr 2026 14:38:17 +0800
Subject: [PATCH] base-files: add dependencies for ipq54xx upgrade tool
 dumpimage

Signed-off-by: Justin.Guo <guoxijun@actiontec.com>
---
 package/base-files/Makefile | 1 +
 1 file changed, 1 insertion(+)

diff --git a/package/base-files/Makefile b/package/base-files/Makefile
index 4532d99e0e..a984afe860 100644
--- a/package/base-files/Makefile
+++ b/package/base-files/Makefile
@@ -41,6 +41,7 @@ define Package/base-files
   CATEGORY:=Base system
   DEPENDS:= \
 	+netifd +libc +jsonfilter +SIGNED_PACKAGES:usign +SIGNED_PACKAGES:openwrt-keyring \
+	+TARGET_ipq54xx:libopenssl \
 	+NAND_SUPPORT:ubi-utils +fstools +fwtool \
 	+SELINUX:procd-selinux +!SELINUX:procd +USE_SECCOMP:procd-seccomp \
 	+SELINUX:busybox-selinux +!SELINUX:busybox
-- 
2.34.1

