--- a/drivers/net/wireless/ath/ath11k/mac.c
+++ b/drivers/net/wireless/ath/ath11k/mac.c
@@ -4785,9 +4785,9 @@ static int ath11k_mac_mgmt_tx(struct ath
 	 */
 	if (is_prb_rsp &&
 	    atomic_read(&ar->num_pending_mgmt_tx) > ATH11K_PRB_RSP_DROP_THRESHOLD) {
-		ath11k_warn(ar->ab,
+		ath11k_dbg(ar->ab, ATH11K_DBG_MAC,
 			    "dropping probe response as pending queue is almost full\n");
-		return -ENOSPC;
+		return -EBUSY;
 	}
 
 	if (skb_queue_len(q) == ATH11K_TX_MGMT_NUM_PENDING_MAX) {
@@ -4820,8 +4820,10 @@ static void ath11k_mac_op_tx(struct ieee
 		is_prb_rsp = ieee80211_is_probe_resp(hdr->frame_control);
 		ret = ath11k_mac_mgmt_tx(ar, skb, is_prb_rsp);
 		if (ret) {
-			ath11k_warn(ar->ab, "failed to queue management frame %d\n",
-				    ret);
+			if (ret != -EBUSY) {
+				ath11k_warn(ar->ab, "failed to queue management frame %d\n",
+				    	    ret);
+			}
 			ieee80211_free_txskb(ar->hw, skb);
 		}
 		return;
