Index: backports-20200628-4.4.60-9a94b73e75/drivers/net/wireless/ath/ath11k/core.c
===================================================================
--- backports-20200628-4.4.60-9a94b73e75.orig/drivers/net/wireless/ath/ath11k/core.c
+++ backports-20200628-4.4.60-9a94b73e75/drivers/net/wireless/ath/ath11k/core.c
@@ -29,6 +29,8 @@ module_param_named(nss_offload, nss_offl
 MODULE_PARM_DESC(nss_offload, "Enable NSS Offload support");
 #endif
 
+struct ath11k_base *ath11k_soc;
+
 extern unsigned int ath11k_frame_mode;
 
 static const struct ath11k_hw_params ath11k_hw_params_list[] = {
@@ -1132,6 +1134,7 @@ struct ath11k_base *ath11k_core_alloc(st
 	timer_setup(&ab->rx_replenish_retry, ath11k_ce_rx_replenish_retry, 0);
 	ab->dev = dev;
 	ab->hif.bus = bus;
+	ath11k_soc = ab;
 
 	return ab;
 
