--- a/drivers/net/wireless/ath/ath11k/core.c
+++ b/drivers/net/wireless/ath/ath11k/core.c
@@ -29,6 +29,8 @@ module_param_named(nss_offload, nss_offl
 MODULE_PARM_DESC(nss_offload, "Enable NSS Offload support");
 #endif
 
+struct ath11k_base *ath11k_soc;
+
 extern unsigned int ath11k_frame_mode;
 
 static const struct ath11k_hw_params ath11k_hw_params[] = {
@@ -1174,6 +1176,7 @@ struct ath11k_base *ath11k_core_alloc(st
 	ab->dev = dev;
 	ab->bus_params = *bus_params;
 	ab->hif.bus = bus;
+	ath11k_soc = ab;
 
 	return ab;
 
