--- a/drivers/net/wireless/ath/ath11k/mac.c
+++ b/drivers/net/wireless/ath/ath11k/mac.c
@@ -4693,7 +4693,8 @@
 }
 
 static struct ieee80211_sta_ht_cap
-ath11k_create_ht_cap(struct ath11k *ar, u32 ar_ht_cap, u32 rate_cap_rx_chainmask)
+ath11k_create_ht_cap(struct ath11k *ar, u32 ar_ht_cap,
+		     u32 rate_cap_tx_chainmask, u32 rate_cap_rx_chainmask)
 {
 	int i;
 	struct ieee80211_sta_ht_cap ht_cap = {0};
@@ -4724,7 +4725,7 @@
 		ht_cap.cap |= smps;
 	}
 
-	if (ar_ht_cap & WMI_HT_CAP_TX_STBC)
+	if ((ar_ht_cap & WMI_HT_CAP_TX_STBC) && (rate_cap_tx_chainmask > 1))
 		ht_cap.cap |= IEEE80211_HT_CAP_TX_STBC;
 
 	if (ar_ht_cap & WMI_HT_CAP_RX_STBC) {
@@ -4903,6 +4904,7 @@
 		if (ht_cap_info)
 			*ht_cap_info = ht_cap;
 		band->ht_cap = ath11k_create_ht_cap(ar, ht_cap,
+						    rate_cap_tx_chainmask,
 						    rate_cap_rx_chainmask);
 
 		band->vht_cap = ath11k_create_vht_cap(ar, rate_cap_tx_chainmask,
@@ -4915,10 +4917,25 @@
 		if (ht_cap_info)
 			*ht_cap_info = ht_cap;
 		band->ht_cap = ath11k_create_ht_cap(ar, ht_cap,
+						    rate_cap_tx_chainmask,
 						    rate_cap_rx_chainmask);
 		band->vht_cap = ath11k_create_vht_cap(ar, rate_cap_tx_chainmask,
 						      rate_cap_rx_chainmask);
 	}
+
+	if (ar->hw->wiphy->bands[NL80211_BAND_5GHZ]) {
+		ar->hw->wiphy->bands[NL80211_BAND_5GHZ]->ht_cap.cap =
+			ar->mac.sbands[NL80211_BAND_5GHZ].ht_cap.cap;
+		ar->hw->wiphy->bands[NL80211_BAND_5GHZ]->vht_cap.cap =
+			ar->mac.sbands[NL80211_BAND_5GHZ].vht_cap.cap;
+	}
+	if (ar->hw->wiphy->bands[NL80211_BAND_2GHZ]) {
+		ar->hw->wiphy->bands[NL80211_BAND_2GHZ]->ht_cap.cap =
+			ar->mac.sbands[NL80211_BAND_2GHZ].ht_cap.cap;
+		ar->hw->wiphy->bands[NL80211_BAND_2GHZ]->vht_cap.cap =
+			ar->mac.sbands[NL80211_BAND_2GHZ].vht_cap.cap;
+	}
+
 }
 
 static int ath11k_check_chain_mask(struct ath11k *ar, u32 ant, bool is_tx_ant)
