diff -Naur a/drivers/net/wireless/ath/ath11k/mac.c b/drivers/net/wireless/ath/ath11k/mac.c
--- a/drivers/net/wireless/ath/ath11k/mac.c	2021-06-09 10:02:12.040840722 -0400
+++ b/drivers/net/wireless/ath/ath11k/mac.c	2021-06-10 10:40:12.094003411 -0400
@@ -3472,13 +3472,14 @@
 		scan_timeout = min_t(u32, arg->max_rest_time *
 				    (arg->chan_list.num_chan - 1) + (req->duration +
 				     ATH11K_SCAN_CHANNEL_SWITCH_WMI_EVT_OVERHEAD) *
-				     arg->chan_list.num_chan, arg->max_scan_time +
-				     ATH11K_MAC_SCAN_TIMEOUT_MSECS);
+				     arg->chan_list.num_chan, arg->max_scan_time);
 	} else {
-		/* Add a 200ms margin to account for event/command processing */
-		scan_timeout = arg->max_scan_time + ATH11K_MAC_SCAN_TIMEOUT_MSECS;
+		scan_timeout = arg->max_scan_time;
 	}
 
+	/* Add a 200ms margin to account for event/command processing */
+	scan_timeout += ATH11K_MAC_SCAN_TIMEOUT_MSECS;
+
 	ret = ath11k_start_scan(ar, arg);
 	if (ret) {
 		ath11k_warn(ar->ab, "failed to start hw scan: %d\n", ret);
