From 8c45f965c30d1cf11e3a5a625a5e2baf3178697f Mon Sep 17 00:00:00 2001
From: John Crispin <john@phrozen.org>
Date: Tue, 12 Nov 2024 10:35:52 +0100
Subject: [PATCH] send on/offline events to cloud discovery service

Signed-off-by: John Crispin <john@phrozen.org>
---
 ubus.c | 5 +++++
 1 file changed, 5 insertions(+)

--- a/main.c
+++ b/main.c
@@ -111,6 +111,7 @@ sul_connect_attempt(struct lws_sorted_us
 {
 	struct per_vhost_data__minimal *vhd;
 
+	system("ubus call cloud offline");
 	vhd = lws_container_of(sul, struct per_vhost_data__minimal, sul);
 
 	vhd->i.context = vhd->context;
@@ -215,6 +216,7 @@ callback_broker(struct lws *wsi, enum lw
 		return 0;
 
 	case LWS_CALLBACK_CLIENT_ESTABLISHED:
+		system("ubus call cloud online");
 		ULOG_INFO("connection established\n");
 		if (!client.selfsigned) {
 			if (!lws_tls_peer_cert_info(wsi, LWS_TLS_CERT_INFO_VALIDITY_TO, &ci, 0))
