--- a/ucentral.h	2026-03-23 12:27:14.701654387 +0800
+++ b/ucentral.h	2026-03-23 12:27:47.153894140 +0800
@@ -38,6 +38,7 @@ struct client_config {
 	int recovery;
 	int telemetry_interval;
 	const char *boot_cause;
+	int hostname_validate;
 };
 extern struct client_config client;
 
--- a/main.c	2026-03-23 12:27:18.769684295 +0800
+++ b/main.c	2026-03-23 12:28:59.830440139 +0800
@@ -49,6 +49,7 @@ struct client_config client = {
 	.serial = "00:11:22:33:44:55",
 	.firmware = "v1.0",
 	.debug = 0,
+	.hostname_validate = 0,
 };
 
 void
@@ -218,7 +219,7 @@ callback_broker(struct lws *wsi, enum lw
 	case LWS_CALLBACK_CLIENT_ESTABLISHED:
 		system("ubus call cloud online");
 		ULOG_INFO("connection established\n");
-		if (!client.selfsigned) {
+		if (!client.selfsigned && client.hostname_validate) {
 			if (!lws_tls_peer_cert_info(wsi, LWS_TLS_CERT_INFO_VALIDITY_TO, &ci, 0))
 				ULOG_INFO(" Peer Cert Valid to  : %s", ctime(&ci.time));
 
@@ -326,6 +327,7 @@ static int print_usage(const char *daemo
 			"\t-P <port>\n"
 			"\t-d <debug>\n"
 			"\t-f <firmware>\n"
+			"\t-h <hostname validation>\n"
 			"\t-r <boot in recovery mode>\n"
 			"\t-v <venue>\n", daemon);
 	return -1;
@@ -340,7 +342,7 @@ int main(int argc, char **argv)
 	int ch;
 	int apply = 1;
 
-	while ((ch = getopt(argc, argv, "S:s:P:v:f:H:dirc:")) != -1) {
+	while ((ch = getopt(argc, argv, "S:s:P:v:f:H:dirc:h")) != -1) {
 		switch (ch) {
 		case 's':
 			client.server = optarg;
@@ -374,6 +376,8 @@ int main(int argc, char **argv)
 			apply = 0;
 			break;
 		case 'h':
+			client.hostname_validate = 1;
+			break;
 		default:
 			return print_usage(*argv);
 		}
