Index: ucentral-client-2022-06-22-90d276fe/config.c
===================================================================
--- ucentral-client-2022-06-22-90d276fe.orig/config.c
+++ ucentral-client-2022-06-22-90d276fe/config.c
@@ -55,7 +55,7 @@ config_apply(uint32_t id)
 		configure_reply(0, "Already applied.", uuid_latest, id);
 		return;
 	}
-	ULOG_INFO("applying cfg:%ld\n", uuid_latest);
+	ULOG_INFO("applying cfg:%"PRIu64"\n", uuid_latest);
 	apply_run(id);
 }
 
@@ -97,7 +97,7 @@ config_init(int apply, uint32_t id)
 
 out:
 	globfree(&gl);
-	ULOG_INFO("config_init latest:%ld active:%ld\n", uuid_latest, uuid_active);
+	ULOG_INFO("config_init latest:%"PRIu64" active:%"PRIu64"\n", uuid_latest, uuid_active);
 }
 
 void
Index: ucentral-client-2022-06-22-90d276fe/cmd.c
===================================================================
--- ucentral-client-2022-06-22-90d276fe.orig/cmd.c
+++ ucentral-client-2022-06-22-90d276fe/cmd.c
@@ -23,7 +23,7 @@ cmd_run_cb(time_t uuid, uint32_t _id)
 
 	ULOG_INFO("running command task\n");
 
-	sprintf(str, "/tmp/ucentral.cmd.%010ld", uuid);
+	sprintf(str, "/tmp/ucentral.cmd.%010"PRIu64"", uuid);
 	snprintf(id, sizeof(id), "%d", _id);
 	execlp("/usr/share/ucentral/cmd.uc", "/usr/share/ucentral/cmd.uc", str, id, NULL);
 	exit(1);
@@ -33,7 +33,7 @@ static void
 cmd_complete_cb(struct task *t, time_t uuid, uint32_t id, int ret)
 {
 	char str[128];
-	sprintf(str, "/tmp/ucentral.cmd.%010ld", uuid);
+	sprintf(str, "/tmp/ucentral.cmd.%010"PRIu64"", uuid);
 	unlink(str);
 	if (t->cancelled)
 		result_send_error(1, "command timed out and was aborted", -1, id);
@@ -63,7 +63,7 @@ cmd_run(struct blob_attr *attr, uint32_t
 		goto out;
 	}
 
-	snprintf(path, sizeof(path), "/tmp/ucentral.cmd.%010ld", t);
+	snprintf(path, sizeof(path), "/tmp/ucentral.cmd.%010"PRIu64"", t);
 	fp = fopen(path, "w+");
 	if (!fp) {
 		ULOG_ERR("failed to open %s\n", path);
Index: ucentral-client-2022-06-22-90d276fe/blink.c
===================================================================
--- ucentral-client-2022-06-22-90d276fe.orig/blink.c
+++ ucentral-client-2022-06-22-90d276fe/blink.c
@@ -11,7 +11,7 @@ blink_run_cb(time_t uuid, uint32_t id)
 
 	ULOG_INFO("running blink task\n");
 
-	snprintf(duration, sizeof(duration), "%ld", uuid);
+	snprintf(duration, sizeof(duration), "%"PRIu64"", uuid);
 	execlp("/usr/libexec/ucentral/ucentral_led_blink.sh", "/usr/libexec/ucentral/ucentral_led_blink.sh", duration, NULL);
 	exit(1);
 }
Index: ucentral-client-2022-06-22-90d276fe/apply.c
===================================================================
--- ucentral-client-2022-06-22-90d276fe.orig/apply.c
+++ ucentral-client-2022-06-22-90d276fe/apply.c
@@ -12,7 +12,7 @@ apply_run_cb(time_t uuid, uint32_t _id)
 
 	ULOG_INFO("running apply task\n");
 
-	sprintf(str, "/etc/ucentral/ucentral.cfg.%010ld", uuid);
+	sprintf(str, "/etc/ucentral/ucentral.cfg.%010"PRIu64"", uuid);
 	sprintf(id, "%d", _id);
 	execlp("/usr/share/ucentral/ucentral.uc", "/usr/share/ucentral/ucentral.uc", str, id, NULL);
 	exit(1);
@@ -29,7 +29,7 @@ apply_complete_cb(struct task *t, time_t
 		return;
 	}
 	uuid_active = uuid_applied = uuid_latest;
-	ULOG_INFO("applied cfg:%ld\n", uuid_latest);
+	ULOG_INFO("applied cfg:%"PRIu64"\n", uuid_latest);
 	//health_run(id, 0);
 }
 
