diff -urN a/package/feeds/packages/openvswitch/Makefile b/package/feeds/packages/openvswitch/Makefile
--- a/package/feeds/packages/openvswitch/Makefile	2020-12-15 14:05:57.848483999 +0530
+++ b/package/feeds/packages/openvswitch/Makefile	2020-12-15 14:05:57.848483999 +0530
@@ -222,7 +222,9 @@
 define ovs_common_install
 	$$(INSTALL_DIR) $$(1)/etc/openvswitch
 	$$(INSTALL_DIR) $$(1)/etc/init.d
+ifndef CONFIG_PACKAGE_opensync
 	$$(INSTALL_BIN) ./files/openvswitch.init $$(1)/etc/init.d/openvswitch
+endif
 	$$(INSTALL_DIR) $$(1)/etc/config
 	$$(INSTALL_DATA) ./files/openvswitch.config $$(1)/etc/config/openvswitch
 	$$(INSTALL_DIR) $$(1)/usr/share/openvswitch/scripts
diff -urN a/package/network/config/firewall/Makefile b/package/network/config/firewall/Makefile
--- a/package/network/config/firewall/Makefile	2020-12-15 14:05:57.844484001 +0530
+++ b/package/network/config/firewall/Makefile	2020-12-15 14:05:57.848483999 +0530
@@ -48,7 +48,9 @@
 	$(INSTALL_DIR) $(1)/sbin
 	$(INSTALL_BIN) $(PKG_BUILD_DIR)/firewall3 $(1)/sbin/fw3
 	$(INSTALL_DIR) $(1)/etc/init.d
+ifndef CONFIG_PACKAGE_opensync
 	$(INSTALL_BIN) ./files/firewall.init $(1)/etc/init.d/firewall
+endif
 	$(INSTALL_DIR) $(1)/etc/hotplug.d/iface
 	$(INSTALL_CONF) ./files/firewall.hotplug $(1)/etc/hotplug.d/iface/20-firewall
 	$(INSTALL_DIR) $(1)/etc/config/
diff -urN a/package/network/services/dnsmasq/Makefile b/package/network/services/dnsmasq/Makefile
--- a/package/network/services/dnsmasq/Makefile	2020-12-15 14:05:57.848483999 +0530
+++ b/package/network/services/dnsmasq/Makefile	2020-12-15 14:05:57.848483999 +0530
@@ -165,7 +165,9 @@
 	$(INSTALL_CONF) ./files/dhcp.conf $(1)/etc/config/dhcp
 	$(INSTALL_CONF) ./files/dnsmasq.conf $(1)/etc/dnsmasq.conf
 	$(INSTALL_DIR) $(1)/etc/init.d
+ifndef CONFIG_PACKAGE_opensync
 	$(INSTALL_BIN) ./files/dnsmasq.init $(1)/etc/init.d/dnsmasq
+endif
 	$(INSTALL_DIR) $(1)/etc/hotplug.d/dhcp
 	$(INSTALL_DIR) $(1)/etc/hotplug.d/neigh
 	$(INSTALL_DIR) $(1)/etc/hotplug.d/ntp
diff -urN a/package/network/utils/iproute2/Makefile b/package/network/utils/iproute2/Makefile
--- a/package/network/utils/iproute2/Makefile	2020-12-15 14:05:57.844484001 +0530
+++ b/package/network/utils/iproute2/Makefile	2020-12-16 12:57:27.867861946 +0530
@@ -155,6 +155,10 @@
 define Package/ip-full/install
 	$(INSTALL_DIR) $(1)/usr/libexec
 	$(INSTALL_BIN) $(PKG_BUILD_DIR)/ip/ip $(1)/usr/libexec/ip-full
+ifdef CONFIG_PACKAGE_opensync
+	$(INSTALL_DIR) $(1)/usr/sbin/
+	$(INSTALL_BIN) $(PKG_BUILD_DIR)/ip/ip $(1)/usr/sbin/ip
+endif
 endef
 
 define Package/tc/install
