--- a/drivers/net/wireless/ath/ath11k/dp.h
+++ b/drivers/net/wireless/ath/ath11k/dp.h
@@ -527,6 +527,14 @@ enum htt_ppdu_stats_tag_type {
 				    BIT(HTT_PPDU_STATS_TAG_TX_MGMTCTRL_PAYLOAD) | \
 				    HTT_PPDU_STATS_TAG_DEFAULT)
 
+#define HTT_PPDU_STATS_TAG_PKTLOG  (BIT(HTT_PPDU_STATS_TAG_USR_MPDU_ENQ_BITMAP_64) \
+				   | BIT(HTT_PPDU_STATS_TAG_USR_MPDU_ENQ_BITMAP_256) \
+				   | BIT(HTT_PPDU_STATS_TAG_USR_COMPLTN_BA_BITMAP_64) \
+				   | BIT(HTT_PPDU_STATS_TAG_USR_COMPLTN_BA_BITMAP_256) \
+				   | BIT(HTT_PPDU_STATS_TAG_INFO) \
+				   | BIT(HTT_PPDU_STATS_TAG_TX_MGMTCTRL_PAYLOAD) \
+				   | HTT_PPDU_STATS_TAG_DEFAULT)
+
 /* HTT_H2T_MSG_TYPE_RX_RING_SELECTION_CFG Message
  *
  * details:
