Index: backports-20191227-4.4.60-b805aa84b44f/drivers/net/wireless/ath/ath11k/qmi.c
===================================================================
--- backports-20191227-4.4.60-b805aa84b44f.orig/drivers/net/wireless/ath/ath11k/qmi.c
+++ backports-20191227-4.4.60-b805aa84b44f/drivers/net/wireless/ath/ath11k/qmi.c
@@ -1813,11 +1813,17 @@ static int ath11k_qmi_send_qdss_config(s
 	int ret;
 
 	snprintf(filename, sizeof(filename),
-		 "%s", ATH11K_QMI_DEFAULT_QDSS_CONFIG_FILE_NAME);
+		 "%s/%s", ab->hw_params.fw.dir, ATH11K_QMI_DEFAULT_QDSS_CONFIG_FILE_NAME);
 	ret = request_firmware(&fw_entry, filename, dev);
 	if (ret) {
-		ath11k_warn(ab, "qmi failed to load QDSS config: %s\n", filename);
-		return ret;
+		/* for backward compatability */
+		snprintf(filename, sizeof(filename),
+			 "%s", ATH11K_QMI_DEFAULT_QDSS_CONFIG_FILE_NAME);
+		ret = request_firmware(&fw_entry, filename, dev);
+		if (ret) {
+			ath11k_warn(ab, "qmi failed to load QDSS config: %s\n", filename);
+			return ret;
+		}
 	}
 
 	ret = wlfw_send_qdss_trace_config_download_req(ab, fw_entry->data,
