--- a/drivers/dax/Kconfig
+++ b/drivers/dax/Kconfig
@@ -4,7 +4,7 @@ config DAX_DRIVER
 	bool
 
 menuconfig DAX
-	tristate "DAX: direct access to differentiated memory"
+	bool "DAX: direct access to differentiated memory"
 	select SRCU
 	default m if NVDIMM_DAX
 
--- a/drivers/md/Kconfig
+++ b/drivers/md/Kconfig
@@ -200,7 +200,7 @@ config BLK_DEV_DM_BUILTIN
 	bool
 
 config BLK_DEV_DM
-	tristate "Device mapper support"
+	bool "Device mapper support"
 	select BLK_DEV_DM_BUILTIN
 	depends on DAX || DAX=n
 	---help---
@@ -486,7 +486,7 @@ config DM_FLAKEY
          A target that intermittently fails I/O for debugging purposes.
 
 config DM_VERITY
-	tristate "Verity target support"
+	bool "Verity target support"
 	depends on BLK_DEV_DM
 	select CRYPTO
 	select CRYPTO_HASH
