From 64a8b8b4d14e84e3b59e95783ae14eac38443acc Mon Sep 17 00:00:00 2001
From: John Crispin <john@phrozen.org>
Date: Fri, 30 Jan 2026 18:51:44 +0100
Subject: [PATCH 07/93] target: introduce default location for external target
 linux generic

Signed-off-by: John Crispin <john@phrozen.org>
---
 include/target.mk | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/include/target.mk b/include/target.mk
index 4bf2aee4d5..a5887f3785 100644
--- a/include/target.mk
+++ b/include/target.mk
@@ -160,7 +160,7 @@ ifeq ($(TARGET_BUILD),1)
   endif
 endif
 
-GENERIC_PLATFORM_DIR := $(TOPDIR)/target/linux/generic
+GENERIC_PLATFORM_DIR := $(firstword $(wildcard $(TOPDIR)/target/linux/feeds/$(BOARD)/../linux/generic $(TOPDIR)/target/linux/generic))
 
 ifneq ($(TARGET_BUILD)$(if $(DUMP),,1),)
   include $(INCLUDE_DIR)/kernel-version.mk
-- 
2.34.1

