From 9c85b5d826a8c781a8053205c16265d21a2a5027 Mon Sep 17 00:00:00 2001
From: John Crispin <john@phrozen.org>
Date: Fri, 30 Jan 2026 19:03:12 +0100
Subject: [PATCH 47/93] ipq807x: do not carry config over during sysupgrade

Signed-off-by: John Crispin <john@phrozen.org>
---
 package/boot/uboot-tools/Makefile | 6 ------
 1 file changed, 6 deletions(-)

diff --git a/package/boot/uboot-tools/Makefile b/package/boot/uboot-tools/Makefile
index 6e0c996f6d..10b7fbba13 100644
--- a/package/boot/uboot-tools/Makefile
+++ b/package/boot/uboot-tools/Makefile
@@ -62,12 +62,6 @@ define Package/uboot-envtools/description
 	bootloader environment.
 endef
 
-define Package/uboot-envtools/conffiles
-	/etc/config/ubootenv
-	/etc/fw_env.config
-	/etc/fw_sys.config
-endef
-
 define Build/Configure
 	$(call Build/Compile/Default,tools-only_defconfig)
 	$(SED) 's/CONFIG_TOOLS_LIBCRYPTO=y/# CONFIG_TOOLS_LIBCRYPTO is not set/' $(PKG_BUILD_DIR)/.config
-- 
2.34.1

