From e10752f0276830410b3fe2fba50f030257f9abaf Mon Sep 17 00:00:00 2001
From: CP Chang <cp.chang@emplustech.com>
Date: Mon, 29 Jun 2026 09:48:28 +0800
Subject: [PATCH 1/1] WAP588M:Fix LAN port can not transmit data

Signed-off-by: CP Chang <cp.chang@emplustech.com>
---
 .../mediatek/dts/mt7981b-emplus-wap588m.dts   | 1 +
 1 file changed, 1 insertion(+)

diff --git a/target/linux/mediatek/dts/mt7981b-emplus-wap588m.dts b/target/linux/mediatek/dts/mt7981b-emplus-wap588m.dts
index e3dfc2e3d5..b143608497 100644
--- a/target/linux/mediatek/dts/mt7981b-emplus-wap588m.dts
+++ b/target/linux/mediatek/dts/mt7981b-emplus-wap588m.dts
@@ -211,6 +211,7 @@
 		phy-handle = <&phy1>;
 		nvmem-cells = <&macaddr_lan>;
 		nvmem-cell-names = "mac-address";
+		managed = "in-band-status";
 	};
 
 	gmac1: mac@1 {
-- 
2.34.1

