From 4b31378553a7d433945772bd2c28689d072ef39a Mon Sep 17 00:00:00 2001
From: lyc-cybertan <yuanchongli@cybertan.com.tw>
Date: Thu, 23 Jul 2026 15:10:57 +0800
Subject: [PATCH] Fix Sonicfi_RAP630W-211G RVR test shows low 5G TX power

---
 package/kernel/mt76/Makefile | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/package/kernel/mt76/Makefile b/package/kernel/mt76/Makefile
index 2272546c0e..5d21e1519f 100644
--- a/package/kernel/mt76/Makefile
+++ b/package/kernel/mt76/Makefile
@@ -238,6 +238,9 @@ define KernelPackage/mt7915e
   DEPENDS+=@PCI_SUPPORT +kmod-mt76-connac +kmod-hwmon-core +kmod-thermal +@DRIVER_11AX_SUPPORT +@KERNEL_RELAY
   FILES:= $(PKG_BUILD_DIR)/mt7915/mt7915e.ko
   AUTOLOAD:=$(call AutoProbe,mt7915e)
+  ifeq ($(CONFIG_TARGET_PROFILE), "DEVICE_sonicfi_rap630w_211g")
+  	MODPARAMS.mt7915e:=wed_enable=1
+  endif
 endef
 
 define KernelPackage/mt7916-firmware
-- 
2.34.1

