From 721b1a5bd935b16ff933ace32a6da36bec473e19 Mon Sep 17 00:00:00 2001
From: Rex Lu <rex.lu@mediatek.com>
Date: Wed, 11 Dec 2024 11:41:49 +0800
Subject: [PATCH] wifi: mt76: mt7915: refactor check for TX BA sesstion status

Signed-off-by: Rex Lu <rex.lu@mediatek.com>
---
 mt76.h            |  3 ++-
 mt76_connac_mac.c | 41 +++++++++++++++++++++++++++++++----------
 2 files changed, 33 insertions(+), 11 deletions(-)

diff --git a/mt76.h b/mt76.h
index 592c3cd..fbddd7e 100644
--- a/mt76.h
+++ b/mt76.h
@@ -348,6 +348,7 @@ DECLARE_EWMA(signal, 10, 8);
 #define MT_WCID_TX_INFO_TXPWR_ADJ	GENMASK(25, 18)
 #define MT_WCID_TX_INFO_SET		BIT(31)
 
+#define ADDBA_RETRY_PERIOD		(5 * HZ)
 struct mt76_wcid {
 	struct mt76_rx_tid __rcu *aggr[IEEE80211_NUM_TIDS];
 
@@ -363,7 +364,7 @@ struct mt76_wcid {
 	 *Bit 16 - 31: ADDBA of TID n succeeded
 	 */
 	unsigned long ampdu_state;
-
+	unsigned long last_addba_req_time[IEEE80211_NUM_TIDS];
 	u16 idx;
 	u8 hw_key_idx;
 	u8 hw_key_idx2;
diff --git a/mt76_connac_mac.c b/mt76_connac_mac.c
index 4d24cc6..c250f9f 100644
--- a/mt76_connac_mac.c
+++ b/mt76_connac_mac.c
@@ -474,6 +474,34 @@ mt76_connac2_mac_write_txwi_80211(struct mt76_dev *dev, __le32 *txwi,
 	}
 }
 
+static void
+mt76_check_tx_ba_status(struct mt76_wcid *wcid, u8 tid)
+{
+	struct ieee80211_sta *sta;
+
+	if (!wcid)
+		return;
+
+	sta = wcid_to_sta(wcid);
+
+	if (!sta ||
+	    !(sta->ht_cap.ht_supported || sta->he_cap.has_he))
+		return;
+
+	if (test_bit(tid, &wcid->ampdu_state)) {
+		ieee80211_refresh_tx_agg_session_timer(sta, tid);
+		return;
+	}
+
+	if (!wcid->last_addba_req_time[tid] ||
+	    time_after(jiffies, wcid->last_addba_req_time[tid] + ADDBA_RETRY_PERIOD)) {
+		set_bit(tid, &wcid->ampdu_state);
+		if (ieee80211_start_tx_ba_session(sta, tid, 0) < 0)
+			clear_bit(tid, &wcid->ampdu_state);
+		wcid->last_addba_req_time[tid] = jiffies;
+	}
+}
+
 void mt76_connac2_mac_write_txwi(struct mt76_dev *dev, __le32 *txwi,
 				 struct sk_buff *skb, struct mt76_wcid *wcid,
 				 struct ieee80211_key_conf *key, int pid,
@@ -617,14 +645,8 @@ bool mt76_connac2_mac_fill_txs(struct mt76_dev *dev, struct mt76_wcid *wcid,
 			le32_get_bits(txs_data[7], MT_TXS7_MPDU_RETRY_CNT);
 
 		if (wcid->sta) {
-			struct ieee80211_sta *sta;
-			u8 tid;
-
-			sta = container_of((void *)wcid, struct ieee80211_sta,
-					   drv_priv);
-			tid = FIELD_GET(MT_TXS0_TID, txs);
-
-			ieee80211_refresh_tx_agg_session_timer(sta, tid);
+			if (!le32_get_bits(txs_data[0], MT_TXS0_FIXED_RATE))
+				mt76_check_tx_ba_status(wcid, FIELD_GET(MT_TXS0_TID, txs));
 		}
 	}
 
@@ -1134,8 +1156,7 @@ void mt76_connac2_tx_check_aggr(struct ieee80211_sta *sta, __le32 *txwi)
 		return;
 
 	wcid = (struct mt76_wcid *)sta->drv_priv;
-	if (!test_and_set_bit(tid, &wcid->ampdu_state))
-		ieee80211_start_tx_ba_session(sta, tid, 0);
+	mt76_check_tx_ba_status(wcid, tid);
 }
 EXPORT_SYMBOL_GPL(mt76_connac2_tx_check_aggr);
 
-- 
2.45.2

