From 1aead9dad4c2eb2592332b402f23f43f4ee79207 Mon Sep 17 00:00:00 2001
From: Peter Chiu <chui-hao.chiu@mediatek.com>
Date: Thu, 22 May 2025 15:42:50 +0800
Subject: [PATCH] wifi: mt76: mt7915: update adie patch in mt7986 power on
 sequence

Update adie init CR.

Signed-off-by: Peter Chiu <chui-hao.chiu@mediatek.com>

diff --git a/mt7915/regs.h b/mt7915/regs.h
index 3452a7e9..28ded58c 100644
--- a/mt7915/regs.h
+++ b/mt7915/regs.h
@@ -941,6 +941,7 @@ enum offs_rev {
 #define MT_ADIE_XTAL_TRIM1_40M_OSC	0x399
 #define MT_ADIE_WRI_CK_SEL		0x4ac
 #define MT_ADIE_RG_STRAP_PIN_IN		0x4fc
+#define MT_ADIE_RG_TOP_XO_08		0x600
 #define MT_ADIE_XTAL_C1			0x654
 #define MT_ADIE_XTAL_C2			0x658
 #define MT_ADIE_RG_XO_01		0x65c
diff --git a/mt7915/soc.c b/mt7915/soc.c
index bb3468a9..c613e77a 100644
--- a/mt7915/soc.c
+++ b/mt7915/soc.c
@@ -506,6 +506,7 @@ static int mt798x_wmac_adie_patch_7976(struct mt7915_dev *dev, u8 adie)
 {
 	u32 id, version, rg_xo_01, rg_xo_03;
 	int ret;
+	bool need_adie_patch = false;
 
 	ret = mt76_wmac_spi_read(dev, adie, MT_ADIE_CHIP_ID, &id);
 	if (ret)
@@ -525,6 +526,7 @@ static int mt798x_wmac_adie_patch_7976(struct mt7915_dev *dev, u8 adie)
 		if (is_mt7981(&dev->mt76)) {
 			rg_xo_01 = 0x1959c80f;
 		} else if (is_mt7986(&dev->mt76)) {
+			need_adie_patch = true;
 			rg_xo_01 = 0x1959f80f;
 		} else {
 			WARN_ON(1);
@@ -537,7 +539,14 @@ static int mt798x_wmac_adie_patch_7976(struct mt7915_dev *dev, u8 adie)
 	if (ret)
 		return ret;
 
-	return mt76_wmac_spi_write(dev, adie, MT_ADIE_RG_XO_03, rg_xo_03);
+	ret = mt76_wmac_spi_write(dev, adie, MT_ADIE_RG_XO_03, rg_xo_03);
+	if (ret)
+		return ret;
+
+	if (!need_adie_patch)
+		return 0;
+
+	return mt76_wmac_spi_write(dev, adie, MT_ADIE_RG_TOP_XO_08, 0);
 }
 
 static int
-- 
2.45.2

