From fcfa851c3dd6782c382dc1890f2623fdc0858f73 Mon Sep 17 00:00:00 2001
From: MeiChia Chiu <meichia.chiu@mediatek.com>
Date: Tue, 8 Aug 2023 11:20:58 +0800
Subject: [PATCH 1032/1052] wifi: mt76: mt7915: Establish BA in VO queue

---
 mt76_connac_mac.c | 2 --
 1 file changed, 2 deletions(-)

diff --git a/mt76_connac_mac.c b/mt76_connac_mac.c
index 6eeea971..c23d266e 100644
--- a/mt76_connac_mac.c
+++ b/mt76_connac_mac.c
@@ -1123,8 +1123,6 @@ void mt76_connac2_tx_check_aggr(struct ieee80211_sta *sta, __le32 *txwi)
 		return;
 
 	tid = le32_get_bits(txwi[1], MT_TXD1_TID);
-	if (tid >= 6) /* skip VO queue */
-		return;
 
 	val = le32_to_cpu(txwi[2]);
 	fc = FIELD_GET(MT_TXD2_FRAME_TYPE, val) << 2 |
-- 
2.18.0

