From 5fa70ee3f31733b9ddee67fdc78fe637df251aca Mon Sep 17 00:00:00 2001
From: Peter Chiu <chui-hao.chiu@mediatek.com>
Date: Mon, 2 Oct 2023 14:00:13 +0800
Subject: [PATCH 1038/1052] wifi: mt76: mt7915: add debuffs knob for protect
 threshold

---
 mt7915/mt7915.h      |  1 +
 mt7915/mtk_debugfs.c | 11 +++++++++++
 2 files changed, 12 insertions(+)

diff --git a/mt7915/mt7915.h b/mt7915/mt7915.h
index bc0f313c..183d59da 100644
--- a/mt7915/mt7915.h
+++ b/mt7915/mt7915.h
@@ -792,6 +792,7 @@ void mt7915_mcu_set_rfeature_starec(void *data, struct mt7915_dev *dev,
 		       struct ieee80211_vif *vif, struct ieee80211_sta *sta);
 int mt7915_mcu_set_rfeature_trig_type(struct mt7915_phy *phy, u8 enable, u8 trig_type);
 int mt7915_mcu_set_mu_dl_ack_policy(struct mt7915_phy *phy, u8 policy_num);
+int mt7915_mcu_set_mu_prot_frame_th(struct mt7915_phy *phy, u32 val);
 void mt7915_mcu_set_ppdu_tx_type(struct mt7915_phy *phy, u8 ppdu_type);
 void mt7915_mcu_set_nusers_ofdma(struct mt7915_phy *phy, u8 type, u8 ofdma_user_cnt);
 void mt7915_mcu_set_mimo(struct mt7915_phy *phy, u8 direction);
diff --git a/mt7915/mtk_debugfs.c b/mt7915/mtk_debugfs.c
index c00184f8..c8efd266 100644
--- a/mt7915/mtk_debugfs.c
+++ b/mt7915/mtk_debugfs.c
@@ -2852,6 +2852,16 @@ static int mt7915_sta_tx_amsdu_set(void *data, u64 tx_amsdu)
 DEFINE_DEBUGFS_ATTRIBUTE(fops_tx_amsdu, NULL,
 			 mt7915_sta_tx_amsdu_set, "%llx\n");
 
+static int mt7915_muru_set_prot_thr(void *data, u64 val)
+{
+	struct mt7915_phy *phy = data;
+
+	return mt7915_mcu_set_mu_prot_frame_th(phy, (u32)val);
+}
+
+DEFINE_DEBUGFS_ATTRIBUTE(fops_muru_set_prot_thr, NULL,
+			 mt7915_muru_set_prot_thr, "%lld\n");
+
 static int mt7915_red_enable_set(void *data, u64 en)
 {
 	struct mt7915_dev *dev = data;
@@ -3879,6 +3889,7 @@ int mt7915_mtk_init_debugfs(struct mt7915_phy *phy, struct dentry *dir)
 	debugfs_create_devm_seqfile(dev->mt76.dev, "fw_wm_info", dir,
 				    mt7915_fw_wm_info_read);
 
+	debugfs_create_file("prot_thr", 0200, dir, phy, &fops_muru_set_prot_thr);
 	debugfs_create_file("red_en", 0600, dir, dev,
 			    &fops_red_en);
 	debugfs_create_file("red_show_sta", 0600, dir, dev,
-- 
2.18.0

