From 3be6f92d47cdfd1ab676c9728cc84560fd2cee91 Mon Sep 17 00:00:00 2001
From: Peter Chiu <chui-hao.chiu@mediatek.com>
Date: Fri, 8 Sep 2023 18:26:21 +0800
Subject: [PATCH 2009/2015] wifi: mt76: mt7915: enable wa log to uart

Signed-off-by: Peter Chiu <chui-hao.chiu@mediatek.com>
---
 mt7915/debugfs.c | 6 +++++-
 1 file changed, 5 insertions(+), 1 deletion(-)

diff --git a/mt7915/debugfs.c b/mt7915/debugfs.c
index 576ec1a5..03ccb51b 100644
--- a/mt7915/debugfs.c
+++ b/mt7915/debugfs.c
@@ -951,7 +951,11 @@ mt7915_fw_debug_wa_set(void *data, u64 val)
 	struct mt7915_dev *dev = data;
 	int ret;
 
-	dev->fw.debug_wa = val ? MCU_FW_LOG_TO_HOST : 0;
+	/* bit 0: log to uart, bit 1: log to Host */
+	if (val > 3)
+		return -EINVAL;
+
+	dev->fw.debug_wa = val;
 
 	ret = mt7915_mcu_fw_log_2_host(dev, MCU_FW_LOG_WA, dev->fw.debug_wa);
 	if (ret)
-- 
2.18.0

