From 24b3e43ac4d858dd17013087e4ea0b4b86002079 Mon Sep 17 00:00:00 2001
From: Peter Chiu <chui-hao.chiu@mediatek.com>
Date: Fri, 8 Sep 2023 18:29:32 +0800
Subject: [PATCH 2010/2015] wifi: mt76: mt7915: add error message when driver
 receive invalid token id

Signed-off-by: Peter Chiu <chui-hao.chiu@mediatek.com>
---
 mt7915/mac.c | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/mt7915/mac.c b/mt7915/mac.c
index 1c8b8732..3f907e61 100644
--- a/mt7915/mac.c
+++ b/mt7915/mac.c
@@ -1007,6 +1007,12 @@ mt7915_mac_tx_free(struct mt7915_dev *dev, void *data, int len)
 			} else {
 				msdu = FIELD_GET(MT_TX_FREE_MSDU_ID, info);
 			}
+
+			/* Todo: check if msdu is sw token */
+			if (msdu > MT7915_TOKEN_SIZE)
+				dev_err(mdev->dev, "Receive invalid token id(%d)\n",
+					msdu);
+
 			count++;
 			txwi = mt76_token_release(mdev, msdu, &wake);
 			if (!txwi)
-- 
2.18.0

