From 03668fecc6bc2826369a16b92e128402ceb2504e Mon Sep 17 00:00:00 2001
From: Shayne Chen <shayne.chen@mediatek.com>
Date: Mon, 25 Nov 2024 14:16:47 +0800
Subject: [PATCH 7/9] mtk: iw: dump per-radio antenna_mask

Signed-off-by: Shayne Chen <shayne.chen@mediatek.com>
---
 info.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/info.c b/info.c
index 18dcd36..091cf38 100644
--- a/info.c
+++ b/info.c
@@ -930,6 +930,9 @@ next:
 					printf("\t\t\tRTS Threshold: %d\n",
 					       nla_get_u32(radio_prop));
 					break;
+				case NL80211_WIPHY_RADIO_ATTR_ANTENNA_MASK:
+					printf("\t\t\tantenna_mask: 0x%08x\n", nla_get_u32(radio_prop));
+					break;
 				default:
 					printf("\t\t\t* <failed to parse>\n");
 				}
-- 
2.45.2

