Index: hostapd-2020-06-08-5a8b3662/src/ap/dfs.c
===================================================================
--- hostapd-2020-06-08-5a8b3662.orig/src/ap/dfs.c
+++ hostapd-2020-06-08-5a8b3662/src/ap/dfs.c
@@ -1032,6 +1032,11 @@ static int hostapd_dfs_start_channel_swi
 	err = 0;
 
 	hostapd_setup_interface_complete(iface, err);
+	if (!err) {
+		wpa_printf(MSG_DEBUG, "Reporting DFS event to ubus");
+		hostapd_ubus_handle_channel_switch_event(iface, HOSTAPD_UBUS_DFS_SWITCH,  channel->freq);
+	}
+
 	return err;
 }
 
