From 6ee6c8762d470281fb778bd51daaa640f7bc36b5 Mon Sep 17 00:00:00 2001
From: Amutha Ravikumar <quic_aravikum@quicinc.com>
Date: Wed, 26 Apr 2023 17:20:46 +0530
Subject: [PATCH] Fix compilation error in upgraded package

Signed-off-by: Amutha Ravikumar <quic_aravikum@quicinc.com>
---
 src/ap/wpa_auth.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/ap/wpa_auth.c b/src/ap/wpa_auth.c
index c5db0ae..00f042d 100644
--- a/src/ap/wpa_auth.c
+++ b/src/ap/wpa_auth.c
@@ -1769,7 +1769,7 @@ void __wpa_send_eapol(struct wpa_authenticator *wpa_auth,
 	}
 
 	wpa_auth_set_eapol(wpa_auth, sm->addr, WPA_EAPOL_inc_EapolFramesTx, 1);
-	wpa_hexdump(MSG_DEBUG, "Send EAPOL-Key msg", hdr, len);
+	wpa_hexdump(MSG_DEBUG, "Send EAPOL-Key msg", (u8 *) hdr, len);
 	wpa_auth_send_eapol(wpa_auth, sm->addr, (u8 *) hdr, len,
 			    sm->pairwise_set);
 	os_free(hdr);
