Index: hostapd-2023-06-22-599d00be/src/ap/ieee802_11.c
===================================================================
--- hostapd-2023-06-22-599d00be.orig/src/ap/ieee802_11.c
+++ hostapd-2023-06-22-599d00be/src/ap/ieee802_11.c
@@ -6797,11 +6797,7 @@ static void handle_assoc_cb(struct hosta
 		}
 	}
 
-	/* WPS not supported on backhaul BSS. Disable 4addr mode on fronthaul */
-	if ((sta->flags & WLAN_STA_WDS) ||
-	    (sta->flags & WLAN_STA_MULTI_AP &&
-	     !(hapd->conf->multi_ap & FRONTHAUL_BSS) &&
-	     !(sta->flags & WLAN_STA_WPS))) {
+	if (sta->flags & (WLAN_STA_WDS | WLAN_STA_MULTI_AP)) {
 		int ret;
 		char ifname_wds[IFNAMSIZ + 1];
 		int aid;
Index: hostapd-2023-06-22-599d00be/src/ap/sta_info.c
===================================================================
--- hostapd-2023-06-22-599d00be.orig/src/ap/sta_info.c
+++ hostapd-2023-06-22-599d00be/src/ap/sta_info.c
@@ -242,10 +242,7 @@ void ap_free_sta(struct hostapd_data *ha
 	ap_sta_set_authorized(hapd, sta, 0);
 	hostapd_set_sta_flags(hapd, sta);
 
-	if ((sta->flags & WLAN_STA_WDS) ||
-	    (sta->flags & WLAN_STA_MULTI_AP &&
-	     !(hapd->conf->multi_ap & FRONTHAUL_BSS) &&
-	     !(sta->flags & WLAN_STA_WPS))) {
+	if (sta->flags & (WLAN_STA_WDS | WLAN_STA_MULTI_AP)) {
 		int aid;
 
 		if (hapd->conf->mld_ap && (sta->flags & WLAN_STA_WDS))
