--- a/src/ap/ieee802_11.c
+++ b/src/ap/ieee802_11.c
@@ -3103,7 +3103,7 @@ static void handle_auth(struct hostapd_d
 		return;
 	}
 #endif /* CONFIG_SAE */
-	{
+	if (hapd->mld) {
 		struct hostapd_ft_over_ds_ml_sta_entry *entry;
 
 		entry = ap_get_ft_ds_ml_sta(hapd, sa);
@@ -5589,7 +5589,7 @@ static void handle_assoc(struct hostapd_
 	if (sta && reassoc)
 		memcpy(sta->origin_ap, mgmt->u.reassoc_req.current_ap, 6);
 #ifdef CONFIG_IEEE80211R_AP
-	if (!sta) {
+	if (!sta && hapd->mld) {
 		wpa_printf(MSG_DEBUG,
 			   "FT over DS: Check for STA entry with ML address");
 		sta = get_sta_from_ft_ds_list(hapd, mgmt, len, reassoc);
