--- a/src/drivers/driver_nl80211.c
+++ b/src/drivers/driver_nl80211.c
@@ -977,11 +977,9 @@ add:
 			break;
 		}
 	}
-	/* if not add it by deleting the older wiphy */
-	if (!found) {
-		dl_list_del(&bss->drv->wiphy_list);
+	/* if not add it */
+	if (!found)
 		dl_list_add(&w->drvs, &bss->drv->wiphy_list);
-	}
 
 	dl_list_add(&w->bsss, &bss->wiphy_list);
 	bss->wiphy_data = w;
@@ -2285,7 +2283,6 @@ static void * wpa_driver_nl80211_drv_ini
 	drv->ctx = ctx;
 	drv->hostapd = !!hostapd;
 	drv->eapol_sock = -1;
-	dl_list_init(&drv->wiphy_list);
 
 	/*
 	 * There is no driver capability flag for this, so assume it is
