--- a/src/drivers/driver_nl80211.c
+++ b/src/drivers/driver_nl80211.c
@@ -14314,6 +14314,8 @@ static int wpa_driver_nl80211_can_share_
 	const u8 *addr;
 	int ret = -1;
 
+	return -1;
+
 	/* create a dummy drv to read the phyname */
 	if (params->global_priv == NULL)
 		return ret;
@@ -14367,7 +14369,7 @@ static int wpa_driver_nl80211_can_share_
 				      params->ifname, addr, ctx, drv_priv,
 				      force_ifname, if_addr,
 				      params->num_bridge && params->bridge[0] ? params->bridge[0] : NULL,
-				      0, 1)) {
+				      1, 1)) {
 		wpa_printf(MSG_DEBUG, "nl80211: Failed to add BSS. Expect issues!");
 		goto free_all;
 	}
