--- a/src/drivers/driver_nl80211.c
+++ b/src/drivers/driver_nl80211.c
@@ -10478,9 +10478,10 @@ static int wpa_driver_nl80211_get_survey
 
 	if (err)
 		wpa_printf(MSG_ERROR, "nl80211: Failed to process survey data");
+	else if (bss->scan_link && bss->scan_link->ctx)
+		wpa_supplicant_event(bss->scan_link->ctx, EVENT_SURVEY, &data);
 	else
-		wpa_supplicant_event(bss->scan_link->ctx ? bss->scan_link->ctx :
-				     bss->ctx, EVENT_SURVEY, &data);
+		wpa_supplicant_event(bss->ctx, EVENT_SURVEY, &data);
 
 	clean_survey_results(arg.survey_results);
 	return err;
