--- a/src/ap/hw_features.c
+++ b/src/ap/hw_features.c
@@ -525,7 +525,7 @@ static void ap_ht40_scan_retry(void *elo
 		wpa_printf(MSG_ERROR,
 			   "Failed to request a scan of neighboring BSSes ret=%d (%s) - try to scan again (attempt %d)",
 			   ret, strerror(-ret), iface->num_ht40_scan_tries);
-		eloop_register_timeout(1, 0, ap_ht40_scan_retry, iface, NULL);
+		eloop_register_timeout(10, 0, ap_ht40_scan_retry, iface, NULL);
 		return;
 	}
 
@@ -582,7 +582,7 @@ static int ieee80211n_check_40mhz(struct
 			   ret, strerror(-ret));
 		iface->num_ht40_scan_tries = 1;
 		eloop_cancel_timeout(ap_ht40_scan_retry, iface, NULL);
-		eloop_register_timeout(1, 0, ap_ht40_scan_retry, iface, NULL);
+		eloop_register_timeout(10, 0, ap_ht40_scan_retry, iface, NULL);
 		return 1;
 	}
 
