Index: linux-5.4.164-qsdk-d5fcb18e5420670c8734c6a659873e73adab6dac/arch/arm64/boot/dts/qcom/ipq8074.dtsi
===================================================================
--- linux-5.4.164-qsdk-d5fcb18e5420670c8734c6a659873e73adab6dac.orig/arch/arm64/boot/dts/qcom/ipq8074.dtsi
+++ linux-5.4.164-qsdk-d5fcb18e5420670c8734c6a659873e73adab6dac/arch/arm64/boot/dts/qcom/ipq8074.dtsi
@@ -1587,7 +1587,7 @@
 		};
 
 		wifi1: wifi1@c0000000 {
-			compatible = "qcom,cnss-qca8074v2", "qcom,ipq8074-wifi";
+			compatible = "qcom,ipq8074-wifi";
 			reg = <0xc000000 0x2000000>;
 			qcom,hw-mode-id = <1>;
 		#ifdef __IPQ_MEM_PROFILE_256_MB__
Index: linux-5.4.164-qsdk-d5fcb18e5420670c8734c6a659873e73adab6dac/arch/arm64/boot/dts/qcom/ipq6018.dtsi
===================================================================
--- linux-5.4.164-qsdk-d5fcb18e5420670c8734c6a659873e73adab6dac.orig/arch/arm64/boot/dts/qcom/ipq6018.dtsi
+++ linux-5.4.164-qsdk-d5fcb18e5420670c8734c6a659873e73adab6dac/arch/arm64/boot/dts/qcom/ipq6018.dtsi
@@ -1561,7 +1561,7 @@
 		};
 
 		wifi0: wifi@c000000 {
-			compatible = "qcom,cnss-qca6018", "qcom,ipq6018-wifi";
+			compatible = "qcom,ipq6018-wifi";
 			reg = <0xc000000 0x1000000>;
 			qcom,hw-mode-id = <1>;
 #ifdef __IPQ_MEM_PROFILE_256_MB__
