Index: linux-5.4.164-qsdk-d5fcb18e5420670c8734c6a659873e73adab6dac/drivers/clk/qcom/clk-branch.c
===================================================================
--- linux-5.4.164-qsdk-d5fcb18e5420670c8734c6a659873e73adab6dac.orig/drivers/clk/qcom/clk-branch.c
+++ linux-5.4.164-qsdk-d5fcb18e5420670c8734c6a659873e73adab6dac/drivers/clk/qcom/clk-branch.c
@@ -67,7 +67,7 @@ static int clk_branch_wait(const struct
 		bool (check_halt)(const struct clk_branch *, bool))
 {
 	bool voted = br->halt_check & BRANCH_VOTED;
-	const char *name = clk_hw_get_name(&br->clkr.hw);
+	//const char *name = clk_hw_get_name(&br->clkr.hw);
 
 	/*
 	 * Skip checking halt bit if we're explicitly ignoring the bit or the
@@ -88,8 +88,8 @@ static int clk_branch_wait(const struct
 				return 0;
 			udelay(1);
 		}
-		WARN(1, "%s status stuck at 'o%s'", name,
-				enabling ? "ff" : "n");
+/*		WARN(1, "%s status stuck at 'o%s'", name,
+				enabling ? "ff" : "n");*/
 		return -EBUSY;
 	}
 	return 0;
