Index: linux-5.4.164-qsdk-d5fcb18e5420670c8734c6a659873e73adab6dac/net/bridge/br_forward.c
===================================================================
--- linux-5.4.164-qsdk-d5fcb18e5420670c8734c6a659873e73adab6dac.orig/net/bridge/br_forward.c
+++ linux-5.4.164-qsdk-d5fcb18e5420670c8734c6a659873e73adab6dac/net/bridge/br_forward.c
@@ -24,7 +24,7 @@ static inline int should_deliver(const s
 	struct net_bridge_vlan_group *vg;
 
 	vg = nbp_vlan_group_rcu(p);
-	return (((p->flags & BR_HAIRPIN_MODE) && !is_multicast_ether_addr(eth_hdr(skb)->h_dest))
+	return (((p->flags & BR_HAIRPIN_MODE))
 		|| (skb->dev != p->dev)) &&
 		br_allowed_egress(vg, skb) && (p->state == BR_STATE_FORWARDING) &&
 		nbp_switchdev_allowed_egress(p, skb) &&
