Index: linux-4.4.60-qsdk-11f09717303ecd83c3a64e9efe23f25921dc1016/drivers/mtd/nand/qcom_nandc.c
===================================================================
--- linux-4.4.60-qsdk-11f09717303ecd83c3a64e9efe23f25921dc1016.orig/drivers/mtd/nand/qcom_nandc.c
+++ linux-4.4.60-qsdk-11f09717303ecd83c3a64e9efe23f25921dc1016/drivers/mtd/nand/qcom_nandc.c
@@ -405,6 +405,9 @@ struct nand_flash_dev qspinand_flash_ids
 	{"MX35UF1GE4AC SPI NAND 1G 1.8V",
 		{ .id = {0xc2, 0x92} },
 		SZ_2K, SZ_128, SZ_128K, 0, 2, 64, NAND_ECC_INFO(4, SZ_512), 0},
+	{"W25N01GW SPI NAND 1.8V 1G-BIT",
+		{ .id = {0xef, 0xba} },
+		SZ_2K, SZ_128, SZ_128K, 0, 2, 64, NAND_ECC_INFO(4, SZ_512), 0},
 	{NULL}
 };
 
