Index: linux-5.4.164-qsdk-d5fcb18e5420670c8734c6a659873e73adab6dac/net/ipv4/Kconfig
===================================================================
--- linux-5.4.164-qsdk-d5fcb18e5420670c8734c6a659873e73adab6dac.orig/net/ipv4/Kconfig
+++ linux-5.4.164-qsdk-d5fcb18e5420670c8734c6a659873e73adab6dac/net/ipv4/Kconfig
@@ -315,7 +315,7 @@ config NET_IPVTI
 	  on top.
 
 config NET_UDP_TUNNEL
-	tristate
+	tristate "IP: UDP tunneling support"
 	select NET_IP_TUNNEL
 	default n
 
