diff -Naur linux-4.4.60-qsdk-11f09717303ecd83c3a64e9efe23f25921dc1016.orig/drivers/mtd/nand/qcom_nandc.c linux-4.4.60-qsdk-11f09717303ecd83c3a64e9efe23f25921dc1016/drivers/mtd/nand/qcom_nandc.c
--- linux-4.4.60-qsdk-11f09717303ecd83c3a64e9efe23f25921dc1016.orig/drivers/mtd/nand/qcom_nandc.c	2022-07-26 10:43:12.428432000 +0800
+++ linux-4.4.60-qsdk-11f09717303ecd83c3a64e9efe23f25921dc1016/drivers/mtd/nand/qcom_nandc.c	2022-07-26 13:46:35.283160000 +0800
@@ -405,6 +405,12 @@
 	{"MX35UF1GE4AC SPI NAND 1G 1.8V",
 		{ .id = {0xc2, 0x92} },
 		SZ_2K, SZ_128, SZ_128K, 0, 2, 64, NAND_ECC_INFO(4, SZ_512), 0},
+	{"MX35UF1GE4AD SPI NAND 1G 1.8V",
+		{ .id = {0xc2, 0x96} },
+		SZ_2K, SZ_128, SZ_128K, 0, 2, 128, NAND_ECC_INFO(8, SZ_512), 0},
+	{"MX35UF2GE4AD SPI NAND 2G 1.8V",
+		{ .id = {0xc2, 0xa6} },
+		SZ_2K, SZ_256, SZ_128K, 0, 2, 128, NAND_ECC_INFO(8, SZ_512), 0},
 	{"W25N01GW SPI NAND 1.8V 1G-BIT",
 		{ .id = {0xef, 0xba} },
 		SZ_2K, SZ_128, SZ_128K, 0, 2, 64, NAND_ECC_INFO(4, SZ_512), 0},
