Index: hostapd-2021-02-20-59e9794c/src/radius/radius_das.c
===================================================================
--- hostapd-2021-02-20-59e9794c.orig/src/radius/radius_das.c
+++ hostapd-2021-02-20-59e9794c/src/radius/radius_das.c
@@ -48,6 +48,8 @@ static struct radius_msg * radius_das_di
 		RADIUS_ATTR_EVENT_TIMESTAMP,
 		RADIUS_ATTR_MESSAGE_AUTHENTICATOR,
 		RADIUS_ATTR_CHARGEABLE_USER_IDENTITY,
+		RADIUS_ATTR_VENDOR_SPECIFIC,
+		RADIUS_ATTR_CALLED_STATION_ID,
 #ifdef CONFIG_IPV6
 		RADIUS_ATTR_NAS_IPV6_ADDRESS,
 #endif /* CONFIG_IPV6 */
@@ -205,9 +207,8 @@ static struct radius_msg * radius_das_co
 		RADIUS_ATTR_EVENT_TIMESTAMP,
 		RADIUS_ATTR_MESSAGE_AUTHENTICATOR,
 		RADIUS_ATTR_CHARGEABLE_USER_IDENTITY,
-#ifdef CONFIG_HS20
 		RADIUS_ATTR_VENDOR_SPECIFIC,
-#endif /* CONFIG_HS20 */
+		RADIUS_ATTR_CALLED_STATION_ID,
 #ifdef CONFIG_IPV6
 		RADIUS_ATTR_NAS_IPV6_ADDRESS,
 #endif /* CONFIG_IPV6 */
