Index: hostapd-2020-07-02-58b384f4/src/ap/x_snoop.c
===================================================================
--- hostapd-2020-07-02-58b384f4.orig/src/ap/x_snoop.c
+++ hostapd-2020-07-02-58b384f4/src/ap/x_snoop.c
@@ -31,12 +31,17 @@ int x_snoop_init(struct hostapd_data *ha
 		return -1;
 	}
 
+	/* This code overwrites the setting done by netifd for isolation mode,
+	 * which makes the clients of wlan bridge port talk to each other even
+	 * when isolation is enabled on the wlan interface. Hence disabling it
+	 * to avoid that race between netifd and hostapd.
 	if (hostapd_drv_br_port_set_attr(hapd, DRV_BR_PORT_ATTR_HAIRPIN_MODE,
 					 1)) {
 		wpa_printf(MSG_DEBUG,
 			   "x_snoop: Failed to enable hairpin_mode on the bridge port");
 		return -1;
 	}
+	*/
 
 	if (hostapd_drv_br_port_set_attr(hapd, DRV_BR_PORT_ATTR_PROXYARP, 1)) {
 		wpa_printf(MSG_DEBUG,
