Index: hostapd-2020-07-02-58b384f4/src/ap/ubus.c
===================================================================
--- hostapd-2020-07-02-58b384f4.orig/src/ap/ubus.c
+++ hostapd-2020-07-02-58b384f4/src/ap/ubus.c
@@ -1016,12 +1016,18 @@ hostapd_switch_chan(struct ubus_context
 		css.freq_params.center_freq1,
 		css.freq_params.center_freq2);
 
-	if ((chan->flag & HOSTAPD_CHAN_RADAR) &&
-	    ((chan->flag & HOSTAPD_CHAN_DFS_MASK)
-	     != HOSTAPD_CHAN_DFS_AVAILABLE)) {
-		wpa_printf(MSG_INFO, "%s: DFS chan need CAC", __func__);
-		hostapd_switch_chan_dfs(iface, &css);
-		return UBUS_STATUS_OK;
+	if (chan->flag & HOSTAPD_CHAN_RADAR) {
+		if ((chan->flag & HOSTAPD_CHAN_DFS_MASK) ==
+		     HOSTAPD_CHAN_DFS_UNAVAILABLE) {
+			wpa_printf(MSG_WARNING, "DFS chan %d not Available",
+				   chan->chan);
+			return UBUS_STATUS_NOT_SUPPORTED;
+		} else if ((chan->flag & HOSTAPD_CHAN_DFS_MASK)
+			!= HOSTAPD_CHAN_DFS_AVAILABLE) {
+			wpa_printf(MSG_DEBUG, "DFS chan do CAC");
+			hostapd_switch_chan_dfs(iface, &css);
+			return UBUS_STATUS_OK;
+		}
 	}
 
 	for (i = 0; i < hapd->iface->num_bss; i++) {
