From 25ca718adeaaa1c43e8f1da3f49d5bd12a1461bb Mon Sep 17 00:00:00 2001
From: mtk23510 <rudra.shahi@mediatek.com>
Date: Fri, 26 May 2023 14:52:35 +0800
Subject: [PATCH 041/236] mtk: hostapd: Add support for gtk rekeying in hostapd
 cli

Signed-off-by: mtk23510 <rudra.shahi@mediatek.com>
---
 hostapd/hostapd_cli.c | 9 +++++++++
 1 file changed, 9 insertions(+)

diff --git a/hostapd/hostapd_cli.c b/hostapd/hostapd_cli.c
index d0c95bfa5..9643bffb0 100644
--- a/hostapd/hostapd_cli.c
+++ b/hostapd/hostapd_cli.c
@@ -1308,6 +1308,13 @@ static int hostapd_cli_cmd_stop_ap(struct wpa_ctrl *ctrl, int argc,
 }
 
 
+static int hostapd_cli_cmd_rekey_gtk(struct wpa_ctrl *ctrl, int argc,
+				      char *argv[])
+{
+	return wpa_ctrl_command(ctrl, "REKEY_GTK");
+}
+
+
 static int hostapd_cli_cmd_vendor(struct wpa_ctrl *ctrl, int argc, char *argv[])
 {
 	char cmd[256];
@@ -1860,6 +1867,8 @@ static const struct hostapd_cli_cmd hostapd_cli_commands[] = {
 	  "= update Beacon frame contents\n"},
 	{ "stop_ap", hostapd_cli_cmd_stop_ap, NULL,
 	  "= stop AP\n"},
+	{ "rekey_gtk", hostapd_cli_cmd_rekey_gtk, NULL,
+	  "= rekey gtk\n"},
 	{ "erp_flush", hostapd_cli_cmd_erp_flush, NULL,
 	  "= drop all ERP keys"},
 	{ "log_level", hostapd_cli_cmd_log_level, NULL,
-- 
2.45.2

