From 9d7d792531d240e25e6d0b95da2fbab3ea658e67 Mon Sep 17 00:00:00 2001
From: Michael-CY Lee <michael-cy.lee@mediatek.com>
Date: Fri, 1 Sep 2023 15:31:24 +0800
Subject: [PATCH 044/236] mtk: hostapd: avoid color switch when beacon is not
 set

Signed-off-by: Michael-CY Lee <michael-cy.lee@mediatek.com>
---
 src/ap/hostapd.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/ap/hostapd.c b/src/ap/hostapd.c
index 256541880..ca499ca92 100644
--- a/src/ap/hostapd.c
+++ b/src/ap/hostapd.c
@@ -5029,7 +5029,7 @@ void hostapd_switch_color(struct hostapd_data *hapd, u64 bitmap)
 {
 	struct os_reltime now;
 
-	if (hapd->cca_in_progress)
+	if (hapd->cca_in_progress || !hapd->beacon_set_done)
 		return;
 
 	if (os_get_reltime(&now))
-- 
2.45.2

