From a801fa644b7035461476a0fe622e16e4f41f0f4e Mon Sep 17 00:00:00 2001
From: mtk25255 <rohit.kamat@mediatek.com>
Date: Thu, 12 Oct 2023 14:29:23 +0800
Subject: [PATCH 047/236] mtk: hostapd: Fix RSNXE Interop issue with STA

---
 src/ap/ieee802_11.c | 16 ++++++++++++++++
 1 file changed, 16 insertions(+)

diff --git a/src/ap/ieee802_11.c b/src/ap/ieee802_11.c
index 89e5b0113..3a3d47976 100644
--- a/src/ap/ieee802_11.c
+++ b/src/ap/ieee802_11.c
@@ -5713,6 +5713,7 @@ static void handle_assoc(struct hostapd_data *hapd,
 	int omit_rsnxe = 0;
 	bool set_beacon = false;
 	bool mld_addrs_not_translated = false;
+	bool sae_pk = false;
 
 	if (len < IEEE80211_HDRLEN + (reassoc ? sizeof(mgmt->u.reassoc_req) :
 				      sizeof(mgmt->u.assoc_req))) {
@@ -5958,6 +5959,7 @@ static void handle_assoc(struct hostapd_data *hapd,
 			       reassoc ? LINK_PARSE_REASSOC : LINK_PARSE_ASSOC);
 	if (resp != WLAN_STATUS_SUCCESS)
 		goto fail;
+
 #ifdef CONFIG_IEEE80211R_AP
 	if (reassoc && sta->auth_alg == WLAN_AUTH_FT)
 		omit_rsnxe = !get_ie(pos, left, WLAN_EID_RSNX);
@@ -5965,6 +5967,20 @@ static void handle_assoc(struct hostapd_data *hapd,
 	if (hapd->conf->rsn_override_omit_rsnxe)
 		omit_rsnxe = 1;
 
+	/* TODO: check if this section is needed */
+	omit_rsnxe = !get_ie(pos, left, WLAN_EID_RSNX);
+#ifdef CONFIG_SAE_PK
+	sae_pk = hostapd_sae_pk_in_use(hapd->conf);
+#endif /* CONFIG_SAE_PK */
+	if (omit_rsnxe) {
+		if (!reassoc && wpa_key_mgmt_sae(hapd->conf->wpa_key_mgmt) &&
+				(hapd->conf->sae_pwe == SAE_PWE_HASH_TO_ELEMENT ||
+				 hapd->conf->sae_pwe == SAE_PWE_BOTH || sae_pk ||
+				 wpa_key_mgmt_sae_ext_key(hapd->conf->wpa_key_mgmt))) {
+			omit_rsnxe = 0;
+		}
+	}
+
 	if (hostapd_get_aid(hapd, sta) < 0) {
 		hostapd_logger(hapd, mgmt->sa, HOSTAPD_MODULE_IEEE80211,
 			       HOSTAPD_LEVEL_INFO, "No room for more AIDs");
-- 
2.45.2

