From caaacec695435e201409a22812138273665469ff Mon Sep 17 00:00:00 2001
From: Evelyn Tsai <evelyn.tsai@mediatek.com>
Date: Wed, 30 Aug 2023 04:23:37 +0800
Subject: [PATCH 048/236] mtk: hostapd: add support for ucode to parse BW320MHz
 info

---
 src/utils/ucode.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/src/utils/ucode.c b/src/utils/ucode.c
index 29c753c32..4b6ed3a94 100644
--- a/src/utils/ucode.c
+++ b/src/utils/ucode.c
@@ -144,6 +144,10 @@ uc_value_t *uc_wpa_freq_info(uc_vm_t *vm, size_t nargs)
 	case 2:
 		chanwidth = CONF_OPER_CHWIDTH_160MHZ;
 		break;
+	case 9:
+		width = 3;
+		chanwidth = CONF_OPER_CHWIDTH_320MHZ;
+		break;
 	default:
 		return NULL;
 	}
-- 
2.45.2

