From c2fe7f12a2e477e9b7219c1c1d576672842c8287 Mon Sep 17 00:00:00 2001
From: Shayne Chen <shayne.chen@mediatek.com>
Date: Thu, 19 Oct 2023 10:48:11 +0800
Subject: [PATCH 058/236] mtk: wpa_supplicant: correctly get assoc frequency

Signed-off-by: Shayne Chen <shayne.chen@mediatek.com>
---
 src/drivers/driver_nl80211_event.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/src/drivers/driver_nl80211_event.c b/src/drivers/driver_nl80211_event.c
index bf95a08ee..57eb45451 100644
--- a/src/drivers/driver_nl80211_event.c
+++ b/src/drivers/driver_nl80211_event.c
@@ -331,6 +331,7 @@ static void mlme_event_assoc(struct wpa_driver_nl80211_data *drv,
 			   wpa_ssid_txt(drv->ssid, drv->ssid_len));
 	}
 
+	drv->assoc_freq = nl80211_get_assoc_freq(drv);
 	event.assoc_info.freq = drv->assoc_freq;
 	drv->first_bss->flink->freq = drv->assoc_freq;
 
-- 
2.45.2

