From 5a922044b09d8087d5b0813d10de297427dcc76e Mon Sep 17 00:00:00 2001
From: Michael-CY Lee <michael-cy.lee@mediatek.com>
Date: Fri, 1 Mar 2024 16:59:53 +0800
Subject: [PATCH 067/236] mtk: hostapd: prevent getting non-MLD STA for other
 links

Signed-off-by: Michael-CY Lee <michael-cy.lee@mediatek.com>
---
 src/ap/sta_info.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/src/ap/sta_info.c b/src/ap/sta_info.c
index a15210154..e8712fcad 100644
--- a/src/ap/sta_info.c
+++ b/src/ap/sta_info.c
@@ -87,7 +87,8 @@ struct sta_info * ap_get_sta(struct hostapd_data *hapd, const u8 *sta)
 			for (s = h->sta_list; s; s = s->next)
 				if ((!os_memcmp(s->setup_link_addr, sta, 6) ||
 				     !os_memcmp(s->addr, sta, 6)) &&
-				     s->flags & WLAN_STA_ASSOC)
+				     s->flags & WLAN_STA_ASSOC &&
+				     s->mld_info.mld_sta)
 					return s;
 		}
 	}
-- 
2.45.2

