From ca6d27a6d83440362fab501461cf12af1666b3db Mon Sep 17 00:00:00 2001
From: Michael-CY Lee <michael-cy.lee@mediatek.com>
Date: Wed, 10 Jul 2024 16:13:24 +0800
Subject: [PATCH 102/236] mtk: hostapd: do not consider ht operation update for
 6G BSS

Signed-off-by: Michael-CY Lee <michael-cy.lee@mediatek.com>
---
 src/ap/ieee802_11.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/src/ap/ieee802_11.c b/src/ap/ieee802_11.c
index bfd86d220..419c93e57 100644
--- a/src/ap/ieee802_11.c
+++ b/src/ap/ieee802_11.c
@@ -4954,7 +4954,8 @@ int ieee80211_ml_process_link(struct hostapd_data *hapd,
 		}
 		hapd->sta_aid[(sta->aid - 1) / 32] |= BIT((sta->aid - 1) % 32);
 		sta->listen_interval = origin_sta->listen_interval;
-		if (update_ht_state(hapd, sta) > 0)
+		if (!is_6ghz_op_class(hapd->iconf->op_class) &&
+		    update_ht_state(hapd, sta) > 0)
 			ieee802_11_update_beacons(hapd->iface);
 	}
 
-- 
2.45.2

