From 49ba34bee90a9843ddd6102e60b02b70b0878ea5 Mon Sep 17 00:00:00 2001
From: Michael-CY Lee <michael-cy.lee@mediatek.com>
Date: Thu, 26 Dec 2024 11:14:14 +0800
Subject: [PATCH 139/236] mtk: hostapd: fix ap_mld access on wrong hapd

Signed-off-by: Michael-CY Lee <michael-cy.lee@mediatek.com>
---
 src/ap/ieee802_11.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/ap/ieee802_11.c b/src/ap/ieee802_11.c
index e1f05e4e0..a0fe13bcb 100644
--- a/src/ap/ieee802_11.c
+++ b/src/ap/ieee802_11.c
@@ -8411,7 +8411,7 @@ static bool hostapd_eid_rnr_bss(struct hostapd_data *hapd,
 
 	ignore_broadcast_ssid = bss->conf->ignore_broadcast_ssid;
 #ifdef CONFIG_IEEE80211BE
-	ap_mld = !!hapd->conf->mld_ap;
+	ap_mld = !!bss->conf->mld_ap;
 	/* FIXME How to exclude the hidden link in beacon? */
 	ignore_broadcast_ssid &= !hostapd_is_ml_partner(bss, reporting_hapd);
 #endif /* CONFIG_IEEE80211BE */
-- 
2.45.2

