From 20b4a05521067c675ee87c0869544fe46e5250bd Mon Sep 17 00:00:00 2001
From: Michael-CY Lee <michael-cy.lee@mediatek.com>
Date: Mon, 9 Dec 2024 11:38:54 +0800
Subject: [PATCH 142/236] mtk: hostapd: WPS: use passphrase over psk file

Signed-off-by: Michael-CY Lee <michael-cy.lee@mediatek.com>
---
 src/ap/wps_hostapd.c | 7 ++++---
 1 file changed, 4 insertions(+), 3 deletions(-)

diff --git a/src/ap/wps_hostapd.c b/src/ap/wps_hostapd.c
index 8a6fc42b2..0429080d1 100644
--- a/src/ap/wps_hostapd.c
+++ b/src/ap/wps_hostapd.c
@@ -1223,9 +1223,10 @@ int hostapd_init_wps(struct hostapd_data *hapd,
 		wps->auth_types |= WPS_AUTH_OPEN;
 	}
 
-	if (conf->ssid.wpa_psk_file) {
-		/* Use per-device PSKs */
-	} else if (conf->ssid.wpa_passphrase) {
+	if (conf->ssid.wpa_psk_file)
+		wpa_printf(MSG_DEBUG, "WPS: allow using per-device PSKs");
+
+	if (conf->ssid.wpa_passphrase) {
 		wps->network_key = (u8 *) os_strdup(conf->ssid.wpa_passphrase);
 		wps->network_key_len = os_strlen(conf->ssid.wpa_passphrase);
 	} else if (conf->ssid.wpa_psk) {
-- 
2.45.2

