From 87eefa8457363925e871b1e761f821b518daf772 Mon Sep 17 00:00:00 2001
From: Shayne Chen <shayne.chen@mediatek.com>
Date: Wed, 30 Apr 2025 18:00:48 +0800
Subject: [PATCH 195/236] mtk: hostapd: tmp fix to adapt to new PMKSA flow

---
 src/ap/ieee802_11.c | 20 +++++---------------
 1 file changed, 5 insertions(+), 15 deletions(-)

diff --git a/src/ap/ieee802_11.c b/src/ap/ieee802_11.c
index 5bc5b9a6d..83fb824d8 100644
--- a/src/ap/ieee802_11.c
+++ b/src/ap/ieee802_11.c
@@ -1164,21 +1164,11 @@ void sae_accept_sta(struct hostapd_data *hapd, struct sta_info *sta)
 	crypto_bignum_deinit(sta->sae->peer_commit_scalar_accepted, 0);
 	sta->sae->peer_commit_scalar_accepted = sta->sae->peer_commit_scalar;
 	sta->sae->peer_commit_scalar = NULL;
-	if (hostapd_is_mld_ap(hapd)) {
-		for_each_mld_link(link, hapd) {
-			if (!link->started)
-				continue;
-			wpa_auth_pmksa_add_sae(link->wpa_auth, sta->addr,
-					sta->sae->pmk, sta->sae->pmk_len,
-					sta->sae->pmkid, sta->sae->akmp,
-					ap_sta_is_mld(hapd, sta));
-		}
-	} else {
-		wpa_auth_pmksa_add_sae(hapd->wpa_auth, sta->addr,
-				sta->sae->pmk, sta->sae->pmk_len,
-				sta->sae->pmkid, sta->sae->akmp,
-				ap_sta_is_mld(hapd, sta));
-	}
+
+	wpa_auth_pmksa_add_sae(hapd->wpa_auth, sta->addr,
+			       sta->sae->pmk, sta->sae->pmk_len,
+			       sta->sae->pmkid, sta->sae->akmp,
+			       ap_sta_is_mld(hapd, sta));
 	sae_sme_send_external_auth_status(hapd, sta, WLAN_STATUS_SUCCESS);
 }
 
-- 
2.45.2

