diff -upr a/hostapd/config_file.c b/hostapd/config_file.c
--- a/hostapd/config_file.c	2025-11-03 01:30:34.636779869 +0800
+++ b/hostapd/config_file.c	2025-11-03 01:30:56.316946970 +0800
@@ -2455,6 +2455,8 @@ static int hostapd_config_fill(struct ho
 			return 1;
 		}
 		conf->driver = driver;
+	} else if (os_strcmp(buf, "uci_section") == 0) {
+		bss->uci_section = os_strdup(pos);
 	} else if (os_strcmp(buf, "driver_params") == 0) {
 		os_free(conf->driver_params);
 		conf->driver_params = os_strdup(pos);
diff -upr a/src/ap/ap_config.c b/src/ap/ap_config.c
--- a/src/ap/ap_config.c	2025-11-03 01:30:34.692780301 +0800
+++ b/src/ap/ap_config.c	2025-11-03 01:30:56.316946970 +0800
@@ -876,6 +876,7 @@ void hostapd_config_free_bss(struct host
 	os_free(conf->radius_req_attr_sqlite);
 	os_free(conf->rsn_preauth_interfaces);
 	os_free(conf->ctrl_interface);
+	os_free(conf->uci_section);
 	os_free(conf->config_id);
 	os_free(conf->ca_cert);
 	os_free(conf->server_cert);
diff -upr a/src/ap/ap_config.h b/src/ap/ap_config.h
--- a/src/ap/ap_config.h	2025-11-03 01:30:34.692780301 +0800
+++ b/src/ap/ap_config.h	2025-11-03 01:30:56.316946970 +0800
@@ -290,6 +290,7 @@ struct hostapd_bss_config {
 	char vlan_bridge[IFNAMSIZ + 1];
 	char wds_bridge[IFNAMSIZ + 1];
 	int bridge_hairpin; /* hairpin_mode on bridge members */
+	char *uci_section;
 
 	enum hostapd_logger_level logger_syslog_level, logger_stdout_level;
 
diff -upr a/src/ap/ubus.c b/src/ap/ubus.c
--- a/src/ap/ubus.c	2025-11-03 01:30:34.692780301 +0800
+++ b/src/ap/ubus.c	2025-11-03 01:30:56.316946970 +0800
@@ -458,6 +458,9 @@ hostapd_bss_get_status(struct ubus_conte
 	if (hapd->signal_mgmt)
 		blobmsg_add_u32(&b, "signal_mgmt", hapd->signal_mgmt);
 
+	if (hapd->conf->uci_section)
+		blobmsg_add_string(&b, "uci_section", hapd->conf->uci_section);
+
 	ubus_send_reply(ctx, req, b.head);
 
 	return 0;
