diff -upr a/hostapd/config_file.c b/hostapd/config_file.c
--- a/hostapd/config_file.c	2025-11-03 01:32:26.353642456 +0800
+++ b/hostapd/config_file.c	2025-11-03 01:32:44.381781983 +0800
@@ -3601,6 +3601,8 @@ static int hostapd_config_fill(struct ho
 		bss->ignore_broadcast_ssid = atoi(pos);
 	} else if (os_strcmp(buf, "no_probe_resp_if_max_sta") == 0) {
 		bss->no_probe_resp_if_max_sta = atoi(pos);
+	} else if (os_strcmp(buf, "dynamic_probe_resp") == 0) {
+		bss->dynamic_probe_resp = atoi(pos);
 #ifdef CONFIG_WEP
 	} else if (os_strcmp(buf, "wep_default_key") == 0) {
 		bss->ssid.wep.idx = atoi(pos);
diff -upr a/src/ap/ap_config.h b/src/ap/ap_config.h
--- a/src/ap/ap_config.h	2025-11-03 01:32:26.409642890 +0800
+++ b/src/ap/ap_config.h	2025-11-03 01:32:44.381781983 +0800
@@ -486,6 +486,7 @@ struct hostapd_bss_config {
 	bool no_disconnect_on_group_keyerror;
 	int ignore_broadcast_ssid;
 	int no_probe_resp_if_max_sta;
+	int dynamic_probe_resp;
 
 	int wmm_enabled;
 	int wmm_uapsd;
diff -upr a/src/ap/beacon.c b/src/ap/beacon.c
--- a/src/ap/beacon.c	2025-11-03 01:32:26.413642921 +0800
+++ b/src/ap/beacon.c	2025-11-03 01:32:44.381781983 +0800
@@ -1548,7 +1548,8 @@ void handle_probe_req(struct hostapd_dat
 	}
 #endif /* CONFIG_P2P */
 
-	if (hapd->conf->ignore_broadcast_ssid && elems.ssid_len == 0 &&
+	if (!hapd->conf->dynamic_probe_resp &&
+	    hapd->conf->ignore_broadcast_ssid && elems.ssid_len == 0 &&
 	    elems.ssid_list_len == 0 && elems.short_ssid_list_len == 0) {
 		wpa_printf(MSG_MSGDUMP, "Probe Request from " MACSTR " for "
 			   "broadcast SSID ignored", MAC2STR(mgmt->sa));
@@ -1595,7 +1596,8 @@ void handle_probe_req(struct hostapd_dat
 		return;
 	}
 
-	if (hapd->conf->ignore_broadcast_ssid && res == WILDCARD_SSID_MATCH) {
+	if (!hapd->conf->dynamic_probe_resp &&
+	    hapd->conf->ignore_broadcast_ssid && res == WILDCARD_SSID_MATCH) {
 		wpa_printf(MSG_MSGDUMP, "Probe Request from " MACSTR " for "
 			   "broadcast SSID ignored", MAC2STR(mgmt->sa));
 		return;
