diff -upr a/src/ap/hostapd.c b/src/ap/hostapd.c
--- a/src/ap/hostapd.c	2025-11-03 01:37:11.748078665 +0800
+++ b/src/ap/hostapd.c	2025-11-03 01:37:24.480315119 +0800
@@ -1316,6 +1316,23 @@ hostapd_das_disconnect(void *ctx, struct
 	struct hostapd_data *hapd = ctx;
 	struct sta_info *sta;
 	int multi;
+	int ubus_resp;
+	struct hostapd_ubus_request req = {
+		.type = HOSTAPD_UBUS_COA,
+		.mgmt_frame = 0,
+		.ssi_signal = 0,
+		.addr = attr->sta_addr,
+	};
+
+	if (hostapd_ubus_handle_event(hapd, &req)) {
+		wpa_printf(MSG_INFO, "DAS: disconnect due approved via ubus");
+		sta = ap_get_sta(hapd, attr->sta_addr);
+		if (sta) {
+			hostapd_drv_sta_deauth(hapd, attr->sta_addr, 2);
+			ap_sta_deauthenticate(hapd, sta, 2);
+		}
+		return RADIUS_DAS_SUCCESS;
+	}
 
 	if (hostapd_das_nas_mismatch(hapd, attr))
 		return RADIUS_DAS_NAS_MISMATCH;
