diff -upr a/src/ap/beacon.c b/src/ap/beacon.c
--- a/src/ap/beacon.c	2025-11-03 01:49:07.522291994 +0800
+++ b/src/ap/beacon.c	2025-11-03 01:49:24.750489055 +0800
@@ -1459,8 +1459,10 @@ void handle_probe_req(struct hostapd_dat
 	};
 
 	if (hapd->iconf->rssi_ignore_probe_request && ssi_signal &&
-	    ssi_signal < hapd->iconf->rssi_ignore_probe_request)
+	    ssi_signal < hapd->iconf->rssi_ignore_probe_request) {
+		hostapd_ubus_notify_rssi(hapd, "rssi-ignore-probe", mgmt->sa, ssi_signal);
 		return;
+	}
 
 	if (len < IEEE80211_HDRLEN)
 		return;
