diff -upr a/src/ap/hostapd.c b/src/ap/hostapd.c
--- a/src/ap/hostapd.c	2025-11-03 01:56:03.518756412 +0800
+++ b/src/ap/hostapd.c	2025-11-03 01:56:23.218956220 +0800
@@ -1909,19 +1909,15 @@ setup_mld:
 		if (x_snoop_init(hapd)) {
 			wpa_printf(MSG_ERROR,
 				   "Generic snooping infrastructure initialization failed");
-			return -1;
-		}
-
-		if (dhcp_snoop_init(hapd)) {
+			conf->proxy_arp = 0;
+		} else if (dhcp_snoop_init(hapd)) {
 			wpa_printf(MSG_ERROR,
 				   "DHCP snooping initialization failed");
-			return -1;
-		}
-
-		if (ndisc_snoop_init(hapd)) {
+			conf->proxy_arp = 0;
+		} else if (ndisc_snoop_init(hapd)) {
 			wpa_printf(MSG_ERROR,
 				   "Neighbor Discovery snooping initialization failed");
-			return -1;
+			conf->proxy_arp = 0;
 		}
 	}
 
