From 910fab53a7a910f60cda7717c31bbbb82e9778c1 Mon Sep 17 00:00:00 2001
From: zhongjia <quic_zhongjia@quicinc.com>
Date: Wed, 18 Oct 2023 14:56:58 +0800
Subject: [PATCH] drivers: net: enable the Werror for the dir of drivers/net

Change-Id: Ib2e66c909d01d1984e5c0529a4826a86d69d8974
Signed-off-by: zhongjia <quic_zhongjia@quicinc.com>
---
 drivers/net/Makefile | 1 -
 1 file changed, 1 deletion(-)

diff --git a/drivers/net/Makefile b/drivers/net/Makefile
index caf0a6b7f97f..6ce076462dbf 100644
--- a/drivers/net/Makefile
+++ b/drivers/net/Makefile
@@ -6,7 +6,6 @@
 #
 # Networking Core Drivers
 #
-KBUILD_CFLAGS := $(filter-out -Werror, $(KBUILD_CFLAGS))
 obj-$(CONFIG_BONDING) += bonding/
 obj-$(CONFIG_IPVLAN) += ipvlan/
 obj-$(CONFIG_IPVTAP) += ipvlan/
-- 
2.34.1

