From 2bd43f099b44fb5a6f79f71051c58f900c54f6f9 Mon Sep 17 00:00:00 2001
From: luozhan <luozhan@cigtech.com>
Date: Thu, 3 Apr 2025 17:45:50 +0800
Subject: [PATCH] Reduce mdc frequency

---
 drivers/net/mdio/mdio-ipq4019.c | 19 ++++++++++++++++++-
 1 file changed, 18 insertions(+), 1 deletion(-)

diff --git a/drivers/net/mdio/mdio-ipq4019.c b/drivers/net/mdio/mdio-ipq4019.c
index 6aa30ed70..c703c8a8d 100644
--- a/drivers/net/mdio/mdio-ipq4019.c
+++ b/drivers/net/mdio/mdio-ipq4019.c
@@ -871,6 +871,7 @@ static int ipq4019_mdio_probe(struct platform_device *pdev)
 	struct mii_bus *bus;
 	struct resource *res;
 	int ret, i;
+	int clk_div = 0xf, flag = 0;
 
 	bus = devm_mdiobus_alloc_size(&pdev->dev, sizeof(*priv));
 	if (!bus)
@@ -891,6 +892,13 @@ static int ipq4019_mdio_probe(struct platform_device *pdev)
 		if (IS_ERR(priv->membase[1]))
 			return PTR_ERR(priv->membase[1]);
 	}
+	
+	if (0 == of_property_read_u32(pdev->dev.of_node, "cig_clk_div", &ret)) //read the mdio clock value from dts
+	{	
+		clk_div = ret;
+		flag = 1;
+		dev_info(&pdev->dev,"CIG clk_div =%x\n",clk_div);
+	}
 
 	for (i = 0; i < MDIO_CLK_CNT; i++) {
 		priv->clk[i] = devm_clk_get_optional(&pdev->dev, ppe_clk_name[i]);
@@ -918,7 +926,16 @@ static int ipq4019_mdio_probe(struct platform_device *pdev)
 	}
 
 	/* MDIO default frequency is 6.25MHz */
-	priv->clk_div = 0xf;
+	//priv->clk_div = 0xf;
+	if(flag)
+	{
+		priv->clk_div = clk_div;
+	}
+	else
+	{
+		priv->clk_div = 0xf;
+	}
+
 	priv->force_c22 = of_property_read_bool(pdev->dev.of_node, "force_clause22");
 
 	priv->preinit = ipq_mii_preinit;
-- 
2.34.1

