From 2bd43f099b44fb5a6f79f71051c58f900c54f6f9 Mon Sep 17 00:00:00 2001
From: luozhan <luozhan@cigtech.com>
Date: Thu, 3 Apr 2025 17:45:50 +0800
Subject: [PATCH] Reduce mdc frequency

---
 drivers/net/mdio/mdio-ipq4019.c | 19 ++++++++++++++++++-
 1 file changed, 18 insertions(+), 1 deletion(-)

--- a/drivers/net/mdio/mdio-ipq4019.c
+++ b/drivers/net/mdio/mdio-ipq4019.c
@@ -919,6 +919,9 @@ static int ipq4019_mdio_probe(struct pla
 
 	/* MDIO default frequency is 6.25MHz */
 	priv->clk_div = 0xf;
+	if (!of_property_read_u32(pdev->dev.of_node, "tip,clk_div", &priv->clk_div))
+		dev_info(&pdev->dev, "overriding clk_div = %x\n", priv->clk_div);
+
 	priv->force_c22 = of_property_read_bool(pdev->dev.of_node, "force_clause22");
 
 	priv->preinit = ipq_mii_preinit;
