--- a/drivers/input/misc/Kconfig	2025-04-22 14:35:28.228629072 +0800
+++ b/drivers/input/misc/Kconfig	2025-04-22 14:42:29.630935581 +0800
@@ -929,4 +929,5 @@
 	  To compile this driver as a module, choose M here: the
 	  module will be called stpmic1_onkey.

+source "drivers/input/misc/lsm303agr/Kconfig"
 endif
--- a/drivers/input/misc/Makefile	2025-04-22 14:35:28.228629072 +0800
+++ b/drivers/input/misc/Makefile	2025-04-22 14:41:23.840267623 +0800
@@ -89,3 +89,4 @@
 obj-$(CONFIG_INPUT_XEN_KBDDEV_FRONTEND)	+= xen-kbdfront.o
 obj-$(CONFIG_INPUT_YEALINK)		+= yealink.o
 obj-$(CONFIG_INPUT_IDEAPAD_SLIDEBAR)	+= ideapad_slidebar.o
+obj-$(CONFIG_INPUT_LSM303AGR)		+= lsm303agr/
--- a/drivers/net/phy/Kconfig	2025-04-22 14:48:04.858191804 +0800
+++ b/drivers/net/phy/Kconfig	2025-04-22 10:17:06.822705335 +0800
@@ -99,6 +99,9 @@
 	tristate "Lantiq XWAY Tantos (PSB6970) Ethernet switch"
 	select SWCONFIG
 
+config RTL8221D_PHY
+	tristate "Driver for Realtek RTL8221D phy"
+
 config RTL8306_PHY
 	tristate "Driver for Realtek RTL8306S switches"
 	select SWCONFIG
--- a/drivers/net/phy/Makefile	2025-04-22 14:48:18.254372527 +0800
+++ b/drivers/net/phy/Makefile	2025-04-22 10:17:10.546732447 +0800
@@ -32,6 +32,7 @@
 obj-$(CONFIG_SWCONFIG_B53)	+= b53/
 obj-$(CONFIG_IP17XX_PHY)	+= ip17xx.o
 obj-$(CONFIG_PSB6970_PHY)	+= psb6970.o
+obj-$(CONFIG_RTL8221D_PHY)	+= rtl8221d.o
 obj-$(CONFIG_RTL8306_PHY)	+= rtl8306.o
 obj-$(CONFIG_RTL8366_SMI)	+= rtl8366_smi.o
 obj-$(CONFIG_RTL8366S_PHY)	+= rtl8366s.o
--- a/drivers/iio/pressure/Kconfig  2025-05-13 15:16:05.019840003 +0800
+++ b/drivers/iio/pressure/Kconfig  2025-05-13 15:17:45.032040654 +0800
@@ -266,4 +266,11 @@
 	tristate
 	select REGMAP_SPI
 
+config ILPS22QS
+	tristate "ILPS22QS pressure support"
+	help
+	  Say Y to enable support ILPS22QS
+
+	  To compile this driver as a module, choose M here: the
+	  module will be called ILPS22QS.
 endmenu

--- a/drivers/iio/pressure/Makefile  2025-05-13 15:17:58.049195788 +0800
+++ b/drivers/iio/pressure/Makefile  2025-05-13 15:19:07.205016058 +0800
@@ -31,6 +31,8 @@
 obj-$(CONFIG_ZPA2326) += zpa2326.o
 obj-$(CONFIG_ZPA2326_I2C) += zpa2326_i2c.o
 obj-$(CONFIG_ZPA2326_SPI) += zpa2326_spi.o
+obj-$(CONFIG_ILPS22QS) += st_ilps22qs.o
+st_ilps22qs-objs := st_ilps22qs_i2c.o st_ilps22qs_core.o
 
 obj-$(CONFIG_IIO_ST_PRESS_I2C) += st_pressure_i2c.o
 obj-$(CONFIG_IIO_ST_PRESS_SPI) += st_pressure_spi.o
