--- a/drivers/net/wwan/Kconfig	2023-05-30 21:03:33.000000000 +0800
+++ b/drivers/net/wwan/Kconfig	2025-08-04 09:51:43.678043884 +0800
@@ -119,6 +119,15 @@
 
 	  If unsure, say N.
 
+config PCIE_MHI_QUECTEL
+        tristate "PCIE MHI driver"
+        help
+          Say Y here if you want to use mhi 
+          devices.
+
+          To compile this driver as a module, choose M here: the 
+          module will be called pcie_mhi.
+
 endif # WWAN
 
 endmenu
--- a/drivers/net/wwan/Makefile	2023-05-30 21:03:33.000000000 +0800
+++ b/drivers/net/wwan/Makefile	2025-08-04 09:51:43.678043884 +0800
@@ -14,3 +14,4 @@
 obj-$(CONFIG_RPMSG_WWAN_CTRL) += rpmsg_wwan_ctrl.o
 obj-$(CONFIG_IOSM) += iosm/
 obj-$(CONFIG_MTK_T7XX) += t7xx/
+obj-$(CONFIG_PCIE_MHI_QUECTEL) += pcie_mhi/
