--- a/drivers/input/misc/lsm303agr/lsm303agr_acc_i2c.c
+++ b/drivers/input/misc/lsm303agr/lsm303agr_acc_i2c.c
@@ -55,10 +55,13 @@ static int lsm303agr_acc_i2c_read(struct
 static int lsm303agr_acc_i2c_write(struct device *dev, u8 reg_addr, int len,
 				   u8 *data)
 {
-	u8 send[len + 1];
+	u8 send[8];
 	struct i2c_msg msg;
 	struct i2c_client *client = to_i2c_client(dev);
 
+	if (len >= sizeof(send))
+		return -EINVAL;
+
 	if (len > 1)
 		reg_addr |= I2C_AUTO_INCREMENT;
 
--- a/drivers/input/misc/lsm303agr/lsm303agr_mag_i2c.c
+++ b/drivers/input/misc/lsm303agr/lsm303agr_mag_i2c.c
@@ -55,10 +55,13 @@ static int lsm303agr_mag_i2c_read(struct
 static int lsm303agr_mag_i2c_write(struct device *dev, u8 reg_addr, int len,
 				   u8 *data)
 {
-	u8 send[len + 1];
+	u8 send[8];
 	struct i2c_msg msg;
 	struct i2c_client *client = to_i2c_client(dev);
 
+	if (len >= sizeof(send))
+		return -EINVAL;
+
 	if (len > 1)
 		reg_addr |= I2C_AUTO_INCREMENT;
 
