--- a/drivers/net/wireless/ath/ath11k/rx_desc.h
+++ b/drivers/net/wireless/ath/ath11k/rx_desc.h
@@ -1375,11 +1375,9 @@ struct hal_rx_desc_ipq8074 {
 	__le32 mpdu_end_tag;
 	struct rx_mpdu_end mpdu_end;
 	u8 rx_padding1[HAL_RX_DESC_PADDING1_BYTES];
-#ifndef CPTCFG_ATH11K_MEM_PROFILE_512M
 	__le32 hdr_status_tag;
 	__le32 phy_ppdu_id;
 	u8 hdr_status[HAL_RX_DESC_HDR_STATUS_LEN];
-#endif
 	u8 msdu_payload[0];
 } __packed;
 
@@ -1396,11 +1394,9 @@ struct hal_rx_desc_qcn9074 {
 	__le32 mpdu_end_tag;
 	struct rx_mpdu_end mpdu_end;
 	u8 rx_padding1[HAL_RX_DESC_PADDING1_BYTES];
-#ifndef CPTCFG_ATH11K_MEM_PROFILE_512M
 	__le32 hdr_status_tag;
 	__le32 phy_ppdu_id;
 	u8 hdr_status[HAL_RX_DESC_HDR_STATUS_LEN];
-#endif
 	u8 msdu_payload[0];
 } __packed;
 
