--- a/drivers/net/wireless/ath/ath11k/peer.c
+++ b/drivers/net/wireless/ath/ath11k/peer.c
@@ -883,6 +883,7 @@ int ath11k_peer_create(struct ath11k *ar
 	struct ieee80211_vif *vif = arvif->vif;
 	struct ath11k_sta *arsta;
 	int ret, fbret;
+	u8 vdev_id = 0;
 
 	lockdep_assert_held(&ar->conf_mutex);
 
@@ -891,20 +892,21 @@ int ath11k_peer_create(struct ath11k *ar
 			    "failed to create peer due to insufficient peer entry resource in firmware\n");
 		return -ENOBUFS;
 	}
+
 	mutex_lock(&ar->ab->tbl_mtx_lock);
 	spin_lock_bh(&ar->ab->base_lock);
 	peer = ath11k_peer_find_by_addr(ar->ab, param->peer_addr);
+	if (peer)
+		vdev_id = peer->vdev_id;
+	spin_unlock_bh(&ar->ab->base_lock);
+	mutex_unlock(&ar->ab->tbl_mtx_lock);
+
 	if (peer) {
-		if (peer->vdev_id == param->vdev_id) {
-			spin_unlock_bh(&ar->ab->base_lock);
-			mutex_unlock(&ar->ab->tbl_mtx_lock);
+		if (vdev_id == param->vdev_id)
 			return -EINVAL;
-		}
-		ath11k_peer_rhash_delete(ar->ab, peer);
-	}
 
-	spin_unlock_bh(&ar->ab->base_lock);
-	mutex_unlock(&ar->ab->tbl_mtx_lock);
+		ath11k_peer_delete(ar, vdev_id, param->peer_addr);
+	}
 
 	ret = ath11k_wmi_send_peer_create_cmd(ar, param);
 	if (ret) {
