From 046d7ac7c26271a215726b28ef68933dfaa34ff0 Mon Sep 17 00:00:00 2001
From: Poovendhan Selvaraj <quic_poovendh@quicinc.com>
Date: Wed, 24 Jan 2024 15:09:41 +0530
Subject: [PATCH 500/500] drivers: qcom: ice: Fix compilation issue for 32 bit

Signed-off-by: Poovendhan Selvaraj <quic_poovendh@quicinc.com>
---
 drivers/soc/qcom/ice.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/drivers/soc/qcom/ice.c b/drivers/soc/qcom/ice.c
index 389f5ebc3b9e..de038e49d074 100644
--- a/drivers/soc/qcom/ice.c
+++ b/drivers/soc/qcom/ice.c
@@ -14,6 +14,7 @@
 #include <linux/of.h>
 #include <linux/of_platform.h>
 #include <linux/platform_device.h>
+#include <linux/slab.h>
 
 #include <linux/firmware/qcom/qcom_scm.h>
 
-- 
2.34.1

