--- a/drivers/net/phy/Kconfig	2025-04-22 14:48:04.858191804 +0800
+++ b/drivers/net/phy/Kconfig	2025-04-22 10:17:06.822705335 +0800
@@ -99,6 +99,9 @@
 	tristate "Lantiq XWAY Tantos (PSB6970) Ethernet switch"
 	select SWCONFIG
 
+config RTL8221D_PHY
+	tristate "Driver for Realtek RTL8221D phy"
+
 config RTL8306_PHY
 	tristate "Driver for Realtek RTL8306S switches"
 	select SWCONFIG
--- a/drivers/net/phy/Makefile	2025-04-22 14:48:18.254372527 +0800
+++ b/drivers/net/phy/Makefile	2025-04-22 10:17:10.546732447 +0800
@@ -32,6 +32,7 @@
 obj-$(CONFIG_SWCONFIG_B53)	+= b53/
 obj-$(CONFIG_IP17XX_PHY)	+= ip17xx.o
 obj-$(CONFIG_PSB6970_PHY)	+= psb6970.o
+obj-$(CONFIG_RTL8221D_PHY)	+= rtl8221d.o
 obj-$(CONFIG_RTL8306_PHY)	+= rtl8306.o
 obj-$(CONFIG_RTL8366_SMI)	+= rtl8366_smi.o
 obj-$(CONFIG_RTL8366S_PHY)	+= rtl8366s.o
