From: aparna.salian <aparna.salian@globaledgesoft.com>
Subject: [PATCH] lib/hostap: Check bridge in dhcp snoop init

Fixes 900-hapd-netlink-ubus-bridge.patch

--- a/src/ap/dhcp_snoop.c
+++ b/src/ap/dhcp_snoop.c
@@ -152,6 +152,9 @@
 
 int dhcp_snoop_init(struct hostapd_data *hapd)
 {
+	if (!strlen(hapd->conf->bridge))
+		return 0;
+
 	hapd->sock_dhcp = x_snoop_get_l2_packet(hapd, handle_dhcp,
 						L2_PACKET_FILTER_DHCP);
 	if (hapd->sock_dhcp == NULL) {
