Index: linux-5.4.213-qsdk-f2ee796db820cd8261889849f043c9c816196452/arch/arm64/boot/dts/qcom/ipq6018.dtsi
===================================================================
--- linux-5.4.213-qsdk-f2ee796db820cd8261889849f043c9c816196452.orig/arch/arm64/boot/dts/qcom/ipq6018.dtsi
+++ linux-5.4.213-qsdk-f2ee796db820cd8261889849f043c9c816196452/arch/arm64/boot/dts/qcom/ipq6018.dtsi
@@ -1882,6 +1882,7 @@
 
 	qti,scm_restart_reason {
 		compatible = "qti_ipq6018,scm_restart_reason";
+		dload_status = <1>;
 	};
 
 	ctx_save: ctx-save {
Index: linux-5.4.213-qsdk-f2ee796db820cd8261889849f043c9c816196452/arch/arm64/boot/dts/qcom/ipq9574.dtsi
===================================================================
--- linux-5.4.213-qsdk-f2ee796db820cd8261889849f043c9c816196452.orig/arch/arm64/boot/dts/qcom/ipq9574.dtsi
+++ linux-5.4.213-qsdk-f2ee796db820cd8261889849f043c9c816196452/arch/arm64/boot/dts/qcom/ipq9574.dtsi
@@ -3678,6 +3678,7 @@
 
 	qti,scm_restart_reason {
 		compatible = "qti_ipq9574,scm_restart_reason";
+		dload_status = <1>;
 	};
 
 	qti,gadget_diag@0 {
