Index: qca-ssdk-775e4dc814580012e1019f53ff5cdb8968ed4ca8/config
===================================================================
--- qca-ssdk-775e4dc814580012e1019f53ff5cdb8968ed4ca8.orig/config
+++ qca-ssdk-775e4dc814580012e1019f53ff5cdb8968ed4ca8/config
@@ -283,6 +283,7 @@ else ifeq (HPPE, $(CHIP_TYPE))
 	IN_SFP_PHY=TRUE
 	IN_MALIBU_PHY=TRUE
 else ifeq (CPPE, $(CHIP_TYPE))
+	IN_QCA803X_PHY=TRUE
 	IN_QCA808X_PHY=TRUE
 	IN_PHY_I2C_MODE=TRUE
 	IN_MALIBU_PHY=TRUE
