From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: John Crispin <john@phrozen.org>
Date: Sat, 31 Jan 2026 00:00:00 +0000
Subject: [PATCH] cmake: bump minimum version to 3.5

CMake has removed compatibility with versions below 3.5.

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -1,4 +1,4 @@
-cmake_minimum_required(VERSION 2.8)
+cmake_minimum_required(VERSION 3.5)

 project(rtty C)

