Index: rtty-7.1.4/src/main.c
===================================================================
--- rtty-7.1.4.orig/src/main.c
+++ rtty-7.1.4/src/main.c
@@ -37,6 +37,8 @@ enum {
     LONG_OPT_HELP = 1
 };
 
+static int force_exit;
+
 static void signal_cb(struct ev_loop *loop, ev_signal *w, int revents)
 {
     if (w->signum == SIGINT) {
@@ -53,6 +55,7 @@ static struct option long_options[] = {
     {"token",       required_argument, NULL, 't'},
     {"verbose",     no_argument,       NULL, 'v'},
     {"version",     no_argument,       NULL, 'V'},
+    {"exit",        required_argument, NULL, 'V'},
     {"help",        no_argument,       NULL, LONG_OPT_HELP},
     {0, 0, 0, 0}
 };
@@ -79,10 +82,16 @@ static void usage(const char *prog)
     exit(1);
 }
 
+static void timeout_cb(EV_P_ ev_timer *w, int revents)
+{
+    exit(1);
+}
+
 int main(int argc, char **argv)
 {
     struct ev_loop *loop = EV_DEFAULT;
     struct ev_signal signal_watcher;
+    struct ev_timer timeout_watcher;
     bool background = false;
     bool verbose = false;
     struct rtty rtty = {
@@ -95,7 +104,7 @@ int main(int argc, char **argv)
     int c;
 
     while (true) {
-        c = getopt_long(argc, argv, "I:h:p:d:asDt:f:RS:vV", long_options, &option_index);
+        c = getopt_long(argc, argv, "I:h:p:d:asDt:f:RS:vVe:", long_options, &option_index);
         if (c == -1)
             break;
 
@@ -146,6 +155,9 @@ int main(int argc, char **argv)
         case LONG_OPT_HELP:
             usage(argv[0]);
             break;
+	case 'e':
+	    force_exit = atoi(optarg);
+	    break;
         default: /* '?' */
             usage(argv[0]);
             break;
@@ -167,6 +179,10 @@ int main(int argc, char **argv)
 
     ev_signal_init(&signal_watcher, signal_cb, SIGINT);
     ev_signal_start(loop, &signal_watcher);
+    if (force_exit) {
+        ev_timer_init(&timeout_watcher, timeout_cb, force_exit, 0.);
+        ev_timer_start(loop, &timeout_watcher);
+    }
 
     if (rtty_start(&rtty) < 0)
         return -1;
