Index: openNDS-8.1.1/linux_openwrt/opennds/files/etc/init.d/opennds
===================================================================
--- openNDS-8.1.1.orig/linux_openwrt/opennds/files/etc/init.d/opennds
+++ openNDS-8.1.1/linux_openwrt/opennds/files/etc/init.d/opennds
@@ -183,6 +183,11 @@ start_service() {
   config_foreach create_instance opennds
 }
 
+service_triggers()
+{
+	procd_add_reload_trigger "opennds"
+}
+
 stop_service() {
   # When procd terminates opennds, it does not exit fast enough.
   # Otherwise procd will restart opennds twice. First time starting
