diff -Nur a/drivers/mtd/ubi/block.c b/drivers/mtd/ubi/block.c
--- a/drivers/mtd/ubi/block.c	2025-11-18 06:29:13.737308414 +0000
+++ b/drivers/mtd/ubi/block.c	2025-11-18 06:29:17.445330040 +0000
@@ -93,6 +93,12 @@
 static DEFINE_MUTEX(devices_mutex);
 static int ubiblock_major;
 
+static char rootfs_volume[256] = "rootfs";
+module_param_string(rootfs_volume, rootfs_volume, sizeof(rootfs_volume), 0444);
+
+static bool no_default_rootdev;
+module_param(no_default_rootdev, bool, 0444);
+
 static int __init ubiblock_set_param(const char *val,
 				     const struct kernel_param *kp)
 {
@@ -430,8 +436,9 @@
 		 dev->ubi_num, dev->vol_id, vi->name);
 	mutex_unlock(&devices_mutex);
 
-	if (!strcmp(vi->name, "rootfs") &&
+	if (!strcmp(vi->name, rootfs_volume) &&
 	    IS_ENABLED(CONFIG_MTD_ROOTFS_ROOT_DEV) &&
+	    !no_default_rootdev &&
 	    ROOT_DEV == 0) {
 		pr_notice("ubiblock: device ubiblock%d_%d (%s) set to be root filesystem\n",
 			  dev->ubi_num, dev->vol_id, vi->name);
@@ -597,7 +604,7 @@
 	int ret, is_ubifs;
 	struct ubi_volume_desc *desc;
 
-	if (strcmp(vi->name, "rootfs") &&
+	if (strcmp(vi->name, rootfs_volume) &&
 	    strcmp(vi->name, "fit"))
 		return;
 
diff -Nur a/init/do_mounts.c b/init/do_mounts.c
--- a/init/do_mounts.c	2025-11-18 06:29:13.737308414 +0000
+++ b/init/do_mounts.c	2025-11-18 06:29:17.441330017 +0000
@@ -409,7 +409,8 @@
 void __init mount_root(char *root_device_name)
 {
 #ifdef CONFIG_MTD_ROOTFS_ROOT_DEV
-	if (!mount_ubi_rootfs())
+	extern bool dual_boot;
+	if (!dual_boot && !mount_ubi_rootfs())
 		return;
 #endif
 
diff -Nur a/kernel/boot_param.c b/kernel/boot_param.c
--- a/kernel/boot_param.c	1970-01-01 00:00:00.000000000 +0000
+++ b/kernel/boot_param.c	2025-11-18 06:29:17.377329644 +0000
@@ -0,0 +1,48 @@
+/* SPDX-License-Identifier: BSD-3-Clause */
+/*
+* Copyright (C) 2022 MediaTek Inc. All rights reserved.
+ *
+ * Author: Weijie Gao <weijie.gao@mediatek.com>
+ */
+
+#include <linux/kernel.h>
+#include <linux/moduleparam.h>
+
+#define BOOT_PARAM_STR_MAX_LEN                 256
+
+bool dual_boot;
+module_param(dual_boot, bool, 0444);
+
+static bool reset_boot_count;
+module_param(reset_boot_count, bool, 0444);
+
+static bool no_split_rootfs_data;
+module_param(no_split_rootfs_data, bool, 0444);
+
+static bool reserve_rootfs_data;
+module_param(reserve_rootfs_data, bool, 0444);
+
+static uint boot_image_slot;
+module_param(boot_image_slot, uint, 0444);
+
+static uint upgrade_image_slot;
+module_param(upgrade_image_slot, uint, 0444);
+
+static char rootfs_data_part[BOOT_PARAM_STR_MAX_LEN];
+module_param_string(rootfs_data_part, rootfs_data_part, BOOT_PARAM_STR_MAX_LEN, 0644);
+
+static char boot_kernel_part[BOOT_PARAM_STR_MAX_LEN];
+module_param_string(boot_kernel_part, boot_kernel_part, BOOT_PARAM_STR_MAX_LEN, 0444);
+
+static char boot_rootfs_part[BOOT_PARAM_STR_MAX_LEN];
+module_param_string(boot_rootfs_part, boot_rootfs_part, BOOT_PARAM_STR_MAX_LEN, 0444);
+
+static char upgrade_kernel_part[BOOT_PARAM_STR_MAX_LEN];
+module_param_string(upgrade_kernel_part, upgrade_kernel_part, BOOT_PARAM_STR_MAX_LEN, 0444);
+
+static char upgrade_rootfs_part[BOOT_PARAM_STR_MAX_LEN];
+module_param_string(upgrade_rootfs_part, upgrade_rootfs_part, BOOT_PARAM_STR_MAX_LEN, 0444);
+
+static char env_part[BOOT_PARAM_STR_MAX_LEN];
+module_param_string(env_part, env_part, BOOT_PARAM_STR_MAX_LEN, 0444);
+
diff -Nur a/kernel/Makefile b/kernel/Makefile
--- a/kernel/Makefile	2025-11-18 06:29:13.633307807 +0000
+++ b/kernel/Makefile	2025-11-18 06:29:17.377329644 +0000
@@ -12,6 +12,8 @@
 	    notifier.o ksysfs.o cred.o reboot.o \
 	    async.o range.o smpboot.o ucount.o regset.o ksyms_common.o
 
+obj-y += boot_param.o
+
 obj-$(CONFIG_USERMODE_DRIVER) += usermode_driver.o
 obj-$(CONFIG_MULTIUSER) += groups.o
 obj-$(CONFIG_VHOST_TASK) += vhost_task.o
