diff --git a/drivers/net/ethernet/mediatek/mtk_eth_soc.c b/drivers/net/ethernet/mediatek/mtk_eth_soc.c
index 721d148..651e2e6 100644
--- a/drivers/net/ethernet/mediatek/mtk_eth_soc.c
+++ b/drivers/net/ethernet/mediatek/mtk_eth_soc.c
@@ -748,18 +759,6 @@ static void mtk_mac_config(struct phylink_config *config, unsigned int mode,
 		mac->id, phy_modes(state->interface), err);
 }
 
-static int mtk_mac_prepare(struct phylink_config *config, unsigned int mode,
-			   phy_interface_t interface)
-{
-	struct mtk_mac *mac = container_of(config, struct mtk_mac,
-					   phylink_config);
-
-	if (mac->pextp && mac->interface != interface)
-		phy_reset(mac->pextp);
-
-	return 0;
-}
-
 static int mtk_mac_finish(struct phylink_config *config, unsigned int mode,
 			  phy_interface_t interface)
 {
@@ -954,6 +953,12 @@ static void mtk_mac_link_up(struct phylink_config *config,
 	struct mtk_mac *mac = container_of(config, struct mtk_mac,
 					   phylink_config);
 
+	/* Repeat pextp setup to tune link */
+	if (mac->pextp) {
+		phy_reset(mac->pextp);
+		phy_set_mode_ext(mac->pextp, PHY_MODE_ETHERNET, interface);
+	}
+
 	if (mtk_is_netsys_v3_or_greater(mac->hw) && mtk_interface_mode_is_xgmii(interface))
 		mtk_xgdm_mac_link_up(mac, phy, mode, interface, speed, duplex,
 				     tx_pause, rx_pause);
@@ -961,15 +966,23 @@ static void mtk_mac_link_up(struct phylink_config *config,
 		mtk_gdm_mac_link_up(mac, phy, mode, interface, speed, duplex,
 				    tx_pause, rx_pause);
 
-	/* Repeat pextp setup to tune link */
-	if (mac->pextp)
-		phy_set_mode_ext(mac->pextp, PHY_MODE_ETHERNET, interface);
+	/* Workaournd for the Realtek RTL8261N */
+	if (phy && (phy->drv->phy_id == 0x001ccaf3)) {
+		mdelay(1000);
+		mdiobus_c45_write(phy->mdio.bus, phy->mdio.addr,
+				  0x1e, 0x141, 0x40);
+		mdiobus_c45_write(phy->mdio.bus, phy->mdio.addr,
+				  0x1e, 0x143, 0x88a0);
+		mdiobus_c45_write(phy->mdio.bus, phy->mdio.addr,
+				  0x1e, 0x141, 0xc0);
+		mdiobus_c45_write(phy->mdio.bus, phy->mdio.addr,
+				  0x1e, 0x143, 0x88a0);
+	}
 }
 
 static const struct phylink_mac_ops mtk_phylink_ops = {
 	.mac_select_pcs = mtk_mac_select_pcs,
 	.mac_config = mtk_mac_config,
-	.mac_prepare = mtk_mac_prepare,
 	.mac_finish = mtk_mac_finish,
 	.mac_link_down = mtk_mac_link_down,
 	.mac_link_up = mtk_mac_link_up,