From 824b3034b05fc0c882bfeb0fada1f7c618d83453 Mon Sep 17 00:00:00 2001
From: Shayne Chen <shayne.chen@mediatek.com>
Date: Tue, 9 Dec 2025 16:19:05 +0800
Subject: [PATCH 14/89] wifi: mt76: mt7996: use mt76_get_txpower()

Signed-off-by: Shayne Chen <shayne.chen@mediatek.com>
---
 mac80211.c    | 21 +++++++++++++++++----
 mt7996/main.c | 28 +---------------------------
 2 files changed, 18 insertions(+), 31 deletions(-)

diff --git a/mac80211.c b/mac80211.c
index 0d5263140..b35913b0e 100644
--- a/mac80211.c
+++ b/mac80211.c
@@ -1770,17 +1770,30 @@ EXPORT_SYMBOL_GPL(mt76_get_power_bound);
 int mt76_get_txpower(struct ieee80211_hw *hw, struct ieee80211_vif *vif,
 		     unsigned int link_id, int *dbm)
 {
-	struct mt76_phy *phy = mt76_vif_phy(hw, vif);
-	int n_chains, delta;
+	struct mt76_phy *phy = hw->priv;
+	struct mt76_dev *dev = phy->dev;
+	struct mt76_vif_link *mlink;
+	int n_chains, delta, ret = -EINVAL;
+
+	mutex_lock(&dev->mutex);
 
+	mlink = mt76_vif_link(dev, vif, link_id);
+	if (!mlink)
+		goto out;
+
+	phy = mt76_vif_link_phy(mlink);
 	if (!phy)
-		return -EINVAL;
+		goto out;
 
 	n_chains = hweight16(phy->chainmask);
 	delta = mt76_tx_power_path_delta(n_chains);
 	*dbm = DIV_ROUND_UP(phy->txpower_cur + delta, 2);
 
-	return 0;
+	ret = 0;
+out:
+	mutex_unlock(&dev->mutex);
+
+	return ret;
 }
 EXPORT_SYMBOL_GPL(mt76_get_txpower);
 
diff --git a/mt7996/main.c b/mt7996/main.c
index 140f62591..155712483 100644
--- a/mt7996/main.c
+++ b/mt7996/main.c
@@ -734,32 +734,6 @@ static void mt7996_configure_filter(struct ieee80211_hw *hw,
 	mutex_unlock(&dev->mt76.mutex);
 }
 
-static int
-mt7996_get_txpower(struct ieee80211_hw *hw, struct ieee80211_vif *vif,
-		   unsigned int link_id, int *dbm)
-{
-	struct mt7996_vif *mvif = (struct mt7996_vif *)vif->drv_priv;
-	struct mt7996_phy *phy = mt7996_vif_link_phy(&mvif->deflink);
-	struct mt7996_dev *dev = mt7996_hw_dev(hw);
-	struct wireless_dev *wdev;
-	int n_chains, delta, i;
-
-	if (!phy) {
-		wdev = ieee80211_vif_to_wdev(vif);
-		for (i = 0; i < hw->wiphy->n_radio; i++)
-			if (wdev->radio_mask & BIT(i))
-				phy = dev->radio_phy[i];
-
-		if (!phy)
-			return -EINVAL;
-	}
-
-	n_chains = hweight16(phy->mt76->chainmask);
-	delta = mt76_tx_power_path_delta(n_chains);
-	*dbm = DIV_ROUND_UP(phy->mt76->txpower_cur + delta, 2);
-
-	return 0;
-}
 static u8
 mt7996_get_rates_table(struct mt7996_phy *phy, struct ieee80211_bss_conf *conf,
 		       bool beacon, bool mcast)
@@ -2401,7 +2375,7 @@ const struct ieee80211_ops mt7996_ops = {
 	.remain_on_channel = mt76_remain_on_channel,
 	.cancel_remain_on_channel = mt76_cancel_remain_on_channel,
 	.release_buffered_frames = mt76_release_buffered_frames,
-	.get_txpower = mt7996_get_txpower,
+	.get_txpower = mt76_get_txpower,
 	.channel_switch_beacon = mt7996_channel_switch_beacon,
 	.post_channel_switch = mt7996_post_channel_switch,
 	.get_stats = mt7996_get_stats,
-- 
2.45.2

