From 1cb9edcc31ee63b6b9f0bc11f7d9fbcf06c71a09 Mon Sep 17 00:00:00 2001
From: Shayne Chen <shayne.chen@mediatek.com>
Date: Fri, 5 Sep 2025 17:04:23 +0800
Subject: [PATCH 33/89] some fixes to compat with internal

---
 mt76.h        | 19 ++++++-------------
 mt7996/main.c | 11 ++++++++++-
 mt7996/mmio.c |  4 ++--
 3 files changed, 18 insertions(+), 16 deletions(-)

diff --git a/mt76.h b/mt76.h
index 29623ea72..fa46bc2fb 100644
--- a/mt76.h
+++ b/mt76.h
@@ -738,25 +738,19 @@ struct mt76_mmio {
 	int npu_type;
 };
 
+/* at most 48 bytes to fit in skb->cb */
 struct mt76_rx_status {
 	union {
 		struct mt76_wcid *wcid;
 		u16 wcid_idx;
 	};
-
 	u32 reorder_time;
-
 	u32 ampdu_ref;
 	u32 timestamp;
-
 	u8 iv[6];
-
-	u8 phy_idx:2;
-	u8 aggr:1;
+	u16 freq:13, phy_idx:2, aggr:1;
 	u8 qos_ctl;
 	u16 seqno;
-
-	u16 freq;
 	u32 flag;
 	u8 enc_flags;
 	u8 encoding:3, bw:4;
@@ -771,14 +765,13 @@ struct mt76_rx_status {
 			u8 gi:2;
 		} eht;
 	};
-
-	u8 amsdu:1, first_amsdu:1, last_amsdu:1;
-	u8 rate_idx;
-	u8 nss:5, band:3;
+	u16 amsdu:1, first_amsdu:1, last_amsdu:1;
+	u16 nss:3, band:3;
+	u16 rate_idx:7;
 	s8 signal;
 	u8 chains;
 	s8 chain_signal[IEEE80211_MAX_CHAINS];
-};
+} __packed;
 
 struct mt76_freq_range_power {
 	const struct cfg80211_sar_freq_ranges *range;
diff --git a/mt7996/main.c b/mt7996/main.c
index 153c58be9..d648785d2 100644
--- a/mt7996/main.c
+++ b/mt7996/main.c
@@ -1816,7 +1816,8 @@ static void mt7996_sta_rate_ctrl_update(void *data, struct ieee80211_sta *sta)
 
 static int
 mt7996_set_bitrate_mask(struct ieee80211_hw *hw, struct ieee80211_vif *vif,
-			const struct cfg80211_bitrate_mask *mask)
+			const struct cfg80211_bitrate_mask *mask,
+			unsigned int link_id)
 {
 	struct mt7996_dev *dev = mt7996_hw_dev(hw);
 	struct mt7996_vif *mvif = (struct mt7996_vif *)vif->drv_priv;
@@ -2355,6 +2356,13 @@ mt7996_set_eml_op_mode(struct ieee80211_hw *hw, struct ieee80211_vif *vif,
 	return ret;
 }
 
+static int
+mt7996_set_qos_map(struct ieee80211_hw *hw, struct ieee80211_vif *vif,
+		   struct cfg80211_qos_map *qos_map)
+{
+	return 0;
+}
+
 const struct ieee80211_ops mt7996_ops = {
 	.add_chanctx = mt76_add_chanctx,
 	.remove_chanctx = mt76_remove_chanctx,
@@ -2419,4 +2427,5 @@ const struct ieee80211_ops mt7996_ops = {
 	.change_sta_links = mt7996_mac_sta_change_links,
 	.reconfig_complete = mt7996_reconfig_complete,
 	.set_eml_op_mode = mt7996_set_eml_op_mode,
+	.set_qos_map = mt7996_set_qos_map,
 };
diff --git a/mt7996/mmio.c b/mt7996/mmio.c
index 4430f37cb..9d0d3f163 100644
--- a/mt7996/mmio.c
+++ b/mt7996/mmio.c
@@ -490,7 +490,7 @@ int mt7996_mmio_wed_init(struct mt7996_dev *dev, void *pdev_ptr,
 				      MT_INT_PCIE1_SOURCE_CSR_EXT;
 		wed->wlan.wpdma_mask = wed->wlan.phy_base +
 				       MT_INT_PCIE1_MASK_CSR;
-		wed->wlan.wpdma_tx = wed->wlan.phy_base + hif1_ofs +
+		wed->wlan.wpdma_tx[0] = wed->wlan.phy_base + hif1_ofs +
 					     MT_TXQ_RING_BASE(0) +
 					     MT7996_TXQ_BAND2 * MT_RING_SIZE;
 		if (mt7996_has_hwrro(dev)) {
@@ -523,7 +523,7 @@ int mt7996_mmio_wed_init(struct mt7996_dev *dev, void *pdev_ptr,
 		wed->wlan.hw_rro = mt7996_has_hwrro(dev);
 		wed->wlan.wpdma_int = wed->wlan.phy_base + MT_INT_SOURCE_CSR;
 		wed->wlan.wpdma_mask = wed->wlan.phy_base + MT_INT_MASK_CSR;
-		wed->wlan.wpdma_tx = wed->wlan.phy_base + MT_TXQ_RING_BASE(0) +
+		wed->wlan.wpdma_tx[0] = wed->wlan.phy_base + MT_TXQ_RING_BASE(0) +
 				     MT7996_TXQ_BAND0 * MT_RING_SIZE;
 
 		wed->wlan.wpdma_rx_glo = wed->wlan.phy_base + MT_WFDMA0_GLO_CFG;
-- 
2.45.2

