From db1db0b3aa03d597ba2346c997858c3aac0f1e0b Mon Sep 17 00:00:00 2001
From: Michael-CY Lee <michael-cy.lee@mediatek.com>
Date: Thu, 26 Feb 2026 11:08:46 +0800
Subject: [PATCH] mtk: mt76: add link_id in TxS

issue severity: High
Signed-off-by: Michael-CY Lee <michael-cy.lee@mediatek.com>
---
 tx.c | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/tx.c b/tx.c
index 80c7ffcd2..a46d0c2bd 100644
--- a/tx.c
+++ b/tx.c
@@ -74,6 +74,11 @@ mt76_tx_status_unlock(struct mt76_dev *dev, struct sk_buff_head *list)
 			} else {
 				status.n_rates = 0;
 			}
+
+			if (status.info && wcid->link_valid) {
+				status.info->status.flags |= IEEE80211_TX_STATUS_LINK_ID_VALID;
+				status.info->status.link_id = wcid->link_id;
+			}
 		}
 
 		hw = mt76_tx_status_get_hw(dev, skb);
-- 
2.45.2

