From 6508729b1be71b305e1a2ac8f529c05724aff699 Mon Sep 17 00:00:00 2001
From: Sagar Jain <sagarj@minim.com>
Date: Tue, 6 Dec 2022 11:21:28 +0530
Subject: [PATCH] r14: enable ssh via web-admin.

SW-2703
---
 .../services/dropbear/files/dropbear.init     | 21 +++++++++++++++++++
 1 file changed, 21 insertions(+)

diff --git a/package/network/services/dropbear/files/dropbear.init b/package/network/services/dropbear/files/dropbear.init
index ea4cad2a8d..7ddb4af549 100755
--- a/package/network/services/dropbear/files/dropbear.init
+++ b/package/network/services/dropbear/files/dropbear.init
@@ -11,6 +11,8 @@ NAME=dropbear
 PIDCOUNT=0
 
 extra_command "killclients" "Kill ${NAME} processes except servers and yourself"
+extra_command "restart_admin" "Temporarily start sshd"
+extra_command "restart_admin_enable" "Enable sshd permanently"
 
 _dropbearkey()
 {
@@ -102,6 +104,14 @@ append_ports()
 	local ipaddrs="$1"
 	local port="$2"
 
+	p=`fw_printenv -n mfg_provisioned 2> /dev/null`
+	s=`fw_printenv -n sshd_enable 2> /dev/null`
+	if [ "$p" == "1" -a ! -f /tmp/sshd_enable -a "$s" != "1" ] ; then
+		# It is already provisioned.
+		# Listen on loopback interface only
+		ipaddrs="127.0.0.1"
+	fi
+
 	[ -z "$ipaddrs" ] && {
 		procd_append_param command -p "$port"
 		return
@@ -227,6 +237,17 @@ shutdown() {
 	killall dropbear
 }
 
+restart_admin() {
+    touch /tmp/sshd_enable
+    stop
+    start
+}
+
+restart_admin_enable() {
+    /usr/sbin/fw_setenv sshd_enable 1
+    restart_admin
+}
+
 killclients()
 {
 	local ignore=''
-- 
2.30.2

