Index: backports-20210222_001-5.4.164-b157d2276/drivers/net/wireless/ath/ath11k/qmi.c
===================================================================
--- backports-20210222_001-5.4.164-b157d2276.orig/drivers/net/wireless/ath/ath11k/qmi.c
+++ backports-20210222_001-5.4.164-b157d2276/drivers/net/wireless/ath/ath11k/qmi.c
@@ -4,6 +4,7 @@
  */
 
 #include <linux/elf.h>
+#include <linux/device.h>
 
 #include "qmi.h"
 #include "core.h"
