Index: linux-5.4.164-qsdk-26349818b464f8c7b52d59ce73579d9f3dd6bd5d/arch/arm64/boot/dts/qcom/ipq8074-memory.dtsi
===================================================================
--- linux-5.4.164-qsdk-26349818b464f8c7b52d59ce73579d9f3dd6bd5d.orig/arch/arm64/boot/dts/qcom/ipq8074-memory.dtsi
+++ linux-5.4.164-qsdk-26349818b464f8c7b52d59ce73579d9f3dd6bd5d/arch/arm64/boot/dts/qcom/ipq8074-memory.dtsi
@@ -129,6 +129,12 @@
 			no-map;
 			reg = <0x0 0x4d900000 0x0 0x100000>;
 		};
+
+		ramoops@4da00000 {
+			compatible = "ramoops";
+			reg = <0x0 0x4da00000 0x0 0x100000>;
+			record-size = <0x1000>;
+		};
 	};
 	#elif __IPQ_MEM_PROFILE_512_MB__
 	/*                      512 MB Profile
@@ -249,6 +255,12 @@
 			no-map;
 			reg = <0x0 0x4e800000 0x0 0x100000>;
 		};
+
+		ramoops@4e900000 {
+			compatible = "ramoops";
+			reg = <0x0 0x4e900000 0x0 0x100000>;
+			record-size = <0x1000>;
+		};
 	};
 	#else
 	/*                   Default Profile
@@ -371,6 +383,11 @@
 			reg = <0x0 0x51000000 0x0 0x100000>;
 		};
 
+		ramoops@51200000 {
+			compatible = "ramoops";
+			reg = <0x0 0x51200000 0x0 0x100000>;
+			record-size = <0x1000>;
+		};
 	};
 	#endif
 };
