--- a/drivers/net/phy/phylink.c	2025-11-28 09:51:11.325681004 +0000
+++ b/drivers/net/phy/phylink.c	2025-11-28 09:47:06.873043674 +0000
@@ -1747,9 +1747,8 @@ static void phylink_resolve(struct work_
 			}
 
 			if (pl->phydev && !(link_state.link & pl->phy_state.link))
-				phylink_printk(KERN_DEBUG, pl,
-					       "resolve link status: system iface=%d, line iface=%d\n",
-					       link_state.link, pl->phy_state.link);
+				pr_info_once("%s: resolve link status: system iface=%d, line iface=%d\n",
+					     phydev_name(pl->phydev), link_state.link, pl->phy_state.link);
 			else if (!link_state.link)
 				phylink_printk(KERN_DEBUG, pl,
 					       "resolve link status: system iface=%d\n",
