From b9281faaf0fba4ca1dbaf377b0f09b7b35a77d35 Mon Sep 17 00:00:00 2001
From: Bo-Cun Chen <bc-bocun.chen@mediatek.com>
Date: Fri, 15 Sep 2023 11:39:30 +0800
Subject: [PATCH 9/9] mtkhnat-ipv6-fix-pskb-expand-head-limitation.patch

---
 net/core/skbuff.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/net/core/skbuff.c b/net/core/skbuff.c
index 886fad128..da6cae59c 100644
--- a/net/core/skbuff.c
+++ b/net/core/skbuff.c
@@ -73,6 +73,7 @@
 #include <net/mctp.h>
 #include <net/page_pool/helpers.h>
 #include <net/dropreason.h>
+#include <net/ra_nat.h>
 
 #include <linux/uaccess.h>
 #include <trace/events/skb.h>
@@ -1871,6 +1872,9 @@ int pskb_expand_head(struct sk_buff *skb, int nhead, int ntail,
 	       skb_shinfo(skb),
 	       offsetof(struct skb_shared_info, frags[skb_shinfo(skb)->nr_frags]));
 
+	/*headroom copy*/
+	memcpy(data, skb->head, FOE_INFO_LEN);
+
 	/*
 	 * if shinfo is shared we must drop the old head gracefully, but if it
 	 * is not we can just drop the old head and let the existing refcount
-- 
2.18.0

