--- a/include/linux/netdevice.h
+++ b/include/linux/netdevice.h
@@ -5076,4 +5076,5 @@ extern struct net_device *blackhole_netd
 		atomic_long_add((VAL), &(DEV)->stats.__##FIELD)
 #define DEV_STATS_READ(DEV, FIELD) atomic_long_read(&(DEV)->stats.__##FIELD)
 
+extern int (*mtk_skb_headroom_copy)(struct sk_buff *new, struct sk_buff *old);
 #endif	/* _LINUX_NETDEVICE_H */
--- a/net/core/dev.c
+++ b/net/core/dev.c
@@ -198,6 +198,9 @@ static DEFINE_READ_MOSTLY_HASHTABLE(napi
 
 static DECLARE_RWSEM(devnet_rename_sem);
 
+int (*mtk_skb_headroom_copy)(struct sk_buff *new, struct sk_buff *old) = NULL;
+EXPORT_SYMBOL(mtk_skb_headroom_copy);
+
 static inline void dev_base_seq_inc(struct net *net)
 {
 	while (++net->dev_base_seq == 0)
@@ -3448,6 +3451,9 @@ static struct sk_buff *validate_xmit_skb
 		if (IS_ERR(segs)) {
 			goto out_kfree_skb;
 		} else if (segs) {
+			if (mtk_skb_headroom_copy)
+				mtk_skb_headroom_copy(segs, skb);
+
 			consume_skb(skb);
 			skb = segs;
 		}
