Index: ath10k-ct-2021-05-22b-54a9ac02/ath10k-5.7/core.c
===================================================================
--- ath10k-ct-2021-05-22b-54a9ac02.orig/ath10k-5.7/core.c
+++ ath10k-ct-2021-05-22b-54a9ac02/ath10k-5.7/core.c
@@ -4146,6 +4146,9 @@ struct ath10k *ath10k_core_create(size_t
 	INIT_WORK(&ar->stop_scan_work, ath10k_wmi_stop_scan_work);
 
 	init_dummy_netdev(&ar->napi_dev);
+	snprintf(ar->napi_dev.name, sizeof(ar->napi_dev.name), "%s",
+		 wiphy_name(ar->hw->wiphy));
+	ar->napi_dev.threaded = 1;
 
 	ret = ath10k_coredump_create(ar);
 	if (ret)
