--- a/kconf/Makefile
+++ b/kconf/Makefile
@@ -2,13 +2,15 @@ CFLAGS=-Wall -Wmissing-prototypes -Wstri
 
 LXDIALOG := lxdialog/checklist.o lxdialog/inputbox.o lxdialog/menubox.o lxdialog/textbox.o lxdialog/util.o lxdialog/yesno.o
 
-conf: conf.o zconf.tab.o
+conf: zconf.tab.c
+	gcc -o conf conf.c zconf.tab.c
+
 mconf_CFLAGS := $(shell ./lxdialog/check-lxdialog.sh -ccflags)
 mconf_LDFLAGS := $(shell ./lxdialog/check-lxdialog.sh -ldflags $(CC))
 mconf: CFLAGS += $(mconf_CFLAGS)
 
 mconf: mconf.o zconf.tab.o $(LXDIALOG)
-	$(CC) -o mconf $^ $(mconf_LDFLAGS)
+	gcc -o mconf $^ $(mconf_LDFLAGS)
 
 .PHONY: clean
 clean:
