Index: backports-20210222_001-4.4.60-b157d2276/drivers/net/wireless/ath/ath11k/mac.c
===================================================================
--- backports-20210222_001-4.4.60-b157d2276.orig/drivers/net/wireless/ath/ath11k/mac.c
+++ backports-20210222_001-4.4.60-b157d2276/drivers/net/wireless/ath/ath11k/mac.c
@@ -3843,13 +3843,14 @@ static int ath11k_mac_op_hw_scan(struct
 		scan_timeout = min_t(u32, arg->max_rest_time *
 				    (arg->chan_list.num_chan - 1) + (req->duration +
 				     ATH11K_SCAN_CHANNEL_SWITCH_WMI_EVT_OVERHEAD) *
-				     arg->chan_list.num_chan, arg->max_scan_time +
-				     ATH11K_MAC_SCAN_TIMEOUT_MSECS);
+				     arg->chan_list.num_chan, arg->max_scan_time);
 	} else {
-		/* Add a 200ms margin to account for event/command processing */
-		scan_timeout = arg->max_scan_time + ATH11K_MAC_SCAN_TIMEOUT_MSECS;
+		scan_timeout = arg->max_scan_time;
 	}
 
+	/* Add a 200ms margin to account for event/command processing */
+	scan_timeout += ATH11K_MAC_SCAN_TIMEOUT_MSECS;
+
 	ret = ath11k_start_scan(ar, arg);
 	if (ret) {
 		ath11k_warn(ar->ab, "failed to start hw scan: %d scan_state %d pdev_idx %d\n",
