--- a/drivers/net/wireless/ath/ath11k/wmi.c	2021-12-31 12:50:46.647341108 -0800
+++ b/drivers/net/wireless/ath/ath11k/wmi.c	2021-12-31 13:16:53.627380659 -0800
@@ -8069,78 +8069,9 @@
 
 static void ath11k_chan_info_event(struct ath11k_base *ab, struct sk_buff *skb)
 {
-	struct wmi_chan_info_event ch_info_ev = {0};
-	struct ath11k *ar;
-	struct survey_info *survey;
-	int idx;
-	/* HW channel counters frequency value in hertz */
-	u32 cc_freq_hz = ab->cc_freq_hz;
-
-	if (ath11k_pull_chan_info_ev(ab, skb->data, skb->len, &ch_info_ev) != 0) {
-		ath11k_warn(ab, "failed to extract chan info event");
-		return;
-	}
-
-	ath11k_dbg(ab, ATH11K_DBG_WMI,
-		   "chan info vdev_id %d err_code %d freq %d cmd_flags %d noise_floor %d rx_clear_count %d cycle_count %d mac_clk_mhz %d\n",
-		   ch_info_ev.vdev_id, ch_info_ev.err_code, ch_info_ev.freq,
-		   ch_info_ev.cmd_flags, ch_info_ev.noise_floor,
-		   ch_info_ev.rx_clear_count, ch_info_ev.cycle_count,
-		   ch_info_ev.mac_clk_mhz);
-
-	if (ch_info_ev.cmd_flags == WMI_CHAN_INFO_END_RESP) {
-		ath11k_dbg(ab, ATH11K_DBG_WMI, "chan info report completed\n");
-		return;
-	}
-
-	rcu_read_lock();
-	ar = ath11k_mac_get_ar_by_vdev_id(ab, ch_info_ev.vdev_id);
-	if (!ar) {
-		ath11k_warn(ab, "invalid vdev id in chan info ev %d",
-			    ch_info_ev.vdev_id);
-		rcu_read_unlock();
-		return;
-	}
-	spin_lock_bh(&ar->data_lock);
-
-	switch (ar->scan.state) {
-	case ATH11K_SCAN_IDLE:
-	case ATH11K_SCAN_STARTING:
-		ath11k_warn(ab, "received chan info event without a scan request, ignoring\n");
-		goto exit;
-	case ATH11K_SCAN_RUNNING:
-	case ATH11K_SCAN_ABORTING:
-		break;
-	}
-
-	idx = freq_to_idx(ar, ch_info_ev.freq);
-	if (idx >= ARRAY_SIZE(ar->survey)) {
-		ath11k_warn(ab, "chan info: invalid frequency %d (idx %d out of bounds)\n",
-			    ch_info_ev.freq, idx);
-		goto exit;
-	}
-
-	/* If FW provides MAC clock frequency in Mhz, overriding the initialized
-	 * HW channel counters frequency value
+	/* Ignore this event. It was clear on read and overriding chan info
+	 * coming from bss_chan_info_event.
 	 */
-	if (ch_info_ev.mac_clk_mhz)
-		cc_freq_hz = (ch_info_ev.mac_clk_mhz * 1000);
-
-	if (ch_info_ev.cmd_flags == WMI_CHAN_INFO_START_RESP) {
-		survey = &ar->survey[idx];
-		memset(survey, 0, sizeof(*survey));
-		survey->noise = ch_info_ev.noise_floor;
-		survey->filled = SURVEY_INFO_NOISE_DBM | SURVEY_INFO_TIME |
-				 SURVEY_INFO_TIME_BUSY;
-		survey->time = div_u64(ch_info_ev.cycle_count, cc_freq_hz);
-		survey->time_busy = div_u64(ch_info_ev.rx_clear_count, cc_freq_hz);
-		if (survey->time)
-			ar->hw->medium_busy = div_u64((survey->time_busy * 100),
-						      survey->time);
-	}
-exit:
-	spin_unlock_bh(&ar->data_lock);
-	rcu_read_unlock();
 }
 
 static void
