Index: mt76-2024-04-03-1e336a85/mt7915/mt7915.h
===================================================================
--- mt76-2024-04-03-1e336a85.orig/mt7915/mt7915.h
+++ mt76-2024-04-03-1e336a85/mt7915/mt7915.h
@@ -71,8 +71,8 @@
 #define MT7915_CFEND_RATE_DEFAULT	0x49	/* OFDM 24M */
 #define MT7915_CFEND_RATE_11B		0x03	/* 11B LP, 11M */
 
-#define MT7915_THERMAL_THROTTLE_MAX	100
-#define MT7915_CDEV_THROTTLE_MAX	99
+#define MT7915_THERMAL_THROTTLE_MAX	80
+#define MT7915_CDEV_THROTTLE_MAX	79
 
 #define MT7915_SKU_RATE_NUM		161
 #define MT7915_SKU_PATH_NUM		185
@@ -86,8 +86,8 @@
 
 #define MT7915_CRIT_TEMP_IDX		0
 #define MT7915_MAX_TEMP_IDX		1
-#define MT7915_CRIT_TEMP		110
-#define MT7915_MAX_TEMP			120
+#define MT7915_CRIT_TEMP		103
+#define MT7915_MAX_TEMP			118
 
 struct mt7915_vif;
 struct mt7915_sta;
