From: John Crispin <john@phrozen.org>
Subject: [PATCH] build: Honour COPTS, OWRT_CFLAGS

--- a/build/flags.mk
+++ b/build/flags.mk
@@ -23,12 +23,13 @@
 # SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 
 # C FLAGS
-CFLAGS   := -Wall -Wextra -Werror
+CFLAGS   ?=
+CFLAGS   += -Wall -Wextra -Werror
 CFLAGS   += -Wno-unused-parameter
 CFLAGS   += -Wno-unused-label
 CFLAGS   += -Wno-strict-aliasing
 CFLAGS   += -fasynchronous-unwind-tables
-
+CFLAGS   += $(OWRT_CFLAGS)
 DEBUGFLAGS := -g
 OPTIMIZE :=
 
