diff -urN a/src/lib/hostap/inc/opensync-hapd.h b/src/lib/hostap/inc/opensync-hapd.h
--- a/src/lib/hostap/inc/opensync-hapd.h	2020-12-28 20:56:22.034792354 +0530
+++ b/src/lib/hostap/inc/opensync-hapd.h	2020-12-28 20:56:59.386736171 +0530
@@ -42,6 +42,7 @@
     int ieee80211ax;
     char htcaps[256];
     char vhtcaps[512];
+    char country[32];
     void (*sta_connected)(struct hapd *hapd, const char *mac, const char *keyid);
     void (*sta_disconnected)(struct hapd *hapd, const char *mac);
     void (*ap_enabled)(struct hapd *hapd);
diff -urN a/src/lib/hostap/src/hapd.c b/src/lib/hostap/src/hapd.c
--- a/src/lib/hostap/src/hapd.c	2020-12-28 13:54:15.224206323 +0530
+++ b/src/lib/hostap/src/hapd.c	2020-12-28 21:15:03.762242284 +0530
@@ -606,6 +606,12 @@
         csnprintf(&buf, &len, "eap_server=%d\n", vconf->wps ? 1 : 0);
     }
 
+    if (strlen(hapd->country)) {
+        csnprintf(&buf, &len, "country_code=%s\n", hapd->country);
+        csnprintf(&buf, &len, "ieee80211d=1\n");
+        csnprintf(&buf, &len, "ieee80211h=1\n");
+    }
+
     WARN_ON(len == 1); /* likely buf was truncated */
 
     return hapd_conf_gen_psk(hapd, vconf);
