diff -urN a/src/lib/hostap/src/hapd.c b/src/lib/hostap/src/hapd.c
--- a/src/lib/hostap/src/hapd.c	2021-01-18 23:41:23.538504896 +0530
+++ b/src/lib/hostap/src/hapd.c	2021-01-18 23:42:54.852904473 +0530
@@ -850,7 +850,7 @@
 static int
 hapd_ctrl_add(struct hapd *hapd)
 {
-    const char *arg = F("bss_config=%s:%s", hapd->ctrl.bss, hapd->confpath);
+    const char *arg = F("bss_config=%s:%s", hapd->phy, hapd->confpath);
     int err = 0;
     /* FIXME: check if I can use hapd->phy instead od hapd->bss above on qca */
     LOGI("%s: adding", hapd->ctrl.bss);
