diff -urN a/src/lib/hostap/src/hapd.c b/src/lib/hostap/src/hapd.c
--- a/src/lib/hostap/src/hapd.c	2021-01-28 19:59:58.152626240 +0530
+++ b/src/lib/hostap/src/hapd.c	2021-01-28 20:03:27.607333444 +0530
@@ -637,6 +637,8 @@
         vstate->ft_psk = atoi(p);
     if ((vstate->btm_exists = (p = ini_geta(conf, "bss_transition"))))
         vstate->btm = atoi(p);
+    if ((vstate->rrm_exists = (p = ini_geta(conf, "rrm_neighbor_report"))))
+        vstate->rrm = atoi(p);
     if ((p = ini_geta(conf, "multi_ap")))
         SCHEMA_SET_STR(vstate->multi_ap, hapd_map_int2str(atoi(p)));
 
