diff -ruN a/src/lib/hostap/src/hapd.c b/src/lib/hostap/src/hapd.c
--- a/src/lib/hostap/src/hapd.c	2021-02-25 08:27:37.644440999 +0530
+++ b/src/lib/hostap/src/hapd.c	2021-02-25 08:02:37.478075301 +0530
@@ -659,6 +659,9 @@
         csnprintf(&buf, &len, "ieee80211h=1\n");
     }
     
+    if (rconf->bcn_int_exists)
+        csnprintf(&buf, &len, "beacon_int=%d\n", rconf->bcn_int);
+
     WARN_ON(len == 1); /* likely buf was truncated */
 
     return hapd_conf_gen_psk(hapd, vconf);
