From: aparna.salian <aparna.salian@globaledgesoft.com>
Subject: [PATCH] wm2: Remove vif state entry for deleted interface

--- a/src/wm2/src/wm2_radio.c
+++ b/src/wm2/src/wm2_radio.c
@@ -780,6 +780,7 @@ wm2_vconf_recalc(const char *ifname, bool force)
     int num_cconfs;
     bool want;
     bool has;
+    int del_vif = 0;
 
     LOGD("%s: recalculating", ifname);
 
@@ -797,6 +798,9 @@ wm2_vconf_recalc(const char *ifname, bool force)
                                        &vstate)))
         wm2_vstate_init(&vstate, ifname);
 
+    if (!want)
+        del_vif = 1;
+
     /* This is workaround to deal with unpatched controller.
      * Having this on device side prevents it from saner 3rd
      * party gw integrations where currently State needs to
@@ -889,6 +893,13 @@ wm2_vconf_recalc(const char *ifname, bool force)
         return;
     }
 
+    if (del_vif) {
+        ovsdb_table_delete_simple(&table_Wifi_VIF_State,
+                                  SCHEMA_COLUMN(Wifi_VIF_State, if_name),
+                                  vconf.if_name);
+        LOGD("%s: VIF state column deleted", vconf.if_name);
+    }
+
     wm2_delayed_recalc_cancel(wm2_vconf_recalc, ifname);
 }
 
