From 800c0cc7a91fc93011d3664471a815d0b73cf964 Mon Sep 17 00:00:00 2001
From: Pankaj Ghosh <g.pankaj@globaledgesoft.com>
Date: Mon, 1 Feb 2021 11:21:44 +0530
Subject: [PATCH] cm2: Stability check fix

Signed-off-by: Pankaj Ghosh <g.pankaj@globaledgesoft.com>
---
 src/cm2/src/cm2_stability.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/cm2/src/cm2_stability.c b/src/cm2/src/cm2_stability.c
index 4b5e91a..b790be6 100644
--- a/src/cm2/src/cm2_stability.c
+++ b/src/cm2/src/cm2_stability.c
@@ -274,7 +274,7 @@ static void cm2_connection_stability_check(void)
 {
     target_connectivity_check_option_t opts;
 
-    if (!cm2_cpu_is_low_loadavg())
+    if (g_state.connected && !cm2_cpu_is_low_loadavg())
         return;
 
     opts = LINK_CHECK | ROUTER_CHECK | NTP_CHECK;
-- 
2.7.4

