From: Lars Kristan <lkristan@plume.com>
Subject: [PATCH] lib/hostap: Fixed a potential crash

Backported from OpenSync 2.0.7.0

--- a/src/lib/hostap/src/wpas.c
+++ b/src/lib/hostap/src/wpas.c
@@ -444,11 +444,14 @@ wpas_bss_get_network(struct schema_Wifi_VIF_State *vstate,
         if ((network = strstr(network, "network={")))
             network += strlen("network={");
 
+    if (!network)
+        network = strdupa("");
+
     map = ini_geta(network, "multi_ap_backhaul_sta");
     psk = ini_geta(network, "psk") ?: "";
 
     /* entry in file looks actually like this: psk="passphrase", so remove the: " */
-    psk++;
+    if (strlen(psk) > 0) psk++;
     if (strlen(psk) > 0) psk[strlen(psk)-1] = 0;
 
     if ((vstate->parent_exists = (bssid != NULL)))
