diff -urN a/src/lib/hostap/inc/opensync-hapd.h b/src/lib/hostap/inc/opensync-hapd.h
--- a/src/lib/hostap/inc/opensync-hapd.h	2021-01-26 20:28:21.057564683 +0530
+++ b/src/lib/hostap/inc/opensync-hapd.h	2021-01-26 20:30:01.223510103 +0530
@@ -55,6 +55,10 @@
     void (*cmd_frame_probe_req)(struct hapd *hapd, const char *event);
     void (*cmd_frame_action)(struct hapd *hapd, const char *event);
     void (*cmd_frame_disconnect)(struct hapd *hapd, const char *event);
+    void (*dfs_event_cac_start)(struct hapd *hapd, const char *event);
+    void (*dfs_event_cac_completed)(struct hapd *hapd, const char *event);
+    void (*dfs_event_radar_detected)(struct hapd *hapd, const char *event);
+    void (*dfs_event_nop_finished)(struct hapd *hapd, const char *event);
     struct ctrl ctrl;
 };
 
diff -urN a/src/lib/hostap/src/hapd.c b/src/lib/hostap/src/hapd.c
--- a/src/lib/hostap/src/hapd.c	2021-01-26 20:28:33.249819819 +0530
+++ b/src/lib/hostap/src/hapd.c	2021-01-26 20:40:37.758631909 +0530
@@ -214,6 +214,30 @@
         return;
     }
 
+    if (!strcmp(event, EV(DFS_EVENT_CAC_START))) {
+        LOGI("%s: dfs event - cac started", hapd->ctrl.bss);
+        if (hapd->dfs_event_cac_start)
+            hapd->dfs_event_cac_start(hapd, args);
+    }
+
+    if (!strcmp(event, EV(DFS_EVENT_CAC_COMPLETED))) {
+        LOGI("%s: dfs event - cac completed", hapd->ctrl.bss);
+        if (hapd->dfs_event_cac_completed)
+            hapd->dfs_event_cac_completed(hapd, args);
+    }
+
+    if (!strcmp(event, EV(DFS_EVENT_RADAR_DETECTED))) {
+        LOGI("%s: dfs event - radar detected", hapd->ctrl.bss);
+        if (hapd->dfs_event_radar_detected)
+            hapd->dfs_event_radar_detected(hapd, args);
+    }
+
+    if (!strcmp(event, EV(DFS_EVENT_NOP_FINISHED))) {
+        LOGI("%s: dfs event - nop finished", hapd->ctrl.bss);
+        if (hapd->dfs_event_nop_finished)
+            hapd->dfs_event_nop_finished(hapd, args);
+    }
+
     LOGI("%s: event: <%d> %s", ctrl->bss, level, buf);
 }
 
