From: aparna.salian <aparna.salian@globaledgesoft.com>
Subject: [PATCH] lib/hostap: Fetch ssid_broadcast information from hostapd conf

--- a/src/lib/hostap/src/hapd.c
+++ b/src/lib/hostap/src/hapd.c
@@ -639,6 +639,8 @@ hapd_bss_get(struct hapd *hapd,
         vstate->btm = atoi(p);
     if ((p = ini_geta(conf, "multi_ap")))
         SCHEMA_SET_STR(vstate->multi_ap, hapd_map_int2str(atoi(p)));
+    if ((vstate->ssid_broadcast_exists = (p = ini_geta(conf, "ignore_broadcast_ssid"))))
+        SCHEMA_SET_STR(vstate->ssid_broadcast, atoi(p) ? "disabled" : "enabled");
 
     if (status) {
         hapd_bss_get_security(vstate, conf, status);
