From: aparna.salian <aparna.salian@globaledgesoft.com>
Subject: [PATCH] lib/hostap: uapsd set/get support

--- a/src/lib/hostap/src/hapd.c
+++ b/src/lib/hostap/src/hapd.c
@@ -411,6 +411,7 @@ hapd_conf_gen(struct hapd *hapd,
     csnprintf(&buf, &len, "wmm_enabled=1\n");
     csnprintf(&buf, &len, "%s", hapd->respect_multi_ap ? "" : "#");
     csnprintf(&buf, &len, "multi_ap=%d\n", hapd_map_str2int(vconf));
+    csnprintf(&buf, &len, "uapsd_advertisement_enabled=%d\n", vconf->uapsd_enable);
 
     /* FIXME: ieee80211n, iee80211ac, ieee80211ax is missing, also min_hw_mode..
      * perhaps some of this needs to be scheduled for wireless api rework
@@ -641,6 +642,8 @@ hapd_bss_get(struct hapd *hapd,
         SCHEMA_SET_STR(vstate->multi_ap, hapd_map_int2str(atoi(p)));
     if ((vstate->ssid_broadcast_exists = (p = ini_geta(conf, "ignore_broadcast_ssid"))))
         SCHEMA_SET_STR(vstate->ssid_broadcast, atoi(p) ? "disabled" : "enabled");
+    if ((vstate->uapsd_enable_exists = (p = ini_geta(conf, "uapsd_advertisement_enabled"))))
+        vstate->uapsd_enable = atoi(p);
 
     if (status) {
         hapd_bss_get_security(vstate, conf, status);
