diff -urN a/src/lib/hostap/inc/opensync-hapd.h b/src/lib/hostap/inc/opensync-hapd.h
--- a/src/lib/hostap/inc/opensync-hapd.h	2021-02-11 13:36:30.755642063 +0530
+++ b/src/lib/hostap/inc/opensync-hapd.h	2021-02-11 14:01:12.294587084 +0530
@@ -59,6 +59,7 @@
     void (*dfs_event_cac_completed)(struct hapd *hapd, const char *event);
     void (*dfs_event_radar_detected)(struct hapd *hapd, const char *event);
     void (*dfs_event_nop_finished)(struct hapd *hapd, const char *event);
+    void (*ap_csa_finished)(struct hapd *hapd, const char *event);
     struct ctrl ctrl;
 };
 
diff -urN a/src/lib/hostap/src/hapd.c b/src/lib/hostap/src/hapd.c
--- a/src/lib/hostap/src/hapd.c	2021-02-11 13:36:36.383584638 +0530
+++ b/src/lib/hostap/src/hapd.c	2021-02-11 14:08:27.122477972 +0530
@@ -238,6 +238,12 @@
             hapd->dfs_event_nop_finished(hapd, args);
     }
 
+    if (!strcmp(event, EV(AP_CSA_FINISHED))) {
+        LOGI("%s: ap csa event", hapd->ctrl.bss);
+        if (hapd->ap_csa_finished)
+            hapd->ap_csa_finished(hapd, args);
+    }
+
     LOGI("%s: event: <%d> %s", ctrl->bss, level, buf);
 }
 
