From: aparna.salian <aparna.salian@globaledgesoft.com>
Subject: [PATCH] lib/hostap: Update regulatory domain in hostapd conf for DFS channel support

--- a/src/lib/hostap/inc/opensync-hapd.h
+++ b/src/lib/hostap/inc/opensync-hapd.h
@@ -42,6 +42,7 @@ struct hapd {
     int ieee80211ax;
     char htcaps[256];
     char vhtcaps[512];
+    char country[32];
     void (*sta_connected)(struct hapd *hapd, const char *mac, const char *keyid);
     void (*sta_disconnected)(struct hapd *hapd, const char *mac);
     void (*ap_enabled)(struct hapd *hapd);
--- a/src/lib/hostap/src/hapd.c
+++ b/src/lib/hostap/src/hapd.c
@@ -607,6 +607,12 @@ hapd_conf_gen(struct hapd *hapd,
         csnprintf(&buf, &len, "eap_server=%d\n", vconf->wps ? 1 : 0);
     }
 
+    if (strlen(hapd->country) > 0) {
+        csnprintf(&buf, &len, "country_code=%s\n", hapd->country);
+        csnprintf(&buf, &len, "ieee80211d=1\n");
+        csnprintf(&buf, &len, "ieee80211h=1\n");
+    }
+
     WARN_ON(len == 1); /* likely buf was truncated */
 
     return hapd_conf_gen_psk(hapd, vconf);
