From: aparna.salian <aparna.salian@globaledgesoft.com>
Subject: [PATCH] lib/hostap: Center frequency fix

--- a/src/lib/common/inc/util.h
+++ b/src/lib/common/inc/util.h
@@ -141,5 +141,6 @@ int    file_put(const char *path, const char *buf);
 char  *file_get(const char *path);
 #define file_geta(path) strdupafree(file_get(path))
 const int *unii_5g_chan2list(int chan, int width);
+int unii_5g_centerfreq(const char *ht_mode, const int channel);
 
 #endif /* UTIL_H_INCLUDED */
--- a/src/lib/common/src/util.c
+++ b/src/lib/common/src/util.c
@@ -1094,3 +1094,21 @@ const int *unii_5g_chan2list(int chan, int width)
 
     return NULL;
 }
+
+int unii_5g_centerfreq(const char *ht_mode, const int channel)
+{
+    const int width = atoi(strlen(ht_mode) > 2 ? ht_mode + 2 : "20");
+    const int *chans = unii_5g_chan2list(channel, width);
+    int sum = 0;
+    int cnt = 0;
+
+    if (!chans)
+        return 0;
+
+    while (*chans) {
+        sum += *chans;
+        cnt++;
+        chans++;
+    }
+    return sum / cnt;
+}
--- a/src/lib/hostap/src/hapd.c
+++ b/src/lib/hostap/src/hapd.c
@@ -479,19 +479,11 @@ hapd_vht_oper_chwidth(const struct schema_Wifi_Radio_Config *rconf)
 static int
 hapd_vht_oper_centr_freq_idx(const struct schema_Wifi_Radio_Config *rconf)
 {
-    const int width = atoi(strlen(rconf->ht_mode) > 2 ? rconf->ht_mode + 2 : "20");
-    const int *chans = unii_5g_chan2list(rconf->channel, width);
-    int sum = 0;
-    int cnt = 0;
-
-    while (*chans) {
-        sum += *chans;
-        cnt++;
-        chans++;
-    }
-    return sum / cnt;
-}
+    if (!rconf->ht_mode_exists || !rconf->channel_exists)
+        return 0;
 
+    return unii_5g_centerfreq(rconf->ht_mode, rconf->channel);
+}
 
 int
 hapd_conf_gen(struct hapd *hapd,
