From: aparna.salian <aparna.salian@globaledgesoft.com>
Subject: [PATCH] lib/hostap: Pass phy name to hostapd while adding a new interface

--- a/src/lib/hostap/src/hapd.c
+++ b/src/lib/hostap/src/hapd.c
@@ -843,7 +843,7 @@ hapd_sta_iter(struct hapd *hapd,
 static int
 hapd_ctrl_add(struct hapd *hapd)
 {
-    const char *arg = F("bss_config=%s:%s", hapd->ctrl.bss, hapd->confpath);
+    const char *arg = F("bss_config=%s:%s", hapd->phy, hapd->confpath);
     int err = 0;
     /* FIXME: check if I can use hapd->phy instead od hapd->bss above on qca */
     LOGI("%s: adding", hapd->ctrl.bss);
