From: aparna.salian <aparna.salian@globaledgesoft.com>
Subject: [PATCH] lib/hostap: Fetch vif state rrm value from hostapd conf

--- a/src/lib/hostap/src/hapd.c
+++ b/src/lib/hostap/src/hapd.c
@@ -772,6 +772,8 @@ hapd_bss_get(struct hapd *hapd,
         vstate->ft_psk = atoi(p);
     if ((vstate->btm_exists = (p = ini_geta(conf, "bss_transition"))))
         vstate->btm = atoi(p);
+    if ((vstate->rrm_exists = (p = ini_geta(conf, "rrm_neighbor_report"))))
+        vstate->rrm = atoi(p);
     if ((p = ini_geta(conf, "multi_ap")))
         SCHEMA_SET_STR(vstate->multi_ap, hapd_map_int2str(atoi(p)));
     if ((vstate->ssid_broadcast_exists = (p = ini_geta(conf, "ignore_broadcast_ssid"))))
