From: Shubham Pawar <p.shubham@globaledgesoft.com>
Subject: [PATCH] lib/hostap: Set beacon interval

--- a/src/lib/hostap/src/hapd.c
+++ b/src/lib/hostap/src/hapd.c
@@ -605,6 +605,9 @@ hapd_conf_gen(struct hapd *hapd,
         csnprintf(&buf, &len, "ieee80211h=1\n");
     }
 
+    if (rconf->bcn_int_exists)
+        csnprintf(&buf, &len, "beacon_int=%d\n", rconf->bcn_int);
+
     WARN_ON(len == 1); /* likely buf was truncated */
 
     return hapd_conf_gen_psk(hapd, vconf);
