From: aparna.salian <aparna.salian@globaledgesoft.com>
Subject: [PATCH] lib/hostap: Callbacks added for hostapd DFS events

--- a/src/lib/hostap/inc/opensync-hapd.h
+++ b/src/lib/hostap/inc/opensync-hapd.h
@@ -55,6 +55,10 @@ struct hapd {
     void (*cmd_frame_probe_req)(struct hapd *hapd, const char *event);
     void (*cmd_frame_action)(struct hapd *hapd, const char *event);
     void (*cmd_frame_disconnect)(struct hapd *hapd, const char *event);
+    void (*dfs_event_cac_start)(struct hapd *hapd, const char *event);
+    void (*dfs_event_cac_completed)(struct hapd *hapd, const char *event);
+    void (*dfs_event_radar_detected)(struct hapd *hapd, const char *event);
+    void (*dfs_event_nop_finished)(struct hapd *hapd, const char *event);
     struct ctrl ctrl;
 };
 
--- a/src/lib/hostap/src/hapd.c
+++ b/src/lib/hostap/src/hapd.c
@@ -214,6 +214,30 @@ hapd_ctrl_cb(struct ctrl *ctrl, int level, const char *buf, size_t len)
         return;
     }
 
+    if (!strcmp(event, EV(DFS_EVENT_CAC_START))) {
+        LOGI("%s: dfs event - cac started", hapd->ctrl.bss);
+        if (hapd->dfs_event_cac_start)
+            hapd->dfs_event_cac_start(hapd, args);
+    }
+
+    if (!strcmp(event, EV(DFS_EVENT_CAC_COMPLETED))) {
+        LOGI("%s: dfs event - cac completed", hapd->ctrl.bss);
+        if (hapd->dfs_event_cac_completed)
+            hapd->dfs_event_cac_completed(hapd, args);
+    }
+
+    if (!strcmp(event, EV(DFS_EVENT_RADAR_DETECTED))) {
+        LOGI("%s: dfs event - radar detected", hapd->ctrl.bss);
+        if (hapd->dfs_event_radar_detected)
+            hapd->dfs_event_radar_detected(hapd, args);
+    }
+
+    if (!strcmp(event, EV(DFS_EVENT_NOP_FINISHED))) {
+        LOGI("%s: dfs event - nop finished", hapd->ctrl.bss);
+        if (hapd->dfs_event_nop_finished)
+            hapd->dfs_event_nop_finished(hapd, args);
+    }
+
     LOGI("%s: event: <%d> %s", ctrl->bss, level, buf);
 }
 
