From: aparna.salian <aparna.salian@globaledgesoft.com>
Subject: [PATCH] lib/hostap: Callback added for AP_CSA_FINISHED event

--- a/src/lib/hostap/inc/opensync-hapd.h
+++ b/src/lib/hostap/inc/opensync-hapd.h
@@ -59,6 +59,7 @@ struct hapd {
     void (*dfs_event_cac_completed)(struct hapd *hapd, const char *event);
     void (*dfs_event_radar_detected)(struct hapd *hapd, const char *event);
     void (*dfs_event_nop_finished)(struct hapd *hapd, const char *event);
+    void (*ap_csa_finished)(struct hapd *hapd, const char *event);
     struct ctrl ctrl;
 };
 
--- a/src/lib/hostap/src/hapd.c
+++ b/src/lib/hostap/src/hapd.c
@@ -238,6 +238,12 @@ hapd_ctrl_cb(struct ctrl *ctrl, int level, const char *buf, size_t len)
             hapd->dfs_event_nop_finished(hapd, args);
     }
 
+    if (!strcmp(event, EV(AP_CSA_FINISHED))) {
+        LOGI("%s: ap csa event", hapd->ctrl.bss);
+        if (hapd->ap_csa_finished)
+            hapd->ap_csa_finished(hapd, args);
+    }
+
     LOGI("%s: event: <%d> %s", ctrl->bss, level, buf);
 }
 
