From: aparna.salian <aparna.salian@globaledgesoft.com>
Subject: [PATCH] lib/hostap: Enable hostapd config option to use driver interface address

--- a/src/lib/hostap/src/hapd.c
+++ b/src/lib/hostap/src/hapd.c
@@ -581,6 +581,7 @@ hapd_conf_gen(struct hapd *hapd,
     csnprintf(&buf, &len, "ieee80211ac=%d\n", hapd_11ac_enabled(rconf));
     csnprintf(&buf, &len, "%s", hapd->ieee80211ax ? "" : "#");
     csnprintf(&buf, &len, "ieee80211ax=%d\n", hapd_11ax_enabled(rconf));
+    csnprintf(&buf, &len, "use_driver_iface_addr=1\n");
 
     /* FIXME: ieee80211n, iee80211ac, ieee80211ax is missing, also min_hw_mode..
      * perhaps some of this needs to be scheduled for wireless api rework
