Index: linux-5.4.164-qsdk-26349818b464f8c7b52d59ce73579d9f3dd6bd5d/arch/arm64/boot/dts/qcom/ipq6018.dtsi
===================================================================
--- linux-5.4.164-qsdk-26349818b464f8c7b52d59ce73579d9f3dd6bd5d.orig/arch/arm64/boot/dts/qcom/ipq6018.dtsi
+++ linux-5.4.164-qsdk-26349818b464f8c7b52d59ce73579d9f3dd6bd5d/arch/arm64/boot/dts/qcom/ipq6018.dtsi
@@ -1867,6 +1867,7 @@
 
 	qti,scm_restart_reason {
 		compatible = "qti_ipq6018,scm_restart_reason";
+		dload_status = <1>;
 	};
 
 	ctx_save: ctx-save {
