Index: linux-5.4.164-qsdk-26349818b464f8c7b52d59ce73579d9f3dd6bd5d/drivers/platform/ipq/bootconfig.c
===================================================================
--- linux-5.4.164-qsdk-26349818b464f8c7b52d59ce73579d9f3dd6bd5d.orig/drivers/platform/ipq/bootconfig.c
+++ linux-5.4.164-qsdk-26349818b464f8c7b52d59ce73579d9f3dd6bd5d/drivers/platform/ipq/bootconfig.c
@@ -366,7 +366,7 @@ static int __init bootconfig_partition_i
 			continue;
 
 		partname_dir[i] = proc_mkdir(part_info[i].name, boot_info_dir);
-		if (partname_dir != NULL) {
+		if (partname_dir[i] != NULL) {
 			proc_create_data("primaryboot", S_IRUGO,
 					   partname_dir[i],
 					   &primaryboot_ops,
