diff --git a/drivers/net/wireless/ath/ath12k/ce.c b/drivers/net/wireless/ath/ath12k/ce.c
index d2c8c425..2ca2f078 100644
--- a/drivers/net/wireless/ath/ath12k/ce.c
+++ b/drivers/net/wireless/ath/ath12k/ce.c
@@ -366,7 +366,8 @@ ath12k_ce_alloc_ring(struct ath12k_base *ab, int nentries, int desc_sz)
 	struct ath12k_ce_ring *ce_ring;
 	dma_addr_t base_addr;
 
-	ce_ring = kzalloc(struct_size(ce_ring, skb, nentries), GFP_KERNEL);
+	ce_ring = kzalloc(struct_size(ce_ring, skb, nentries), GFP_KERNEL,
+			  ath12k_ce_ring);
 	if (!ce_ring)
 		return ERR_PTR(-ENOMEM);
 
@@ -438,7 +439,8 @@ static int ath12k_ce_alloc_pipe(struct ath12k_base *ab, int ce_id)
 	/* As ce_stats are for debug purpose, bring up should not fail
 	 * in case of ce_stats allocation failure.
 	 */
-	ce_stats = kzalloc(sizeof(*pipe->ce_stats), GFP_KERNEL);
+	ce_stats = kzalloc(sizeof(*pipe->ce_stats), GFP_KERNEL,
+			   ath12k_ce_stats);
 	if (!ce_stats)
 		pipe->ce_stats = NULL;
 	else
diff --git a/drivers/net/wireless/ath/ath12k/core.c b/drivers/net/wireless/ath/ath12k/core.c
index b3b03225..301581ba 100644
--- a/drivers/net/wireless/ath/ath12k/core.c
+++ b/drivers/net/wireless/ath/ath12k/core.c
@@ -2716,7 +2716,9 @@ key_add:
 				return;
 
 			if (!arvif->is_created) {
-				key_conf = kzalloc(sizeof(*key_conf), GFP_ATOMIC);
+				key_conf = kzalloc(sizeof(*key_conf),
+						   GFP_ATOMIC,
+						   ath12k_key_conf);
 
 				if (!key_conf)
 					return;
@@ -3979,7 +3981,7 @@ static struct ath12k_hw_group *ath12k_core_hw_group_alloc(struct ath12k_base *ab
 	list_for_each_entry(ag, &ath12k_hw_group_list, list)
 		count++;
 
-	ag = kzalloc(sizeof(*ag), GFP_KERNEL);
+	ag = kzalloc(sizeof(*ag), GFP_KERNEL, ath12k_hw_group);
 	if (!ag)
 		return NULL;
 
@@ -4780,7 +4782,8 @@ int ath12k_wsi_load_info_init(struct ath12k_base *ab)
 		return ret;
 
 	if (!ag->wsi_load_info) {
-		wsi_load_info = kzalloc(sizeof(*wsi_load_info), GFP_KERNEL);
+		wsi_load_info = kzalloc(sizeof(*wsi_load_info), GFP_KERNEL,
+					ath12k_mlo_wsi_load_info);
 		if (!wsi_load_info)
 			return -ENOMEM;
 		ag->wsi_load_info = wsi_load_info;
@@ -5105,7 +5108,7 @@ struct ath12k_base *ath12k_core_alloc(struct device *dev, size_t priv_size,
 	struct ath12k_base *ab;
 	u32 addr;
 
-	ab = kzalloc(sizeof(*ab) + priv_size, GFP_KERNEL);
+	ab = kzalloc(sizeof(*ab) + priv_size, GFP_KERNEL, ath12k_base);
 	if (!ab)
 		return NULL;
 
diff --git a/drivers/net/wireless/ath/ath12k/coredump.c b/drivers/net/wireless/ath/ath12k/coredump.c
index 7e021d64..1b780639 100644
--- a/drivers/net/wireless/ath/ath12k/coredump.c
+++ b/drivers/net/wireless/ath/ath12k/coredump.c
@@ -230,7 +230,7 @@ static int ath12k_coredump_build_elf32(struct ath12k_base *ab,
 		cur_off += chunks[i].len;
 	}
 
-	st = kzalloc(sizeof(*st), GFP_KERNEL);
+	st = kzalloc(sizeof(*st), GFP_KERNEL, ath12k_pci_elf_coredump_state);
 	if (!st) {
 		vfree(elf_hdr);
 		kfree(chunks[0].vaddr);
@@ -573,7 +573,7 @@ void ath12k_coredump_download_rddm(struct ath12k_base *ab)
 
        len = num_seg * sizeof(*segment);
 
-       segment = kzalloc(len, GFP_NOWAIT);
+       segment = kzalloc(len, GFP_NOWAIT, ath12k_dump_segment);
        if (!segment) {
 		ath12k_err(ab, " Failed to allocate memory for segment for rddm download\n");
                return;
@@ -914,7 +914,7 @@ void ath12k_coredump_ahb_collect(struct ath12k_base *ab)
         phnum++;
 
         seg_info_len = phnum * sizeof(*segment);
-        segment = kzalloc(seg_info_len, GFP_NOWAIT);
+        segment = kzalloc(seg_info_len, GFP_NOWAIT, ath12k_ahb_dump_segment);
         if (!segment) {
                 ath12k_warn(ab, "Memory unavailable\n");
                 return;
diff --git a/drivers/net/wireless/ath/ath12k/dbring.c b/drivers/net/wireless/ath/ath12k/dbring.c
index d46d2ae1..33ce295e 100644
--- a/drivers/net/wireless/ath/ath12k/dbring.c
+++ b/drivers/net/wireless/ath/ath12k/dbring.c
@@ -137,7 +137,7 @@ static int ath12k_dbring_fill_bufs(struct ath12k *ar,
 	size = sizeof(*buff) + ring->buf_sz + align - 1;
 
 	while (num_remain > 0) {
-		buff = kzalloc(size, gfp);
+		buff = kzalloc(size, gfp, ath12k_dbring_element);
 		if (!buff)
 			break;
 
diff --git a/drivers/net/wireless/ath/ath12k/debugfs.c b/drivers/net/wireless/ath/ath12k/debugfs.c
index fa0d2fbd..c358b86c 100644
--- a/drivers/net/wireless/ath/ath12k/debugfs.c
+++ b/drivers/net/wireless/ath/ath12k/debugfs.c
@@ -2499,7 +2499,8 @@ static ssize_t ath12k_write_nrp_mac(struct file *file,
 		snprintf(fname + i, sizeof(fname) - i, "%02x:", mac[j]);
 	}
 
-	param = kzalloc(sizeof(*param), GFP_KERNEL);
+	param = kzalloc(sizeof(*param), GFP_KERNEL,
+			ath12k_set_neighbor_rx_params);
 	if (!param) {
 		ret = -ENOMEM;
 		goto exit;
@@ -2539,7 +2540,7 @@ static ssize_t ath12k_write_nrp_mac(struct file *file,
 		}
 		spin_unlock_bh(&dp->dp_lock);
 
-		nrp = kzalloc(sizeof(*nrp), GFP_KERNEL);
+		nrp = kzalloc(sizeof(*nrp), GFP_KERNEL, ath12k_neighbor_peer);
 		if (!nrp) {
 			ret = -ENOMEM;
 			goto err_free;
@@ -3808,7 +3809,7 @@ static ssize_t ath12k_read_vdev_tid_stats(struct file *file,
 	if (!buf)
 		return -ENOMEM;
 
-	tstats = kzalloc(sizeof(*tstats), GFP_KERNEL);
+	tstats = kzalloc(sizeof(*tstats), GFP_KERNEL, netdev_tid_stats);
 	if (!tstats) {
 		vfree(buf);
 		return -ENOMEM;
@@ -4020,7 +4021,7 @@ int ath12k_parse_qos_map(struct ath12k *ar, u8 *values, u8 len)
 	u8 num_des, des_len, i;
 	int ret;
 
-	qos_map = kzalloc(sizeof(*qos_map), GFP_KERNEL);
+	qos_map = kzalloc(sizeof(*qos_map), GFP_KERNEL, ath12k_qos_map);
 	if (!qos_map)
 		return -ENOMEM;
 
diff --git a/drivers/net/wireless/ath/ath12k/debugfs_htt_stats.c b/drivers/net/wireless/ath/ath12k/debugfs_htt_stats.c
index d5afa4a6..482d10bb 100644
--- a/drivers/net/wireless/ath/ath12k/debugfs_htt_stats.c
+++ b/drivers/net/wireless/ath/ath12k/debugfs_htt_stats.c
@@ -11819,7 +11819,8 @@ static int ath12k_open_htt_stats(struct inode *inode,
 		goto err_unlock;
 	}
 
-	stats_req = kzalloc(sizeof(*stats_req) + ATH12K_HTT_STATS_BUF_SIZE, GFP_KERNEL);
+	stats_req = kzalloc(sizeof(*stats_req) + ATH12K_HTT_STATS_BUF_SIZE,
+			    GFP_KERNEL, debug_htt_stats_req);
 	if (!stats_req) {
 		ret = -ENOMEM;
 		goto err_unlock;
diff --git a/drivers/net/wireless/ath/ath12k/dp_htt.c b/drivers/net/wireless/ath/ath12k/dp_htt.c
index 4a8e119b..4390ce3c 100644
--- a/drivers/net/wireless/ath/ath12k/dp_htt.c
+++ b/drivers/net/wireless/ath/ath12k/dp_htt.c
@@ -736,7 +736,8 @@ struct htt_ppdu_stats_info *ath12k_dp_htt_get_ppdu_desc(struct ath12k_pdev_dp *d
 		}
 	}
 
-	ppdu_info = kzalloc(sizeof(*ppdu_info), GFP_ATOMIC);
+	ppdu_info = kzalloc(sizeof(*ppdu_info), GFP_ATOMIC,
+			    htt_ppdu_stats_info);
 	if (!ppdu_info)
 		return NULL;
 
diff --git a/drivers/net/wireless/ath/ath12k/dp_mon.c b/drivers/net/wireless/ath/ath12k/dp_mon.c
index d82d0800..8cdd4f8f 100644
--- a/drivers/net/wireless/ath/ath12k/dp_mon.c
+++ b/drivers/net/wireless/ath/ath12k/dp_mon.c
@@ -495,7 +495,8 @@ ath12k_dp_mon_tx_get_ppdu_info(struct ath12k_mon_data *pmon,
 	}
 
 	/* allocate new tx_ppdu_info */
-	tx_ppdu_info = kzalloc(sizeof(*tx_ppdu_info), GFP_ATOMIC);
+	tx_ppdu_info = kzalloc(sizeof(*tx_ppdu_info), GFP_ATOMIC,
+			       dp_mon_tx_ppdu_info);
 	if (!tx_ppdu_info)
 		return NULL;
 
@@ -741,7 +742,7 @@ ath12k_dp_mon_tx_process_status_tlv(u32 tlv_status,
 	case HAL_TX_MON_MPDU_START:
 		struct dp_mon_mpdu *mon_mpdu = tx_ppdu_info->tx_mon_mpdu;
 
-		mon_mpdu = kzalloc(sizeof(*mon_mpdu), GFP_ATOMIC);
+		mon_mpdu = kzalloc(sizeof(*mon_mpdu), GFP_ATOMIC, dp_mon_mpdu);
 		break;
 	case HAL_TX_MON_MPDU_END:
 		list_add_tail(&tx_ppdu_info->tx_mon_mpdu->list,
@@ -1516,7 +1517,7 @@ int ath12k_dp_mon_init(struct ath12k_dp *dp)
 {
 	struct ath12k_dp_mon *dp_mon;
 
-	dp_mon = kzalloc(sizeof(*dp_mon), GFP_KERNEL);
+	dp_mon = kzalloc(sizeof(*dp_mon), GFP_KERNEL, ath12k_dp_mon);
 	if (!dp_mon)
 		return -ENOMEM;
 
@@ -1539,7 +1540,8 @@ int ath12k_dp_mon_pdev_alloc(struct ath12k_pdev_dp *dp_pdev)
 {
 	struct ath12k_pdev_mon_dp *dp_mon_pdev;
 
-	dp_mon_pdev = kzalloc(sizeof(*dp_mon_pdev), GFP_KERNEL);
+	dp_mon_pdev = kzalloc(sizeof(*dp_mon_pdev), GFP_KERNEL,
+			      ath12k_pdev_mon_dp);
 	if (!dp_mon_pdev)
 		return -ENOMEM;
 
diff --git a/drivers/net/wireless/ath/ath12k/dp_peer.c b/drivers/net/wireless/ath/ath12k/dp_peer.c
index 0d9e4a15..cfdade45 100644
--- a/drivers/net/wireless/ath/ath12k/dp_peer.c
+++ b/drivers/net/wireless/ath/ath12k/dp_peer.c
@@ -219,7 +219,7 @@ void ath12k_peer_map_event(struct ath12k_base *ab, u8 vdev_id, u16 peer_id,
 	spin_lock_bh(&dp->dp_lock);
 	peer = ath12k_dp_link_peer_find_by_vdev_id_and_addr(dp, vdev_id, mac_addr);
 	if (!peer) {
-		peer = kzalloc(sizeof(*peer), GFP_ATOMIC);
+		peer = kzalloc(sizeof(*peer), GFP_ATOMIC, ath12k_dp_link_peer);
 		if (!peer)
 			goto exit;
 
@@ -315,7 +315,7 @@ static int ath12k_dp_link_peer_rhash_addr_tbl_init(struct ath12k_dp *dp)
 		return 0;
 
 	size = sizeof(*dp->rhead_peer_addr);
-	rhash_addr_tbl = kzalloc(size, GFP_KERNEL);
+	rhash_addr_tbl = kzalloc(size, GFP_KERNEL, rhashtable);
 	if (!rhash_addr_tbl)
 		return -ENOMEM;
 
@@ -598,7 +598,7 @@ int ath12k_dp_peer_create(struct ath12k_dp_hw *dp_hw, u8 *addr,
 
 	spin_unlock_bh(&dp_hw->peer_lock);
 
-	dp_peer = kzalloc(sizeof(*dp_peer), GFP_ATOMIC);
+	dp_peer = kzalloc(sizeof(*dp_peer), GFP_ATOMIC, ath12k_dp_peer);
 	if (!dp_peer)
 		return -ENOMEM;
 
@@ -921,7 +921,7 @@ ath12k_dp_peer_qos_alloc(struct ath12k_dp *dp,
 	if (!peer)
 		return NULL;
 
-	qos = kzalloc(sizeof(*qos), GFP_ATOMIC);
+	qos = kzalloc(sizeof(*qos), GFP_ATOMIC, ath12k_dp_peer_qos);
 	if (!qos)
 		return NULL;
 
@@ -946,7 +946,7 @@ bool ath12k_dp_qos_stats_alloc(struct ath12k *ar,
 	    !ath12k_debugfs_is_qos_stats_enabled(ar))
 		return false;
 
-	qos_stats = kzalloc(sizeof(*qos_stats), GFP_ATOMIC);
+	qos_stats = kzalloc(sizeof(*qos_stats), GFP_ATOMIC, ath12k_qos_stats);
 	if (!qos_stats) {
 		ath12k_err(ar->ab, "Peer QoS stats allocation failed for peer: %pM link_id: %u\n",
 			   peer->addr, peer->link_id);
diff --git a/drivers/net/wireless/ath/ath12k/dp_rx.c b/drivers/net/wireless/ath/ath12k/dp_rx.c
index 36b2d03a..ec1f6070 100644
--- a/drivers/net/wireless/ath/ath12k/dp_rx.c
+++ b/drivers/net/wireless/ath/ath12k/dp_rx.c
@@ -1217,7 +1217,7 @@ struct dp_rx_fst *ath12k_dp_rx_fst_attach(struct ath12k_base *ab)
 	if (!ab->hw_params->support_fse)
 		return NULL;
 
-	fst = kzalloc(sizeof(*fst), GFP_KERNEL);
+	fst = kzalloc(sizeof(*fst), GFP_KERNEL, dp_rx_fst);
 	if (!fst)
 		return NULL;
 
diff --git a/drivers/net/wireless/ath/ath12k/mac.c b/drivers/net/wireless/ath/ath12k/mac.c
index be64f152..154ae3b0 100644
--- a/drivers/net/wireless/ath/ath12k/mac.c
+++ b/drivers/net/wireless/ath/ath12k/mac.c
@@ -1958,7 +1958,8 @@ static void ath12k_mac_nrp_delete(struct ath12k *ar)
 		return;
 	}
 
-	param = kzalloc(sizeof(*param) * nrp_pdev_count, GFP_KERNEL);
+	param = kzalloc(sizeof(*param) * nrp_pdev_count, GFP_KERNEL,
+			ath12k_set_neighbor_rx_params);
 	if (!param) {
 		/* Return entries to the original list */
 		spin_lock_bh(&dp->dp_lock);
@@ -4958,8 +4959,9 @@ ieee80211_link_sta *ath12k_mac_inherit_radio_cap(struct ath12k *ar,
 
 	sta = container_of((void *)arsta->ahsta, struct ieee80211_sta, drv_priv);
 
-	link_sta = (struct ieee80211_link_sta *)
-		   kzalloc(sizeof(struct ieee80211_link_sta), GFP_ATOMIC);
+	link_sta = (struct ieee80211_link_sta *)kzalloc(sizeof(struct ieee80211_link_sta),
+							GFP_ATOMIC,
+							ieee80211_link_sta);
 
 	if (!link_sta)
 		return NULL;
@@ -5578,8 +5580,9 @@ ath12k_mac_assign_link_vif(struct ath12k_hw *ah, struct ieee80211_vif *vif,
 	if (!ahvif->links_map && link_id < ATH12K_DEFAULT_SCAN_LINK) {
 		arvif = &ahvif->deflink;
 	} else {
-		arvif = (struct ath12k_link_vif *)
-		kzalloc(sizeof(struct ath12k_link_vif), GFP_KERNEL);
+		arvif = (struct ath12k_link_vif *)kzalloc(sizeof(struct ath12k_link_vif),
+							  GFP_KERNEL,
+							  ath12k_link_vif);
 		if (!arvif)
 			return NULL;
 	}
@@ -8567,7 +8570,7 @@ static int ath12k_mac_initiate_hw_scan(struct ieee80211_hw *hw,
 	if (ret)
 		goto exit;
 
-	arg = kzalloc(sizeof(*arg), GFP_KERNEL);
+	arg = kzalloc(sizeof(*arg), GFP_KERNEL, ath12k_wmi_scan_req_arg);
 	if (!arg) {
 		ret = -ENOMEM;
 		goto exit;
@@ -9038,7 +9041,8 @@ static int ath12k_mac_update_key_cache(struct ath12k_vif_cache *cache,
 	}
 
 	if (cmd == SET_KEY) {
-		key_conf = kzalloc(sizeof(*key_conf), GFP_KERNEL);
+		key_conf = kzalloc(sizeof(*key_conf), GFP_KERNEL,
+				   ath12k_key_conf);
 
 		if (!key_conf)
 			return -ENOMEM;
@@ -10669,9 +10673,9 @@ static int ath12k_mac_set_peer_ch_switch_data(struct ath12k_link_vif *arvif,
 	peer_data = arvif->peer_ch_width_switch_data;
 
 	if (!peer_data) {
-		peer_data = kzalloc(struct_size(peer_data, peer_arg,
-						num_sta_count),
-				    GFP_KERNEL);
+		peer_data = kzalloc(struct_size(peer_data, peer_arg, num_sta_count),
+				    GFP_KERNEL,
+				    ath12k_peer_ch_width_switch_data);
 		if (!peer_data)
 			return -ENOMEM;
 
@@ -12424,7 +12428,7 @@ static struct ath12k_link_sta *ath12k_mac_alloc_assign_link_sta(struct ath12k_hw
 		return ahsta->link[link_id];
 	}
 
-	arsta = kmalloc(sizeof(*arsta), GFP_KERNEL);
+	arsta = kmalloc(sizeof(*arsta), GFP_KERNEL, ath12k_link_sta);
 	if (!arsta)
 		return NULL;
 
@@ -13033,7 +13037,8 @@ void ath12k_mac_op_set_dscp_tid(struct ieee80211_hw *hw,
 	if (link_id >= IEEE80211_MLD_MAX_NUM_LINKS)
 		return;
 
-	new_qos_map = kzalloc(sizeof(*new_qos_map), GFP_KERNEL);
+	new_qos_map = kzalloc(sizeof(*new_qos_map), GFP_KERNEL,
+			      ath12k_qos_map);
 	if (!new_qos_map) {
 		return;
 	}
@@ -13391,7 +13396,8 @@ ath12k_mac_get_link_migr_peer_node(struct ath12k_dp_peer *ml_peer,
 
 	hw_link_id = arvif->ar->pdev->hw_link_id;
 
-	node = kzalloc(sizeof(*node), GFP_ATOMIC);
+	node = kzalloc(sizeof(*node), GFP_ATOMIC,
+		       ath12k_mac_pri_link_migr_peer_node);
 	if (!node)
 		return NULL;
 
@@ -16953,7 +16959,8 @@ ppe_vp_config:
 		struct ath12k_vlan_iface *vlan_iface;
 		int ret;
 
-		vlan_iface = kzalloc(sizeof(*vlan_iface), GFP_ATOMIC);
+		vlan_iface = kzalloc(sizeof(*vlan_iface), GFP_ATOMIC,
+				     ath12k_vlan_iface);
 		if (!vlan_iface) {
 			ret = -ENOMEM;
 			if (ahvif->dp_vif.ppe_vp_type == PPE_VP_USER_TYPE_DS)
@@ -22492,7 +22499,8 @@ static int ath12k_mac_setup_iface_combinations(struct ath12k_hw *ah)
 	struct ath12k *ar;
 	int i, ret;
 
-	combinations = kzalloc(sizeof(*combinations), GFP_KERNEL);
+	combinations = kzalloc(sizeof(*combinations), GFP_KERNEL,
+			       ieee80211_iface_combination);
 	if (!combinations)
 		return -ENOMEM;
 
@@ -22517,7 +22525,8 @@ static int ath12k_mac_setup_iface_combinations(struct ath12k_hw *ah)
 	}
 
 	for_each_ar(ah, ar, i) {
-		comb = kzalloc(sizeof(*comb), GFP_KERNEL);
+		comb = kzalloc(sizeof(*comb), GFP_KERNEL,
+			       ieee80211_iface_combination);
 		if (!comb) {
 			ret = -ENOMEM;
 			goto err_free_radios;
@@ -22840,8 +22849,8 @@ static int ath12k_alloc_per_hw_mac_addr(struct ath12k_hw *ah)
 	int i;
 	ar = ah->radio;
 
-	addresses = kzalloc(sizeof(*addresses) * ah->num_radio,
-			    GFP_KERNEL);
+	addresses = kzalloc(sizeof(*addresses) * ah->num_radio, GFP_KERNEL,
+			    mac_address);
 	if(!addresses)
 		return -ENOMEM;
 
diff --git a/drivers/net/wireless/ath/ath12k/peer.c b/drivers/net/wireless/ath/ath12k/peer.c
index 01bafcc2..11afa4c4 100644
--- a/drivers/net/wireless/ath/ath12k/peer.c
+++ b/drivers/net/wireless/ath/ath12k/peer.c
@@ -578,7 +578,7 @@ static int ath12k_link_sta_rhash_addr_tbl_init(struct ath12k_base *ab)
 		return 0;
 
 	size = sizeof(*ab->rhead_sta_addr);
-	rhash_addr_tbl = kzalloc(size, GFP_KERNEL);
+	rhash_addr_tbl = kzalloc(size, GFP_KERNEL, rhashtable);
 	if (!rhash_addr_tbl)
 		return -ENOMEM;
 
diff --git a/drivers/net/wireless/ath/ath12k/qmi.c b/drivers/net/wireless/ath/ath12k/qmi.c
index 224a5f4f..a912c75e 100644
--- a/drivers/net/wireless/ath/ath12k/qmi.c
+++ b/drivers/net/wireless/ath/ath12k/qmi.c
@@ -2792,11 +2792,13 @@ int ath12k_qmi_mem_read(struct ath12k_base *ab, u32 mem_addr, void *mem_value,si
 	struct qmi_txn txn = {};
 	int ret;
 
-	req = kzalloc(sizeof(*req), GFP_KERNEL);
+	req = kzalloc(sizeof(*req), GFP_KERNEL,
+		      qmi_wlanfw_mem_read_req_msg_v01);
 	if (!req)
 		return -ENOMEM;
 
-	resp = kzalloc(sizeof(*resp), GFP_KERNEL);
+	resp = kzalloc(sizeof(*resp), GFP_KERNEL,
+		       qmi_wlanfw_mem_read_resp_msg_v01);
 	if (!resp) {
 		kfree(req);
 		return -ENOMEM;
@@ -2860,11 +2862,13 @@ int ath12k_qmi_mem_write(struct ath12k_base *ab, u32 mem_addr, void* mem_value,
 	struct qmi_txn txn = {};
 	int ret;
 
-	req = kzalloc(sizeof(*req), GFP_KERNEL);
+	req = kzalloc(sizeof(*req), GFP_KERNEL,
+		      qmi_wlanfw_mem_write_req_msg_v01);
 	if (!req)
 		return -ENOMEM;
 
-	resp = kzalloc(sizeof(*resp), GFP_KERNEL);
+	resp = kzalloc(sizeof(*resp), GFP_KERNEL,
+		       qmi_wlanfw_mem_write_resp_msg_v01);
 	if (!resp) {
 		kfree(req);
 		return -ENOMEM;
@@ -3602,11 +3606,13 @@ static int ath12k_qmi_fw_ind_register_send(struct ath12k_base *ab)
 	struct qmi_txn txn;
 	int ret;
 
-	req = kzalloc(sizeof(*req), GFP_KERNEL);
+	req = kzalloc(sizeof(*req), GFP_KERNEL,
+		      qmi_wlanfw_ind_register_req_msg_v01);
 	if (!req)
 		return -ENOMEM;
 
-	resp = kzalloc(sizeof(*resp), GFP_KERNEL);
+	resp = kzalloc(sizeof(*resp), GFP_KERNEL,
+		       qmi_wlanfw_ind_register_resp_msg_v01);
 	if (!resp) {
 		ret = -ENOMEM;
 		goto resp_out;
@@ -3681,7 +3687,8 @@ int ath12k_qmi_respond_fw_mem_request(struct ath12k_base *ab)
 	int ret = 0, i;
 	bool delayed;
 
-	req = kzalloc(sizeof(*req), GFP_KERNEL);
+	req = kzalloc(sizeof(*req), GFP_KERNEL,
+		      qmi_wlanfw_respond_mem_req_msg_v01);
 	if (!req)
 		return -ENOMEM;
 
@@ -4623,7 +4630,8 @@ static int ath12k_qmi_load_file_target_mem(struct ath12k_base *ab,
 	int ret = 0;
 	u32 remaining = len;
 
-	req = kzalloc(sizeof(*req), GFP_KERNEL);
+	req = kzalloc(sizeof(*req), GFP_KERNEL,
+		      qmi_wlanfw_bdf_download_req_msg_v01);
 	if (!req)
 		return -ENOMEM;
 
@@ -5027,7 +5035,8 @@ static int ath12k_qmi_wlanfw_wlan_cfg_send(struct ath12k_base *ab)
 	ce_cfg	= (struct ce_pipe_config *)ab->qmi.ce_cfg.tgt_ce;
 	svc_cfg	= (struct service_to_pipe *)ab->qmi.ce_cfg.svc_to_ce_map;
 
-	req = kzalloc(sizeof(*req), GFP_KERNEL);
+	req = kzalloc(sizeof(*req), GFP_KERNEL,
+		      qmi_wlanfw_wlan_cfg_req_msg_v01);
 	if (!req)
 		return -ENOMEM;
 
@@ -5213,11 +5222,13 @@ int ath12k_qmi_m3_dump_upload_done_ind_send(struct ath12k_base *ab,
         struct qmi_txn txn;
         int ret;
 
-        req = kzalloc(sizeof(*req), GFP_KERNEL);
+        req = kzalloc(sizeof(*req), GFP_KERNEL,
+                      qmi_wlanfw_m3_dump_upload_done_req_msg_v01);
         if (!req)
                 return -ENOMEM;
 
-        resp = kzalloc(sizeof(*resp), GFP_KERNEL);
+        resp = kzalloc(sizeof(*resp), GFP_KERNEL,
+                       qmi_wlanfw_m3_dump_upload_done_resp_msg_v01);
         if (!resp) {
                 kfree(req);
                 return -ENOMEM;
@@ -5322,7 +5333,7 @@ ath12k_qmi_driver_event_post(struct ath12k_qmi *qmi,
 {
 	struct ath12k_qmi_driver_event *event;
 
-	event = kzalloc(sizeof(*event), GFP_ATOMIC);
+	event = kzalloc(sizeof(*event), GFP_ATOMIC, ath12k_qmi_driver_event);
 	if (!event)
 		return -ENOMEM;
 
@@ -5433,11 +5444,13 @@ static int ath12k_qmi_fw_cfg_send_sync(struct ath12k_base *ab,
 	const u8 *temp = data;
 	unsigned int remaining = len;
 
-	req = kzalloc(sizeof(*req), GFP_KERNEL);
+	req = kzalloc(sizeof(*req), GFP_KERNEL,
+		      wlanfw_cfg_download_req_msg_v01);
 	if (!req)
 		return -ENOMEM;
 
-	resp = kzalloc(sizeof(*resp), GFP_KERNEL);
+	resp = kzalloc(sizeof(*resp), GFP_KERNEL,
+		       wlanfw_cfg_download_resp_msg_v01);
 	if (!resp) {
 		kfree(req);
 		return -ENOMEM;
@@ -5789,7 +5802,8 @@ static void ath12k_qmi_m3_dump_upload_req_ind_cb(struct qmi_handle *qmi_hdl,
 
         ath12k_dbg(ab, ATH12K_DBG_QMI, "qmi m3 dump memory request\n");
 
-        event_data = kzalloc(sizeof(*event_data), GFP_KERNEL);
+        event_data = kzalloc(sizeof(*event_data), GFP_KERNEL,
+                             ath12k_qmi_m3_dump_upload_req_data);
         if (!event_data)
                 return;
 
@@ -5808,7 +5822,7 @@ int ath12k_enable_fwlog(struct ath12k_base *ab)
 	struct qmi_txn txn = {};
 	int ret = 0;
 
-	req = kzalloc(sizeof(*req), GFP_KERNEL);
+	req = kzalloc(sizeof(*req), GFP_KERNEL, wlfw_ini_req_msg_v01);
 	if (!req)
 		return -ENOMEM;
 
diff --git a/drivers/net/wireless/ath/ath12k/reg.c b/drivers/net/wireless/ath/ath12k/reg.c
index 97bf3343..8d78995f 100644
--- a/drivers/net/wireless/ath/ath12k/reg.c
+++ b/drivers/net/wireless/ath/ath12k/reg.c
@@ -455,7 +455,7 @@ int ath12k_regd_update(struct ath12k *ar, bool init)
 	regd_len = sizeof(*regd) + (regd->n_reg_rules *
 		sizeof(struct ieee80211_reg_rule));
 
-	regd_copy = kzalloc(regd_len, GFP_ATOMIC);
+	regd_copy = kzalloc(regd_len, GFP_ATOMIC, ieee80211_regdomain);
 	if (regd_copy)
 		ath12k_copy_regd(regd, regd_copy);
 
@@ -994,18 +994,16 @@ ath12k_reg_build_regd(struct ath12k_base *ab,
 	 * rule in ab and use this to intersect with afc rules each time we get
 	 * afc power info.
 	 */
-	sp_rule = kzalloc(sizeof(*sp_rule) +
-			  (num_6ghz_sp_rules * sizeof(struct ieee80211_reg_rule)),
-			  GFP_ATOMIC);
+	sp_rule = kzalloc(sizeof(*sp_rule) + (num_6ghz_sp_rules * sizeof(struct ieee80211_reg_rule)),
+			  GFP_ATOMIC, ath12k_6ghz_sp_reg_rule);
 
 	if (!sp_rule)
 		return new_regd;
 
 	sp_rule->num_6ghz_sp_rule = num_6ghz_sp_rules;
 
-	new_regd = kzalloc(sizeof(*new_regd) +
-			   (num_rules * sizeof(struct ieee80211_reg_rule)),
-			   GFP_ATOMIC);
+	new_regd = kzalloc(sizeof(*new_regd) + (num_rules * sizeof(struct ieee80211_reg_rule)),
+			   GFP_ATOMIC, ieee80211_regdomain);
 	if (!new_regd) {
 		kfree(sp_rule);
 		goto ret;
@@ -1985,9 +1983,8 @@ ath12k_reg_intersect_with_partner_afc_info(struct ath12k *ar,
 		return 0;
 	}
 
-	partner_intersected_ranges =
-	    kzalloc(num_partner_afc_rules *
-		    sizeof(*partner_intersected_ranges), GFP_ATOMIC);
+	partner_intersected_ranges = kzalloc(num_partner_afc_rules * sizeof(*partner_intersected_ranges),
+					     GFP_ATOMIC, ath12k_afc_freq_obj);
 
 	if (!partner_intersected_ranges) {
 		ath12k_err(ab, "Failed to alloc partner_intersected_ranges\n");
@@ -2389,9 +2386,8 @@ ath12k_reg_build_new_regd_for_afc(struct ath12k *ar,
 				    num_regd_rules + num_old_sp_rules);
 	}
 
-	new_regd = kzalloc(sizeof(*new_regd) +
-			   (sizeof(*new_regd_rules) * new_reg_rule_cnt),
-			   GFP_ATOMIC);
+	new_regd = kzalloc(sizeof(*new_regd) + (sizeof(*new_regd_rules) * new_reg_rule_cnt),
+			   GFP_ATOMIC, ieee80211_regdomain);
 	if (!new_regd) {
 		ath12k_err(ab, "Failed to alloc new_regd\n");
 		return -ENOMEM;
@@ -3088,7 +3084,8 @@ ath12k_reg_fill_afc_opclass_obj_lst(struct ath12k_afc_host_request *p_afc_req)
 	if (ret)
 		return NULL;
 
-	opclass_obj_lst =  kzalloc(sizeof(*opclass_obj_lst), GFP_ATOMIC);
+	opclass_obj_lst = kzalloc(sizeof(*opclass_obj_lst), GFP_ATOMIC,
+				  ath12k_afc_opclass_obj_list);
 	if (!opclass_obj_lst) {
 		ath12k_free_6g_opclasses_and_channels(num_opclasses,
 						      opclass_lst, chansize_lst,
@@ -3352,7 +3349,8 @@ static struct ath12k_afc_frange_list *ath12k_reg_fill_freq_lst(struct ath12k *ar
 		return NULL;
 	}
 
-	p_frange_lst_local = kzalloc(sizeof(*p_frange_lst_local), GFP_ATOMIC);
+	p_frange_lst_local = kzalloc(sizeof(*p_frange_lst_local), GFP_ATOMIC,
+				     ath12k_afc_frange_list);
 	if (!p_frange_lst_local)
 		return NULL;
 
@@ -3376,7 +3374,8 @@ static struct ath12k_afc_location *ath12k_reg_fill_afc_location_obj(struct ath12
 {
 	struct ath12k_afc_location *p_afc_location;
 
-	p_afc_location = kzalloc(sizeof(*p_afc_location), GFP_ATOMIC);
+	p_afc_location = kzalloc(sizeof(*p_afc_location), GFP_ATOMIC,
+				 ath12k_afc_location);
 	if (!p_afc_location)
 		return NULL;
 
diff --git a/drivers/net/wireless/ath/ath12k/wifi7/dp.c b/drivers/net/wireless/ath/ath12k/wifi7/dp.c
index d34cd2b4..e2b70354 100644
--- a/drivers/net/wireless/ath/ath12k/wifi7/dp.c
+++ b/drivers/net/wireless/ath/ath12k/wifi7/dp.c
@@ -320,7 +320,8 @@ static struct ath12k_dp_hw_group *ath12k_wifi7_dp_hw_group_alloc(void)
 {
 	struct ath12k_dp_hw_group *dp_hw_grp;
 
-	dp_hw_grp = kzalloc(sizeof(*dp_hw_grp), GFP_KERNEL);
+	dp_hw_grp = kzalloc(sizeof(*dp_hw_grp), GFP_KERNEL,
+			    ath12k_dp_hw_group);
 	if (!dp_hw_grp) {
 		pr_err("failed to allocate dp_hw_group\n");
 		return NULL;
@@ -363,7 +364,7 @@ struct ath12k_dp *ath12k_wifi7_dp_init(struct ath12k_base *ab)
 	struct ath12k_dp *dp;
 	int ret;
 
-	dp = kzalloc(sizeof(*dp), GFP_KERNEL);
+	dp = kzalloc(sizeof(*dp), GFP_KERNEL, ath12k_dp);
 	if (!dp)
 		return NULL;
 
diff --git a/drivers/net/wireless/ath/ath12k/wifi7/dp_mon1.c b/drivers/net/wireless/ath/ath12k/wifi7/dp_mon1.c
index 9285bf43..dbd0871a 100644
--- a/drivers/net/wireless/ath/ath12k/wifi7/dp_mon1.c
+++ b/drivers/net/wireless/ath/ath12k/wifi7/dp_mon1.c
@@ -1086,7 +1086,8 @@ void ath12k_wifi7_dp_mon_rx_dest_process(struct ath12k_pdev_dp *dp_pdev, int mac
 		}
 
 		if (head_msdu && tail_msdu) {
-			tmp_mpdu = kzalloc(sizeof(*tmp_mpdu), GFP_ATOMIC);
+			tmp_mpdu = kzalloc(sizeof(*tmp_mpdu), GFP_ATOMIC,
+					   dp_mon_mpdu);
 			if (!tmp_mpdu)
 				break;
 
diff --git a/drivers/net/wireless/ath/ath12k/wifi7/dp_rx.c b/drivers/net/wireless/ath/ath12k/wifi7/dp_rx.c
index 41136241..7ec32951 100644
--- a/drivers/net/wireless/ath/ath12k/wifi7/dp_rx.c
+++ b/drivers/net/wireless/ath/ath12k/wifi7/dp_rx.c
@@ -139,7 +139,7 @@ int ath12k_wifi7_dp_reo_cmd_send(struct ath12k_base *ab,
 	 * for tid delete command to free up the resource on the command status
 	 * indication?
 	 */
-	dp_cmd = kzalloc(sizeof(*dp_cmd), GFP_ATOMIC);
+	dp_cmd = kzalloc(sizeof(*dp_cmd), GFP_ATOMIC, ath12k_dp_rx_reo_cmd);
 
 	if (!dp_cmd)
 		return -ENOMEM;
@@ -255,7 +255,8 @@ void ath12k_wifi7_dp_rx_tid_del_func(struct ath12k_dp *dp, void *ctx,
 	}
 	spin_unlock_bh(&dp->reo_cmd_update_rx_queue_lock);
 
-	elem = kzalloc(sizeof(*elem), GFP_ATOMIC);
+	elem = kzalloc(sizeof(*elem), GFP_ATOMIC,
+		       ath12k_dp_rx_reo_cache_flush_elem);
 	if (!elem)
 		goto free_desc;
 
@@ -367,7 +368,7 @@ void ath12k_wifi7_dp_rx_peer_tid_delete(struct ath12k *ar,
 	if (!rx_tid->active)
 		return;
 
-	elem = kzalloc(sizeof(*elem), GFP_ATOMIC);
+	elem = kzalloc(sizeof(*elem), GFP_ATOMIC, dp_reo_update_rx_queue_elem);
 	if (!elem)
 		return;
 
diff --git a/drivers/net/wireless/ath/ath12k/wifi7/hal_rx.c b/drivers/net/wireless/ath/ath12k/wifi7/hal_rx.c
index f4d42c25..baf1181b 100644
--- a/drivers/net/wireless/ath/ath12k/wifi7/hal_rx.c
+++ b/drivers/net/wireless/ath/ath12k/wifi7/hal_rx.c
@@ -1383,7 +1383,7 @@ ath12k_wifi7_hal_rx_fst_attach(struct ath12k_base *ab)
 	struct hal_rx_fst *fst;
 	u32 alloc_size;
 
-	fst = kzalloc(sizeof(*fst), GFP_KERNEL);
+	fst = kzalloc(sizeof(*fst), GFP_KERNEL, hal_rx_fst);
 	if (!fst)
 		return NULL;
 
diff --git a/drivers/net/wireless/ath/ath12k/wifi7/hw.c b/drivers/net/wireless/ath/ath12k/wifi7/hw.c
index d3f44309..fc060477 100644
--- a/drivers/net/wireless/ath/ath12k/wifi7/hw.c
+++ b/drivers/net/wireless/ath/ath12k/wifi7/hw.c
@@ -1851,7 +1851,8 @@ int ath12k_wifi7_hw_init(struct ath12k_base *ab)
 
 	if (ab->hif.bus == ATH12K_BUS_PCI &&
 	    ab->msi.config->total_vectors == ATH12K_MSI_16) {
-		hw_params_msi8 = kzalloc(sizeof(struct ath12k_hw_params), GFP_KERNEL);
+		hw_params_msi8 = kzalloc(sizeof(struct ath12k_hw_params),
+					 GFP_KERNEL, ath12k_hw_params);
 		if (!hw_params_msi8)
 			return -ENOMEM;
 		memcpy(hw_params_msi8, hw_params, sizeof(struct ath12k_hw_params));
diff --git a/drivers/net/wireless/ath/ath12k/wmi.c b/drivers/net/wireless/ath/ath12k/wmi.c
index cafa49b6..974b4f45 100644
--- a/drivers/net/wireless/ath/ath12k/wmi.c
+++ b/drivers/net/wireless/ath/ath12k/wmi.c
@@ -3553,7 +3553,8 @@ int ath12k_wmi_update_scan_chan_list(struct ath12k *ar,
 		return -EOPNOTSUPP;
 	}
 
-	arg = kzalloc(struct_size(arg, channel, num_channels), GFP_KERNEL);
+	arg = kzalloc(struct_size(arg, channel, num_channels), GFP_KERNEL,
+		      ath12k_wmi_scan_chan_list_arg);
 
 	if (!arg)
 		return -ENOMEM;
@@ -6751,7 +6752,8 @@ ath12k_wmi_pdev_update_muedca_params_status_event(struct ath12k_base *ab,
 		   "update MU-EDCA parameters for radio index %u\n",
 		   ar->radio_idx);
 
-	params = kzalloc(sizeof(*params), GFP_ATOMIC);
+	params = kzalloc(sizeof(*params), GFP_ATOMIC,
+			 ieee80211_mu_edca_param_set);
 	if (!params) {
 		ath12k_warn(ab,
 			    "Failed to allocate memory for updated MU-EDCA Parameters");
@@ -7574,7 +7576,8 @@ static int ath12k_copy_afc_power_event_fixed_info(struct ath12k_base *ab,
 	struct ath12k_afc_sp_reg_info *afc_reg_info;
 
 	afc_pwr_param = (struct wmi_afc_power_event_param *)ptr;
-	afc_reg_info = kzalloc(sizeof(*afc_reg_info), GFP_ATOMIC);
+	afc_reg_info = kzalloc(sizeof(*afc_reg_info), GFP_ATOMIC,
+			       ath12k_afc_sp_reg_info);
 
 	if (!afc_reg_info)
 		return -ENOMEM;
@@ -7619,7 +7622,7 @@ static int ath12k_wmi_afc_fill_freq_obj(struct ath12k_base *ab,
 	ath12k_dbg(ab, ATH12K_DBG_AFC, "Num afc freq obj received %d\n",
 		   afc_reg_info->num_freq_objs);
 	freq_obj = kzalloc(afc_reg_info->num_freq_objs * sizeof(*freq_obj),
-			   GFP_ATOMIC);
+			   GFP_ATOMIC, ath12k_afc_freq_obj);
 	if (!freq_obj)
 		return -ENOMEM;
 
@@ -7664,7 +7667,7 @@ static int ath12k_wmi_afc_fill_chan_obj(struct ath12k_base *ab,
 	ath12k_dbg(ab, ATH12K_DBG_AFC, "Num chan objects received %d\n",
 		   afc_reg_info->num_chan_objs);
 	chan_obj = kzalloc(afc_reg_info->num_chan_objs * sizeof(*chan_obj),
-			   GFP_ATOMIC);
+			   GFP_ATOMIC, ath12k_afc_chan_obj);
 	if (!chan_obj)
 		return -ENOMEM;
 
@@ -7678,7 +7681,7 @@ static int ath12k_wmi_afc_fill_chan_obj(struct ath12k_base *ab,
 			   "Chan obj %d  global_opclass : %d num_chans %d\n", i,
 			   chan_obj[i].global_opclass, chan_obj[i].num_chans);
 		eirp_info = kzalloc(chan_obj[i].num_chans * sizeof(*eirp_info),
-				    GFP_ATOMIC);
+				    GFP_ATOMIC, ath12k_chan_eirp_obj);
 		if (!eirp_info)
 			return -ENOMEM;
 
@@ -7988,7 +7991,7 @@ static struct ath12k_reg_rule
 	u32 count;
 
 	reg_rule_ptr = kzalloc((num_reg_rules * sizeof(*reg_rule_ptr)),
-			       GFP_ATOMIC);
+			       GFP_ATOMIC, ath12k_reg_rule);
 
 	if (!reg_rule_ptr)
 		return NULL;
@@ -9807,7 +9810,7 @@ static int ath12k_reg_chan_list_event(struct ath12k_base *ab, struct sk_buff *sk
        struct ath12k_reg_info *reg_info;
        int ret, i, j;
 
-       reg_info = kzalloc(sizeof(*reg_info), GFP_ATOMIC);
+       reg_info = kzalloc(sizeof(*reg_info), GFP_ATOMIC, ath12k_reg_info);
        if (!reg_info)
                return -ENOMEM;
 
@@ -11379,7 +11382,7 @@ static int ath12k_wmi_tlv_fw_stats_data_parse(struct ath12k_base *ab,
 
 		data += sizeof(*src);
 		len -= sizeof(*src);
-		dst = kzalloc(sizeof(*dst), GFP_ATOMIC);
+		dst = kzalloc(sizeof(*dst), GFP_ATOMIC, ath12k_fw_stats_vdev);
 		if (!dst)
 			continue;
 		ath12k_wmi_pull_vdev_stats(src, dst);
@@ -11398,7 +11401,7 @@ static int ath12k_wmi_tlv_fw_stats_data_parse(struct ath12k_base *ab,
 
 		data += sizeof(*src);
 		len -= sizeof(*src);
-		dst = kzalloc(sizeof(*dst), GFP_ATOMIC);
+		dst = kzalloc(sizeof(*dst), GFP_ATOMIC, ath12k_fw_stats_bcn);
 		if (!dst)
 			continue;
 		ath12k_wmi_pull_bcn_stats(src, dst);
@@ -11420,7 +11423,7 @@ static int ath12k_wmi_tlv_fw_stats_data_parse(struct ath12k_base *ab,
 		data += sizeof(*src);
 		len -= sizeof(*src);
 
-		dst = kzalloc(sizeof(*dst), GFP_ATOMIC);
+		dst = kzalloc(sizeof(*dst), GFP_ATOMIC, ath12k_fw_stats_pdev);
 		if (!dst)
 			continue;
 
@@ -11973,7 +11976,8 @@ ath12k_wmi_dcs_wlan_interference_event(struct ath12k_base *ab,
 	}
 	spin_unlock_bh(&ar->data_lock);
 
-	dcs_wlan_intf = kzalloc(sizeof(*dcs_wlan_intf), GFP_ATOMIC);
+	dcs_wlan_intf = kzalloc(sizeof(*dcs_wlan_intf), GFP_ATOMIC,
+				ath12k_dcs_wlan_interference);
 	if (!dcs_wlan_intf)
 		goto exit;
 
@@ -13095,13 +13099,16 @@ int wmi_print_ctrl_path_pdev_tx_stats_tlv(struct ath12k_base *ab, u16 len, const
 	struct wmi_ctrl_path_stats_ev_parse_param *stats_buff = (struct wmi_ctrl_path_stats_ev_parse_param *)data;
 	struct wmi_ctrl_path_pdev_stats_params *pdev_stats_skb = (struct wmi_ctrl_path_pdev_stats_params *)ptr;
 	struct wmi_ctrl_path_pdev_stats_params *pdev_stats = NULL;
-	struct wmi_ctrl_path_stats_list *stats = kzalloc(sizeof(struct wmi_ctrl_path_stats_list), GFP_ATOMIC);
+	struct wmi_ctrl_path_stats_list *stats = kzalloc(sizeof(struct wmi_ctrl_path_stats_list),
+							 GFP_ATOMIC,
+							 wmi_ctrl_path_stats_list);
 	struct ath12k *ar = NULL;
 
 	if (!stats)
 		return -ENOMEM;
 
-	pdev_stats = kzalloc(sizeof(*pdev_stats), GFP_ATOMIC);
+	pdev_stats = kzalloc(sizeof(*pdev_stats), GFP_ATOMIC,
+			     wmi_ctrl_path_pdev_stats_params);
 	if (!pdev_stats) {
 		kfree(stats);
 		return -ENOMEM;
@@ -13135,13 +13142,16 @@ int wmi_print_ctrl_path_cal_stats_tlv(struct ath12k_base *ab, u16 len,
 	struct wmi_ctrl_path_stats_ev_parse_param *stats_buff = (struct wmi_ctrl_path_stats_ev_parse_param *)data;
 	struct wmi_ctrl_path_cal_stats *cal_stats_skb = (struct wmi_ctrl_path_cal_stats *)ptr;
 	struct wmi_ctrl_path_cal_stats *cal_stats = NULL;
-	struct wmi_ctrl_path_stats_list *stats = kzalloc(sizeof(struct wmi_ctrl_path_stats_list), GFP_ATOMIC);
+	struct wmi_ctrl_path_stats_list *stats = kzalloc(sizeof(struct wmi_ctrl_path_stats_list),
+							 GFP_ATOMIC,
+							 wmi_ctrl_path_stats_list);
 	struct ath12k *ar = NULL;
 
 	if (!stats)
 		return -ENOMEM;
 
-	cal_stats = kzalloc(sizeof(*cal_stats), GFP_ATOMIC);
+	cal_stats = kzalloc(sizeof(*cal_stats), GFP_ATOMIC,
+			    wmi_ctrl_path_cal_stats);
 	if (!cal_stats) {
 		kfree(stats);
 		return -ENOMEM;
@@ -13180,11 +13190,12 @@ int wmi_print_ctrl_path_btcoex_stats_tlv(struct ath12k_base *ab, u16 len,
 	struct wmi_ctrl_path_stats_list *stats;
 	struct ath12k *ar = NULL;
 
-	stats = kzalloc(sizeof(*stats), GFP_ATOMIC);
+	stats = kzalloc(sizeof(*stats), GFP_ATOMIC, wmi_ctrl_path_stats_list);
 	if (!stats)
 		return -ENOMEM;
 
-	btcoex_stats = kzalloc(sizeof(*btcoex_stats), GFP_ATOMIC);
+	btcoex_stats = kzalloc(sizeof(*btcoex_stats), GFP_ATOMIC,
+			       wmi_ctrl_path_btcoex_stats);
 	if (!btcoex_stats) {
 		kfree(stats);
 		return -ENOMEM;
@@ -13255,11 +13266,12 @@ int wmi_print_ctrl_path_awgn_stats_tlv(struct ath12k_base *ab, u16 len,
 			break;
 	}
 
-	stats = kzalloc(sizeof(*stats), GFP_ATOMIC);
+	stats = kzalloc(sizeof(*stats), GFP_ATOMIC, wmi_ctrl_path_stats_list);
 	if (!stats)
 		return -ENOMEM;
 
-	awgn_stats = kzalloc(sizeof(*awgn_stats), GFP_ATOMIC);
+	awgn_stats = kzalloc(sizeof(*awgn_stats), GFP_ATOMIC,
+			     wmi_ctrl_path_awgn_stats);
 
 	if (!awgn_stats) {
 		kfree(stats);
@@ -13295,11 +13307,13 @@ int wmi_print_ctrl_path_mem_stats_tlv(struct ath12k_base *ab, u16 len,
 		ar = ab->pdevs[i].ar;
 		if (!ar || !ar->ctrl_mem_stats)
 			continue;
-		stats = kzalloc(sizeof(*stats), GFP_ATOMIC);
+		stats = kzalloc(sizeof(*stats), GFP_ATOMIC,
+				wmi_ctrl_path_stats_list);
 		if (!stats)
 			return -ENOMEM;
 
-		mem_stats = kzalloc(sizeof(*mem_stats), GFP_ATOMIC);
+		mem_stats = kzalloc(sizeof(*mem_stats), GFP_ATOMIC,
+				    wmi_ctrl_path_mem_stats_params);
 		if (!mem_stats) {
 			kfree(stats);
 			return -ENOMEM;
@@ -13352,11 +13366,12 @@ int wmi_print_ctrl_path_afc_stats_tlv(struct ath12k_base *ab, u16 len,
 		return -EINVAL;
 	}
 
-	stats = kzalloc(sizeof(*stats), GFP_ATOMIC);
+	stats = kzalloc(sizeof(*stats), GFP_ATOMIC, wmi_ctrl_path_stats_list);
 	if (!stats)
 		return -ENOMEM;
 
-	afc_stats = kzalloc(sizeof(*afc_stats), GFP_ATOMIC);
+	afc_stats = kzalloc(sizeof(*afc_stats), GFP_ATOMIC,
+			    wmi_ctrl_path_afc_stats);
 
 	if (!afc_stats) {
 		kfree(stats);
@@ -13382,14 +13397,17 @@ int wmi_print_ctrl_path_pmlo_stats_tlv(struct ath12k_base *ab, u16 len, const vo
        struct wmi_ctrl_path_stats_ev_parse_param *stats_buff = (struct wmi_ctrl_path_stats_ev_parse_param *)data;
        struct wmi_ctrl_path_pmlo_telemetry_stats *pmlo_stats_skb = (struct wmi_ctrl_path_pmlo_telemetry_stats *)ptr;
        struct wmi_ctrl_path_pmlo_telemetry_stats *pmlo_stats = NULL;
-       struct wmi_ctrl_path_stats_list *stats = kzalloc(sizeof(struct wmi_ctrl_path_stats_list), GFP_ATOMIC);
+       struct wmi_ctrl_path_stats_list *stats = kzalloc(sizeof(struct wmi_ctrl_path_stats_list),
+                                                        GFP_ATOMIC,
+                                                        wmi_ctrl_path_stats_list);
        struct ath12k *ar = NULL;
        u32 value;
 
        if (!stats)
                return -ENOMEM;
 
-	pmlo_stats = kzalloc(sizeof(*pmlo_stats), GFP_ATOMIC);
+	pmlo_stats = kzalloc(sizeof(*pmlo_stats), GFP_ATOMIC,
+                             wmi_ctrl_path_pmlo_telemetry_stats);
        if (!pmlo_stats) {
                kfree(stats);
                return -ENOMEM;
diff --git a/drivers/net/wireless/ath/ath12k/wow.c b/drivers/net/wireless/ath/ath12k/wow.c
index 371c3e8d..d93fd6f6 100644
--- a/drivers/net/wireless/ath/ath12k/wow.c
+++ b/drivers/net/wireless/ath/ath12k/wow.c
@@ -392,7 +392,8 @@ static int ath12k_wow_vif_set_wakeups(struct ath12k_link_vif *arvif,
 			struct wmi_pno_scan_req_arg *pno;
 			int ret;
 
-			pno = kzalloc(sizeof(*pno), GFP_KERNEL);
+			pno = kzalloc(sizeof(*pno), GFP_KERNEL,
+				      wmi_pno_scan_req_arg);
 			if (!pno)
 				return -ENOMEM;
 
@@ -500,7 +501,7 @@ static int ath12k_wow_vdev_clean_nlo(struct ath12k *ar, u32 vdev_id)
 	if (!ar->nlo_enabled)
 		return 0;
 
-	pno = kzalloc(sizeof(*pno), GFP_KERNEL);
+	pno = kzalloc(sizeof(*pno), GFP_KERNEL, wmi_pno_scan_req_arg);
 	if (!pno)
 		return -ENOMEM;
 
@@ -742,7 +743,8 @@ static int ath12k_wow_arp_ns_offload(struct ath12k *ar, bool enable)
 
 	lockdep_assert_wiphy(ath12k_ar_to_hw(ar)->wiphy);
 
-	offload = kmalloc(sizeof(*offload), GFP_KERNEL);
+	offload = kmalloc(sizeof(*offload), GFP_KERNEL,
+			  wmi_arp_ns_offload_arg);
 	if (!offload)
 		return -ENOMEM;
 
