--- a/src/init/ssdk_hppe.c
+++ b/src/init/ssdk_hppe.c
@@ -38,7 +38,7 @@ static sw_error_t qca_hppe_fdb_hw_init(a
 	SW_RTN_ON_NULL(p_api->adpt_port_bridge_txmac_set);
 
 	for(port = SSDK_PHYSICAL_PORT0; port <= SSDK_PHYSICAL_PORT7; port++) {
-		fal_fdb_port_learning_ctrl_set(dev_id, port, A_TRUE, FAL_MAC_FRWRD);
+		fal_fdb_port_learning_ctrl_set(dev_id, port, A_FALSE, FAL_MAC_FRWRD);
 		fal_fdb_port_stamove_ctrl_set(dev_id, port, A_TRUE, FAL_MAC_FRWRD);
 		fal_portvlan_member_update(dev_id, port, 0x7f);
 		if (port == SSDK_PHYSICAL_PORT0 || port == SSDK_PHYSICAL_PORT7) {
@@ -50,7 +50,7 @@ static sw_error_t qca_hppe_fdb_hw_init(a
 	}
 
 	fal_fdb_aging_ctrl_set(dev_id, A_TRUE);
-	fal_fdb_learning_ctrl_set(dev_id, A_TRUE);
+	fal_fdb_learning_ctrl_set(dev_id, A_FALSE);
 
 	return SW_OK;
 }
