Index: qca-ssdk/include/init/ssdk_plat.h
===================================================================
--- qca-ssdk.orig/include/init/ssdk_plat.h
+++ qca-ssdk/include/init/ssdk_plat.h
@@ -160,7 +160,7 @@
 #define AR8327_NUM_PHYS                              5
 #define AR8327_PORT_CPU                              0
 #define AR8327_NUM_PORTS                             7
-#define AR8327_MAX_VLANS                             128
+#define AR8327_MAX_VLANS                             4095
 
 #define MII_PHYADDR_C45                              (1<<30)
 
