--- a/drivers/net/wireless/ath/ath11k/qmi.c
+++ b/drivers/net/wireless/ath/ath11k/qmi.c
@@ -2516,10 +2516,9 @@ static int ath11k_qmi_load_bdf_qmi(struc
 		fw_entry = ath11k_core_firmware_request(ab, ATH11K_DEFAULT_CAL_FILE);
 		if (IS_ERR(fw_entry)) {
 			ret = PTR_ERR(fw_entry);
-			ath11k_warn(ab,
-				    "qmi failed to load CAL data file:%s\n",
-				    filename);
-			goto out;
+			ath11k_warn(ab,
+				    "qmi failed to load CAL data file:%s booting with minimal performance\n",filename);
+			return 0;
 		}
 success:
 		fw_size = min_t(u32, ab->hw_params.fw.board_size, fw_entry->size);
