From: Karthikeyan Periyasamy <periyasa@codeaurora.org>
Subject: [PATCH] ath11k: Add debug logs for the CR2849877 and CR2854245

Signed-off-by: Karthikeyan Periyasamy <periyasa@codeaurora.org>
---
--- a/drivers/net/wireless/ath/ath11k/dp_rx.c
+++ b/drivers/net/wireless/ath/ath11k/dp_rx.c
@@ -2714,6 +2714,7 @@ static void ath11k_dp_rx_h_ppdu(struct a
 	u8 channel_num;
 	u32 center_freq, meta_data;
 	struct ieee80211_channel *channel;
+	int no_chan_info = -1;
 
 	rx_status->freq = 0;
 	rx_status->rate_idx = 0;
@@ -2731,20 +2732,40 @@ static void ath11k_dp_rx_h_ppdu(struct a
 	    center_freq <= ATH11K_MAX_6G_FREQ) {
 		rx_status->band = NL80211_BAND_6GHZ;
 		rx_status->freq = center_freq;
+		no_chan_info = 1;
 	} else if (channel_num >= 1 && channel_num <= 14) {
 		rx_status->band = NL80211_BAND_2GHZ;
+		no_chan_info = 1;
 	} else if (channel_num >= 36 && channel_num <= 173) {
 		rx_status->band = NL80211_BAND_5GHZ;
 	} else {
+		no_chan_info = 1;
+	}
+
+	if (unlikely(no_chan_info == -1)) {
+		ath11k_warn(ar->ab, "[Test] not in range channel_num %d center_freq %d\n", channel_num, center_freq);
+		ath11k_warn(ar->ab, "[Test] wiphy-bandinfo %p\n", ar->hw->wiphy->bands[rx_status->band]);
+		ath11k_warn(ar->ab, "[Test] msdu is first msdu %d\n", ath11k_dp_rx_h_msdu_end_first_msdu(ar->ab, rx_desc));
+		ath11k_warn(ar->ab, "[Test] msdu is last msdu %d\n", ath11k_dp_rx_h_msdu_end_first_msdu(ar->ab, rx_desc));
+		ath11k_warn(ar->ab, "[Test] msdu len %d\n", ath11k_dp_rx_h_msdu_start_msdu_len(ar->ab, rx_desc));
+	}
+
+	if (!ar->hw->wiphy->bands[rx_status->band]) {
+		ath11k_warn(ar->ab, "[Test] band info NULL for band %d\n", rx_status->band);
 		spin_lock_bh(&ar->data_lock);
 		channel = ar->rx_channel;
 		if (channel) {
+			ath11k_warn(ar->ab, "[Test] band %d center_freq %d from rx_channel\n", channel->band, channel->center_freq);
 			rx_status->band = channel->band;
 			channel_num =
 				ieee80211_frequency_to_channel(channel->center_freq);
+			if (rx_status->band == NL80211_BAND_6GHZ) {
+				ath11k_warn(ar->ab, "[Test] filling centre_freq=%d for 6Ghz\n", channel->center_freq);
+				rx_status->freq = channel->center_freq;
+			}
 		}
 		spin_unlock_bh(&ar->data_lock);
-		ath11k_dbg_dump(ar->ab, ATH11K_DBG_DATA, NULL, "rx_desc: ",
+		ath11k_dbg_dump(ar->ab, ATH11K_DBG_TEST, NULL, "[Test] rx_desc: ",
 				rx_desc, sizeof(struct hal_rx_desc));
 	}
 
@@ -3585,6 +3606,12 @@ static int ath11k_dp_rx_reap_mon_status_
 						&cookie, &rbm);
 		if (paddr) {
 			buf_id = FIELD_GET(DP_RXDMA_BUF_COOKIE_BUF_ID, cookie);
+			ar->cookie_pdev_id = FIELD_GET(DP_RXDMA_BUF_COOKIE_PDEV_ID, cookie);
+
+			if (mac_id != ar->cookie_pdev_id) {
+				ath11k_warn(ab, "invalid local mac_id %d pdev_id %d\n",
+					    mac_id, ar->cookie_pdev_id);
+			}
 
 			spin_lock_bh(&rx_ring->idr_lock);
 			skb = idr_find(&rx_ring->bufs_idr, buf_id);
@@ -5743,6 +5770,11 @@ void ath11k_dp_rx_mon_dest_process(struc
 	spin_unlock_bh(&pmon->mon_lock);
 
 	if (rx_bufs_used) {
+		if (ar->cookie_pdev_id != dp->mac_id || mac_id != dp->mac_id) {
+			ath11k_warn(ar->ab, "invalid mac_id for rxbuf replenish pdev id %d mac id %d dp mac id %d\n",
+				    ar->cookie_pdev_id, mac_id, dp->mac_id);
+		}
+
 		rx_mon_stats->dest_ppdu_done++;
 		hal_params = ar->ab->hw_params.hal_params;
 
--- a/drivers/net/wireless/ath/ath11k/mac.c
+++ b/drivers/net/wireless/ath/ath11k/mac.c
@@ -3911,7 +3911,8 @@ static int ath11k_mac_op_hw_scan(struct
 
 	ret = ath11k_start_scan(ar, arg);
 	if (ret) {
-		ath11k_warn(ar->ab, "failed to start hw scan: %d\n", ret);
+		ath11k_warn(ar->ab, "failed to start hw scan: %d scan_state %d pdev_idx %d\n",
+			    ret, ar->scan.state, ar->pdev_idx);
 		spin_lock_bh(&ar->data_lock);
 		ar->scan.state = ATH11K_SCAN_IDLE;
 		spin_unlock_bh(&ar->data_lock);
@@ -7295,7 +7296,7 @@ static void ath11k_mac_op_remove_chanctx
 	struct ath11k *ar = hw->priv;
 	struct ath11k_base *ab = ar->ab;
 
-	ath11k_dbg(ab, ATH11K_DBG_MAC,
+	ath11k_dbg(ab, ATH11K_DBG_TEST,
 		   "mac chanctx remove freq %u width %d ptr %pK\n",
 		   ctx->def.chan->center_freq, ctx->def.width, ctx);
 
--- a/net/mac80211/rx.c
+++ b/net/mac80211/rx.c
@@ -4994,12 +4994,16 @@ void ieee80211_rx_list(struct ieee80211_
 		}
 	}
 
-	if (WARN_ON(status->band >= NUM_NL80211_BANDS))
+	if (WARN_ON(status->band >= NUM_NL80211_BANDS)) {
+		printk("invalid status band %d freq %d\n", status->band, status->freq);
 		goto drop;
+	}
 
 	sband = local->hw.wiphy->bands[status->band];
-	if (WARN_ON(!sband))
+	if (WARN_ON(!sband)) {
+		printk("NULL sband status band %d freq %d\n", status->band, status->freq);
 		goto drop;
+	}
 
 	/*
 	 * If we're suspending, it is possible although not too likely
--- a/net/mac80211/scan.c
+++ b/net/mac80211/scan.c
@@ -647,9 +647,15 @@ static void ieee80211_scan_state_send_pr
 	int i;
 	struct ieee80211_sub_if_data *sdata;
 	struct cfg80211_scan_request *scan_req;
-	enum nl80211_band band = local->hw.conf.chandef.chan->band;
+	enum nl80211_band band;
 	u32 flags = 0, tx_flags;
 
+	if (!local->hw.conf.chandef.chan) {
+		printk("channel NULL\n");
+		WARN_ON(1);
+	}
+
+	band = local->hw.conf.chandef.chan->band;
 	scan_req = rcu_dereference_protected(local->scan_req,
 					     lockdep_is_held(&local->mtx));
 
--- a/drivers/net/wireless/ath/ath11k/core.h
+++ b/drivers/net/wireless/ath/ath11k/core.h
@@ -743,6 +743,7 @@ struct ath11k {
 	u8 ftm_msgref;
 	int ap_ps_enabled;
 	enum ath11k_ap_ps_state ap_ps_state;
+	int cookie_pdev_id;
 	/* protected by conf_mutex */
 	u8 ps_state_enable;
 	u8 ps_timekeeper_enable;
--- a/net/mac80211/cfg.c
+++ b/net/mac80211/cfg.c
@@ -3671,6 +3671,7 @@ int ieee80211_attach_ack_skb(struct ieee
 
 	if (id < 0) {
 		kfree_skb(ack_skb);
+		printk(KERN_ERR "attach ack fail %d\n", id);
 		return -ENOMEM;
 	}
 
--- a/net/mac80211/tx.c
+++ b/net/mac80211/tx.c
@@ -2537,6 +2537,7 @@ static u16 ieee80211_store_ack_skb(struc
 			}
 		} else {
 			kfree_skb(ack_skb);
+			printk(KERN_ERR "store ack fail %d\n", id);
 		}
 	}
 
--- a/drivers/net/wireless/ath/ath11k/core.c
+++ b/drivers/net/wireless/ath/ath11k/core.c
@@ -24,7 +24,7 @@ module_param_named(nss_offload, nss_offl
 MODULE_PARM_DESC(nss_offload, "Enable NSS Offload support");
 #endif
 
-unsigned int ath11k_debug_mask;
+unsigned int ath11k_debug_mask = ATH11K_DBG_TEST;
 EXPORT_SYMBOL(ath11k_debug_mask);
 module_param_named(debug_mask, ath11k_debug_mask, uint, 0644);
 MODULE_PARM_DESC(debug_mask, "Debugging mask");
--- a/drivers/net/wireless/ath/ath11k/debug.h
+++ b/drivers/net/wireless/ath/ath11k/debug.h
@@ -25,6 +25,7 @@ enum ath11k_debug_mask {
 	ATH11K_DBG_PCI		= 0x00001000,
 	ATH11K_DBG_DP_TX	= 0x00001000,
 	ATH11K_DBG_DP_RX	= 0x00002000,
+	ATH11K_DBG_TEST		= 0x00080000,
 	ATH11K_DBG_ANY		= 0xffffffff,
 };
 
