From ae444bf877fd94256a110d03582ddb045c541525 Mon Sep 17 00:00:00 2001
From: Aditya Kumar Singh <quic_adisi@quicinc.com>
Date: Wed, 24 Aug 2022 09:25:31 -0800
Subject: [PATCH] mac80211: fix compilation issues with Pkg Upgrd

Signed-off-by: Aditya Kumar Singh <quic_adisi@quicinc.com>
---
 net/wireless/nl80211.c | 6 ++++++
 1 file changed, 6 insertions(+)

--- a/net/mac80211/cfg.c
+++ b/net/mac80211/cfg.c
@@ -540,10 +540,12 @@ static struct ieee80211_key *
 ieee80211_lookup_key(struct ieee80211_sub_if_data *sdata,
 		     u8 key_idx, bool pairwise, const u8 *mac_addr)
 {
-	struct ieee80211_local *local = sdata->local;
+	struct ieee80211_local *local;
 	struct ieee80211_key *key;
 	struct sta_info *sta;
 
+	local = sdata->local;
+
 	if (mac_addr) {
 		sta = sta_info_get_bss(sdata, mac_addr);
 		if (!sta)
--- a/drivers/net/wireless/ath/ath12k/mhi.c
+++ b/drivers/net/wireless/ath/ath12k/mhi.c
@@ -365,7 +365,9 @@ int ath12k_mhi_register(struct ath12k_pc
 	mhi_ctrl->cntrl_dev = ab->dev;
 	mhi_ctrl->fw_image = ab_pci->amss_path;
 	mhi_ctrl->regs = ab->mem;
+#if LINUX_VERSION_IS_GEQ(5,10,0)
 	mhi_ctrl->reg_len = ab->mem_len;
+#endif
 
 	ret = ath12k_mhi_get_msi(ab_pci);
 	if (ret) {
--- a/drivers/net/wireless/ath/ath12k/pci.c
+++ b/drivers/net/wireless/ath/ath12k/pci.c
@@ -666,7 +666,11 @@ static int ath12k_pci_msi_alloc(struct a
 	}
 
 	ab_pci->msi_ep_base_data = msi_desc->msg.data;
+#if LINUX_VERSION_IS_GEQ(5,10,0)
 	if (msi_desc->pci.msi_attrib.is_64)
+#else
+	if (msi_desc->msi_attrib.is_64)
+#endif
 		set_bit(ATH12K_PCI_FLAG_IS_MSI_64, &ab_pci->flags);
 
 	ath12k_dbg(ab, ATH12K_DBG_PCI, "msi base data is %d\n", ab_pci->msi_ep_base_data);
--- a/drivers/net/wireless/ath/ath11k/pci.c
+++ b/drivers/net/wireless/ath/ath11k/pci.c
@@ -441,7 +441,11 @@ static int ath11k_pci_alloc_msi(struct a
 	pci_read_config_dword(pci_dev, pci_dev->msi_cap + PCI_MSI_ADDRESS_LO,
 			      &ab->pci.msi.addr_lo);
 
+#if LINUX_VERSION_IS_GEQ(5,10,0)
 	if (msi_desc->pci.msi_attrib.is_64) {
+#else
+	if (msi_desc->msi_attrib.is_64) {
+#endif
 		pci_read_config_dword(pci_dev, pci_dev->msi_cap + PCI_MSI_ADDRESS_HI,
 				      &ab->pci.msi.addr_hi);
 	} else {
