From d89ae85deceb51eede2d3aa9a914b4f6ceee1c13 Mon Sep 17 00:00:00 2001
From: Ramya Gnanasekar <quic_rgnanase@quicinc.com>
Date: Sat, 26 Feb 2022 00:55:49 +0530
Subject: [PATCH] ath12k: WMI fixes

Signed-off-by: Ramya Gnanasekar <quic_rgnanase@quicinc.com>

--- a/drivers/net/wireless/ath/ath12k/dp_rx.c
+++ b/drivers/net/wireless/ath/ath12k/dp_rx.c
@@ -4101,6 +4101,7 @@ int ath12k_dp_rxdma_ring_sel_config_wcn7
 						       DP_RXDMA_REFILL_RING_SIZE,
 						       &tlv_filter);
 	}
+#if 0
 
 	return ret;
 }
@@ -4119,6 +4120,7 @@ int ath12k_dp_rx_htt_setup(struct ath12k
 			    ret);
 		return ret;
 	}
+#endif
 
 	if (ab->hw_params->rx_mac_buf_ring) {
 		for (i = 0; i < ab->hw_params->num_rxmda_per_pdev; i++) {
