From d89ae85deceb51eede2d3aa9a914b4f6ceee1c13 Mon Sep 17 00:00:00 2001
From: Ramya Gnanasekar <quic_rgnanase@quicinc.com>
Date: Sat, 26 Feb 2022 00:55:49 +0530
Subject: [PATCH] ath12k: WMI fixes

Signed-off-by: Ramya Gnanasekar <quic_rgnanase@quicinc.com>

diff --git a/drivers/net/wireless/ath/ath12k/dp.c b/drivers/net/wireless/ath/ath12k/dp.c
index 1e7e73f..c6bf16a 100644
--- a/drivers/net/wireless/ath/ath12k/dp.c
+++ b/drivers/net/wireless/ath/ath12k/dp.c
@@ -1378,8 +1378,6 @@ static int ath12k_dp_cc_init(struct ath12k_base *ab)
 	int i, ret = 0;
 	u32 cmem_base;
 
-	return 0;
-
 	INIT_LIST_HEAD(&dp->rx_desc_free_list);
 	INIT_LIST_HEAD(&dp->rx_desc_used_list);
 	spin_lock_init(&dp->rx_desc_lock);
diff --git a/drivers/net/wireless/ath/ath12k/dp_rx.c b/drivers/net/wireless/ath/ath12k/dp_rx.c
index 01fc36a..6165e6f 100644
--- a/drivers/net/wireless/ath/ath12k/dp_rx.c
+++ b/drivers/net/wireless/ath/ath12k/dp_rx.c
@@ -4065,6 +4065,7 @@ int ath12k_dp_rx_htt_setup(struct ath12k_base *ab)
 			    ret);
 		return ret;
 	}
+#if 0
 
 	ring_id = dp->rxdma_mon_buf_ring.refill_buf_ring.ring_id;
 	ret = ath12k_dp_tx_htt_srng_setup(ab, ring_id,
@@ -4083,6 +4084,7 @@ int ath12k_dp_rx_htt_setup(struct ath12k_base *ab)
 			    ret);
 		return ret;
 	}
+#endif
 
 	ret = ath12k_dp_rxdma_ring_sel_config(ab);
 	if (ret) {
-- 
2.17.1

