From 232788c73cafd57fe58d62070a2b59327833c0cd Mon Sep 17 00:00:00 2001
From: Sharadanand Karanjkar <sharadanandk@gmail.com>
Date: Thu, 14 Jul 2022 15:55:26 +0200
Subject: [PATCH] Adding support for bootconfig module.

Signed-off-by: Sharadanand Karanjkar <sharadanandk@gmail.com>
---
 package/kernel/linux/modules/other.mk | 16 ++++++++++++++++
 1 file changed, 16 insertions(+)

diff --git a/package/kernel/linux/modules/other.mk b/package/kernel/linux/modules/other.mk
index 8e3642ab9a..b302a296c5 100644
--- a/package/kernel/linux/modules/other.mk
+++ b/package/kernel/linux/modules/other.mk
@@ -1303,3 +1303,19 @@ define KernelPackage/f71808e-wdt/description
 endef
 
 $(eval $(call KernelPackage,f71808e-wdt))
+
+define KernelPackage/bootconfig
+  SUBMENU:=$(OTHER_MENU)
+  TITLE:=Bootconfig partition for failsafe
+  KCONFIG:=CONFIG_BOOTCONFIG_PARTITION
+  FILES:=$(LINUX_DIR)/drivers/platform/msm/bootconfig.ko@lt4.4 \
+        $(LINUX_DIR)/drivers/platform/ipq/bootconfig.ko@ge4.4
+  AUTOLOAD:=$(call AutoLoad,56,bootconfig,1)
+endef
+
+define KernelPackage/bootconfig/description
+  Bootconfig partition for failsafe
+endef
+
+$(eval $(call KernelPackage,bootconfig))
+
-- 
2.30.2

