From b2cea8e521532dfa4f4156933bbd0f2a8c560644 Mon Sep 17 00:00:00 2001
From: John Crispin <john@phrozen.org>
Date: Mon, 23 Jan 2023 11:58:34 +0100
Subject: [PATCH] ubox: update to latest HEAD

Signed-off-by: John Crispin <john@phrozen.org>
---
 package/system/ubox/Makefile | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/package/system/ubox/Makefile b/package/system/ubox/Makefile
index 9ba4abdd58..766ca1c2e7 100644
--- a/package/system/ubox/Makefile
+++ b/package/system/ubox/Makefile
@@ -5,9 +5,9 @@ PKG_RELEASE:=2
 
 PKG_SOURCE_PROTO:=git
 PKG_SOURCE_URL=$(PROJECT_GIT)/project/ubox.git
+PKG_MIRROR_HASH:=aeabb4f0b0bb163fa62e2672cb3f0fd0bed3622bd7ad95d13a36e00978b4f864
 PKG_SOURCE_DATE:=2020-10-25
-PKG_SOURCE_VERSION:=9ef886819dd48303d8ced4cdbc9afbf32682535c
-PKG_MIRROR_HASH:=e2d93b798b91de98cb003f7b3be97d3d8a2413c8612b527d096c86ac5365cbdd
+PKG_SOURCE_VERSION:=cc34fb7b922f183e6ece5fa0fec31d4eb95c5823
 CMAKE_INSTALL:=1
 
 PKG_LICENSE:=GPL-2.0
-- 
2.34.1

