From 7b703f7557779fa47bb17dc692bd6140d89a3cdc Mon Dec 5 02:58:36 2022 +0000
From: Sohail Ahmad <sohail@indionetworks.com>
Date: Mon Dec 5 02:58:36 2022 +0000
Subject: [PATCH] dnsmasq: enabled EDNS configuration options- CPE ID and subnet

Signed-off-by: Sohail Ahmad <sohail@indionetworks.com>

diff --git a/package/network/services/dnsmasq/files/dnsmasq.init b/package/network/services/dnsmasq/files/dnsmasq.init
index d154d2566d..bacb9cc669 100644
--- a/package/network/services/dnsmasq/files/dnsmasq.init
+++ b/package/network/services/dnsmasq/files/dnsmasq.init
@@ -964,6 +964,10 @@ dnsmasq_start()
 	append_parm "$cfg" "max_cache_ttl" "--max-cache-ttl"
 	append_parm "$cfg" "pxe_prompt" "--pxe-prompt"
 	append_parm "$cfg" "tftp_unique_root" "--tftp-unique-root"
+
+	append_parm "$cfg" "addcpe" "--add-cpe-id"
+	append_parm "$cfg" "addsubnet" "--add-subnet"
+
 	config_list_foreach "$cfg" "pxe_service" append_pxe_service
 	config_get DOMAIN "$cfg" domain
 
