From ab0d397712ed67af2b1d9fb875c9dec419f56daa Mon Sep 17 00:00:00 2001
From: P Praneesh <quic_ppranees@quicinc.com>
Date: Mon, 3 Jan 2022 12:59:10 +0530
Subject: [PATCH] ath12k: WAR AP bringup crashes

Signed-off-by: P Praneesh <quic_ppranees@quicinc.com>
---
 drivers/net/wireless/ath/ath12k/mac.c | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/drivers/net/wireless/ath/ath12k/mac.c b/drivers/net/wireless/ath/ath12k/mac.c
index d1d2ac9..02c7c55 100644
--- a/drivers/net/wireless/ath/ath12k/mac.c
+++ b/drivers/net/wireless/ath/ath12k/mac.c
@@ -5228,6 +5228,8 @@ static void ath12k_mac_vif_unref(struct ath12k_dp *dp, struct ieee80211_vif *vif
 	struct sk_buff *skb;
 	int i;
 
+	return;
+
 	for (i = 0; i < ATH12K_HW_MAX_QUEUES; i++) {
 		spin_lock_bh(&dp->tx_desc_lock[i]);
 
@@ -5345,6 +5347,7 @@ static void ath12k_mac_op_configure_filter(struct ieee80211_hw *hw,
 	*total_flags &= SUPPORTED_FILTERS;
 	ar->filter_flags = *total_flags;
 
+#if 0
 	/* For monitor mode */
 	reset_flag = !(ar->filter_flags & FIF_BCN_PRBRESP_PROMISC);
 
@@ -5358,6 +5361,9 @@ static void ath12k_mac_op_configure_filter(struct ieee80211_hw *hw,
 		ath12k_warn(ar->ab,
 			    "fail to set monitor filter: %d\n", ret);
 	}
+#else
+	ath12k_warn(ar->ab, "fail to set monitor filter: %d\n", ret);
+#endif
 	ath12k_dbg(ar->ab, ATH12K_DBG_MAC,
 		   "changed_flags:0x%x, total_flags:0x%x, reset_flag:%d\n",
 		   changed_flags, *total_flags, reset_flag);
-- 
2.17.1

