From 3f674a6f2414a7c3e2e06ea561c04137355603a3 Mon Sep 17 00:00:00 2001
From: StanleyYP Wang <StanleyYP.Wang@mediatek.com>
Date: Wed, 5 Jul 2023 10:47:20 +0800
Subject: [PATCH 34/35] hostapd: mtk: Fix hostapd_dfs_start_cac log

Signed-off-by: StanleyYP Wang <StanleyYP.Wang@mediatek.com>
---
 src/ap/dfs.c | 6 ++++--
 1 file changed, 4 insertions(+), 2 deletions(-)

diff --git a/src/ap/dfs.c b/src/ap/dfs.c
index d688ecb..db8375e 100644
--- a/src/ap/dfs.c
+++ b/src/ap/dfs.c
@@ -1627,9 +1627,11 @@ int hostapd_dfs_start_cac(struct hostapd_iface *iface, int freq,
 	/* TODO: How to check CAC time for ETSI weather channels? */
 	iface->dfs_cac_ms = 60000;
 	wpa_msg(iface->bss[0]->msg_ctx, MSG_INFO, DFS_EVENT_CAC_START
-		"freq=%d chan=%d chan_offset=%d width=%d seg0=%d "
+		"freq=%d chan=%d chan_offset=%d width=%s seg0=%d "
 		"seg1=%d cac_time=%ds%s",
-		freq, (freq - 5000) / 5, chan_offset, chan_width, cf1, cf2,
+		freq, (freq - 5000) / 5, chan_offset,
+		channel_width_to_string(chan_width),
+		(cf1 - 5000) / 5, cf2 ? (cf2 - 5000) / 5 : 0,
 		iface->dfs_cac_ms / 1000,
 		hostapd_dfs_is_background_event(iface, freq) ?
 		" (background)" : "");
-- 
2.18.0

