From d927ebf4a9373eabd383514bad580dc9d7fba3a4 Mon Sep 17 00:00:00 2001
From: Shayne Chen <shayne.chen@mediatek.com>
Date: Tue, 20 Sep 2022 19:33:45 +0800
Subject: [PATCH 02/35] hostapd: mtk: print sae groups by hostapd ctrl

---
 hostapd/ctrl_iface.c | 13 +++++++++++++
 1 file changed, 13 insertions(+)

diff --git a/hostapd/ctrl_iface.c b/hostapd/ctrl_iface.c
index c2a2822..bc690c5 100644
--- a/hostapd/ctrl_iface.c
+++ b/hostapd/ctrl_iface.c
@@ -1412,6 +1412,19 @@ static int hostapd_ctrl_iface_get(struct hostapd_data *hapd, char *cmd,
 		if (os_snprintf_error(buflen, res))
 			return -1;
 		return res;
+	} else if (os_strcmp(cmd, "sae_group_capability") == 0) {
+#ifdef CONFIG_SAE
+		/* see sae_set_group() */
+		res = os_snprintf(buf, buflen, "%s%s%s%s19 20 21",
+				  dh_groups_get(15) ? "15 ": "",
+				  dh_groups_get(16) ? "16 ": "",
+				  dh_groups_get(17) ? "17 ": "",
+				  dh_groups_get(18) ? "18 ": "");
+
+		if (os_snprintf_error(buflen, res))
+			return -1;
+		return res;
+#endif /* CONFIG_SAE */
 	}
 
 	return -1;
-- 
2.18.0

