From d11dea1d3dee3577be404bfb6f7dc2460858242d Mon Sep 17 00:00:00 2001
From: mtk20656 <chank.chen@mediatek.com>
Date: Wed, 13 Sep 2023 19:29:51 +0800
Subject: [PATCH] [hostapd][mt76]6g bss connect do not consider ht operation

Signed-off-by: mtk20656 <chank.chen@mediatek.com>
---
 src/ap/ieee802_11.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/src/ap/ieee802_11.c b/src/ap/ieee802_11.c
old mode 100755
new mode 100644
index ef520c8..904b1b5
--- a/src/ap/ieee802_11.c
+++ b/src/ap/ieee802_11.c
@@ -5434,7 +5434,8 @@ static void handle_assoc(struct hostapd_data *hapd,
 			ieee802_11_set_beacons(hapd->iface);
 	}
 
-	update_ht_state(hapd, sta);
+	if (!is_6ghz_op_class(hapd->iconf->op_class))
+		update_ht_state(hapd, sta);
 
 	hostapd_logger(hapd, sta->addr, HOSTAPD_MODULE_IEEE80211,
 		       HOSTAPD_LEVEL_DEBUG,
-- 
2.18.0

