From c585424eaba09fcce7a17250fc7eef84c63f7a3a Mon Sep 17 00:00:00 2001
From: "himanshu.goyal" <himanshu.goyal@mediatek.com>
Date: Tue, 24 Jan 2023 19:06:44 +0800
Subject: [PATCH 15/35] hostapd: mtk: Add vendor command attribute for RTS BW
 signaling.

Signed-off-by: himanshu.goyal <himanshu.goyal@mediatek.com>
---
 src/common/mtk_vendor.h | 1 +
 1 file changed, 1 insertion(+)

diff --git a/src/common/mtk_vendor.h b/src/common/mtk_vendor.h
index 97c5a47..62d79cb 100644
--- a/src/common/mtk_vendor.h
+++ b/src/common/mtk_vendor.h
@@ -169,6 +169,7 @@ enum mtk_vendor_attr_wireless_ctrl {
 	MTK_VENDOR_ATTR_WIRELESS_CTRL_MIMO,
 	MTK_VENDOR_ATTR_WIRELESS_CTRL_AMSDU,
 	MTK_VENDOR_ATTR_WIRELESS_CTRL_CERT,
+	MTK_VENDOR_ATTR_WIRELESS_CTRL_RTS_SIGTA,
 
 	/* keep last */
 	NUM_MTK_VENDOR_ATTRS_WIRELESS_CTRL,
-- 
2.18.0

