Index: qca-ssdk-9e687f44ee13ce0904f291c9771172b488b7f9ed/src/adpt/appe/adpt_appe_vxlan.c
===================================================================
--- qca-ssdk-9e687f44ee13ce0904f291c9771172b488b7f9ed.orig/src/adpt/appe/adpt_appe_vxlan.c
+++ qca-ssdk-9e687f44ee13ce0904f291c9771172b488b7f9ed/src/adpt/appe/adpt_appe_vxlan.c
@@ -150,7 +150,7 @@ sw_error_t
 adpt_appe_vxlan_entry_add(a_uint32_t dev_id, fal_vxlan_type_t type,
 						fal_tunnel_udp_entry_t * entry)
 {
-	a_uint32_t idx, entry_idx, entry_sign;
+	a_uint32_t idx, entry_idx = 0, entry_sign;
 	a_bool_t entry_inuse;
 	a_int32_t vxlan_port_bitmap;
 	fal_tunnel_udp_entry_t temp_entry;
Index: qca-ssdk-9e687f44ee13ce0904f291c9771172b488b7f9ed/src/adpt/appe/adpt_appe_geneve.c
===================================================================
--- qca-ssdk-9e687f44ee13ce0904f291c9771172b488b7f9ed.orig/src/adpt/appe/adpt_appe_geneve.c
+++ qca-ssdk-9e687f44ee13ce0904f291c9771172b488b7f9ed/src/adpt/appe/adpt_appe_geneve.c
@@ -30,7 +30,7 @@
 sw_error_t
 adpt_appe_geneve_entry_add(a_uint32_t dev_id, fal_tunnel_udp_entry_t * entry)
 {
-	a_uint32_t idx, entry_idx, entry_sign;
+	a_uint32_t idx, entry_idx = 0, entry_sign;
 	a_bool_t entry_inuse;
 	a_int32_t geneve_port_bitmap;
 	fal_tunnel_udp_entry_t temp_entry;
Index: qca-ssdk-9e687f44ee13ce0904f291c9771172b488b7f9ed/src/adpt/appe/adpt_appe_tunnel_program.c
===================================================================
--- qca-ssdk-9e687f44ee13ce0904f291c9771172b488b7f9ed.orig/src/adpt/appe/adpt_appe_tunnel_program.c
+++ qca-ssdk-9e687f44ee13ce0904f291c9771172b488b7f9ed/src/adpt/appe/adpt_appe_tunnel_program.c
@@ -502,7 +502,7 @@ sw_error_t
 adpt_appe_tunnel_program_entry_add(a_uint32_t dev_id,
 		fal_tunnel_program_type_t type, fal_tunnel_program_entry_t * entry)
 {
-	a_uint32_t idx, entry_idx, entry_sign;
+	a_uint32_t idx, entry_idx = 0, entry_sign;
 	a_int32_t program_port_bitmap, program_id;
 	fal_tunnel_program_entry_t temp_entry = {0};
 
@@ -746,7 +746,7 @@ sw_error_t
 adpt_appe_tunnel_program_udf_add(a_uint32_t dev_id,
 		fal_tunnel_program_type_t type, fal_tunnel_program_udf_t * udf)
 {
-	a_uint32_t idx, entry_idx;
+	a_uint32_t idx, entry_idx = 0;
 	a_int32_t program_id, temp_program_id;
 	fal_tunnel_program_udf_t temp_udf = {0};
 	a_bool_t entry_sign = A_FALSE;
