--- a/drivers/net/wireless/ath/ath12k/core.c
+++ b/drivers/net/wireless/ath/ath12k/core.c
@@ -48,7 +48,7 @@ unsigned int ath12k_ftm_mode;
 module_param_named(ftm_mode, ath12k_ftm_mode, uint, 0444);
 MODULE_PARM_DESC(ftm_mode, "Boots up in factory test mode");
 
-unsigned int ath12k_mlo_capable = true;
+unsigned int ath12k_mlo_capable = true;
 module_param_named(mlo_capable, ath12k_mlo_capable, uint, 0644);
 MODULE_PARM_DESC(mlo_capable, "mlo capable: 0-disable, 1-enable");
 
