--- a/src/ap/wps_hostapd.c
+++ b/src/ap/wps_hostapd.c
@@ -1178,8 +1178,7 @@ int hostapd_init_wps(struct hostapd_data
 					  WPA_CIPHER_GCMP_256)) {
 			wps->encr_types |= WPS_ENCR_AES;
 			wps->encr_types_rsn |= WPS_ENCR_AES;
-		}
-		if (conf->rsn_pairwise & WPA_CIPHER_TKIP) {
+		} else if (conf->rsn_pairwise & WPA_CIPHER_TKIP) {
 #ifdef CONFIG_NO_TKIP
 			wpa_printf(MSG_INFO, "WPS: TKIP not supported");
 			goto fail;
@@ -1188,9 +1187,7 @@ int hostapd_init_wps(struct hostapd_data
 			wps->encr_types_rsn |= WPS_ENCR_TKIP;
 #endif /* CONFIG_NO_TKIP */
 		}
-	}
-
-	if (conf->wpa & WPA_PROTO_WPA) {
+	} else if (conf->wpa & WPA_PROTO_WPA) {
 #ifdef CONFIG_NO_TKIP
 		if (!(conf->wpa & WPA_PROTO_RSN)) {
 			wpa_printf(MSG_INFO, "WPS: WPA(v1) not supported");
@@ -1206,8 +1203,7 @@ int hostapd_init_wps(struct hostapd_data
 		if (conf->wpa_pairwise & WPA_CIPHER_CCMP) {
 			wps->encr_types |= WPS_ENCR_AES;
 			wps->encr_types_wpa |= WPS_ENCR_AES;
-		}
-		if (conf->wpa_pairwise & WPA_CIPHER_TKIP) {
+		} else if (conf->wpa_pairwise & WPA_CIPHER_TKIP) {
 			wps->encr_types |= WPS_ENCR_TKIP;
 			wps->encr_types_wpa |= WPS_ENCR_TKIP;
 		}
