From cd7a788163c723228248ed211a9a0977fe7b11e4 Mon Sep 17 00:00:00 2001
From: leiwei <leiwei@codeaurora.org>
Date: Tue, 5 Jan 2021 14:15:07 +0800
Subject: [PATCH] hostap: Support 802.1x plug re-establish macsec link

Signed-off-by: leiwei <leiwei@codeaurora.org>
---
 src/pae/ieee802_1x_kay.c | 8 +++++---
 1 file changed, 5 insertions(+), 3 deletions(-)

diff --git a/src/pae/ieee802_1x_kay.c b/src/pae/ieee802_1x_kay.c
index 3184fae..ad5a34b 100644
--- a/src/pae/ieee802_1x_kay.c
+++ b/src/pae/ieee802_1x_kay.c
@@ -2597,8 +2597,10 @@ static void ieee802_1x_participant_timer(void *eloop_ctx, void *timeout_ctx)
 			participant->lrx = false;
 			participant->otx = false;
 			participant->orx = false;
-			participant->is_key_server = false;
-			participant->is_elected = false;
+			if(participant->mode == PSK) {
+				participant->is_key_server = false;
+				participant->is_elected = false;
+			}
 			kay->authenticated = false;
 			kay->secured = false;
 			kay->failed = false;
@@ -2644,7 +2646,7 @@ static void ieee802_1x_participant_timer(void *eloop_ctx, void *timeout_ctx)
 	}
 
 	if (participant->retry_count < MAX_RETRY_CNT ||
-	    participant->mode == PSK) {
+	    participant->mode == PSK || participant->mode == EAP_EXCHANGE) {
 		ieee802_1x_participant_send_mkpdu(participant);
 		participant->retry_count++;
 	}
-- 
2.7.4

