From 67a8f56b6182bd6fe6fd47d296447f6322152290 Mon Sep 17 00:00:00 2001
From: "himanshu.goyal" <himanshu.goyal@mediatek.com>
Date: Fri, 17 Mar 2023 17:36:01 +0800
Subject: [PATCH 14/19] nl80211: mtk: Mark DFS channel as available for CSA.

---
 net/wireless/nl80211.c | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/net/wireless/nl80211.c b/net/wireless/nl80211.c
index c41675f..0917713 100644
--- a/net/wireless/nl80211.c
+++ b/net/wireless/nl80211.c
@@ -9616,6 +9616,11 @@ skip_beacons:
 	if (err)
 		goto free;
 
+	/* Use RADAR_BACKGROUND attribute here for skipping CAC */
+	if (info->attrs[NL80211_ATTR_RADAR_BACKGROUND]) {
+		cfg80211_set_dfs_state(&rdev->wiphy, &params.chandef, NL80211_DFS_AVAILABLE);
+	}
+
 	if (!cfg80211_reg_can_beacon_relax(&rdev->wiphy, &params.chandef,
 					   wdev->iftype)) {
 		err = -EINVAL;
-- 
2.18.0

