Index: hostapd-2021-02-20-59e9794c/src/ap/ubus.c
===================================================================
--- hostapd-2021-02-20-59e9794c.orig/src/ap/ubus.c
+++ hostapd-2021-02-20-59e9794c/src/ap/ubus.c
@@ -1987,7 +1987,8 @@ int hostapd_ubus_handle_rt_event(struct
 	session_id = sta->cl_session_id;
 
 	/* find by session id */
-	rec = avl_find_element(&hapd->ubus.rt_events, &session_id, rec, avl);
+	if (session_id)
+		rec = avl_find_element(&hapd->ubus.rt_events, &session_id, rec, avl);
 
 	/* prepare rec if not found */
 	if (!rec) {
