Index: qca-nss-dp-cs1/edma/edma_data_plane.h
===================================================================
--- qca-nss-dp-cs1.orig/edma/edma_data_plane.h
+++ qca-nss-dp-cs1/edma/edma_data_plane.h
@@ -27,7 +27,7 @@
 #define EDMA_RX_PREHDR_SIZE		(sizeof(struct edma_rx_preheader))
 #define EDMA_TX_PREHDR_SIZE		(sizeof(struct edma_tx_preheader))
 #define EDMA_RING_SIZE			128
-#define EDMA_NAPI_WORK			100
+#define EDMA_NAPI_WORK			NAPI_POLL_WEIGHT
 #define EDMA_START_GMACS		NSS_DP_START_PHY_PORT
 #define EDMA_MAX_GMACS			NSS_DP_MAX_PHY_PORTS
 #define EDMA_TX_PKT_MIN_SIZE		33
Index: qca-nss-dp-cs1/edma/edma_tx_rx.c
===================================================================
--- qca-nss-dp-cs1.orig/edma/edma_tx_rx.c
+++ qca-nss-dp-cs1/edma/edma_tx_rx.c
@@ -433,7 +433,7 @@ int edma_napi(struct napi_struct *napi,
 
 	for (i = 0; i < ehw->rxdesc_rings; i++) {
 		rxdesc_ring = &ehw->rxdesc_ring[i];
-		work_done += edma_clean_rx(ehw, budget, rxdesc_ring);
+		edma_clean_rx(ehw, budget, rxdesc_ring);
 	}
 
 	for (i = 0; i < ehw->txcmpl_rings; i++) {
@@ -501,7 +501,7 @@ int edma_napi(struct napi_struct *napi,
 						edma_hw.rxfill_intr_mask);
 		}
 	}
-	return work_done;
+	return work_done > budget ? budget : work_done;
 }
 
 /*
