--- ucentral-schema-2022-05-29-e9529eab.orig/system/state.uc.orig	2026-03-30 10:49:01.424586758 +0800
+++ ucentral-schema-2022-05-29-e9529eab/system/state.uc	2026-03-30 10:49:25.832781649 +0800
@@ -374,7 +374,7 @@ cursor.foreach("network", "interface", f
 	if (!length(status))
 		return;
 
-	if (devices && length(devices[role]["bridge-members"]))
+	if (devices?.[role] && length(devices[role]["bridge-members"]))
 		iface_ports = devices[role]["bridge-members"];
 	iface.uptime = status.uptime || 0;
 
