--- a/schema/metrics.wifi-frames.yml
+++ b/schema/metrics.wifi-frames.yml
@@ -19,3 +19,5 @@ properties:
         - key-mismatch
         - beacon-report
         - radar-detected
+        - sta-authorized
+        - ft-finish
--- a/schemareader.uc
+++ b/schemareader.uc
@@ -9022,9 +9022,9 @@ properties:
 					if (type(value) != "string")
 						push(errors, [ location, "must be of type string" ]);

-					if (!(value in [ "probe", "auth", "assoc", "disassoc", "deauth", "local-deauth", "inactive-deauth", "key-mismatch", "beacon-report", "radar-detected" ]))
-						push(errors, [ location, "must be one of \"probe\", \"auth\", \"assoc\", \"disassoc\", \"deauth\", \"local-deauth\", \"inactive-deauth\", \"key-mismatch\", \"beacon-report\" or \"radar-detected\"" ]);
+					if (!(value in [ "probe", "auth", "assoc", "disassoc", "deauth", "local-deauth", "inactive-deauth", "key-mismatch", "beacon-report", "radar-detected", "sta-authorized", "ft-finish" ]))
+						push(errors, [ location, "must be one of \"probe\", \"auth\", \"assoc\", \"disassoc\", \"deauth\", \"local-deauth\", \"inactive-deauth\", \"key-mismatch\", \"beacon-report\", \"radar-detected\", \"sta-authorized\" or \"ft-finish\"" ]);

 					return value;
 				}

Signed-off-by: Sujeet Kumar <sujeet.kumar@inventum.net>
