--- a/system/capabilities.uc	2023-09-19 13:31:15.000000000 +0530
+++ b/system/capabilities.uc	2026-06-23 13:05:38.980755710 +0530
@@ -5,14 +5,17 @@

 let ubus = require("ubus");
 let fs = require("fs");
-
+let default_config = {
+	country: 'US'
+};
 let boardfile = fs.open("/etc/board.json", "r");
 let board = json(boardfile.read("all"));
 boardfile.close();
 let restrictfile = fs.open("/etc/ucentral/restrictions.json", "r");

 capa = {
-	'secure-rtty': true
+	'secure-rtty': true,
+	'default-config': true,
 };
 if (restrictfile) {
 	capa.restrictions = json(restrictfile.read("all")) || {};
@@ -24,11 +27,57 @@
 	else
 		capa.developer = false;
 }
+
+let version = json(fs.readfile('/etc/ucentral/version.json') || '{}');
+let schema = json(fs.readfile('/etc/ucentral/schema.json') || '{}');
+let version_vendor = json(fs.readfile('/etc/ucentral/version.vendor.json') || '{}');
+let schema_vendor = json(fs.readfile('/etc/ucentral/schema.vendor.json') || '{}');
+
+if (length(version_vendor))
+	version.vendor = version_vendor;
+
+if (length(schema_vendor))
+	schema.vendor = schema_vendor;
+
 ctx = ubus.connect();
+
+if (board.wifi?.country) {
+	let countries = split(board.wifi.country, ' ');
+	default_config.country = countries[0];
+	capa.country_codes = countries;
+}
+
+if (fs.stat('/tmp/squashfs')) {
+	let pipe = fs.popen('fw_printenv -n country');
+	let country = replace(pipe.read("all"), '\n', '');
+	pipe.close();
+	if (country)
+		default_config.country = country;
+	else
+		system('fw_setenv country ' + default_config.country);
+	fs.writefile('/etc/ucentral/ucentral.defaults', default_config);
+} else {
+	if (!fs.stat('/certificates/ucentral.defaults')) {
+		fs.writefile('/certificates/ucentral.defaults', default_config);
+	}
+	if (fs.stat('/certificates/ucentral.defaults')) {
+		let defaults = fs.readfile('/certificates/ucentral.defaults');
+		if (json(defaults))
+			default_config = json(defaults);
+	}
+	fs.writefile('/etc/ucentral/ucentral.defaults', default_config);
+}
+capa.country = default_config.country || 'US';
+
 let wifi = require("wifi.phy");
 capa.compatible = replace(board.model.id, ',', '_');
 capa.model = board.model.name;

+capa.version = {
+	'ap': version,
+	schema
+};
+
 if (board.bridge && board.bridge.name == "switch")
 	capa.platform = "switch";
 else if (length(wifi))
@@ -96,6 +145,10 @@
 if (length(wifi))
 	capa.wifi = wifi;

-capafile = fs.open("/etc/ucentral/capabilities.json", "w");
+capa.compress_cmd = true;
+
+/* capafile = fs.open("/etc/ucentral/capabilities.json", "w");
 capafile.write(capa);
-capafile.close();
+capafile.close(); */
+
+fs.writefile('/etc/ucentral/capabilities.json', capa);

Signed-off-by: Sujeet Kumar <sujeet.kumar@inventum.net>
