--- ucentral-schema-2026-01-24-8491119c.orig/renderer/templates/interface/ssid.uc	2026-04-01 15:37:35.732785746 +0800
+++ ucentral-schema-2026-01-24-8491119c/renderer/templates/interface/ssid.uc	2026-04-01 16:24:33.397694699 +0800
@@ -239,9 +239,14 @@
 		if (!ssid.encryption || ssid.encryption.proto in [ "none" ])
 			return 0;
 
-		if (ssid.encryption.proto in [ "sae-mixed", "wpa3-mixed", "sae", "wpa3", "wpa3-192" ])
+		if (ssid.encryption.proto in [ "sae", "wpa3", "wpa3-192" ])
 			return 2;
 
+		if (ssid.encryption.proto in [ "sae-mixed", "wpa3-mixed" ]) {
+			let val = index([ "disabled", "optional", "required" ], ssid.encryption.ieee80211w);
+			return (val <= 0) ? 1 : val;
+		}
+
 		return index([ "disabled", "optional", "required" ], ssid.encryption.ieee80211w);
 	}
 
