--- ucentral-schema-2026.01.24~8491119c.orig/renderer/templates/services/dhcp_inject.uc
+++ ucentral-schema-2026.01.24~8491119c/renderer/templates/services/dhcp_inject.uc
@@ -1,12 +1,12 @@
 {% let ifaces = services.lookup_interfaces_by_ssids("dhcp-inject") %}
-{% if (!length(ifaces)) return %}
 {% let upstreams = [] %}
 {% for (let iface in ifaces): %}
 {%    if (iface.role == "upstream") { push(upstreams, iface) } %}
 {% endfor %}
 
 {% let enabled = length(upstreams) %}
 {% services.set_enabled("dhcpinject", enabled) %}
+{% if (!enabled) return %}
 
 {% for (let upstream in upstreams): %}
 
