--- a/arch/arm64/boot/dts/qcom/ipq6018.dtsi
+++ b/arch/arm64/boot/dts/qcom/ipq6018.dtsi
@@ -1628,14 +1628,11 @@
 			qcom,tgt-mem-mode = <0>;
 #endif
 
-#if defined(__CNSS2__)
 			mem-region = <&q6_region>;
 			qcom,bdf-addr = <0x4ABC0000 0x4ABC0000 0x4ABC0000 0x0 0x0>;
 			qcom,caldb-addr = <0x4B500000 0x4B500000 0x4B500000 0x0 0x0>;
 			qcom,caldb-size = <0x480000>;
-#else
 			memory-region = <&q6_region>;
-#endif
 			qcom,rproc = <&q6v5_wcss>;
 			interrupts = <0 320 1>, /* o_wcss_apps_intr[0] = */
 						<0 319 1>,
