Index: linux-5.4.164-qsdk-26349818b464f8c7b52d59ce73579d9f3dd6bd5d/arch/arm64/boot/dts/qcom/ipq6018-memory.dtsi
===================================================================
--- linux-5.4.164-qsdk-26349818b464f8c7b52d59ce73579d9f3dd6bd5d.orig/arch/arm64/boot/dts/qcom/ipq6018-memory.dtsi
+++ linux-5.4.164-qsdk-26349818b464f8c7b52d59ce73579d9f3dd6bd5d/arch/arm64/boot/dts/qcom/ipq6018-memory.dtsi
@@ -132,6 +132,12 @@
 			reg = <0x0 0x4d400000 0x0 0x100000>;
 		};
 
+		ramoops@4d500000 {
+			compatible = "ramoops";
+			reg = <0x0 0x4d500000 0x0 0x100000>;
+			record-size = <0x1000>;
+		};
+
 		rpm_msg_ram: rpm_msg_ram@0x60000 {
 			no-map;
 			reg = <0x0 0x60000 0x0 0x6000>;
@@ -256,6 +262,12 @@
 			reg = <0x0 0x4e300000 0x0 0x100000>;
 		};
 
+		ramoops@4e400000 {
+			compatible = "ramoops";
+			reg = <0x0 0x4e400000 0x0 0x100000>;
+			record-size = <0x1000>;
+		};
+
 		rpm_msg_ram: rpm_msg_ram@0x60000 {
 			no-map;
 			reg = <0x0 0x60000 0x0 0x6000>;
@@ -383,6 +395,12 @@
 			reg = <0x0 0x50100000 0x0 0x100000>;
 		};
 
+		ramoops@50200000 {
+			compatible = "ramoops";
+			reg = <0x0 0x50200000 0x0 0x100000>;
+			record-size = <0x1000>;
+		};
+
 		rpm_msg_ram: rpm_msg_ram@0x60000 {
 			no-map;
 			reg = <0x0 0x60000 0x0 0x6000>;
