--- a/src/fal/fal_port_ctrl.c
+++ b/src/fal/fal_port_ctrl.c
@@ -430,7 +430,7 @@ _fal_port_hibernate_get (a_uint32_t dev_
 
 static sw_error_t
 _fal_port_cdt (a_uint32_t dev_id, fal_port_t port_id, a_uint32_t mdi_pair,
-	       a_uint32_t * cable_status, a_uint32_t * cable_len)
+	       fal_cable_status_t * cable_status, a_uint32_t * cable_len)
 {
   sw_error_t rv;
   hsl_api_t *p_api;
@@ -2495,7 +2495,7 @@ fal_port_hibernate_get (a_uint32_t dev_i
  */
 sw_error_t
 fal_port_cdt (a_uint32_t dev_id, fal_port_t port_id, a_uint32_t mdi_pair,
-	      a_uint32_t * cable_status, a_uint32_t * cable_len)
+	      fal_cable_status_t * cable_status, a_uint32_t * cable_len)
 {
   sw_error_t rv;
 
--- a/src/fal/fal_portvlan.c
+++ b/src/fal/fal_portvlan.c
@@ -1701,7 +1701,7 @@ fal_netisolate_get(a_uint32_t dev_id, a_
  * @return SW_OK or error code
  */
 sw_error_t
-fal_eg_trans_filter_bypass_en_set(a_uint32_t dev_id, a_bool_t enable)
+fal_eg_trans_filter_bypass_en_set(a_uint32_t dev_id, a_uint32_t enable)
 {
     sw_error_t rv;
 
@@ -1718,7 +1718,7 @@ fal_eg_trans_filter_bypass_en_set(a_uint
  * @return SW_OK or error code
  */
 sw_error_t
-fal_eg_trans_filter_bypass_en_get(a_uint32_t dev_id, a_bool_t* enable)
+fal_eg_trans_filter_bypass_en_get(a_uint32_t dev_id, a_uint32_t* enable)
 {
     sw_error_t rv;
 
