From 25df1c3e41f274f70e4fbf5fdc10e4290ba019f5 Mon Sep 17 00:00:00 2001
From: John Crispin <john@phrozen.org>
Date: Thu, 20 Jan 2022 10:48:35 +0100
Subject: [PATCH] fixes

---
 lib/nl80211.c | 92 +++++++++++++++++++++++++++++++++++++++++++++++----
 lib/rtnl.c    |  1 +
 2 files changed, 87 insertions(+), 6 deletions(-)

Index: ucode-2022-04-07-33f1e0b0/lib/nl80211.c
===================================================================
--- ucode-2022-04-07-33f1e0b0.orig/lib/nl80211.c
+++ ucode-2022-04-07-33f1e0b0/lib/nl80211.c
@@ -51,6 +51,8 @@ limitations under the License.
 
 #define NL80211_ATTR_NOT_IMPLEMENTED 0x10000
 
+#define NL80211_ATTR_NOT_IMPLEMENTED 0x10000
+
 static struct {
 	int code;
 	char *msg;
@@ -263,6 +265,14 @@ static const uc_nl_nested_spec_t nl80211
 	}
 };
 
+#ifndef NL80211_MESHCONF_NOLEARN
+#define NL80211_MESHCONF_NOLEARN NL80211_ATTR_NOT_IMPLEMENTED
+#endif
+
+#ifndef NL80211_MESHCONF_CONNECTED_TO_AS
+#define NL80211_MESHCONF_CONNECTED_TO_AS NL80211_ATTR_NOT_IMPLEMENTED
+#endif
+
 static const uc_nl_nested_spec_t nl80211_mesh_params_nla = {
 	.headsize = 0,
 	.nattrs = 29,
@@ -354,6 +364,14 @@ static const uc_nl_nested_spec_t nl80211
 	}
 };
 
+#ifndef NL80211_PMSR_FTM_REQ_ATTR_LMR_FEEDBACK
+#define NL80211_PMSR_FTM_REQ_ATTR_LMR_FEEDBACK NL80211_ATTR_NOT_IMPLEMENTED
+#endif
+
+#ifndef NL80211_PMSR_FTM_REQ_ATTR_BSS_COLOR
+#define NL80211_PMSR_FTM_REQ_ATTR_BSS_COLOR NL80211_ATTR_NOT_IMPLEMENTED
+#endif
+
 static const uc_nl_nested_spec_t nl80211_peer_measurements_peers_req_data_ftm_nla = {
 	.headsize = 0,
 	.nattrs = 13,
@@ -503,6 +521,26 @@ static const uc_nl_nested_spec_t nl80211
 	}
 };
 
+#ifndef NL80211_FREQUENCY_ATTR_1MHZ
+#define NL80211_FREQUENCY_ATTR_1MHZ NL80211_ATTR_NOT_IMPLEMENTED
+#endif
+
+#ifndef NL80211_FREQUENCY_ATTR_2MHZ
+#define NL80211_FREQUENCY_ATTR_2MHZ NL80211_ATTR_NOT_IMPLEMENTED
+#endif
+
+#ifndef NL80211_FREQUENCY_ATTR_4MHZ
+#define NL80211_FREQUENCY_ATTR_4MHZ NL80211_ATTR_NOT_IMPLEMENTED
+#endif
+
+#ifndef NL80211_FREQUENCY_ATTR_8MHZ
+#define NL80211_FREQUENCY_ATTR_8MHZ NL80211_ATTR_NOT_IMPLEMENTED
+#endif
+
+#ifndef NL80211_FREQUENCY_ATTR_16MHZ
+#define NL80211_FREQUENCY_ATTR_16MHZ NL80211_ATTR_NOT_IMPLEMENTED
+#endif
+
 static const uc_nl_nested_spec_t nl80211_wiphy_bands_freqs_nla = {
 	.headsize = 0,
 	.nattrs = 25,
@@ -544,6 +582,10 @@ static const uc_nl_nested_spec_t nl80211
 	}
 };
 
+#ifndef NL80211_BAND_IFTYPE_ATTR_VENDOR_ELEMS
+#define NL80211_BAND_IFTYPE_ATTR_VENDOR_ELEMS NL80211_ATTR_NOT_IMPLEMENTED
+#endif
+
 static const uc_nl_nested_spec_t nl80211_wiphy_bands_iftype_data_nla = {
 	.headsize = 0,
 	.nattrs = 7,
@@ -714,6 +756,10 @@ static const uc_nl_nested_spec_t nl80211
 	}
 };
 
+#ifndef NL80211_STA_INFO_CONNECTED_TO_AS
+#define NL80211_STA_INFO_CONNECTED_TO_AS NL80211_ATTR_NOT_IMPLEMENTED
+#endif
+
 static const uc_nl_nested_spec_t nl80211_sta_info_nla = {
 	.headsize = 0,
 	.nattrs = 35,
