Index: ucode-2022-04-07-7fa59ce4/lib/uloop.c
===================================================================
--- ucode-2022-04-07-7fa59ce4.orig/lib/uloop.c
+++ ucode-2022-04-07-7fa59ce4/lib/uloop.c
@@ -971,6 +971,8 @@ uc_uloop_task(uc_vm_t *vm, size_t nargs)
 		err_return(errno);
 
 	if (pid == 0) {
+		uloop_done();
+
 		patch_devnull(0, false);
 		patch_devnull(1, true);
 		patch_devnull(2, true);
