Index: hostapd-2025.03.10~95ad7115/src/ap/ap_config.h
===================================================================
--- hostapd-2025.03.10~95ad7115.orig/src/ap/ap_config.h
+++ hostapd-2025.03.10~95ad7115/src/ap/ap_config.h
@@ -1426,7 +1426,7 @@ hostapd_set_oper_centr_freq_seg0_idx(str
 	if (conf->ieee80211be)
 		conf->eht_oper_centr_freq_seg0_idx = oper_centr_freq_seg0_idx;
 	if (is_6ghz_op_class(conf->op_class) &&
-	    center_idx_to_bw_6ghz(oper_centr_freq_seg0_idx) == 4)
+		is_6ghz_freq(conf->channel) && (center_idx_to_bw_6ghz(oper_centr_freq_seg0_idx) == 4))
 		oper_centr_freq_seg0_idx +=
 			conf->channel > oper_centr_freq_seg0_idx ? 16 : -16;
 
