Index: hostapd-2025.03.10~95ad7115/src/drivers/driver_nl80211.c
===================================================================
--- hostapd-2025.03.10~95ad7115.orig/src/drivers/driver_nl80211.c
+++ hostapd-2025.03.10~95ad7115/src/drivers/driver_nl80211.c
@@ -15794,6 +15794,8 @@ static bool wpa_driver_nl80211_can_share
 	struct i802_bss *bss;
 	bool ret = false;
 
+	return false;
+
 	if (!params->global_priv)
 		return false;
 
