Index: hostapd-2025.03.10~95ad7115/src/drivers/driver_nl80211.c
===================================================================
--- hostapd-2025.03.10~95ad7115.orig/src/drivers/driver_nl80211.c
+++ hostapd-2025.03.10~95ad7115/src/drivers/driver_nl80211.c
@@ -2444,7 +2444,6 @@ static void * wpa_driver_nl80211_drv_ini
 	drv->ctx = ctx;
 	drv->hostapd = !!hostapd;
 	drv->eapol_sock = -1;
-	dl_list_init(&drv->wiphy_list);
 
 	/*
 	 * There is no driver capability flag for this, so assume it is
