Index: hostapd-2025.03.10~95ad7115/src/ap/vlan_init.c
===================================================================
--- hostapd-2025.03.10~95ad7115.orig/src/ap/vlan_init.c
+++ hostapd-2025.03.10~95ad7115/src/ap/vlan_init.c
@@ -23,7 +23,8 @@ static int vlan_if_add(struct hostapd_da
 		       int existsok)
 {
 	bool vlan_exists = iface_exists(vlan->ifname);
-	int ret;
+	int ret = 0;
+
 #ifdef CONFIG_WEP
 	int i;
 
@@ -38,7 +39,7 @@ static int vlan_if_add(struct hostapd_da
 #endif /* CONFIG_WEP */
 
 	if (!hapd->driver || !hapd->driver->if_add) {
-		char *dot = strstr(vlan->ifname, ".");
+		char *dot = strrchr(vlan->ifname, '.');
 		if (dot)
 			*dot = '\0';
 		ret = 0;
@@ -59,7 +60,7 @@ static int vlan_if_add(struct hostapd_da
 	if (hapd->wpa_auth)
 		ret = wpa_auth_ensure_group(hapd->wpa_auth, vlan->vlan_id);
 
-	if (!ret && !vlan_exists && hapd->driver->if_add)
+	if (!ret && !vlan_exists)
 		hostapd_ubus_add_vlan(hapd, vlan);
 
 	if (ret == 0)
