Index: hostapd-2025.03.10~95ad7115/src/ap/hostapd.c
===================================================================
--- hostapd-2025.03.10~95ad7115.orig/src/ap/hostapd.c
+++ hostapd-2025.03.10~95ad7115/src/ap/hostapd.c
@@ -6189,14 +6189,11 @@ void hostapd_switch_color_timeout_handle
 	}
 
 	if (i == HE_OPERATION_BSS_COLOR_MAX) {
-		/* There are no free colors so turn BSS coloring off */
+		/* There are no free colors so do not change color and wait for next check */
 		wpa_printf(MSG_INFO,
-			   "No free colors left, turning off BSS coloring");
+			   "No free colors left, stay at the same color");
 		hapd->iface->conf->he_op.he_bss_color_disabled = 1;
-		hapd->iface->conf->he_op.he_bss_color = os_random() % 63 + 1;
 		hapd->no_free_color = 1;
-		for (b = 0; b < hapd->iface->num_bss; b++)
-			ieee802_11_set_beacon(hapd->iface->bss[b]);
 
 		 /* Enabling for next check after timeout*/
 		 hapd->iface->conf->he_op.he_bss_color_disabled = 0;
