Index: hostapd-2025.03.10~95ad7115/hostapd/config_file.c
===================================================================
--- hostapd-2025.03.10~95ad7115.orig/hostapd/config_file.c
+++ hostapd-2025.03.10~95ad7115/hostapd/config_file.c
@@ -3044,7 +3044,7 @@ static int hostapd_config_fill(struct ho
 	} else if (os_strcmp(buf, "ft_psk_generate_local") == 0) {
 		bss->ft_psk_generate_local = atoi(pos);
 	} else if (os_strcmp(buf, "ft_key") == 0) {
-		strncpy(bss->ft_key, pos, sizeof(bss->ft_key));
+		strncpy((char *)bss->ft_key, pos, sizeof(bss->ft_key));
 	} else if (os_strcmp(buf, "ft_l2_refresh") == 0) {
 		bss->ft_l2_refresh = atoi(pos);
 #endif /* CONFIG_IEEE80211R_AP */
@@ -5314,15 +5314,15 @@ struct hostapd_config * hostapd_config_r
 	for (i = 0; i < conf->num_bss; i++) {
 		if (*conf->bss[i]->ft_key) {
 			u8 buffer[128];
-			sprintf(buffer, "%02X:%02X:%02X:%02X:%02X:%02X %02X%02X%02X%02X%02X%02X %s", MAC2STR(conf->bss[i]->bssid), MAC2STR(conf->bss[i]->bssid), conf->bss[i]->ft_key);
-			add_r0kh(conf->bss[i], buffer);
-			sprintf(buffer, "%02X:%02X:%02X:%02X:%02X:%02X %02X:%02X:%02X:%02X:%02X:%02X %s", MAC2STR(conf->bss[i]->bssid), MAC2STR(conf->bss[i]->bssid), conf->bss[i]->ft_key);
-			add_r1kh(conf->bss[i], buffer);
-			sprintf(buffer, "ff:ff:ff:ff:ff:ff * %s", conf->bss[i]->ft_key);
-			add_r0kh(conf->bss[i], buffer);
-			sprintf(buffer, "00:00:00:00:00:00 00:00:00:00:00:00 %s", conf->bss[i]->ft_key);
-			add_r1kh(conf->bss[i], buffer);
-			hexstr2bin(conf->bss[i]->bssid, conf->bss[i]->r1_key_holder, FT_R1KH_ID_LEN);
+			sprintf((char *)buffer, "%02X:%02X:%02X:%02X:%02X:%02X %02X%02X%02X%02X%02X%02X %s", MAC2STR(conf->bss[i]->bssid), MAC2STR(conf->bss[i]->bssid), conf->bss[i]->ft_key);
+			add_r0kh(conf->bss[i], (char *)buffer);
+			sprintf((char *)buffer, "%02X:%02X:%02X:%02X:%02X:%02X %02X:%02X:%02X:%02X:%02X:%02X %s", MAC2STR(conf->bss[i]->bssid), MAC2STR(conf->bss[i]->bssid), conf->bss[i]->ft_key);
+			add_r1kh(conf->bss[i], (char *) buffer);
+			sprintf((char *)buffer, "ff:ff:ff:ff:ff:ff * %s", conf->bss[i]->ft_key);
+			add_r0kh(conf->bss[i], (char *) buffer);
+			sprintf((char *)buffer, "00:00:00:00:00:00 00:00:00:00:00:00 %s", conf->bss[i]->ft_key);
+			add_r1kh(conf->bss[i], (char *) buffer);
+			hexstr2bin((char *)conf->bss[i]->bssid, conf->bss[i]->r1_key_holder, FT_R1KH_ID_LEN);
 			conf->bss[i]->r0_key_holder_bssid = 1;
 		}
 		hostapd_set_security_params(conf->bss[i], 1);
Index: hostapd-2025.03.10~95ad7115/hostapd/radius.c
===================================================================
--- hostapd-2025.03.10~95ad7115.orig/hostapd/radius.c
+++ hostapd-2025.03.10~95ad7115/hostapd/radius.c
@@ -167,10 +167,9 @@ load_userfile(struct radius_state *s)
 		[USERDATA_PHASE1] = { "phase1", BLOBMSG_TYPE_TABLE },
 		[USERDATA_PHASE2] = { "phase2", BLOBMSG_TYPE_TABLE },
 	};
-	struct blob_attr *tb[__USERDATA_MAX], *cur;
+	struct blob_attr *tb[__USERDATA_MAX];
 	static struct blob_buf b;
 	struct stat st;
-	int rem;
 
 	if (stat(s->user_file, &st))
 		return;
@@ -276,7 +275,6 @@ radius_count_attrs(struct blob_attr **tb
 
 	blobmsg_for_each_attr(cur, data, rem) {
 		struct radius_parse_attr_data *data;
-		size_t prev = *attr_size;
 
 		data = radius_parse_attr(cur);
 		if (!data)
@@ -340,7 +338,6 @@ static void
 radius_parse_attrs(struct blob_attr **tb, struct radius_parse_attr_state *state)
 {
 	struct blob_attr *data = tb[USER_ATTR_RADIUS];
-	struct hostapd_radius_attr *prev = NULL;
 	struct blob_attr *cur;
 	int len, rem;
 	void *val;
@@ -372,7 +369,6 @@ radius_parse_attrs(struct blob_attr **tb
 	blobmsg_for_each_attr(cur, data, rem) {
 		struct radius_parse_attr_data *data;
 		void *val;
-		int size;
 
 		data = radius_parse_attr(cur);
 		if (!data)
@@ -453,13 +449,15 @@ radius_user_get_state(struct radius_user
 	struct blob_attr *tb[__USER_ATTR_MAX], *cur;
 	char *password_buf, *salt_buf, *name_buf;
 	struct radius_parse_attr_state astate = {};
-	struct hostapd_radius_attr *attr;
 	struct radius_user_state *state;
 	int pw_len = 0, salt_len = 0;
 	struct eap_user *eap;
-	struct wpabuf *val;
 	size_t attrsize = 0;
+#if 0
 	void *attrdata;
+	struct hostapd_radius_attr *attr;
+	struct wpabuf *val;
+#endif
 	int n_attr = 0;
 
 	state = avl_find_element(&u->user_state, id, state, node);
@@ -483,18 +481,18 @@ radius_user_get_state(struct radius_user
 			 &astate.buf, n_attr * sizeof(*astate.buf),
 			 &astate.attrdata, attrsize);
 	eap = &state->data;
-	eap->salt = salt_len ? salt_buf : NULL;
+	eap->salt = salt_len ? (u8 *)salt_buf : NULL;
 	eap->salt_len = salt_len;
-	eap->password = pw_len ? password_buf : NULL;
+	eap->password = pw_len ? (u8 *)password_buf : NULL;
 	eap->password_len = pw_len;
 	eap->force_version = -1;
 
 	if ((cur = tb[USER_ATTR_SALT]) != NULL)
-		hexstr2bin(blobmsg_get_string(cur), salt_buf, salt_len);
+		hexstr2bin(blobmsg_get_string(cur), (u8 *)salt_buf, salt_len);
 	if ((cur = tb[USER_ATTR_PASSWORD]) != NULL)
 		memcpy(password_buf, blobmsg_get_string(cur), pw_len);
 	else if ((cur = tb[USER_ATTR_HASH]) != NULL) {
-		hexstr2bin(blobmsg_get_string(cur), password_buf, pw_len);
+		hexstr2bin(blobmsg_get_string(cur), (u8 *)password_buf, pw_len);
 		eap->password_hash = 1;
 	}
 	radius_user_parse_methods(eap, tb[USER_ATTR_METHODS]);
@@ -510,9 +508,11 @@ radius_user_get_state(struct radius_user
 
 	return &state->data;
 
+#if 0
 free:
 	free(state);
 	return NULL;
+#endif
 }
 
 static int radius_get_eap_user(void *ctx, const u8 *identity,
@@ -595,6 +595,8 @@ static int radius_init(struct radius_sta
 	memset(s, 0, sizeof(*s));
 	radius_userdata_init(&s->phase1);
 	radius_userdata_init(&s->phase2);
+
+	return 0;
 }
 
 static void radius_deinit(struct radius_state *s)
@@ -613,6 +615,7 @@ static int usage(const char *progname)
 {
 	fprintf(stderr, "Usage: %s <options>\n",
 		progname);
+	return 0;
 }
 
 int radius_main(int argc, char **argv)
@@ -655,7 +658,7 @@ int radius_main(int argc, char **argv)
 			config.tls.dh_file = optarg;
 			break;
 		case 'i':
-			state.eap.server_id = optarg;
+			state.eap.server_id = (u8 *)optarg;
 			state.eap.server_id_len = strlen(optarg);
 			break;
 		case 'k':
Index: hostapd-2025.03.10~95ad7115/src/ap/wpa_auth_glue.c
===================================================================
--- hostapd-2025.03.10~95ad7115.orig/src/ap/wpa_auth_glue.c
+++ hostapd-2025.03.10~95ad7115/src/ap/wpa_auth_glue.c
@@ -157,7 +157,7 @@ static void hostapd_wpa_auth_conf(struct
 	os_memcpy(wconf->mobility_domain, conf->mobility_domain,
 		  MOBILITY_DOMAIN_ID_LEN);
 	if (1 || conf->r0_key_holder_bssid) {
-		sprintf(wconf->r0_key_holder, "%02X%02X%02X%02X%02X%02X", MAC2STR(conf->bssid));
+		sprintf((char *)wconf->r0_key_holder, "%02X%02X%02X%02X%02X%02X", MAC2STR(conf->bssid));
 		wconf->r0_key_holder_len = 12;
 	} else if (conf->nas_identifier &&
 	    os_strlen(conf->nas_identifier) <= FT_R0KH_ID_MAX_LEN) {
Index: hostapd-2025.03.10~95ad7115/src/ap/ieee802_1x.c
===================================================================
--- hostapd-2025.03.10~95ad7115.orig/src/ap/ieee802_1x.c
+++ hostapd-2025.03.10~95ad7115/src/ap/ieee802_1x.c
@@ -2007,7 +2007,7 @@ static int ieee802_1x_update_wispr(struc
 {
 	memset(sta->bandwidth, 0, sizeof(sta->bandwidth));
 
-	if (radius_msg_get_wispr(msg, &sta->bandwidth))
+	if (radius_msg_get_wispr(msg, sta->bandwidth))
 		return 0;
 
 	if (!sta->bandwidth[0] && !sta->bandwidth[1])
Index: hostapd-2025.03.10~95ad7115/src/ap/sta_info.c
===================================================================
--- hostapd-2025.03.10~95ad7115.orig/src/ap/sta_info.c
+++ hostapd-2025.03.10~95ad7115/src/ap/sta_info.c
@@ -2013,12 +2013,6 @@ bool ap_sta_set_authorized_flag(struct h
 	return true;
 }
 
-void ap_sta_set_authorized_event(struct hostapd_data *hapd, 
-				struct sta_info *sta, int authorized)
-{
-	__ap_sta_set_authorized_event(hapd, sta, authorized, 0);
-}
-
 void __ap_sta_set_authorized_event(struct hostapd_data *hapd,
 				 struct sta_info *sta, int authorized, int reassoc)
 {
@@ -2141,6 +2135,12 @@ void __ap_sta_set_authorized_event(struc
 #endif /* CONFIG_FST */
 }
 
+void ap_sta_set_authorized_event(struct hostapd_data *hapd, 
+				struct sta_info *sta, int authorized)
+{
+	__ap_sta_set_authorized_event(hapd, sta, authorized, 0);
+}
+
 bool __ap_sta_set_authorized(struct hostapd_data *hapd, struct sta_info *sta,
 			   int authorized, int reassoc)
 {
@@ -2153,7 +2153,7 @@ bool __ap_sta_set_authorized(struct host
 void ap_sta_set_authorized(struct hostapd_data *hapd, struct sta_info *sta,
 			   int authorized)
 {
-	_ap_sta_set_authorized(hapd, sta, authorized, 0);
+	__ap_sta_set_authorized(hapd, sta, authorized, 0);
 }
 
 void ap_sta_disconnect(struct hostapd_data *hapd, struct sta_info *sta,
Index: hostapd-2025.03.10~95ad7115/src/ap/sta_info.h
===================================================================
--- hostapd-2025.03.10~95ad7115.orig/src/ap/sta_info.h
+++ hostapd-2025.03.10~95ad7115/src/ap/sta_info.h
@@ -480,9 +480,9 @@ bool ap_sta_set_authorized_flag(struct h
 				int authorized);
 void ap_sta_set_authorized_event(struct hostapd_data *hapd,
 				 struct sta_info *sta, int authorized);
-bool ap_sta_set_authorized(struct hostapd_data *hapd,
+void ap_sta_set_authorized(struct hostapd_data *hapd,
 			   struct sta_info *sta, int authorized);
-void _ap_sta_set_authorized(struct hostapd_data *hapd,
+bool __ap_sta_set_authorized(struct hostapd_data *hapd,
 			    struct sta_info *sta, int authorized,
 			    int reassoc);
 
Index: hostapd-2025.03.10~95ad7115/src/ap/beacon.c
===================================================================
--- hostapd-2025.03.10~95ad7115.orig/src/ap/beacon.c
+++ hostapd-2025.03.10~95ad7115/src/ap/beacon.c
@@ -1563,7 +1563,9 @@ void handle_probe_req(struct hostapd_dat
 
 	if (hapd->iconf->rssi_ignore_probe_request && ssi_signal &&
 	    ssi_signal < hapd->iconf->rssi_ignore_probe_request) {
+#if 0
 		hostapd_ubus_notify_rssi(hapd, "rssi-ignore-probe", mgmt->sa, ssi_signal);
+#endif
 		return;
 	}
 
Index: hostapd-2025.03.10~95ad7115/src/ap/ubus.c
===================================================================
--- hostapd-2025.03.10~95ad7115.orig/src/ap/ubus.c
+++ hostapd-2025.03.10~95ad7115/src/ap/ubus.c
@@ -1817,8 +1817,6 @@ int hostapd_ubus_handle_event(struct hos
 		ubus_notify(ctx, &hapd->ubus.obj, type, b.head, -1);
 		return WLAN_STATUS_SUCCESS;
 	}
-	if (reassoc)
-		blobmsg_add_macaddr(&b, "origin", sta->origin_ap);
 	blobmsg_printf(&b, "target", MACSTR, MAC2STR(hapd->conf->bssid));
 
 	if (ubus_notify_async(ctx, &hapd->ubus.obj, type, b.head, &ureq.nreq))
Index: hostapd-2025.03.10~95ad7115/src/ap/wpa_auth_ft.c
===================================================================
--- hostapd-2025.03.10~95ad7115.orig/src/ap/wpa_auth_ft.c
+++ hostapd-2025.03.10~95ad7115/src/ap/wpa_auth_ft.c
@@ -446,7 +446,6 @@ static size_t wpa_ft_rate_limit_lin(cons
 				    u8 *start, u8 *endpos)
 {
 	size_t tlv_len;
-	int i, len;
 	struct ft_rrb_tlv *hdr;
 	u8 *pos = start;
 
@@ -574,7 +573,7 @@ static int wpa_ft_rrb_build(const u8 *ke
 			    const struct tlv_list *tlvs_auth,
 			    const struct vlan_description *vlan,
 			    const u8 *src_addr, u8 type,
-			    u8 **packet, size_t *packet_len
+			    u8 **packet, size_t *packet_len,
 				const struct rate_description *rate)
 {
 	u8 *plain = NULL, *auth = NULL, *pos, *tmp;
@@ -1147,7 +1146,7 @@ wpa_ft_rrb_seq_req(struct wpa_authentica
 
 	if (wpa_ft_rrb_build(key, key_len, NULL, NULL, seq_req_auth, NULL,
 			     wpa_auth->addr, FT_PACKET_R0KH_R1KH_SEQ_REQ,
-			     &packet, &packet_len, NUL, NULLL) < 0) {
+			     &packet, &packet_len, NULL) < 0) {
 		item = NULL; /* some other seq resp might still accept this */
 		goto err;
 	}
@@ -3574,7 +3573,7 @@ static int wpa_ft_local_derive_pmk_r1(st
 				      const u8 **radius_cui,
 				      size_t *radius_cui_len,
 				      int *out_session_timeout,
-				      size_t *pmk_r1_len
+				      size_t *pmk_r1_len,
 					  struct rate_description *rate)
 {
 	struct wpa_auth_config *conf = &wpa_auth->conf;
@@ -3611,7 +3610,7 @@ static int wpa_ft_local_derive_pmk_r1(st
 			    out_pmk_r1_name,
 			    sm->pairwise, r0->vlan, expires_in, session_timeout,
 			    r0->identity, r0->identity_len,
-			    r0->radius_cui, r0->radius_cui_len
+			    r0->radius_cui, r0->radius_cui_len,
 				r0->rate);
 
 	*out_pairwise = sm->pairwise;
Index: hostapd-2025.03.10~95ad7115/src/ap/hostapd.h
===================================================================
--- hostapd-2025.03.10~95ad7115.orig/src/ap/hostapd.h
+++ hostapd-2025.03.10~95ad7115/src/ap/hostapd.h
@@ -639,6 +639,7 @@ struct hostapd_data {
 	u64 scan_cookie; /* Scan instance identifier for the ongoing HT40 scan
 			  */
 	bool disable_cu;
+	int signal_mgmt;
 	struct hostapd_wmm_ac_params prev_wmm[4];
 	size_t mbssid_idx;
 	struct hostapd_multi_mbssid_group *mbssid_group;
Index: hostapd-2025.03.10~95ad7115/src/ap/hostapd.c
===================================================================
--- hostapd-2025.03.10~95ad7115.orig/src/ap/hostapd.c
+++ hostapd-2025.03.10~95ad7115/src/ap/hostapd.c
@@ -5295,7 +5295,7 @@ void hostapd_new_assoc_sta(struct hostap
 	 * IEEE 802.1X/WPA code will start accounting after the station has
 	 * been authorized. */
 	if (!hapd->conf->ieee802_1x && !hapd->conf->wpa) {
-		if (ap_sta_set_authorized(hapd, sta, 1)) {
+		if (__ap_sta_set_authorized(hapd, sta, 1, 0)) {
 			/* Update driver authorized flag for the STA to cover
 			 * the case where AP SME is in the driver and there is
 			 * no separate event for handling TX status event for
Index: hostapd-2025.03.10~95ad7115/src/ap/ieee802_11.c
===================================================================
--- hostapd-2025.03.10~95ad7115.orig/src/ap/ieee802_11.c
+++ hostapd-2025.03.10~95ad7115/src/ap/ieee802_11.c
@@ -7907,7 +7907,7 @@ static void handle_assoc_cb(struct hosta
 		 * Open, static WEP, FT protocol, or FILS; no separate
 		 * authorization step.
 		 */
-		_ap_sta_set_authorized(hapd, sta, 1, reassoc);
+		__ap_sta_set_authorized(hapd, sta, 1, reassoc);
 	}
 
 	if (sta->auth_alg == WLAN_AUTH_FT) {
