Index: hostapd-2025.03.10~95ad7115/src/drivers/driver_nl80211_scan.c
===================================================================
--- hostapd-2025.03.10~95ad7115.orig/src/drivers/driver_nl80211_scan.c
+++ hostapd-2025.03.10~95ad7115/src/drivers/driver_nl80211_scan.c
@@ -457,6 +457,9 @@ int wpa_driver_nl80211_scan(struct i802_
 			}
 #endif /* CONFIG_IEEE80211BE */
 
+			if (ret == -EBUSY)
+				goto fail;
+
 			/*
 			 * mac80211 does not allow scan requests in AP mode, so
 			 * try to do this in station mode.
