Index: hostapd-2025.03.10~95ad7115/src/ap/wpa_auth_ft.c
===================================================================
--- hostapd-2025.03.10~95ad7115.orig/src/ap/wpa_auth_ft.c
+++ hostapd-2025.03.10~95ad7115/src/ap/wpa_auth_ft.c
@@ -3664,6 +3664,7 @@ static int wpa_ft_process_auth_req(struc
 	const u8 *identity, *radius_cui;
 	size_t identity_len = 0, radius_cui_len = 0;
 	size_t pmk_r1_len, kdk_len, len, bmle_len;
+	struct os_reltime now;
 	int retval = WLAN_STATUS_UNSPECIFIED_FAILURE;
 
 	*resp_ies = NULL;
@@ -3818,16 +3819,25 @@ pmk_r1_derived:
 
 	if (!sm->ANonce_generated) {
 		/* Generate a new ANonce for the first FT authentication request */
-		if (random_get_bytes(sm->ANonce, WPA_NONCE_LEN)) {
-			wpa_printf(MSG_DEBUG, "FT: Failed to get random data for "
-				   "ANonce");
-			goto out;
+		if (os_get_reltime(&now) < 0 ||
+				os_reltime_expired(&now, &sm->ANonce_time, 1)) {
+			if (random_get_bytes(sm->ANonce, WPA_NONCE_LEN)) {
+				wpa_printf(MSG_DEBUG, "FT: Failed to get random data for "
+						"ANonce");
+				return WLAN_STATUS_UNSPECIFIED_FAILURE;
+			}
+			sm->ANonce_time.sec = now.sec;
+			sm->ANonce_time.usec = now.usec;
+			wpa_printf(MSG_INFO, "FT: ANonce was randomized");
+		} else {
+			wpa_printf(MSG_INFO, "FT: ANonce has not expired");
+
 		}
 
 		sm->ANonce_generated = true;
 	} else  {
 		wpa_hexdump(MSG_DEBUG, "FT: Reusing existing ANonce",
-			    sm->ANonce, WPA_NONCE_LEN);
+				sm->ANonce, WPA_NONCE_LEN);
 	}
 
 	/* Now that we know the correct PMK-R1 length and as such, the length
Index: hostapd-2025.03.10~95ad7115/src/ap/wpa_auth_i.h
===================================================================
--- hostapd-2025.03.10~95ad7115.orig/src/ap/wpa_auth_i.h
+++ hostapd-2025.03.10~95ad7115/src/ap/wpa_auth_i.h
@@ -54,6 +54,7 @@ struct wpa_state_machine {
 	bool MICVerified;
 	bool GUpdateStationKeys;
 	u8 ANonce[WPA_NONCE_LEN];
+	struct os_reltime ANonce_time;
 	u8 SNonce[WPA_NONCE_LEN];
 	u8 alt_SNonce[WPA_NONCE_LEN];
 	u8 alt_replay_counter[WPA_REPLAY_COUNTER_LEN];
