Index: hostapd-2025.03.10~95ad7115/src/ap/beacon.c
===================================================================
--- hostapd-2025.03.10~95ad7115.orig/src/ap/beacon.c
+++ hostapd-2025.03.10~95ad7115/src/ap/beacon.c
@@ -1562,8 +1562,10 @@ void handle_probe_req(struct hostapd_dat
 	};
 
 	if (hapd->iconf->rssi_ignore_probe_request && ssi_signal &&
-	    ssi_signal < hapd->iconf->rssi_ignore_probe_request)
+	    ssi_signal < hapd->iconf->rssi_ignore_probe_request) {
+		hostapd_ubus_notify_rssi(hapd, "rssi-ignore-probe", mgmt->sa, ssi_signal);
 		return;
+	}
 
 	if (len < IEEE80211_HDRLEN)
 		return;
