Index: qca-nss-ppe-ffc8f2447dbb2d4a29d1fca4fad516132d5848fe/drv/ppe_drv/tun/ppe_drv_tun.c
===================================================================
--- qca-nss-ppe-ffc8f2447dbb2d4a29d1fca4fad516132d5848fe.orig/drv/ppe_drv/tun/ppe_drv_tun.c
+++ qca-nss-ppe-ffc8f2447dbb2d4a29d1fca4fad516132d5848fe/drv/ppe_drv/tun/ppe_drv_tun.c
@@ -114,7 +114,7 @@ static void ppe_drv_tun_free(struct kref
  */
 static bool ppe_drv_tun_deref(struct ppe_drv_tun *ptun)
 {
-	uint8_t tun_idx = ptun->tun_idx;
+	//uint8_t tun_idx = ptun->tun_idx;
 
 	ppe_drv_assert(kref_read(&ptun->ref), "%p: ref count under run for tun", ptun);
 
Index: qca-nss-ppe-ffc8f2447dbb2d4a29d1fca4fad516132d5848fe/drv/ppe_drv/tun/ppe_drv_tun_encap.c
===================================================================
--- qca-nss-ppe-ffc8f2447dbb2d4a29d1fca4fad516132d5848fe.orig/drv/ppe_drv/tun/ppe_drv_tun_encap.c
+++ qca-nss-ppe-ffc8f2447dbb2d4a29d1fca4fad516132d5848fe/drv/ppe_drv/tun/ppe_drv_tun_encap.c
@@ -95,7 +95,6 @@ void ppe_drv_tun_encap_free(struct kref
  */
 bool ppe_drv_tun_encap_deref(struct ppe_drv_tun_encap *ptec)
 {
-	uint8_t tun_idx = ptec->tun_idx;
 
 	ppe_drv_assert(kref_read(&ptec->ref), "%p: ref count under run for ptec", ptec);
 
