diff --git a/src/wm2/src/wm2_radio.c b/src/wm2/src/wm2_radio.c
old mode 100644
new mode 100755
index b01a88f..5148636
--- a/src/wm2/src/wm2_radio.c
+++ b/src/wm2/src/wm2_radio.c
@@ -874,7 +874,7 @@ wm2_vconf_recalc(const char *ifname, bool force)
     if (has && strlen(SCHEMA_KEY_VAL(vconf.security, "key")) < 8) {
         LOGD("%s: overriding 'ssid' and 'security' for onboarding", ifname);
         vconf.ssid_exists = vstate.ssid_exists;
-        STRSCPY(vconf.ssid, vstate.ssid);
+//        STRSCPY(vconf.ssid, vstate.ssid);
-        memcpy(vconf.security, vstate.security, sizeof(vconf.security));
+//        memcpy(vconf.security, vstate.security, sizeof(vconf.security));
     }
 
@@ -1132,7 +1132,7 @@ wm2_rconf_recalc(const char *ifname, bool force)
         if (rconf.channel_exists && rconf.vif_configs_len == 0) {
             LOGD("%s: ignoring rconf channel %d: no vifs available yet",
                   rconf.if_name, rconf.channel);
-            rconf.channel = rstate.channel;
+//            rconf.channel = rstate.channel;
         }
 
     if (want) {
