From a772694c6914fa7b4f349486d16a7933328d05a4 Mon Sep 17 00:00:00 2001
From: Shayne Chen <shayne.chen@mediatek.com>
Date: Mon, 25 Nov 2024 14:16:47 +0800
Subject: [PATCH 14/17] mtk: iw: dump per-radio antenna_mask

Signed-off-by: Shayne Chen <shayne.chen@mediatek.com>
---
 info.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/info.c b/info.c
index 5ce3f78..ea6c066 100644
--- a/info.c
+++ b/info.c
@@ -931,6 +931,9 @@ next:
 					printf("\t\t\tRTS Threshold: %d\n",
 					       nla_get_u32(radio_prop));
 					break;
+				case NL80211_WIPHY_RADIO_ATTR_ANTENNA_MASK:
+					printf("\t\t\tantenna_mask: 0x%08x\n", nla_get_u32(radio_prop));
+					break;
 				default:
 					printf("\t\t\t* <failed to parse>\n");
 				}
-- 
2.45.2

