--- a/src/ap/drv_callbacks.c
+++ b/src/ap/drv_callbacks.c
@@ -2477,6 +2477,14 @@
	bool radio_matched = false;
	int i;

+	if (hapd->iface->conf->he_mu_edca.he_mu_ac_be_param[2] != 255 ||
+		hapd->iface->conf->he_mu_edca.he_mu_ac_bk_param[2] != 255 ||
+		hapd->iface->conf->he_mu_edca.he_mu_ac_vi_param[2] != 255 ||
+		hapd->iface->conf->he_mu_edca.he_mu_ac_vo_param[2] != 255) {
+		wpa_printf(MSG_DEBUG, "User-configured MU-EDCA parameters detected, ignoring firmware update");
+		return;
+	}
+
	if (hapd->conf->mld_ap) {
		for_each_mld_link(selected_hapd, hapd) {
			if (!selected_hapd->iface ||
