From daaf7595379630f7c887b87cfdf3609c26e4b5f6 Mon Sep 17 00:00:00 2001
From: John Crispin <john@phrozen.org>
Date: Tue, 22 Sep 2020 18:02:55 +0200
Subject: [PATCH 01/21] .gitignore: add the /profiles folder

Signed-off-by: John Crispin <john@phrozen.org>
---
 .gitignore | 1 +
 1 file changed, 1 insertion(+)

diff --git a/.gitignore b/.gitignore
index a80cdbbea3..77cd4e730c 100644
--- a/.gitignore
+++ b/.gitignore
@@ -17,6 +17,7 @@
 /overlay
 /package/feeds
 /package/openwrt-packages
+/profiles
 /*.patch
 /llvm-bpf*
 key-build*
-- 
2.25.1

