Index: usteer-2021-04-19-f42bf296/local_node.c
===================================================================
--- usteer-2021-04-19-f42bf296.orig/local_node.c
+++ usteer-2021-04-19-f42bf296/local_node.c
@@ -476,8 +476,8 @@ usteer_local_node_status_cb(struct ubus_
 		node->freq = blobmsg_get_u32(tb[MSG_FREQ]);
 	if (tb[MSG_CHANNEL])
 		node->channel = blobmsg_get_u32(tb[MSG_CHANNEL]);
-	if (tb[MSG_FREQ])
-		node->op_class = blobmsg_get_u32(tb[MSG_OP_CLASS]);	
+	if (tb[MSG_OP_CLASS])
+		node->op_class = blobmsg_get_u32(tb[MSG_OP_CLASS]);
 }
 
 static void
