Index: mesh11sd-1.2.0/linux_openwrt/mesh11sd/files/etc/init.d/mesh11sd
===================================================================
--- mesh11sd-1.2.0.orig/linux_openwrt/mesh11sd/files/etc/init.d/mesh11sd
+++ mesh11sd-1.2.0/linux_openwrt/mesh11sd/files/etc/init.d/mesh11sd
@@ -11,3 +11,11 @@ start_service() {
 	procd_close_instance
 }
 
+reload_service() {
+        wifi down
+        sleep 5
+        wifi up
+}
+service_triggers() {
+        procd_add_reload_trigger "mesh11sd"
+}
