--- a/src/mesh11sd	2024-09-06 17:13:52.075070001 +0800
+++ b/src/mesh11sd	2024-09-06 17:47:36.326914438 +0800
@@ -434,7 +434,6 @@
 		fi
 
 		lastmode=$mode
-		meshindex=0
 
 		if [ "$enabled" = 1 ]; then
 			#get list of mesh configs
@@ -443,16 +442,19 @@
 			if [ ! -z "$meshconfigs" ]; then
 				for meshconfig in $meshconfigs; do
 					ifname=$(uci get $meshconfig.ifname 2> /dev/null)
-
-					if [ -z "$ifname" ] || [ "$ifname" != "mesh$meshindex" ]; then
-						# No interface name in config, so add one
-						ucibatch="set $meshconfig.ifname='mesh$meshindex'"
-						echo "$ucibatch" | uci batch
-						changed=1
-						syslogmessage="Setting mesh interface name to [ mesh$meshindex ]"
-						write_to_syslog
+					device=$(uci -q get $meshconfig.device)
+					device_type=$(uci -q get wireless.$device.type)
+					new_ifname="halow_mesh"
+					if [ "$device_type" == "morse" ]; then
+						if [ -z "$ifname" ] || [ "$ifname" != "$new_ifname" ]; then
+							# No interface name in config, so add one
+							ucibatch="set $meshconfig.ifname='$new_ifname'"
+							echo "$ucibatch" | uci batch
+							changed=1
+							syslogmessage="Setting mesh interface name to [ $new_ifname ]"
+							write_to_syslog
+						fi
 					fi
-					meshindex=$(($meshindex+1))
 				done
 
 				if [ "$changed" -eq 1 ]; then
