Index: miniupnpd-1.9.20150609/miniupnpd.c
===================================================================
--- miniupnpd-1.9.20150609.orig/miniupnpd.c	2015-06-09 21:06:21.000000000 +0530
+++ miniupnpd-1.9.20150609/miniupnpd.c	2016-06-20 21:16:42.442494356 +0530
@@ -1046,6 +1046,7 @@
 	struct lan_addr_s * lan_addr;
 	struct lan_addr_s * lan_addr2;
 
+	v->clean_ruleset_threshold = 0;
 	/* only print usage if -h is used */
 	for(i=1; i<argc; i++)
 	{
