Index: hostapd-2023-02-21-ath12.3-cs/wpa_supplicant/wpa_supplicant.c
===================================================================
--- hostapd-2023-02-21-ath12.3-cs.orig/wpa_supplicant/wpa_supplicant.c
+++ hostapd-2023-02-21-ath12.3-cs/wpa_supplicant/wpa_supplicant.c
@@ -2744,6 +2744,9 @@ void ibss_mesh_setup_freq(struct wpa_sup
         bool is_24ghz, is_6ghz;
 
         freq->freq = ssid->frequency;
+	if (ssid->fixed_freq) {
+		obss_scan = 0;
+	}
 
         for (j = 0; j < wpa_s->last_scan_res_used; j++) {
                 struct wpa_bss *bss = wpa_s->last_scan_res[j];
