Index: hostapd-2023-02-21-ath12.3-cs/hostapd/config_file.c
===================================================================
--- hostapd-2023-02-21-ath12.3-cs.orig/hostapd/config_file.c
+++ hostapd-2023-02-21-ath12.3-cs/hostapd/config_file.c
@@ -2271,6 +2271,8 @@ static int hostapd_config_fill(struct ho
 			return 1;
 		}
 		conf->driver = driver;
+	} else if (os_strcmp(buf, "uci_section") == 0) {
+		bss->uci_section = os_strdup(pos);
 	} else if (os_strcmp(buf, "driver_params") == 0) {
 		os_free(conf->driver_params);
 		conf->driver_params = os_strdup(pos);
Index: hostapd-2023-02-21-ath12.3-cs/src/ap/ap_config.h
===================================================================
--- hostapd-2023-02-21-ath12.3-cs.orig/src/ap/ap_config.h
+++ hostapd-2023-02-21-ath12.3-cs/src/ap/ap_config.h
@@ -287,6 +287,7 @@ struct hostapd_bss_config {
 	char snoop_iface[IFNAMSIZ + 1];
 	char vlan_bridge[IFNAMSIZ + 1];
 	char wds_bridge[IFNAMSIZ + 1];
+	char *uci_section;
 
 #ifdef CONFIG_IEEE80211BE
 	unsigned int mlo_link_id;
Index: hostapd-2023-02-21-ath12.3-cs/src/ap/ap_config.c
===================================================================
--- hostapd-2023-02-21-ath12.3-cs.orig/src/ap/ap_config.c
+++ hostapd-2023-02-21-ath12.3-cs/src/ap/ap_config.c
@@ -825,6 +825,7 @@ void hostapd_config_free_bss(struct host
 	os_free(conf->radius_server_clients);
 	os_free(conf->radius);
 	os_free(conf->radius_das_shared_secret);
+	os_free(conf->uci_section);
 	hostapd_config_free_vlan(conf);
 	os_free(conf->time_zone);
 
