diff -uNr hostapd-2023-02-21-ath12.3-cs.orig/src/ap/ieee802_11.c hostapd-2023-02-21-ath12.3-cs/src/ap/ieee802_11.c
--- hostapd-2023-02-21-ath12.3-cs.orig/src/ap/ieee802_11.c	2024-09-10 13:29:43.000000000 +0800
+++ hostapd-2023-02-21-ath12.3-cs/src/ap/ieee802_11.c	2024-09-09 18:15:59.000000000 +0800
@@ -5965,6 +5965,7 @@
 	wpa_printf(MSG_DEBUG, "disassocation: STA=" MACSTR " reason_code=%d",
 		   MAC2STR(mgmt->sa),
 		   le_to_host16(mgmt->u.disassoc.reason_code));
+	hostapd_ubus_notify(hapd, "disassoc", mgmt->sa);
 
 	sta = ap_get_sta(hapd, mgmt->sa);
 	if (sta == NULL) {
@@ -6165,7 +6166,6 @@
 		sta->last_seq_ctrl = seq_ctrl;
 		sta->last_subtype = WLAN_FC_STYPE_ACTION;
 	}
-	hostapd_ubus_notify(hapd, "disassoc", mgmt->sa);
 
 	switch (mgmt->u.action.category) {
 #ifdef CONFIG_IEEE80211R_AP
