From 5bccdb5e506dddf3f9c85868faad56c6994d80c7 Mon Sep 17 00:00:00 2001
From: Brett Mastbergen <brettm@minim.com>
Date: Mon, 9 Jan 2023 13:31:48 -0500
Subject: [PATCH] mac80211: Add support for 5ghz channel 177

SW-3051
---
 mac80211.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/mac80211.c b/mac80211.c
index bc3a7514..d7830d6b 100644
--- a/mac80211.c
+++ b/mac80211.c
@@ -75,6 +75,7 @@ static const struct ieee80211_channel mt76_channels_5ghz[] = {
 	CHAN5G(165, 5825),
 	CHAN5G(169, 5845),
 	CHAN5G(173, 5865),
+	CHAN5G(177, 5885),
 };
 
 static const struct ieee80211_channel mt76_channels_6ghz[] = {
-- 
2.30.2

