Index: udevmand-20220112-065f75cb/bridge.c
===================================================================
--- udevmand-20220112-065f75cb.orig/bridge.c
+++ udevmand-20220112-065f75cb/bridge.c
@@ -17,6 +17,8 @@
 
 #define BR_MAX_ENTRY	2048
 
+#define GLOB_ONLYDIR	0
+
 static struct uloop_timeout bridge_timer;
 static struct vlist_tree bridge_mac;
 static LIST_HEAD(bridge_if);
Index: udevmand-20220112-065f75cb/udevmand.h
===================================================================
--- udevmand-20220112-065f75cb.orig/udevmand.h
+++ udevmand-20220112-065f75cb/udevmand.h
@@ -102,7 +102,7 @@ struct bridge_mac {
 
 int avl_mac_cmp(const void *k1, const void *k2, void *ptr);
 
-struct avl_tree mac_tree;
+extern struct avl_tree mac_tree;
 int mac_dump_all(void);
 void mac_dump(struct mac *mac, int interface);
 struct mac* mac_find(uint8_t *addr);
@@ -116,7 +116,7 @@ bool nl_status_socket(struct nl_socket *
 		     int (*cb)(struct nl_msg *msg, void *arg), void *priv);
 int genl_send_and_recv(struct nl_socket *ev, struct nl_msg * msg);
 
-struct blob_buf b;
+extern struct blob_buf b;
 void blobmsg_add_iface(struct blob_buf *bbuf, char *name, int index);
 void blobmsg_add_iftype(struct blob_buf *bbuf, const char *name, const uint32_t iftype);
 void blobmsg_add_ipv4(struct blob_buf *bbuf, const char *name, const uint8_t* addr);
