--- a/drivers/net/wireless/ath/ath11k/mac.c
+++ b/drivers/net/wireless/ath/ath11k/mac.c
@@ -7225,12 +7225,8 @@ static int ath11k_mac_op_add_interface(s
 	ath11k_debugfs_add_interface(arvif);
 
 	if (vif->type != NL80211_IFTYPE_MONITOR &&
-	    test_bit(ATH11K_FLAG_MONITOR_CONF_ENABLED, &ar->monitor_flags)) {
-		ret = ath11k_mac_monitor_vdev_create(ar);
-		if (ret)
-			ath11k_warn(ar->ab, "failed to create monitor vdev during add interface: %d",
-				    ret);
-	}
+	    test_bit(ATH11K_FLAG_MONITOR_CONF_ENABLED, &ar->monitor_flags))
+		ath11k_mac_monitor_vdev_create(ar);
 
 	ret = ath11k_mac_ap_ps_recalc(ar);
 	if (ret)
