From a08869e704f38a117dcd7ae274ca9be7576184fb Mon Sep 17 00:00:00 2001
From: Pavithra R <pavir@codeaurora.org>
Date: Sun, 1 Oct 2017 02:46:35 -0700
Subject: [PATCH] [tcrypt] Add test case for authenc(hmac(sha256),cbc(aes))

Change-Id: Icd585298e034277dee7a97f51771978845e0f805
Signed-off-by: Pavithra R <pavir@codeaurora.org>
Signed-off-by: Tushar Ganatra <quic_tganatra@quicinc.com>
---
 crypto/tcrypt.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/crypto/tcrypt.c b/crypto/tcrypt.c
index 1a648b9c0bfb..10c3cc462e6d 100644
--- a/crypto/tcrypt.c
+++ b/crypto/tcrypt.c
@@ -1802,6 +1802,10 @@ static int do_test(const char *alg, u32 type, u32 mask, int m, u32 num_mb)
 		ret = min(ret, tcrypt_test("cmac(sm4)"));
 		break;
 
+	case 180:
+		ret = min(ret, tcrypt_test("authenc(hmac(sha256),cbc(aes))"));
+		break;
+
 	case 181:
 		ret = min(ret, tcrypt_test("authenc(hmac(sha1),cbc(des))"));
 		break;
-- 
2.34.1

