Index: linux-5.4.238/drivers/net/phy/Kconfig
===================================================================
--- linux-5.4.238.orig/drivers/net/phy/Kconfig	2023-12-19 12:20:34.714805000 +0800
+++ linux-5.4.238/drivers/net/phy/Kconfig	2023-12-25 10:11:11.328554000 +0800
@@ -337,6 +337,8 @@
 
 source "drivers/net/phy/mtk/mt753x/Kconfig"
 
+source "drivers/net/phy/airoha/an8855/Kconfig"
+
 comment "MII PHY device drivers"
 
 config SFP
Index: linux-5.4.238/drivers/net/phy/Makefile
===================================================================
--- linux-5.4.238.orig/drivers/net/phy/Makefile	2023-12-19 12:20:34.718809000 +0800
+++ linux-5.4.238/drivers/net/phy/Makefile	2023-12-25 10:13:11.891535000 +0800
@@ -121,5 +121,6 @@
 obj-$(CONFIG_VITESSE_PHY)	+= vitesse.o
 obj-$(CONFIG_XILINX_GMII2RGMII) += xilinx_gmii2rgmii.o
 obj-$(CONFIG_MT753X_GSW)        += mtk/mt753x/
+obj-$(CONFIG_AN8855_GSW)        += airoha/an8855/
 obj-$(CONFIG_RTL8367S_GSW)	+= rtk/
 
