--- a/drivers/char/hw_random/mtk-rng.c
+++ b/drivers/char/hw_random/mtk-rng.c
@@ -231,6 +231,10 @@ static const struct mtk_rng_of_data mt79
 	.rng_version = 2,
 };
 
+static const struct mtk_rng_of_data mt7987_rng_data = {
+	.rng_version = 2,
+};
+
 static const struct mtk_rng_of_data mt7988_rng_data = {
 	.rng_version = 2,
 };
@@ -242,6 +246,7 @@ static const struct mtk_rng_of_data mt76
 static const struct of_device_id mtk_rng_match[] = {
 	{ .compatible = "mediatek,mt7981-rng", .data = &mt7981_rng_data },
 	{ .compatible = "mediatek,mt7986-rng", .data = &mt7986_rng_data },
+	{ .compatible = "mediatek,mt7987-rng", .data = &mt7987_rng_data },
 	{ .compatible = "mediatek,mt7988-rng", .data = &mt7988_rng_data },
 	{ .compatible = "mediatek,mt7623-rng", .data = &mt7623_rng_data },
 	{},
