Index: linux-5.4.238/drivers/net/dsa/Kconfig
===================================================================
--- linux-5.4.238.orig/drivers/net/dsa/Kconfig	2023-12-26 15:31:49.427259000 +0800
+++ linux-5.4.238/drivers/net/dsa/Kconfig	2023-12-26 15:46:52.655226000 +0800
@@ -48,6 +48,8 @@
 	  This enables support for the Marvell 88E6060 ethernet switch
 	  chip.
 
+source "drivers/net/dsa/airoha/an8855/Kconfig"
+
 source "drivers/net/dsa/microchip/Kconfig"
 
 source "drivers/net/dsa/mv88e6xxx/Kconfig"
Index: linux-5.4.238/drivers/net/dsa/Makefile
===================================================================
--- linux-5.4.238.orig/drivers/net/dsa/Makefile	2023-12-26 15:32:08.081306000 +0800
+++ linux-5.4.238/drivers/net/dsa/Makefile	2023-12-26 15:47:59.858217000 +0800
@@ -18,6 +18,7 @@
 obj-$(CONFIG_NET_DSA_VITESSE_VSC73XX) += vitesse-vsc73xx-core.o
 obj-$(CONFIG_NET_DSA_VITESSE_VSC73XX_PLATFORM) += vitesse-vsc73xx-platform.o
 obj-$(CONFIG_NET_DSA_VITESSE_VSC73XX_SPI) += vitesse-vsc73xx-spi.o
+obj-y				+= airoha/an8855/
 obj-y				+= b53/
 obj-y				+= microchip/
 obj-y				+= mv88e6xxx/
