--- a/drivers/net/phy/phylink.c
+++ b/drivers/net/phy/phylink.c
@@ -686,7 +686,10 @@ static void phylink_link_up(struct phyli
 		 * the link_state) to the interface speed, and will send
 		 * pause frames to the MAC to limit its transmission speed.
 		 */
-		speed = phylink_interface_max_speed(link_state.interface);
+		/* For tunnel HW offload, we need to get true link rate to
+		 * set QDMA rate limit as link rate.
+		 */
+		// speed = phylink_interface_max_speed(link_state.interface);
 		duplex = DUPLEX_FULL;
 		rx_pause = true;
 		break;
