Index: linux-5.4.238/drivers/net/ethernet/mediatek/mtk_hnat/Makefile
===================================================================
--- linux-5.4.238.orig/drivers/net/ethernet/mediatek/mtk_hnat/Makefile	2023-09-05 21:55:25.000000000 +0800
+++ linux-5.4.238/drivers/net/ethernet/mediatek/mtk_hnat/Makefile	2023-12-18 11:26:54.110684000 +0800
@@ -2,4 +2,8 @@
 
 obj-$(CONFIG_NET_MEDIATEK_HNAT)         += mtkhnat.o
 mtkhnat-objs := hnat.o hnat_nf_hook.o hnat_debugfs.o hnat_mcast.o
+ifeq ($(CONFIG_NET_DSA_AN8855), y)
+mtkhnat-y	+= hnat_stag.o
+else
 mtkhnat-$(CONFIG_NET_DSA_MT7530)	+= hnat_stag.o
+endif
