From 648532e848587a525aa02add84b119844a4aabc3 Mon Sep 17 00:00:00 2001
From: Sam Shih <sam.shih@mediatek.com>
Date: Fri, 2 Jun 2023 13:06:14 +0800
Subject: [PATCH] [spi-and-storage][999-2328-mtd-add-nmbm-support.patch]

---
 drivers/mtd/Kconfig  | 2 ++
 drivers/mtd/Makefile | 2 ++
 2 files changed, 4 insertions(+)

diff --git a/drivers/mtd/Kconfig b/drivers/mtd/Kconfig
index 503f19763..3ed42b402 100644
--- a/drivers/mtd/Kconfig
+++ b/drivers/mtd/Kconfig
@@ -228,6 +228,8 @@ source "drivers/mtd/ubi/Kconfig"
 
 source "drivers/mtd/hyperbus/Kconfig"
 
+source "drivers/mtd/nmbm/Kconfig"
+
 source "drivers/mtd/composite/Kconfig"
 
 endif # MTD
diff --git a/drivers/mtd/Makefile b/drivers/mtd/Makefile
index 147e79596..f27f66784 100644
--- a/drivers/mtd/Makefile
+++ b/drivers/mtd/Makefile
@@ -33,5 +33,7 @@ obj-$(CONFIG_MTD_SPI_NOR)	+= spi-nor/
 obj-$(CONFIG_MTD_UBI)		+= ubi/
 obj-$(CONFIG_MTD_HYPERBUS)	+= hyperbus/
 
+obj-y				+= nmbm/
+
 # Composite drivers must be loaded last
 obj-y				+= composite/
-- 
2.34.1

