Index: linux-5.4.238/net/dsa/Makefile
===================================================================
--- linux-5.4.238.orig/net/dsa/Makefile	2023-12-09 09:43:04.335694000 +0800
+++ linux-5.4.238/net/dsa/Makefile	2023-12-09 10:24:27.672514000 +0800
@@ -16,3 +16,4 @@
 obj-$(CONFIG_NET_DSA_TAG_QCA) += tag_qca.o
 obj-$(CONFIG_NET_DSA_TAG_SJA1105) += tag_sja1105.o
 obj-$(CONFIG_NET_DSA_TAG_TRAILER) += tag_trailer.o
+obj-$(CONFIG_NET_DSA_TAG_AIROHA) += tag_arht.o
Index: linux-5.4.238/net/dsa/Kconfig
===================================================================
--- linux-5.4.238.orig/net/dsa/Kconfig	2023-12-09 09:43:04.332694000 +0800
+++ linux-5.4.238/net/dsa/Kconfig	2023-12-09 10:26:13.596504000 +0800
@@ -74,6 +74,12 @@
 	  Say Y or M if you want to enable support for tagging frames for
 	  Mediatek switches.
 
+config NET_DSA_TAG_AIROHA
+	tristate "Tag driver for Airoha switches"
+	help
+	  Say Y or M if you want to enable support for tagging frames for
+	  Airoha switches.
+
 config NET_DSA_TAG_KSZ
 	tristate "Tag driver for Microchip 8795/9477/9893 families of switches"
 	help
Index: linux-5.4.238/include/net/dsa.h
===================================================================
--- linux-5.4.238.orig/include/net/dsa.h	2023-12-09 09:43:17.940694000 +0800
+++ linux-5.4.238/include/net/dsa.h	2023-12-09 10:30:06.432504000 +0800
@@ -43,6 +43,7 @@
 #define DSA_TAG_PROTO_SJA1105_VALUE		13
 #define DSA_TAG_PROTO_KSZ8795_VALUE		14
 #define DSA_TAG_PROTO_RTL4_A_VALUE		17
+#define DSA_TAG_PROTO_ARHT_VALUE		28
 
 enum dsa_tag_protocol {
 	DSA_TAG_PROTO_NONE		= DSA_TAG_PROTO_NONE_VALUE,
@@ -61,6 +62,7 @@
 	DSA_TAG_PROTO_SJA1105		= DSA_TAG_PROTO_SJA1105_VALUE,
 	DSA_TAG_PROTO_KSZ8795		= DSA_TAG_PROTO_KSZ8795_VALUE,
 	DSA_TAG_PROTO_RTL4_A		= DSA_TAG_PROTO_RTL4_A_VALUE,
+	DSA_TAG_PROTO_ARHT		= DSA_TAG_PROTO_ARHT_VALUE,
 };
 
 struct packet_type;
