--- a/drivers/net/ethernet/mediatek/mtk_eth_soc.c
+++ b/drivers/net/ethernet/mediatek/mtk_eth_soc.c
@@ -5860,6 +5860,11 @@ static int mtk_probe(struct platform_dev
 		mtk_qdma_debugfs_init(eth);
 	}
 
+	if (of_machine_is_compatible("edgecore,eap111") || of_machine_is_compatible("edgecore,eap112")) {
+		sprintf(eth->netdev[0]->name, "eth1");
+		sprintf(eth->netdev[1]->name, "eth0");
+	}
+
 	for (i = 0; i < MTK_MAX_DEVS; i++) {
 		if (!eth->netdev[i])
 			continue;
