From 6fa8802185f08391dd3eb6e0609268e1108a7f57 Mon Sep 17 00:00:00 2001
From: Sam Shih <sam.shih@mediatek.com>
Date: Fri, 2 Jun 2023 13:06:15 +0800
Subject: [PATCH] [spi-and-storage][999-2332-mtd-add-mtk-snand-driver.patch]

---
 drivers/mtd/Kconfig  | 2 ++
 drivers/mtd/Makefile | 2 ++
 2 files changed, 4 insertions(+)

diff --git a/drivers/mtd/Kconfig b/drivers/mtd/Kconfig
index 3ed42b402..4b6f9d851 100644
--- a/drivers/mtd/Kconfig
+++ b/drivers/mtd/Kconfig
@@ -230,6 +230,8 @@ source "drivers/mtd/hyperbus/Kconfig"
 
 source "drivers/mtd/nmbm/Kconfig"
 
+source "drivers/mtd/mtk-snand/Kconfig"
+
 source "drivers/mtd/composite/Kconfig"
 
 endif # MTD
diff --git a/drivers/mtd/Makefile b/drivers/mtd/Makefile
index f27f66784..df786f994 100644
--- a/drivers/mtd/Makefile
+++ b/drivers/mtd/Makefile
@@ -35,5 +35,7 @@ obj-$(CONFIG_MTD_HYPERBUS)	+= hyperbus/
 
 obj-y				+= nmbm/
 
+obj-$(CONFIG_MTK_SPI_NAND)	+= mtk-snand/
+
 # Composite drivers must be loaded last
 obj-y				+= composite/
-- 
2.34.1

