--- a/drivers/iio/pressure/Kconfig
+++ b/drivers/iio/pressure/Kconfig
@@ -71,6 +71,17 @@ config ZTS8032
 	help
 	  Support for the Zilltek ZTS8032 barometric pressure sensor.
 
+config ZTS8232
+        tristate "Zilltek ZTS8232 pressure and temperature sensor"
+        depends on I2C
+        select REGMAP_I2C
+        help
+          Support for the Zilltek ZTS8232 digital barometric pressure sensor.
+          It can be accessed over I2C bus.
+
+          This driver can also be built as a module.  If so, the module will be
+          called zts8232.
+
 config HID_SENSOR_PRESS
 	depends on HID_SENSOR_HUB
 	select IIO_BUFFER
--- a/drivers/iio/pressure/Makefile
+++ b/drivers/iio/pressure/Makefile
@@ -11,6 +11,7 @@ obj-$(CONFIG_BMP280_I2C) += bmp280-i2c.o
 obj-$(CONFIG_BMP280_SPI) += bmp280-spi.o
 obj-$(CONFIG_DPS310) += dps310.o
 obj-$(CONFIG_ZTS8032) += zts8032.o
+obj-$(CONFIG_ZTS8232) += zts8232.o
 obj-$(CONFIG_IIO_CROS_EC_BARO) += cros_ec_baro.o
 obj-$(CONFIG_HID_SENSOR_PRESS)   += hid-sensor-press.o
 obj-$(CONFIG_HP03) += hp03.o
